---
title: "Revision plots and analysis"
output: html_document
---

<!-- I am using this Rmd for all the extra plots and analysis required for the paper revision -->
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#Upload libraries
library(ggplot2)
library(gridExtra)
library(cowplot)
library(reshape2)
library(hashmap)
library(gtools)
library(ggtree)
library(ape)
library(dplyr)
library(stringr)
library(ggforce)
library(pheatmap)
library(impute)
library(ggpubr)
library(lattice)
library(grid)
library(tidyverse)

#Set paths for input/output directories
home = "/Users/federica/mnt/"
metadata_dir = paste0(home, "projects/bilaterian_GE/data/samples_metadata/")
evo_distances_dir = paste0(home, "projects/bilaterian_GE/data/evo_distances/")
pca_dir = paste0(home, "projects/bilaterian_GE/data/pca_analysis_all_norms/")
splsda_dir = paste0(home, "projects/bilaterian_GE/data/splsda_analysis/")
broccoli_revision_dir = paste0(home, "projects/bilaterian_GE/data/broccoli_revision/")
pfam_transfers_dir = paste0(home, "projects/bilaterian_GE/data/PFAM_transfers/revision/")
ts_gains_losses_dir = paste0(home, "projects/bilaterian_GE/data/ts_gains_losses/species_QN_5_TPM_v6/All_version2/inferences/")
ts_gains_losses_raw_dir = paste0(home, "projects/bilaterian_GE/data/ts_gains_losses/species_QN_5_TPM_v6/All_version2/")
tau_revision_dir = paste0(home, "projects/bilaterian_GE/data/tau_revision/")
tau_original_dir = paste0(home, "projects/bilaterian_GE/data/ts_call/species_QN_5_TPM_taus/All_version2/")

version = "All_version2"
category = "STRICT"
clade = "Bilateria"
evo_type = "conserved"

#Generate variables
original_species = c("Hs2", "Mm2", "Bt2", "Mdo", "Gga", "Xtr", "Dre", "Cmi", "Bla", "Sp2", "Dme", "Eba", "Aae", "Bmo", "Tca", "Ame", "Bge", "Cdi", "Sma", "Obi")
all_species = c("Hsa", "Mmu", "Bta", "Mdo", "Gga", "Xtr", "Dre", "Cmi", "Bla", "Spu", "Dme", "Eba", "Aae", "Bmo", "Tca", "Ame", "Bge", "Cdi", "Sma", "Obi")
ordered_species = all_species
#Bilateria = all_species
#Vertebrata = c("Hsa", "Mmu", "Bta", "Mdo", "Gga", "Xtr", "Dre", "Cmi")
#Insecta = c("Dme", "Eba", "Aae", "Bmo", "Tca", "Ame", "Bge", "Cdi")
Outgroups = c("Bla", "Spu", "Sma", "Obi")
all_tissues=c("Adipose", "DigestiveTract", "Kidney", "Epithelial", "Muscle", "Neural", "Ovary", "Testis")
ordered_tissues = c("Neural", "Testis", "Ovary", "Muscle", "Kidney", "Epithelial", "DigestiveTract", "Adipose")

####### Variables neeeded to plot the tree
Bilateria = all_species
Euarchontoglires = c("Hs2", "Mm2")
Eutheria = c(Euarchontoglires, "Bt2")
Mammalia = c(Eutheria, "Mdo")
Amniota = c(Mammalia, "Gga")
Tetrapoda = c(Amniota, "Xtr")
Euteleostomi = c(Tetrapoda, "Dre")
Vertebrata = c(Euteleostomi, "Cmi")
Chordata = c(Vertebrata, "Bla")
Deuterostoma = c(Chordata, "Sp2")
Cyclorrapha = c("Dme", "Eba")
Diptera = c(Cyclorrapha, "Aae")
Panorpidae = c(Diptera, "Bmo")
Oligoneoptera = c(Panorpidae, "Tca")
Holometabola = c(Oligoneoptera, "Ame")
Neoptera = c(Holometabola, "Bge")
Insecta = c(Neoptera, "Cdi")
Arthropoda = c(Insecta, "Sma")
Protostoma = c(Arthropoda, "Obi")


#Set aestetics vectors
tissue_colors = c("Adipose"="firebrick1", "Neural"="mediumblue", 
                  "Muscle"="springgreen3", "Ovary"="darkorchid", 
                  "Kidney"="gold", "Testis"="violet", 
                  "DigestiveTract"="chocolate1", "Epithelial"="steelblue1")

tissue_letters = c("Neural" = utf8ToInt("N"), "Testis"=utf8ToInt("T"),
                   "Ovary" = utf8ToInt("O"), "Muscle"=utf8ToInt("M"), 
                   "Kidney" = utf8ToInt("X"), "Epithelial"=utf8ToInt("E"),
                   "DigestiveTract"=utf8ToInt("D"), "Adipose"=utf8ToInt("A"))

species_colors = c("Hsa"="#00BFFF", "Mmu"="#159DFF", "Bta"="#1873CD", "Mdo"="#174B8B", "Gga"="#3F3F8B", "Xtr"="#64398B",
                 "Dre"="#82359D", "Cmi"="#9932CC", "Bla"="#2E8B57", "Spu"="#96B12B", rev(c("Obi"="#FFFF00",
                 "Sma"="#FFCC33", "Cdi"="goldenrod", "Bge"="burlywood2", "Ame"="#FF9966", "Tca"="#FF6600",
                 "Bmo"="#CC6600", "Aae"="#993300", "Eba"="firebrick3", "Dme"="red")))

species_shapes = c("Hsa"=0, "Mmu"=1, "Bta"=2, "Mdo"=3, "Gga"=4, "Xtr"=5, 
                   "Dre"=6, "Cmi"=7, "Bla"=9, "Spu"=8, "Dme"=15, "Eba"=18, "Aae"= 23, 
                   "Bmo"=25, "Tca"=17, "Bge"=10, "Ame"=16, "Cdi"=13, "Sma"=14, "Obi"=11)
species_alphas = c("Hs2"=1, "Mm2"=1, "Bt2"=1, "Mdo"=1, "Gga"=1, "Xtr"=1, "Dre"=1, 
                   "Cmi"=1, "Bla"=0.5, "Sp2"=0.5, "Sma"=0.5, "Ame"=0.5, "Dme"=0.5, "Cdi"=0.5, 
                   "Obi"=0.5, "Bge"=0.5, "BmA"=0.5, "Tca"=0.5)
clade_colors = c("Vertebrata"= "mediumorchid", "Deuterostoma"="olivedrab3", 
                 "Insecta"="tan3", "Protostoma"="yellow", "Bilateria"="firebrick1", 
                 "Outgroups"="dimgray")

scientific_names_vector = c("Hsa"="Homo sapiens (Hsa)", "Mmu"="Mus musculus (Mmu)", "Bta"="Bos taurus (Bta)", "Mdo"="Monodelphis domestica (Mdo)", 
                            "Gga"="Gallus gallus (Gga)", "Xtr"="Xenopus tropicalis (Xtr)", "Dre"="Danio rerio (Dre)", "Cmi"="Callorhinchus milii (Cmi)", 
                            "Bla"="Branchiostoma lanceolatum (Bla)", "Spu"="Strongylocentrotus purpuratus (Spu)", "Dme"="Drosophila melanogaster (Dme)",
                            "Eba"="Episyrphus balteatus (Eba)", "Aae"="Aaedes aegypti (Aae)", "Bmo"="Bombyx mori (Bmo)", "Tca"="Tribolium castaneum (Tca)",
                            "Ame"="Apis mellifera (Ame)", "Bge"="Blattella germanica (Bge)", "Cdi"="Cloeon dipterum (Cdi)",
                            "Sma"="Strigamia marittima (Sma)", "Obi"="Octopus bimaculoides (Obi)")

all_ancestors_colors = c("Euarchontoglires"="#159DFF",
                         "Eutheria" = "#1873CD",
                         "Mammalia" = "#174B8B",
                         "Amniota" = "#3F3F8B",  
                         "Tetrapoda" =  "#64398B",
                         "Euteleostomi" = "#82359D",
                         "Vertebrata" = "#9932CC",
                         "Chordata" = "#2E8B57",
                         "Deuterostoma" = "#96B12B",
                         "Cyclorrapha" = "firebrick3",
                         "Diptera" = "#993300",
                         "Panorpidae" = "#CC6600",
                         "Oligoneoptera" = "#FF6600", 
                         "Holometabola" = "#FF9966",
                         "Neoptera" = "burlywood2",
                         "Insecta" = "goldenrod",
                         "Arthropoda" = "#FFCC33",
                         "Protostoma" = "#FFFF00",
                         "species\nspecific" = "gray71")

################################################
####### SPECIES DISTANCES ######################
################################################

species_distances = c("Hsa_Mmu"=1, "Hsa_Bta"=2, "Hsa_Mdo"=3, "Hsa_Gga"=4, "Hsa_Xtr"=5, "Hsa_Dre"=6, "Hsa_Cmi"=7, "Hsa_Bla"=8, "Hsa_Spu"=9, 
                      "Hsa_Dme"=10, "Hsa_Eba"=10, "Hsa_Aae"=10, "Hsa_Bmo"=10, "Hsa_Tca"=10, "Hsa_Ame"=10, "Hsa_Bge"=10, "Hsa_Cdi"=10, "Hsa_Sma"=10, "Hsa_Obi"=10,
                      "Mmu_Bta"=2, "Mmu_Mdo"=3, "Mmu_Gga"=4, "Mmu_Xtr"=5, "Mmu_Dre"=6, "Mmu_Cmi"=7, "Mmu_Bla"=8, "Mmu_Spu"=9,
                      "Mmu_Dme"=10, "Mmu_Eba"=10, "Mmu_Aae"=10, "Mmu_Bmo"=10, "Mmu_Tca"=10, "Mmu_Ame"=10, "Mmu_Bge"=10, "Mmu_Cdi"=10, "Mmu_Sma"=10, "Mmu_Obi"=10, 
                      "Bta_Mdo"=3, "Bta_Gga"=4, "Bta_Xtr"=5, "Bta_Dre"=6, "Bta_Cmi"=7, "Bta_Bla"=8, "Bta_Spu"=9,
                      "Bta_Dme"=10, "Bta_Eba"=10, "Bta_Aae"=10, "Bta_Bmo"=10, "Bta_Tca"=10, "Bta_Ame"=10, "Bta_Bge"=10, "Bta_Cdi"=10, "Bta_Sma"=10, "Bta_Obi"=10,
                      "Mdo_Gga"=4, "Mdo_Xtr"=5, "Mdo_Dre"=6, "Mdo_Cmi"=7, "Mdo_Bla"=8, "Mdo_Spu"=9,
                      "Mdo_Dme"=10, "Mdo_Eba"=10, "Mdo_Aae"=10, "Mdo_Bmo"=10, "Mdo_Tca"=10, "Mdo_Ame"=10, "Mdo_Bge"=10, "Mdo_Cdi"=10, "Mdo_Sma"=10, "Mdo_Obi"=10,
                      "Gga_Xtr"=5, "Gga_Dre"=6 ,"Gga_Cmi"=7, "Gga_Bla"=8, "Gga_Spu"=9, 
                      "Gga_Dme"=10, "Gga_Eba"=10, "Gga_Aae"=10, "Gga_Bmo"=10, "Gga_Tca"=10, "Gga_Ame"=10, "Gga_Bge"=10, "Gga_Cdi"=10, "Gga_Sma"=10, "Gga_Obi"=10,
                      "Xtr_Dre"=6, "Xtr_Cmi"=7, "Xtr_Bla"=8, "Xtr_Spu"=9, 
                      "Xtr_Dme"=10, "Xtr_Eba"=10, "Xtr_Aae"=10, "Xtr_Bmo"=10, "Xtr_Tca"=10, "Xtr_Ame"=10, "Xtr_Bge"=10, "Xtr_Cdi"=10, "Xtr_Sma"=10, "Xtr_Obi"=10,
                      "Dre_Cmi"=7, "Dre_Bla"=8, "Dre_Spu"=9,
                      "Dre_Dme"=10, "Dre_Eba"=10, "Dre_Aae"=10, "Dre_Bmo"=10, "Dre_Tca"=10, "Dre_Ame"=10, "Dre_Bge"=10, "Dre_Cdi"=10, "Dre_Sma"=10, "Dre_Obi"=10,
                      "Cmi_Bla"=8, "Cmi_Spu"=9,
                      "Cmi_Dme"=10, "Cmi_Eba"=10, "Cmi_Aae"=10, "Cmi_Bmo"=10, "Cmi_Tca"=10, "Cmi_Ame"=10, "Cmi_Bge"=10, "Cmi_Cdi"=10, "Cmi_Sma"=10, "Cmi_Obi"=10,
                      "Bla_Spu"=9,
                      "Bla_Dme"=10, "Bla_Eba"=10, "Bla_Aae"=10, "Bla_Bmo"=10, "Bla_Tca"=10, "Bla_Ame"=10, "Bla_Bge"=10, "Bla_Cdi"=10, "Bla_Sma"=10, "Bla_Obi"=10,
                      "Spu_Dme"=10, "Spu_Eba"=10, "Spu_Aae"=10, "Spu_Bmo"=10, "Spu_Tca"=10, "Spu_Ame"=10, "Spu_Bge"=10, "Spu_Cdi"=10, "Spu_Sma"=10, "Spu_Obi"=10,
                      "Dme_Eba"=1, "Dme_Aae"=2, "Dme_Bmo"=3, "Dme_Tca"=4, "Dme_Ame"=5, "Dme_Bge"=6, "Dme_Cdi"=7, "Dme_Sma"=8, "Dme_Obi"=9,
                      "Eba_Aae"=2, "Eba_Bmo"=3, "Eba_Tca"=4, "Eba_Ame"=5, "Eba_Bge"=6, "Eba_Cdi"=7, "Eba_Sma"=8, "Eba_Obi"=9,
                      "Aae_Bmo"=3, "Aae_Tca"=4, "Aae_Ame"=5, "Aae_Bge"=6, "Aae_Cdi"=7, "Aae_Sma"=8, "Aae_Obi"=9,
                      "Bmo_Tca"=4, "Bmo_Ame"=5, "Bmo_Bge"=6, "Bmo_Cdi"=7, "Bmo_Sma"=8, "Bmo_Obi"=9,
                      "Tca_Ame"=5, "Tca_Bge"=6, "Tca_Cdi"=7, "Tca_Sma"=8, "Tca_Obi"=9, 
                      "Ame_Bge"=6, "Ame_Cdi"=7, "Ame_Sma"=8, "Ame_Obi"=9,
                      "Bge_Cdi"=7, "Bge_Sma"=8, "Bge_Obi"=9,
                      "Cdi_Sma"=8, "Cdi_Obi"=9, 
                      "Sma_Obi"=9)
species_distances_reverted = species_distances
names(species_distances_reverted) = paste0(gsub(".*_", "", names(species_distances_reverted)), "_", gsub("_.*", "", names(species_distances_reverted)))
same_species_distances = rep(0, length(all_species)); names(same_species_distances) = paste0(all_species, "_", all_species)
all_species_distances = c(species_distances, species_distances_reverted, same_species_distances)
all_species_distances_df = as.data.frame(all_species_distances) %>% rownames_to_column("Species_combo")

deuterostoma_y_labels = c("Euarchontoglires", "Eutheria", "Mammalia", "Amniota", "Tetrapoda", "Euteleostomi", "Vertebrata", "Chordata", "Deuterostoma", "Bilateria")
protostoma_y_labels = c("Cyclorrapha", "Diptera", "Panorpidae", "Oligoneoptera", "Holometabola", "Neoptera", "Insecta", "Arthropoda", "Protostoma", "Bilateria")

#Build node level dict
ancestor_node_level_dict = hashmap(c(deuterostoma_y_labels[deuterostoma_y_labels!="Bilateria"], 
                                     protostoma_y_labels[protostoma_y_labels!="Bilateria"],
                                     original_species,
                                     "Bilateria"), 
                                   c(rep(rev(seq(1,9)),2), 
                                     rep(10, length(original_species)),
                                     0))

all_ancestors = c(rev(deuterostoma_y_labels[deuterostoma_y_labels!="Bilateria"]), rev(protostoma_y_labels[protostoma_y_labels!="Bilateria"]))
#### Generate name vector with ancestor and corresponding earliest diverging species
earliest_diverging_species_df = data.frame(Diverging_species = c(Deuterostoma[3:length(Deuterostoma)], 
                                                                    Protostoma[3:length(Protostoma)]),
                                              Ancestor = c(deuterostoma_y_labels[2:(length(deuterostoma_y_labels)-1)],
                                                           protostoma_y_labels[2:(length(protostoma_y_labels)-1)]))

### TAU
tau_cutoff = 0.75
```

```{r functions, echo=FALSE, echo=FALSE, warning=FALSE, dependson=c("setup", "functions")}
######################################################
######### PCA functions ##############################
######################################################

PCA_basic_plot = function(expr_df, features_table, species_set, cluster_number) {
    imputation_out = impute.knn(as.matrix(expr_df), k = 10) #impute missing values.
    imputation_df = as.data.frame(imputation_out$data) #transform to a dataframe.
    #compute PCA. I need the genes on columns in the input table.
    PCA_res = prcomp(as.matrix(t(imputation_df)), center = TRUE, scale = TRUE) #center and scale the PCs.
    ordered_PCA_res = as.data.frame(PCA_res$x[order(rownames(PCA_res$x)),]) 
    #filter the sample only for the samples that made it after the sva correction
    filtered_features_table = features_table[rownames(features_table) %in% colnames(imputation_df),]
    ordered_features_table = filtered_features_table[order(rownames(filtered_features_table)),] #order the table
    all(rownames(ordered_features_table) == rownames(ordered_PCA_res))
    #compute the percentage of variance explained
    eigs <- PCA_res$sdev^2
    first_PC_var = round(eigs[1]/sum(eigs), 3)*100
    second_PC_var = round(eigs[2]/sum(eigs), 3)*100
    third_PC_var = round(eigs[3]/sum(eigs), 3)*100
    fourth_PC_var = round(eigs[4]/sum(eigs), 3)*100
    #filter shapes vector
    my_filtered_shapes_vector=species_shapes[names(species_shapes) %in% unique(as.vector(ordered_features_table$species))] #filter the shape vector
    #filter ordered species
    filt_ordered_species = ordered_species[ordered_species %in% unique(as.vector(ordered_features_table$species))]
    ordered_features_table$species <- factor(ordered_features_table$species, levels=filt_ordered_species) #change the order of the levels in the factor
    #plot the first two PC
    ordered_PCA_res$tissues = ordered_features_table$tissues
    ordered_PCA_res$species = ordered_features_table$species
    
    p = ggplot() + geom_point(data=as.data.frame(ordered_PCA_res), 
                              aes(x=PC1, y=PC2, 
                                  color=tissues, 
                                  fill=tissues,
                                  shape=species), size=2, stroke=0.8) +
      # stat_ellipse(data=as.data.frame(ordered_PCA_res), geom="polygon", aes(x=PC1, y=PC2, color=tissues, fill = tissues),
      #              alpha = 0.1,
      #              show.legend = FALSE,
      #              level = 0.80) +
      ggtitle(paste0(species_set, " ~ ", cluster_number)) + 
      xlab(paste0("PC1: ", first_PC_var, "%")) + 
      ylab(paste0("PC2: ", second_PC_var, "%")) + theme_bw() +
      theme(axis.title = element_text(color="black", size=14), 
            axis.text = element_text(color="black", size=12),
            legend.text = element_text(color="black", size=10), 
            legend.title = element_text(color="black", size=12),
            legend.spacing.y = unit(0, 'cm')) +
      scale_colour_manual(values=tissue_colors) + #this is great because I can have named vectors
      scale_fill_manual(values=tissue_colors) +
      scale_shape_manual(values=my_filtered_shapes_vector, breaks=filt_ordered_species) + 
      guides(color=guide_legend(title="Tissues", byrow=TRUE), 
             fill=guide_legend(title="Tissues", byrow=TRUE), 
             shape=guide_legend(title="Species", byrow=TRUE))
    
    p1 = ggplot() + geom_point(data=as.data.frame(ordered_PCA_res), 
                               aes(x=PC2, y=PC3, 
                                   color=tissues,
                                   fill=tissues,
                                   shape=species), size=2, stroke=0.8) +
      # stat_ellipse(data=as.data.frame(ordered_PCA_res), geom="polygon", aes(x=PC2, y=PC3, color=tissues, fill = tissues),
      #        alpha = 0.1,
      #        show.legend = FALSE,
      #        level = 0.80) +
      ggtitle(paste0(species_set, " ~ ", cluster_number)) + 
      xlab(paste0("PC2: ", second_PC_var, "%")) + 
      ylab(paste0("PC3: ", third_PC_var, "%")) + 
      theme_bw() +
      theme(axis.title = element_text(color="black", size=14), 
            axis.text = element_text(color="black", size=12),
            legend.text = element_text(color="black", size=10), 
            legend.title = element_text(color="black", size=12),
            legend.spacing.y = unit(0, 'cm')) +
      scale_colour_manual(values=tissue_colors) + #this is great because I can have named vectors
      scale_fill_manual(values=tissue_colors) +
      scale_shape_manual(values=my_filtered_shapes_vector, breaks=filt_ordered_species) + 
      guides(color=guide_legend(title="Tissues", byrow=TRUE), 
             fill=guide_legend(title="Tissues", byrow=TRUE), 
             shape=guide_legend(title="Species", byrow=TRUE))
    
    p2 = ggplot() + geom_point(data=as.data.frame(ordered_PCA_res), 
                               aes(x=PC3, y=PC4, 
                                   color=tissues,
                                   fill=tissues,
                                   shape=species), size=2, stroke=0.8) +
        # stat_ellipse(data=as.data.frame(ordered_PCA_res), geom="polygon", aes(x=PC3, y=PC4, color=tissues, fill = tissues),
        #        alpha = 0.1,
        #        show.legend = FALSE,
        #        level = 0.80) +
      ggtitle(paste0(species_set, " ~ ", cluster_number)) + xlab(paste0("PC3: ", third_PC_var, "%")) + ylab(paste0("PC4: ", fourth_PC_var, "%")) + theme_bw() +
      theme(axis.title = element_text(color="black", size=14), 
            axis.text = element_text(color="black", size=12),
            legend.text = element_text(color="black", size=10), 
            legend.title = element_text(color="black", size=12),
            legend.spacing.y = unit(0, 'cm')) +
      scale_colour_manual(values=tissue_colors) + #this is great because I can have named vectors
      scale_fill_manual(values=tissue_colors) +
      scale_shape_manual(values=my_filtered_shapes_vector, breaks=filt_ordered_species) + 
      guides(color=guide_legend(title="Tissues", byrow=TRUE), 
             fill=guide_legend(title="Tissues", byrow=TRUE), 
             shape=guide_legend(title="Species", byrow=TRUE))
    
    res = list()
    res[[1]] = p
    res[[2]] = p1
    res[[3]] = p2
    
    my_plots = plot_grid(p, p1, p2, rel_widths=c(1,1,1), align="h", ncol=3)
    return(list("plots" = my_plots, "results"=res))
}

PCA_basic_ellipses_plot = function(expr_df, features_table, species_set, cluster_number) {
    imputation_out = impute.knn(as.matrix(expr_df), k = 10) #impute missing values.
    imputation_df = as.data.frame(imputation_out$data) #transform to a dataframe.
    #compute PCA. I need the genes on columns in the input table.
    PCA_res = prcomp(as.matrix(t(imputation_df)), center = TRUE, scale = TRUE) #center and scale the PCs.
    ordered_PCA_res = as.data.frame(PCA_res$x[order(rownames(PCA_res$x)),]) 
    #filter the sample only for the samples that made it after the sva correction
    filtered_features_table = features_table[rownames(features_table) %in% colnames(imputation_df),]
    ordered_features_table = filtered_features_table[order(rownames(filtered_features_table)),] #order the table
    all(rownames(ordered_features_table) == rownames(ordered_PCA_res))
    #compute the percentage of variance explained
    eigs <- PCA_res$sdev^2
    first_PC_var = round(eigs[1]/sum(eigs), 3)*100
    second_PC_var = round(eigs[2]/sum(eigs), 3)*100
    third_PC_var = round(eigs[3]/sum(eigs), 3)*100
    fourth_PC_var = round(eigs[4]/sum(eigs), 3)*100
    #filter shapes vector
    my_filtered_shapes_vector=species_shapes[names(species_shapes) %in% unique(as.vector(ordered_features_table$species))] #filter the shape vector
    #filter ordered species
    filt_ordered_species = ordered_species[ordered_species %in% unique(as.vector(ordered_features_table$species))]
    ordered_features_table$species <- factor(ordered_features_table$species, levels=filt_ordered_species) #change the order of the levels in the factor
    #plot the first two PC
    ordered_PCA_res$tissues = ordered_features_table$tissues
    ordered_PCA_res$species = ordered_features_table$species
    
    p = ggplot() + geom_point(data=as.data.frame(ordered_PCA_res), 
                              aes(x=PC1, y=PC2, 
                                  color=tissues, 
                                  fill=tissues,
                                  shape=species), size=2, stroke=0.8) +
      stat_ellipse(data=as.data.frame(ordered_PCA_res), geom="polygon", aes(x=PC1, y=PC2, color=tissues, fill = tissues),
                   alpha = 0.1,
                   show.legend = FALSE,
                   level = 0.80) +
      ggtitle(paste0(species_set, " ~ ", cluster_number)) + 
      xlab(paste0("PC1: ", first_PC_var, "%")) + 
      ylab(paste0("PC2: ", second_PC_var, "%")) + theme_bw() +
      theme(axis.title = element_text(color="black", size=14), 
            axis.text = element_text(color="black", size=12),
            legend.text = element_text(color="black", size=10), 
            legend.title = element_text(color="black", size=12),
            legend.spacing.y = unit(0, 'cm')) +
      scale_colour_manual(values=tissue_colors) + #this is great because I can have named vectors
      scale_fill_manual(values=tissue_colors) +
      scale_shape_manual(values=my_filtered_shapes_vector, breaks=filt_ordered_species) + 
      guides(color=guide_legend(title="Tissues", byrow=TRUE), 
             fill=guide_legend(title="Tissues", byrow=TRUE), 
             shape=guide_legend(title="Species", byrow=TRUE))
    
    p1 = ggplot() + geom_point(data=as.data.frame(ordered_PCA_res), 
                               aes(x=PC2, y=PC3, 
                                   color=tissues,
                                   fill=tissues,
                                   shape=species), size=2, stroke=0.8) +
      stat_ellipse(data=as.data.frame(ordered_PCA_res), geom="polygon", aes(x=PC2, y=PC3, color=tissues, fill = tissues),
             alpha = 0.1,
             show.legend = FALSE,
             level = 0.80) +
      ggtitle(paste0(species_set, " ~ ", cluster_number)) + 
      xlab(paste0("PC2: ", second_PC_var, "%")) + 
      ylab(paste0("PC3: ", third_PC_var, "%")) + 
      theme_bw() +
      theme(axis.title = element_text(color="black", size=14), 
            axis.text = element_text(color="black", size=12),
            legend.text = element_text(color="black", size=10), 
            legend.title = element_text(color="black", size=12),
            legend.spacing.y = unit(0, 'cm')) +
      scale_colour_manual(values=tissue_colors) + #this is great because I can have named vectors
      scale_fill_manual(values=tissue_colors) +
      scale_shape_manual(values=my_filtered_shapes_vector, breaks=filt_ordered_species) + 
      guides(color=guide_legend(title="Tissues", byrow=TRUE), 
             fill=guide_legend(title="Tissues", byrow=TRUE), 
             shape=guide_legend(title="Species", byrow=TRUE))
    
    p2 = ggplot() + geom_point(data=as.data.frame(ordered_PCA_res), 
                               aes(x=PC3, y=PC4, 
                                   color=tissues,
                                   fill=tissues,
                                   shape=species), size=2, stroke=0.8) +
        stat_ellipse(data=as.data.frame(ordered_PCA_res), geom="polygon", aes(x=PC3, y=PC4, color=tissues, fill = tissues),
               alpha = 0.1,
               show.legend = FALSE,
               level = 0.80) +
      ggtitle(paste0(species_set, " ~ ", cluster_number)) + xlab(paste0("PC3: ", third_PC_var, "%")) + ylab(paste0("PC4: ", fourth_PC_var, "%")) + theme_bw() +
      theme(axis.title = element_text(color="black", size=14), 
            axis.text = element_text(color="black", size=12),
            legend.text = element_text(color="black", size=10), 
            legend.title = element_text(color="black", size=12),
            legend.spacing.y = unit(0, 'cm')) +
      scale_colour_manual(values=tissue_colors) + #this is great because I can have named vectors
      scale_fill_manual(values=tissue_colors) +
      scale_shape_manual(values=my_filtered_shapes_vector, breaks=filt_ordered_species) + 
      guides(color=guide_legend(title="Tissues", byrow=TRUE), 
             fill=guide_legend(title="Tissues", byrow=TRUE), 
             shape=guide_legend(title="Species", byrow=TRUE))
    
    res = list()
    res[[1]] = p
    res[[2]] = p1
    res[[3]] = p2
    
    my_plots = plot_grid(p, p1, p2, rel_widths=c(1,1,1), align="h", ncol=3)
    return(list("plots" = my_plots, "results"=res))
}

PCA_basic_by_species_plot = function(expr_df, features_table, species_set, cluster_number) {
    imputation_out = impute.knn(as.matrix(expr_df), k = 10) #impute missing values.
    imputation_df = as.data.frame(imputation_out$data) #transform to a dataframe.
    #compute PCA. I need the genes on columns in the input table.
    PCA_res = prcomp(as.matrix(t(imputation_df)), center = TRUE, scale = TRUE) #center and scale the PCs.
    ordered_PCA_res = as.data.frame(PCA_res$x[order(rownames(PCA_res$x)),]) 
    #filter the sample only for the samples that made it after the sva correction
    filtered_features_table = features_table[rownames(features_table) %in% colnames(imputation_df),]
    ordered_features_table = filtered_features_table[order(rownames(filtered_features_table)),] #order the table
    all(rownames(ordered_features_table) == rownames(ordered_PCA_res))
    #compute the percentage of variance explained
    eigs <- PCA_res$sdev^2
    first_PC_var = round(eigs[1]/sum(eigs), 3)*100
    second_PC_var = round(eigs[2]/sum(eigs), 3)*100
    third_PC_var = round(eigs[3]/sum(eigs), 3)*100
    fourth_PC_var = round(eigs[4]/sum(eigs), 3)*100
    #filter shapes vector
    my_filtered_shapes_vector=species_shapes[names(species_shapes) %in% unique(as.vector(ordered_features_table$species))] #filter the shape vector
    #filter ordered species
    filt_ordered_species = ordered_species[ordered_species %in% unique(as.vector(ordered_features_table$species))]
    ordered_features_table$species <- factor(ordered_features_table$species, levels=filt_ordered_species) #change the order of the levels in the factor
    #plot the first two PC
    ordered_PCA_res$tissues = ordered_features_table$tissues
    ordered_PCA_res$species = ordered_features_table$species
    
    p = ggplot() + geom_point(data=as.data.frame(ordered_PCA_res), 
                              aes(x=PC1, y=PC2, 
                                  color=species, 
                                  fill=species,
                                  shape=tissues), size=3, stroke=0.8) +
      # stat_ellipse(data=as.data.frame(ordered_PCA_res), geom="polygon", aes(x=PC1, y=PC2, color=tissues, fill = tissues),
      #              alpha = 0.1,
      #              show.legend = FALSE,
      #              level = 0.80) +
      ggtitle(paste0(species_set, " ~ ", cluster_number)) + 
      xlab(paste0("PC1: ", first_PC_var, "%")) + 
      ylab(paste0("PC2: ", second_PC_var, "%")) + theme_bw() +
      theme(axis.title = element_text(color="black", size=14), 
            axis.text = element_text(color="black", size=12),
            legend.text = element_text(color="black", size=10), 
            legend.title = element_text(color="black", size=12),
            legend.spacing.y = unit(0, 'cm')) +
      scale_colour_manual(values=species_colors) + #this is great because I can have named vectors
      scale_fill_manual(values=species_colors) +
      scale_shape_manual(values=tissue_letters, breaks=ordered_tissues) + 
      guides(color=guide_legend(title="Species", byrow=TRUE), 
             fill=guide_legend(title="Species", byrow=TRUE), 
             shape=guide_legend(title="Tissues", byrow=TRUE))
    
    p1 = ggplot() + geom_point(data=as.data.frame(ordered_PCA_res), 
                               aes(x=PC2, y=PC3, 
                                   color=species,
                                   fill=species,
                                   shape=tissues), size=3, stroke=0.8) +
      # stat_ellipse(data=as.data.frame(ordered_PCA_res), geom="polygon", aes(x=PC2, y=PC3, color=tissues, fill = tissues),
      #        alpha = 0.1,
      #        show.legend = FALSE,
      #        level = 0.80) +
      ggtitle(paste0(species_set, " ~ ", cluster_number)) + 
      xlab(paste0("PC2: ", second_PC_var, "%")) + 
      ylab(paste0("PC3: ", third_PC_var, "%")) + 
      theme_bw() +
      theme(axis.title = element_text(color="black", size=14), 
            axis.text = element_text(color="black", size=12),
            legend.text = element_text(color="black", size=10), 
            legend.title = element_text(color="black", size=12),
            legend.spacing.y = unit(0, 'cm')) +
      scale_colour_manual(values=species_colors) + #this is great because I can have named vectors
      scale_fill_manual(values=species_colors) +
      scale_shape_manual(values=tissue_letters, breaks=ordered_tissues) + 
      guides(color=guide_legend(title="Species", byrow=TRUE), 
             fill=guide_legend(title="Species", byrow=TRUE), 
             shape=guide_legend(title="Tissues", byrow=TRUE))
    
    p2 = ggplot() + geom_point(data=as.data.frame(ordered_PCA_res), 
                               aes(x=PC3, y=PC4, 
                                   color=species,
                                   fill=species,
                                   shape=tissues), size=3, stroke=0.8) +
        # stat_ellipse(data=as.data.frame(ordered_PCA_res), geom="polygon", aes(x=PC3, y=PC4, color=tissues, fill = tissues),
        #        alpha = 0.1,
        #        show.legend = FALSE,
        #        level = 0.80) +
      ggtitle(paste0(species_set, " ~ ", cluster_number)) + xlab(paste0("PC3: ", third_PC_var, "%")) + ylab(paste0("PC4: ", fourth_PC_var, "%")) + theme_bw() +
      theme(axis.title = element_text(color="black", size=14), 
            axis.text = element_text(color="black", size=12),
            legend.text = element_text(color="black", size=10), 
            legend.title = element_text(color="black", size=12),
            legend.spacing.y = unit(0, 'cm')) +
      scale_colour_manual(values=species_colors) + #this is great because I can have named vectors
      scale_fill_manual(values=species_colors) +
      scale_shape_manual(values=tissue_letters, breaks=ordered_tissues) + 
      guides(color=guide_legend(title="Species", byrow=TRUE), 
             fill=guide_legend(title="Species", byrow=TRUE), 
             shape=guide_legend(title="Tissues", byrow=TRUE))
    
    res = list()
    res[[1]] = p
    res[[2]] = p1
    res[[3]] = p2
    
    my_plots = plot_grid(p, p1, p2, rel_widths=c(1,1,1), align="h", ncol=3)
    return(list("plots" = my_plots, "results"=res))
}

PCA_basic_by_species_ellipses_plot = function(expr_df, features_table, species_set, cluster_number) {
    imputation_out = impute.knn(as.matrix(expr_df), k = 10) #impute missing values.
    imputation_df = as.data.frame(imputation_out$data) #transform to a dataframe.
    #compute PCA. I need the genes on columns in the input table.
    PCA_res = prcomp(as.matrix(t(imputation_df)), center = TRUE, scale = TRUE) #center and scale the PCs.
    ordered_PCA_res = as.data.frame(PCA_res$x[order(rownames(PCA_res$x)),]) 
    #filter the sample only for the samples that made it after the sva correction
    filtered_features_table = features_table[rownames(features_table) %in% colnames(imputation_df),]
    ordered_features_table = filtered_features_table[order(rownames(filtered_features_table)),] #order the table
    all(rownames(ordered_features_table) == rownames(ordered_PCA_res))
    #compute the percentage of variance explained
    eigs <- PCA_res$sdev^2
    first_PC_var = round(eigs[1]/sum(eigs), 3)*100
    second_PC_var = round(eigs[2]/sum(eigs), 3)*100
    third_PC_var = round(eigs[3]/sum(eigs), 3)*100
    fourth_PC_var = round(eigs[4]/sum(eigs), 3)*100
    #filter shapes vector
    my_filtered_shapes_vector=species_shapes[names(species_shapes) %in% unique(as.vector(ordered_features_table$species))] #filter the shape vector
    #filter ordered species
    filt_ordered_species = ordered_species[ordered_species %in% unique(as.vector(ordered_features_table$species))]
    ordered_features_table$species <- factor(ordered_features_table$species, levels=filt_ordered_species) #change the order of the levels in the factor
    #plot the first two PC
    ordered_PCA_res$tissues = ordered_features_table$tissues
    ordered_PCA_res$species = ordered_features_table$species
    
    p = ggplot() + geom_point(data=as.data.frame(ordered_PCA_res), 
                              aes(x=PC1, y=PC2, 
                                  color=species, 
                                  fill=species,
                                  shape=tissues), size=3, stroke=0.8) +
      stat_ellipse(data=as.data.frame(ordered_PCA_res), geom="polygon", aes(x=PC1, y=PC2, color=species, fill = species),
                   alpha = 0.1,
                   show.legend = FALSE,
                   level = 0.80) +
      ggtitle(paste0(species_set, " ~ ", cluster_number)) + 
      xlab(paste0("PC1: ", first_PC_var, "%")) + 
      ylab(paste0("PC2: ", second_PC_var, "%")) + theme_bw() +
      theme(axis.title = element_text(color="black", size=14), 
            axis.text = element_text(color="black", size=12),
            legend.text = element_text(color="black", size=10), 
            legend.title = element_text(color="black", size=12),
            legend.spacing.y = unit(0, 'cm')) +
      scale_colour_manual(values=species_colors) + #this is great because I can have named vectors
      scale_fill_manual(values=species_colors) +
      scale_shape_manual(values=tissue_letters, breaks=ordered_tissues) + 
      guides(color=guide_legend(title="Species", byrow=TRUE), 
             fill=guide_legend(title="Species", byrow=TRUE), 
             shape=guide_legend(title="Tissues", byrow=TRUE))
    
    p1 = ggplot() + geom_point(data=as.data.frame(ordered_PCA_res), 
                               aes(x=PC2, y=PC3, 
                                   color=species,
                                   fill=species,
                                   shape=tissues), size=3, stroke=0.8) +
      stat_ellipse(data=as.data.frame(ordered_PCA_res), geom="polygon", aes(x=PC2, y=PC3, color=species, fill = species),
             alpha = 0.1,
             show.legend = FALSE,
             level = 0.80) +
      ggtitle(paste0(species_set, " ~ ", cluster_number)) + 
      xlab(paste0("PC2: ", second_PC_var, "%")) + 
      ylab(paste0("PC3: ", third_PC_var, "%")) + 
      theme_bw() +
      theme(axis.title = element_text(color="black", size=14), 
            axis.text = element_text(color="black", size=12),
            legend.text = element_text(color="black", size=10), 
            legend.title = element_text(color="black", size=12),
            legend.spacing.y = unit(0, 'cm')) +
      scale_colour_manual(values=species_colors) + #this is great because I can have named vectors
      scale_fill_manual(values=species_colors) +
      scale_shape_manual(values=tissue_letters, breaks=ordered_tissues) + 
      guides(color=guide_legend(title="Species", byrow=TRUE), 
             fill=guide_legend(title="Species", byrow=TRUE), 
             shape=guide_legend(title="Tissues", byrow=TRUE))
    
    p2 = ggplot() + geom_point(data=as.data.frame(ordered_PCA_res), 
                               aes(x=PC3, y=PC4, 
                                   color=species,
                                   fill=species,
                                   shape=tissues), size=3, stroke=0.8) +
        stat_ellipse(data=as.data.frame(ordered_PCA_res), geom="polygon", aes(x=PC3, y=PC4, color=species, fill = species),
               alpha = 0.1,
               show.legend = FALSE,
               level = 0.80) +
      ggtitle(paste0(species_set, " ~ ", cluster_number)) + xlab(paste0("PC3: ", third_PC_var, "%")) + ylab(paste0("PC4: ", fourth_PC_var, "%")) + theme_bw() +
      theme(axis.title = element_text(color="black", size=14), 
            axis.text = element_text(color="black", size=12),
            legend.text = element_text(color="black", size=10), 
            legend.title = element_text(color="black", size=12),
            legend.spacing.y = unit(0, 'cm')) +
      scale_colour_manual(values=species_colors) + #this is great because I can have named vectors
      scale_fill_manual(values=species_colors) +
      scale_shape_manual(values=tissue_letters, breaks=ordered_tissues) + 
      guides(color=guide_legend(title="Species", byrow=TRUE), 
             fill=guide_legend(title="Species", byrow=TRUE), 
             shape=guide_legend(title="Tissues", byrow=TRUE))
    
    res = list()
    res[[1]] = p
    res[[2]] = p1
    res[[3]] = p2
    
    my_plots = plot_grid(p, p1, p2, rel_widths=c(1,1,1), align="h", ncol=3)
    return(list("plots" = my_plots, "results"=res))
}

PCA_basic_by_clade_plot = function(expr_df, features_table, species_set, cluster_number) {
    imputation_out = impute.knn(as.matrix(expr_df), k = 10) #impute missing values.
    imputation_df = as.data.frame(imputation_out$data) #transform to a dataframe.
    #compute PCA. I need the genes on columns in the input table.
    PCA_res = prcomp(as.matrix(t(imputation_df)), center = TRUE, scale = TRUE) #center and scale the PCs.
    ordered_PCA_res = as.data.frame(PCA_res$x[order(rownames(PCA_res$x)),]) 
    #filter the sample only for the samples that made it after the sva correction
    filtered_features_table = features_table[rownames(features_table) %in% colnames(imputation_df),]
    ordered_features_table = filtered_features_table[order(rownames(filtered_features_table)),] #order the table
    all(rownames(ordered_features_table) == rownames(ordered_PCA_res))
    #compute the percentage of variance explained
    eigs <- PCA_res$sdev^2
    first_PC_var = round(eigs[1]/sum(eigs), 3)*100
    second_PC_var = round(eigs[2]/sum(eigs), 3)*100
    third_PC_var = round(eigs[3]/sum(eigs), 3)*100
    fourth_PC_var = round(eigs[4]/sum(eigs), 3)*100
    #filter shapes vector
    my_filtered_shapes_vector=species_shapes[names(species_shapes) %in% unique(as.vector(ordered_features_table$species))] #filter the shape vector
    #filter ordered species
    filt_ordered_species = ordered_species[ordered_species %in% unique(as.vector(ordered_features_table$species))]
    ordered_features_table$species <- factor(ordered_features_table$species, levels=filt_ordered_species) #change the order of the levels in the factor
    #plot the first two PC
    ordered_PCA_res$tissues = ordered_features_table$tissues
    ordered_PCA_res$species = ordered_features_table$species
    #Add clade info
    ordered_PCA_res$clade = "Vertebrate"
    ordered_PCA_res$clade[!(ordered_PCA_res$species %in% Vertebrata)]= "Not_vertebrate"
    
    p = ggplot() + geom_point(data=as.data.frame(ordered_PCA_res), 
                              aes(x=PC1, y=PC2, 
                                  color=clade, 
                                  fill=clade,
                                  shape=tissues), size=3, stroke=0.8) +
      # stat_ellipse(data=as.data.frame(ordered_PCA_res), geom="polygon", aes(x=PC1, y=PC2, color=clade, fill = clade),
      #              alpha = 0.1,
      #              show.legend = FALSE,
      #              level = 0.95) +
      ggtitle(paste0(species_set, " ~ ", cluster_number)) + 
      xlab(paste0("PC1: ", first_PC_var, "%")) + 
      ylab(paste0("PC2: ", second_PC_var, "%")) + theme_bw() +
      theme(axis.title = element_text(color="black", size=14), 
            axis.text = element_text(color="black", size=12),
            legend.text = element_text(color="black", size=10), 
            legend.title = element_text(color="black", size=12),
            legend.spacing.y = unit(0, 'cm')) +
      scale_colour_manual(values=c("Vertebrate"="#82359D", "Not_vertebrate"="goldenrod")) +
      scale_fill_manual(values=c("Vertebrate"="#82359D", "Not_vertebrate"="goldenrod")) +
      scale_shape_manual(values=tissue_letters, breaks=ordered_tissues) + 
      guides(color=guide_legend(title="Species", byrow=TRUE), 
             fill=guide_legend(title="Species", byrow=TRUE), 
             shape=guide_legend(title="Tissues", byrow=TRUE))
    
    p1 = ggplot() + geom_point(data=as.data.frame(ordered_PCA_res), 
                               aes(x=PC2, y=PC3, 
                                   color=clade,
                                   fill=clade,
                                   shape=tissues), size=3, stroke=0.8) +
      # stat_ellipse(data=as.data.frame(ordered_PCA_res), geom="polygon", aes(x=PC2, y=PC3, color=clade, fill = clade),
      #        alpha = 0.1,
      #        show.legend = FALSE,
      #        level = 0.95) +
      ggtitle(paste0(species_set, " ~ ", cluster_number)) + 
      xlab(paste0("PC2: ", second_PC_var, "%")) + 
      ylab(paste0("PC3: ", third_PC_var, "%")) + 
      theme_bw() +
      theme(axis.title = element_text(color="black", size=14), 
            axis.text = element_text(color="black", size=12),
            legend.text = element_text(color="black", size=10), 
            legend.title = element_text(color="black", size=12),
            legend.spacing.y = unit(0, 'cm')) +
      scale_colour_manual(values=c("Vertebrate"="#82359D", "Not_vertebrate"="goldenrod")) +
      scale_fill_manual(values=c("Vertebrate"="#82359D", "Not_vertebrate"="goldenrod")) +
      scale_shape_manual(values=tissue_letters, breaks=ordered_tissues) + 
      guides(color=guide_legend(title="Species", byrow=TRUE), 
             fill=guide_legend(title="Species", byrow=TRUE), 
             shape=guide_legend(title="Tissues", byrow=TRUE))
    
    p2 = ggplot() + geom_point(data=as.data.frame(ordered_PCA_res), 
                               aes(x=PC3, y=PC4, 
                                   color=clade,
                                   fill=clade,
                                   shape=tissues), size=3, stroke=0.8) +
        # stat_ellipse(data=as.data.frame(ordered_PCA_res), geom="polygon", aes(x=PC3, y=PC4, color=clade, fill = clade),
        #        alpha = 0.1,
        #        show.legend = FALSE,
        #        level = 0.95) +
      ggtitle(paste0(species_set, " ~ ", cluster_number)) + xlab(paste0("PC3: ", third_PC_var, "%")) + ylab(paste0("PC4: ", fourth_PC_var, "%")) + theme_bw() +
      theme(axis.title = element_text(color="black", size=14), 
            axis.text = element_text(color="black", size=12),
            legend.text = element_text(color="black", size=10), 
            legend.title = element_text(color="black", size=12),
            legend.spacing.y = unit(0, 'cm')) +
      scale_colour_manual(values=c("Vertebrate"="#82359D", "Not_vertebrate"="goldenrod")) +
      scale_fill_manual(values=c("Vertebrate"="#82359D", "Not_vertebrate"="goldenrod")) +
      scale_shape_manual(values=tissue_letters, breaks=ordered_tissues) + 
      guides(color=guide_legend(title="Species", byrow=TRUE), 
             fill=guide_legend(title="Species", byrow=TRUE), 
             shape=guide_legend(title="Tissues", byrow=TRUE))
    
    res = list()
    res[[1]] = p
    res[[2]] = p1
    res[[3]] = p2
    
    my_plots = plot_grid(p, p1, p2, rel_widths=c(1,1,1), align="h", ncol=3)
    return(list("plots" = my_plots, "results"=res))
}

anova_plot = function(expr_df, features_table, species_set) {
  imputation_out = impute.knn(as.matrix(expr_df), k = 10) #impute NAs
  imputation_df = as.data.frame(imputation_out$data)
  ##### compute PCA. I need the genes on columns in the input table
  PCA_res = prcomp(as.matrix(t(imputation_df)), center = TRUE, scale = TRUE)
  ordered_PCA_res = as.data.frame(PCA_res$x[order(rownames(PCA_res$x)),])
  #filter feature table as neeeded
  filtered_features_table = features_table[rownames(features_table) %in% colnames(imputation_df),]
  ordered_features_table = filtered_features_table[order(rownames(filtered_features_table)),]
  all(rownames(ordered_features_table) == rownames(ordered_PCA_res))
  
  #merge results
  my_merged_table = merge(ordered_PCA_res, ordered_features_table, by=0)
  
  #Create the categories relative to the tissues.
  for (element in all_tissues) { #cycle on the tissues
    local_tissues = all_tissues
    local_tissues[local_tissues != element] = paste0("not_", element)
    my_dict <- hashmap(all_tissues,  local_tissues)
    my_merged_table$element =  as.factor(my_dict$find(my_merged_table$tissues))
    names(my_merged_table)[names(my_merged_table) == "element"] = element
  }
    
  #This is something relative to the species.
  my_species = unique(as.vector(my_merged_table$species))
  for (element in my_species) {
    local_species = my_species
    local_species[local_species != element] = paste0("not_", element)
    my_dict <- hashmap(my_species, local_species)
    my_merged_table$element =  as.factor(my_dict$find(my_merged_table$species))
    names(my_merged_table)[names(my_merged_table) == "element"] = element
  }
  
  colnames(my_merged_table)[1] = "sample"
  
  PCs = colnames(ordered_PCA_res)[1:15]
  pvalues_matrix = vector()
  species_pvalues_matrix = vector()
  tissues_pvalues_matrix = vector()
  for(PC in PCs) {
    pvalues_vector = vector() #create empty pvalue vector
    species_pvalues_vector = vector()
    tissues_pvalues_vector = vector()
    for(category in c("species", "tissues")) { #cycle on the previously defined categories
      res.aov <- aov(as.formula(paste0(PC, " ~ ", category)), data = my_merged_table) #run the anova test
      pvalue = summary(res.aov)[[1]][["Pr(>F)"]][1] #get the general pvalue from the anova test
      pvalues_vector = c(pvalues_vector, pvalue) #build the pvalue vector from the different anova test (one per category)
    }
    for (category in my_species) {
      res.aov <- aov(as.formula(paste0(PC, " ~ ", category)), data = my_merged_table) #run the anova test
      pvalue = summary(res.aov)[[1]][["Pr(>F)"]][1] #get the general pvalue from the anova test
      species_pvalues_vector = c(species_pvalues_vector, pvalue)
    }
    for (category in all_tissues) {
      res.aov <- aov(as.formula(paste0(PC, " ~ ", category)), data = my_merged_table) #run the anova test
      pvalue = summary(res.aov)[[1]][["Pr(>F)"]][1] #get the general pvalue from the anova test
      tissues_pvalues_vector = c(tissues_pvalues_vector, pvalue)
    }
    pvalues_matrix = rbind(pvalues_matrix, pvalues_vector) #build pvalue matrix (each row refers to a different PC)
    species_pvalues_matrix = rbind(species_pvalues_matrix, species_pvalues_vector) #build species matrix (each row refers to a different PC)
    tissues_pvalues_matrix = rbind(tissues_pvalues_matrix, tissues_pvalues_vector) #build tissues matrix (each row refers to a different PC)
  }
  
  #transform rownames and colnames
  rownames(pvalues_matrix) = PCs
  colnames(pvalues_matrix) = c("species", "tissues")
  rownames(species_pvalues_matrix) = PCs
  colnames(species_pvalues_matrix) = my_species
  rownames(tissues_pvalues_matrix) = PCs
  colnames(tissues_pvalues_matrix) = all_tissues
  
  #Correct the pvalues
  tot_tests = dim(pvalues_matrix)[1]*dim(pvalues_matrix)[2] + 
    dim(species_pvalues_matrix)[1]*dim(species_pvalues_matrix)[2] + 
    dim(tissues_pvalues_matrix)[1]*dim(tissues_pvalues_matrix)[2]
  
  pvalues_matrix = apply(pvalues_matrix, c(1,2), function(x) x*tot_tests)
  species_pvalues_matrix = apply(species_pvalues_matrix, c(1,2), function(x) x*tot_tests)
  tissues_pvalues_matrix = apply(tissues_pvalues_matrix, c(1,2), function(x) x*tot_tests)
  pvalues_matrix[pvalues_matrix > 1] = 1
  species_pvalues_matrix[species_pvalues_matrix > 1] = 1
  tissues_pvalues_matrix[tissues_pvalues_matrix > 1] = 1
  
  ######## compute max and min between all tables in order to scale all the plots accordingly.
  my_max = max(max(apply(pvalues_matrix, c(1,2), 
                         function(x) -log10(x))), 
               max(apply(species_pvalues_matrix, c(1,2), 
                         function(x) -log10(x))), 
               max(apply(tissues_pvalues_matrix, c(1,2), 
                         function(x) -log10(x))))
  my_min = 0
  
  ######### work on the general pvalues matrix
  log10_pvalues_matrix = apply(pvalues_matrix, c(1,2), function(x) -log10(x))
  log10_pvalues_table = as.data.frame(log10_pvalues_matrix)
  log10_pvalues_table$PC = rownames(log10_pvalues_table)
  log10_pvalues_table_long = melt(log10_pvalues_table, id.vars="PC")
  
  my_palette <- colorRampPalette(c("white", "springgreen", "forestgreen"))(n = 100)
  my_palette_1 = colorRampPalette(c("white", "springgreen"))(n=20)
  my_palette_2 = colorRampPalette(c("springgreen", "forestgreen"))(n=80)
  my_palette = c(my_palette_1, my_palette_2)
  
  log10_pvalues_table_long$PC = factor(log10_pvalues_table_long$PC, levels=rev(log10_pvalues_table$PC))
  p = ggplot() + geom_tile(data=log10_pvalues_table_long, aes(x=variable, y=PC, fill=value)) + 
    ggtitle(paste0(species_set, " ~ Anova on PCs")) +
    theme_bw() + 
    theme(axis.text.x = element_text(color="black", angle=30, hjust=1, size=12),
          axis.text.y = element_text(color="black", size=12), axis.title.y = element_blank(),
          axis.title.x = element_blank(), legend.position="bottom", axis.ticks = element_blank(),
          legend.text=element_text(size=12), legend.title=element_text(size=13),
          plot.title = element_text(size = 16), legend.key.size = unit(3, "mm")) +
    scale_fill_gradientn(colours=my_palette, limits = c(my_min, my_max)) + 
    guides(fill=guide_legend(title="-log10(pvalue)"))
  
  ######## work on species comparison matrix
  #the values should already be -log10(pvalue)
  species_pvalues_matrix = apply(species_pvalues_matrix, c(1,2), function(x) -log10(x))
  species_pvalues_table = as.data.frame(species_pvalues_matrix)
  species_pvalues_table$PC = rownames(species_pvalues_table)
  species_pvalues_table_long = melt(species_pvalues_table, id.vars="PC")
  
  #my_palette <- colorRampPalette(c("white", "chocolate1", "orangered3"))(n = 100)
  my_palette_1 = colorRampPalette(c("white", "chocolate1"))(n=20)
  my_palette_2 = colorRampPalette(c("chocolate1", "orangered3"))(n=80)
  my_palette = c(my_palette_1, my_palette_2)
  
  #create factors to have ordered axis
  species_pvalues_table_long$PC = factor(species_pvalues_table_long$PC, levels=rev(species_pvalues_table$PC))
  species_pvalues_table_long$variable = factor(species_pvalues_table_long$variable, levels=all_species)
  
  #plot
  p1 = ggplot() + geom_tile(data=species_pvalues_table_long, aes(x=variable, y=PC, fill=value)) + 
    theme_bw() + 
    theme(axis.text.x = element_text(color="black", angle=45, hjust=1, size=12), 
          axis.title.x = element_blank(), 
          legend.position="bottom", axis.ticks = element_blank(), 
          axis.title.y=element_blank(), 
          legend.text=element_text(size=12), 
          legend.title=element_text(size=13),
          axis.text.y = element_blank(),
          plot.title = element_text(size = 16), legend.key.size = unit(3, "mm")) +
    scale_fill_gradientn(colours=my_palette, limits = c(my_min, my_max)) + 
    ggtitle(paste0(species_set, " ~ species sep by Anova")) + guides(fill=guide_legend(title="-log10(pvalue)"))
  
  ###### work on tissues comparison matrix
  #the values should already be -log10(pvalue)
  tissues_pvalues_matrix = apply(tissues_pvalues_matrix, c(1,2), function(x) -log10(x))
  tissues_pvalues_table = as.data.frame(tissues_pvalues_matrix)
  tissues_pvalues_table$PC = rownames(tissues_pvalues_table)
  tissues_pvalues_table_long = melt(tissues_pvalues_table, id.vars="PC")
  
  #my_palette <- colorRampPalette(c("white", "chocolate1", "orangered3"))(n = 100)
  my_palette_1 = colorRampPalette(c("white", "deepskyblue1"))(n=20)
  my_palette_2 = colorRampPalette(c("deepskyblue1", "mediumblue"))(n=80)
  my_palette = c(my_palette_1, my_palette_2)
  
  #create factors to have ordered axis
  tissues_pvalues_table_long$PC = factor(tissues_pvalues_table_long$PC, levels=rev(tissues_pvalues_table$PC))
  tissues_pvalues_table_long$variable = factor(tissues_pvalues_table_long$variable, levels=all_tissues)
  
  #plot
  p2 = ggplot() + geom_tile(data=tissues_pvalues_table_long, aes(x=variable, y=PC, fill=value)) + 
    theme_bw() + 
    theme(axis.text.x = element_text(color="black", angle=45, hjust=1, size=12), 
          axis.title.x = element_blank(), 
          legend.position="bottom", axis.ticks = element_blank(), 
          axis.title.y=element_blank(), 
          legend.text=element_text(size=12), 
          legend.title=element_text(size=13),
          axis.text.y = element_blank(),
          plot.title = element_text(size = 16),
          legend.key.size = unit(3, "mm")) +
    scale_fill_gradientn(colours=my_palette, limits = c(my_min, my_max)) + 
    ggtitle(paste0(species_set, " ~ tissues sep by Anova")) + guides(fill=guide_legend(title="-log10(pvalue)"))
  
  eigs <- PCA_res$sdev^2
  my_eigs = eigs/sum(eigs)*100
  my_eigs_df = as.data.frame(my_eigs[1:15])
  my_eigs_df$PC = paste0("PC", seq(1,nrow(my_eigs_df)))
  my_eigs_df$PC = factor(my_eigs_df$PC, levels=rev(my_eigs_df$PC))
  colnames(my_eigs_df) = c("my_eigs", "PC")
  
  p3 = ggplot() + geom_bar(data=my_eigs_df, aes(x=PC, y=my_eigs, fill="random"), stat="identity") + coord_flip(ylim=c(0,100)) +
    theme_bw() + 
    theme(axis.text.x = element_text(color="black", size=12), 
          axis.title.y=element_blank(), axis.text.y = element_blank(),
          axis.title.x = element_text(color="black", size=13),
          axis.line = element_blank(), axis.ticks = element_blank(),
          panel.border = element_blank(), panel.background = element_blank(), 
          legend.text=element_text(size=12), legend.title=element_text(size=13),
          panel.grid.minor = element_blank(), panel.grid.major = element_blank(), legend.position="bottom", legend.key.size = unit(3, "mm")) + 
    scale_fill_manual(values=c("random"="violet")) +
    ylab("Variance explained (%)") +
    ggtitle(paste0(species_set, " ~ PC variance"))

  res = list()
  res[[1]] = p
  res[[2]] = p1
  res[[3]] = p2
  res[[4]] = p3
  
  my_plots = plot_grid(p, p1, p2, p3, rel_widths=c(2,2,2,0.75), align="h", ncol=4)
  return(list("plots" = my_plots, "results"=res))
}

## This is to plot just PC1 vs PC2
plot_PCA = function(expr_df, features_table, species_set, cluster_number) {
  imputation_out = impute.knn(as.matrix(expr_df), k = 10) #impute missing values.
  imputation_df = as.data.frame(imputation_out$data) #transform to a dataframe.
  #compute PCA. I need the genes on columns in the input table.
  PCA_res = prcomp(as.matrix(t(imputation_df)), center = TRUE, scale = TRUE) #center and scale the PCs.
  ordered_PCA_res = as.data.frame(PCA_res$x[order(rownames(PCA_res$x)),]) 
  #filter the sample only for the samples that made it after the sva correction
  filtered_features_table = features_table[rownames(features_table) %in% colnames(imputation_df),]
  ordered_features_table = filtered_features_table[order(rownames(filtered_features_table)),] #order the table
  all(rownames(ordered_features_table) == rownames(ordered_PCA_res))
  #compute the percentage of variance explained
  eigs <- PCA_res$sdev^2
  first_PC_var = round(eigs[1]/sum(eigs), 3)*100
  second_PC_var = round(eigs[2]/sum(eigs), 3)*100
  third_PC_var = round(eigs[3]/sum(eigs), 3)*100
  fourth_PC_var = round(eigs[4]/sum(eigs), 3)*100
  #filter shapes vector
  my_filtered_shapes_vector = species_shapes[unique(as.vector(ordered_features_table$species))] #filter the shape vector
  #filter ordered species
  #filt_ordered_species = ordered_species[ordered_species %in% unique(as.vector(ordered_features_table$species))]
  ordered_features_table$species <- factor(ordered_features_table$species, levels=ordered_species) #change the order of the levels in the factor
  #plot the first two PC
  ordered_PCA_res = as.data.frame(ordered_PCA_res)
  ordered_PCA_res$tissues = ordered_features_table$tissues
  ordered_PCA_res$species = ordered_features_table$species
  print(levels(ordered_PCA_res$species))
  
  p = ggplot() + geom_point(data=ordered_PCA_res, 
                            aes(x=PC1, y=PC2, 
                                color=species, 
                                #fill=species,
                                shape=tissues), size=2.5, stroke=0.8) +
    xlab(paste0("PC1: ", first_PC_var, "%")) + 
    ylab(paste0("PC2: ", second_PC_var, "%")) + theme_bw() +
    theme(axis.title = element_text(color="black", size=12), 
          axis.text = element_text(color="black", size=12),
          legend.text = element_text(color="black", size=12), 
          legend.title = element_text(color="black", size=12, face="bold")) +
    scale_colour_manual(name="Species", values=species_colors, labels=ordered_species) + #this is great because I can have named vectors
    scale_shape_manual(name="Tissues", values=tissue_letters, labels=ordered_tissues)
  
  res = list()
  res[[1]] = p
  
  my_plots = plot_grid(p, p, rel_widths=c(1,1), align="h", ncol=2)
  return(list("plots" = my_plots, "results"=res))
}

######################################################
######################################################
######################################################

PCA_basic_plot_test = function(expr_df, features_table, species_set, cluster_number) {
    imputation_out = impute.knn(as.matrix(expr_df), k = 10) #impute missing values.
    imputation_df = as.data.frame(imputation_out$data) #transform to a dataframe.
    #compute PCA. I need the genes on columns in the input table.
    PCA_res = prcomp(as.matrix(t(imputation_df)), center = TRUE, scale = TRUE) #center and scale the PCs.
    ordered_PCA_res = as.data.frame(PCA_res$x[order(rownames(PCA_res$x)),]) 
    #filter the sample only for the samples that made it after the sva correction
    filtered_features_table = features_table[rownames(features_table) %in% colnames(imputation_df),]
    ordered_features_table = filtered_features_table[order(rownames(filtered_features_table)),] #order the table
    all(rownames(ordered_features_table) == rownames(ordered_PCA_res))
    #compute the percentage of variance explained
    eigs <- PCA_res$sdev^2
    first_PC_var = round(eigs[51]/sum(eigs), 3)*100
    second_PC_var = round(eigs[52]/sum(eigs), 3)*100
    third_PC_var = round(eigs[53]/sum(eigs), 3)*100
    fourth_PC_var = round(eigs[54]/sum(eigs), 3)*100
    #filter shapes vector
    my_filtered_shapes_vector=species_shapes[names(species_shapes) %in% unique(as.vector(ordered_features_table$species))] #filter the shape vector
    #filter ordered species
    filt_ordered_species = ordered_species[ordered_species %in% unique(as.vector(ordered_features_table$species))]
    ordered_features_table$species <- factor(ordered_features_table$species, levels=filt_ordered_species) #change the order of the levels in the factor
    #plot the first two PC
    ordered_PCA_res$tissues = ordered_features_table$tissues
    ordered_PCA_res$species = ordered_features_table$species
    
    p = ggplot() + geom_point(data=as.data.frame(ordered_PCA_res), 
                              aes(x=PC51, y=PC52, 
                                  color=tissues, 
                                  fill=tissues,
                                  shape=species), size=2, stroke=0.8) +
      # stat_ellipse(data=as.data.frame(ordered_PCA_res), geom="polygon", aes(x=PC1, y=PC2, color=tissues, fill = tissues),
      #              alpha = 0.1,
      #              show.legend = FALSE,
      #              level = 0.80) +
      ggtitle(paste0(species_set, " ~ ", cluster_number)) + 
      xlab(paste0("PC1: ", first_PC_var, "%")) + 
      ylab(paste0("PC2: ", second_PC_var, "%")) + theme_bw() +
      theme(axis.title = element_text(color="black", size=14), 
            axis.text = element_text(color="black", size=12),
            legend.text = element_text(color="black", size=10), 
            legend.title = element_text(color="black", size=12),
            legend.spacing.y = unit(0, 'cm')) +
      scale_colour_manual(values=tissue_colors) + #this is great because I can have named vectors
      scale_fill_manual(values=tissue_colors) +
      scale_shape_manual(values=my_filtered_shapes_vector, breaks=filt_ordered_species) + 
      guides(color=guide_legend(title="Tissues", byrow=TRUE), 
             fill=guide_legend(title="Tissues", byrow=TRUE), 
             shape=guide_legend(title="Species", byrow=TRUE))
    
    p1 = ggplot() + geom_point(data=as.data.frame(ordered_PCA_res), 
                               aes(x=PC52, y=PC53, 
                                   color=tissues,
                                   fill=tissues,
                                   shape=species), size=2, stroke=0.8) +
      # stat_ellipse(data=as.data.frame(ordered_PCA_res), geom="polygon", aes(x=PC2, y=PC3, color=tissues, fill = tissues),
      #        alpha = 0.1,
      #        show.legend = FALSE,
      #        level = 0.80) +
      ggtitle(paste0(species_set, " ~ ", cluster_number)) + 
      xlab(paste0("PC2: ", second_PC_var, "%")) + 
      ylab(paste0("PC3: ", third_PC_var, "%")) + 
      theme_bw() +
      theme(axis.title = element_text(color="black", size=14), 
            axis.text = element_text(color="black", size=12),
            legend.text = element_text(color="black", size=10), 
            legend.title = element_text(color="black", size=12),
            legend.spacing.y = unit(0, 'cm')) +
      scale_colour_manual(values=tissue_colors) + #this is great because I can have named vectors
      scale_fill_manual(values=tissue_colors) +
      scale_shape_manual(values=my_filtered_shapes_vector, breaks=filt_ordered_species) + 
      guides(color=guide_legend(title="Tissues", byrow=TRUE), 
             fill=guide_legend(title="Tissues", byrow=TRUE), 
             shape=guide_legend(title="Species", byrow=TRUE))
    
    p2 = ggplot() + geom_point(data=as.data.frame(ordered_PCA_res), 
                               aes(x=PC53, y=PC54, 
                                   color=tissues,
                                   fill=tissues,
                                   shape=species), size=2, stroke=0.8) +
        # stat_ellipse(data=as.data.frame(ordered_PCA_res), geom="polygon", aes(x=PC3, y=PC4, color=tissues, fill = tissues),
        #        alpha = 0.1,
        #        show.legend = FALSE,
        #        level = 0.80) +
      ggtitle(paste0(species_set, " ~ ", cluster_number)) + xlab(paste0("PC3: ", third_PC_var, "%")) + ylab(paste0("PC4: ", fourth_PC_var, "%")) + theme_bw() +
      theme(axis.title = element_text(color="black", size=14), 
            axis.text = element_text(color="black", size=12),
            legend.text = element_text(color="black", size=10), 
            legend.title = element_text(color="black", size=12),
            legend.spacing.y = unit(0, 'cm')) +
      scale_colour_manual(values=tissue_colors) + #this is great because I can have named vectors
      scale_fill_manual(values=tissue_colors) +
      scale_shape_manual(values=my_filtered_shapes_vector, breaks=filt_ordered_species) + 
      guides(color=guide_legend(title="Tissues", byrow=TRUE), 
             fill=guide_legend(title="Tissues", byrow=TRUE), 
             shape=guide_legend(title="Species", byrow=TRUE))
    
    res = list()
    res[[1]] = p
    res[[2]] = p1
    res[[3]] = p2
    
    my_plots = plot_grid(p, p1, p2, rel_widths=c(1,1,1), align="h", ncol=3)
    return(list("plots" = my_plots, "results"=res))
}

######### FUNCTION TO ADD A COLUMN IF MISSING
add_cols <- function(df, cols) {
  add <- cols[!cols %in% names(df)]
  if(length(add) !=0 ) df[add] <- 0
  return(df)
}
```

```{r extra_functions, echo=FALSE, warning=FALSE, message=FALSE, dependson=c("setup", "functions")}
####### These are functions borrowed from the PCAtools package.
getComponents <- function(
  pcaobj,
  components = NULL)
{
  if (!is.null(components)) {
    return(pcaobj$components[components])
  } else if (is.null(components)) {
    return(pcaobj$components)
  }
}


eigencorplot = function(
  pcaobj,
  metadata,
  #components = getComponents(pcaobj, seq_len(10)),
  components = colnames(pcaobj)[1:10],
  metavars,
  titleX = '',
  cexTitleX = 1.0,
  rotTitleX = 0,
  colTitleX = 'black',
  fontTitleX = 2,
  titleY = '',
  cexTitleY = 1.0,
  rotTitleY = 0,
  colTitleY = 'black',
  fontTitleY = 2,
  cexLabX = 1.0,
  rotLabX = 0,
  colLabX = 'black',
  fontLabX = 2,
  cexLabY = 1.0,
  rotLabY = 0,
  colLabY = 'black',
  fontLabY = 2,
  posLab = 'bottomleft',
  col = c('blue4', 'blue3', 'blue2', 'blue1', 'white',
    'red1', 'red2', 'red3', 'red4'),
  posColKey = 'right',
  cexLabColKey = 1.0,
  cexCorval = 1.0,
  colCorval = 'black',
  fontCorval = 1,
  scale = TRUE,
  main = '',
  cexMain = 2,
  rotMain = 0,
  colMain = 'black',
  fontMain = 2,
  corFUN = 'pearson',
  corUSE = 'pairwise.complete.obs',
  corMultipleTestCorrection = 'none',
  signifSymbols = c('***', '**', '*', ''),
  signifCutpoints = c(0, 0.001, 0.01, 0.05, 1),
  colFrame = 'white',
  plotRsquared = FALSE,
  returnPlot = TRUE) 

  {
  data <- pcaobj
  #metadata <- pcaobj$metadata

  # issue warning if any columns to use are not numeric --- This is kind of annoying
  # --- and there should be an option to force numeric conversion as anyway I'm doing it in advance
  
  for (i in seq_len(length(components))) {
    if(!is.numeric(data[,components[i]])) {
      warning(components[i],
        ' is not numeric - please check the source data',
        ' as everything will be converted to a matrix')
    }
  }
  for (i in seq_len(length(metavars))) {
    if(!is.numeric(metadata[,metavars[i]])) {
      warning(metavars[i],
        ' is not numeric - please check the source data',
        ' as non-numeric variables will be coerced to numeric')
    }
  }

  # convert the data for x and y to data matrix
  #	NAs are left NA
  #	Character (A-Z a-z) are converted to NA
  #	Character numbers are converted to integers
  #	Factors are converted to numbers based on level ordering
  xvals <- data.matrix(data[,which(colnames(data) %in% components)])
  yvals <- metadata[,which(colnames(metadata) %in% metavars)]

  ### code courtesy of aleighbrown
    # let's make sure that anything that isnt' numeric becomes a numeric
    # find all the character columns
    # ---select all the columns which are characters
    # using base R now for dependencies sake 
      chararcter_columns = unlist(lapply(yvals, is.numeric))  
      # negate it - basically if it
      chararcter_columns = !chararcter_columns
      # select only the names that are true 
      chararcter_columns = names(which(chararcter_columns))
      for (c in chararcter_columns) {
        yvals[, eval(quote(c))] = as.numeric(as.factor(yvals[, eval(quote(c))]))
      }
  ### END
    
  yvals <- data.matrix(yvals)

  # create correlation table
  corvals <- cor(xvals, yvals, use = corUSE, method = corFUN)

  # create a new df with same dimensions as corvals and fill with P values
  # total number of tests we perform
  N <- ncol(xvals) * ncol(yvals)
  pvals <- data.frame(pval = numeric(N),
                       i = numeric(N),
                       j = numeric(N))
  k <- 0
  for (i in seq_len(ncol(xvals))) {
    for (j in seq_len(ncol(yvals))) { 
      k <- k + 1
      pvals[k,'pval'] <- cor.test(xvals[,i],
                             yvals[,j],
                             use = corUSE,
                             method = corFUN)$p.value
      pvals[k,"i"] <- colnames(xvals)[i]
      pvals[k,"j"] <- colnames(yvals)[j]

    }
  }

  ### code courtesy of aleighbrown
    # -----if you want to adjust the p-values for multiple testing
    if(corMultipleTestCorrection != "none"){
      pvals$pval <- p.adjust(pvals$pval, method = corMultipleTestCorrection)
    }

    pvals <- reshape2::dcast(pvals, i ~ j, value.var = "pval")
    # -----make sure the pvals matchs the order of corrvals table
    rownames(pvals) <- pvals$i
    pvals$i <- NULL
    pvals <- pvals[match(rownames(corvals), rownames(pvals)), ]
    # ---make sure the columns are in the correct order
    pvals <- pvals[colnames(corvals)]
    # ------
  ### END

  # are we plotting R^2 values?
  if (plotRsquared==TRUE) {
    corvals <- corvals ^ 2
  }

  # determine max and min correlation values in order to define the range
  if (scale == FALSE && plotRsquared == TRUE) {
    iUpperRange <- 1
    iLowerRange <- 0
  } else if (scale == FALSE && plotRsquared == FALSE) {
    iUpperRange <- 1
    iLowerRange <- -1
  } else if (scale == TRUE) {
    max <- max(corvals)
    min <- min(corvals)
    if(abs(max) > abs(min)) {
      iUpperRange <- max + 0.01
      iLowerRange <- (max * (-1)) - 0.01
    } else {
      iUpperRange <- abs(min) + 0.01
      iLowerRange <- min - 0.01
    }
    if (plotRsquared==TRUE) {
      iUpperRange <- max + 0.1
      iLowerRange <- 0
    }
  }

  # define the colour scheme/palette
  cols <- colorRampPalette(col)

  # create a new df with same dimensions as corvals
  # fill with significances encoded with asterisks
  signif <- corvals
  for (i in seq_len(ncol(pvals))) {
    signif[,i] <- c(symnum(pvals[,i],
      corr = FALSE,
      na = FALSE,
      cutpoints = signifCutpoints,
      symbols = signifSymbols))
  }

  # create a new df with same dimensions as corvals
  # fill with r values merged with the encoded significances
  plotLabels <- corvals
  for (i in seq_len(nrow(corvals))) {
    for(j in seq_len(ncol(corvals))) {
      plotLabels[i,j] <- paste(round(corvals[i,j], 2),
        signif[i,j],
        sep='')
      colnames(plotLabels)[j] <- colnames(corvals)[j]
    }

    rownames(plotLabels)[i] <- rownames(corvals)[i]
  }

  # position of axis ticks
  if (posLab == 'bottomleft') {
    posLab = 1
    axisTicks = c(1,0)
  } else if (posLab == 'topright') {
    posLab = 2
    axisTicks = c(0,1)
  } else if (posLab == 'all') {
    posLab = 3
    axisTicks = c(1,1)
  } else if (posLab == 'none') {
    posLab = 0
    axisTicks = c(0,0)
  }

  # define a panel function for adding labels
  # labels are passed with z as a third dimension
  labels <- function(x, y, z, ...) {
    panel.levelplot(x, y, z, ...)
    ltext(x, y,
      labels = plotLabels,
      cex = cexCorval,
      col = colCorval,
      font = fontCorval)
  }

  # produce the levelplot
  l <- levelplot(
    data.matrix(corvals),
    xlab = list(label = titleX,
      cex = cexTitleX,
      rot = rotTitleX,
      col = colTitleX,
      font = fontTitleX),
    ylab = list(label = titleY,
      cex = cexTitleY,
      rot = rotTitleY,
      col = colTitleY,
      font = fontTitleY),
    panel = labels,
    pretty = TRUE,
    par.settings = list(panel.background = list(col = colFrame)),
    scales = list(
      x = list(cex = cexLabX,
        rot = rotLabX,
        col = colLabX,
        font = fontLabX),
      y = list(cex = cexLabY,
        rot = rotLabY,
        col = colLabY,
        font = fontLabY),
      tck = axisTicks,
      alternating = posLab),
    aspect = 'fill',
    col.regions = cols,
    cuts = 100,
    at = seq(iLowerRange, iUpperRange, 0.01),
    main = list(label = main,
      cex = cexMain,
      rot = rotMain,
      col = colMain,
      font = fontMain),
    colorkey = list(space = posColKey,
      labels = list(cex = cexLabColKey)))

  # return plot?
  if (returnPlot == TRUE) {
    return(l)
  } else if (returnPlot == FALSE) {
    l
  }
}
```


<!-- Running the PCA plots with orthogroups in the ancestral tissue-specific modules
We expect to get very well defined tissue groups -->
<!-- I am also adding chuncks for other types of PCAs we might use in our rebuttal -->
```{r generate_metadata_table, echo=FALSE, warning=FALSE, message=FALSE, dependson="setup"}
pca_metadata_df = vector()
for (species in all_species) {
  file_species = species
  if (species == "BmA") {file_species = "Bmo"}
  if (species == "Hsa") {file_species = "Hs2"}
  if (species == "Mmu") {file_species = "Mm2"}
  if (species == "Bta") {file_species = "Bt2"}
  if (species == "Spu") {file_species = "Sp2"}
  metadata_df = unique(read.delim(paste0(metadata_dir, file_species, "_samples_info.tab"), header=TRUE)[,1:3])
  colnames(metadata_df) = c("species", "tissues", "metasample")
  metadata_df$species = rep(species, nrow(metadata_df)) #This is necessary only to translate Bmo to BmA
  rownames(metadata_df) = paste0(rep(species, nrow(metadata_df)), "_", as.vector(metadata_df$metasample))
  metadata_df$metasample = NULL
  pca_metadata_df = rbind(pca_metadata_df, metadata_df)
}
colnames(pca_metadata_df) = c("species", "tissues")
```


<!-- ########################################################### -->
<!-- ############## New PCA combo ############################## -->
<!-- ########################################################### -->
```{r PCA_clades_single_copy_orthologs, warning=FALSE, message=FALSE, echo=FALSE, dependson=c("setup", "functions"), fig.width=16, fig.height=13}
for (my_clade in "bilateria") {
  ##############################################
  ##### GENERATE INPUT TABLE ###################
  ##############################################
  
  ######### Upload table
  expr_file = paste0(broccoli_revision_dir, my_clade, "/expression_tables/", my_clade, "_conserved-metasamples_median_expr-NOSVA-log2-TPMs-single_copies-NORM.tab")
  expr_df = read.delim(expr_file, sep="\t", header=TRUE, row=1)
  expr_df = expr_df[complete.cases(expr_df),] #I am using only the cases without NAs, so that I don't have to do any imputation.
  #Update colnames
  colnames(expr_df) = gsub("BmA", "Bmo", gsub("Hs2", "Hsa", gsub("Mm2", "Mmu", gsub("Bt2", "Bta", gsub("Sp2", "Spu", colnames(expr_df))))))
  
  ##############################################
  ##### PLOT ###################################
  ##############################################
  
  
  orthogroups_number=nrow(expr_df)
  res_PCA = PCA_basic_plot(expr_df, pca_metadata_df, clade, orthogroups_number)
  res_anova = anova_plot(expr_df, pca_metadata_df, clade)
  #plots
  first_row = annotate_figure(ggarrange(res_PCA$results[[1]] + ggtitle(""), 
                                        res_PCA$results[[2]] + ggtitle(""), 
                                        res_PCA$results[[3]] + ggtitle(""), 
                                        common.legend=TRUE, legend="right", 
                                        ncol=3), 
                          fig.lab=paste0(str_to_title(my_clade), " PC analysis ~ ", orthogroups_number, " orthogroups"), fig.lab.pos = "top.left", fig.lab.size=18)
  second_row = plot_grid(res_anova$results[[1]] + ggtitle("Anova on PCs"), 
                         res_anova$results[[2]] + ggtitle("Anova ~ species separation"), 
                         res_anova$results[[3]] + ggtitle("Anova  ~ tissues separation"), 
                         res_anova$results[[4]] + ggtitle("Variance explained"), 
                         rel_widths=c(1.25,2,2,1), align="h", ncol=4)
  
  final_plot_1 = ggarrange(first_row, second_row, nrow=2)
  print(final_plot_1)
  
  ##############################################
  ##### PLOT ###################################
  ##############################################
  orthogroups_number=nrow(expr_df)
  res_PCA = PCA_basic_by_species_plot(expr_df, pca_metadata_df, clade, orthogroups_number)
  res_anova = anova_plot(expr_df, pca_metadata_df, clade)
  #plots
  first_row = annotate_figure(ggarrange(res_PCA$results[[1]] + ggtitle(""), 
                                        res_PCA$results[[2]] + ggtitle(""), 
                                        res_PCA$results[[3]] + ggtitle(""), 
                                        common.legend=TRUE, legend="right", 
                                        ncol=3), 
                          fig.lab=paste0(str_to_title(my_clade), " PC analysis ~ ", orthogroups_number, " orthogroups"), fig.lab.pos = "top.left", fig.lab.size=18)
  second_row = plot_grid(res_anova$results[[1]] + ggtitle("Anova on PCs"), 
                         res_anova$results[[2]] + ggtitle("Anova ~ species separation"), 
                         res_anova$results[[3]] + ggtitle("Anova  ~ tissues separation"), 
                         res_anova$results[[4]] + ggtitle("Variance explained"), 
                         rel_widths=c(1.25,2,2,1), align="h", ncol=4)
  
  final_plot_2 = ggarrange(first_row, second_row, nrow=2)
  print(final_plot_2)
  
  pdf(paste0("/Users/federica/Documents/CRG_PhD/bilaterian_GE/NatGen-EcoEvo_submission/Revision/Images/Figures_from_R/PCA-", my_clade, "-single_copies.pdf"), height = 13, width=16)
  print(final_plot_1)
  print(final_plot_2)
  dev.off()
}
```

```{r PCA_bilateria_average_expr_orthologs, warning=FALSE, message=FALSE, echo=FALSE, dependson=c("setup", "functions"), fig.width=16, fig.height=13}
##############################################
##### GENERATE INPUT TABLE ###################
##############################################
my_clade = "bilateria"
######### Upload table
expr_file = paste0(broccoli_revision_dir, my_clade, "/expression_tables/", my_clade, "_conserved-metasamples_median_expr-NOSVA-log2-TPMs-paralogs_average-NORM.tab")
expr_df = read.delim(expr_file, sep="\t", header=TRUE, row=1)
expr_df = expr_df[complete.cases(expr_df),] #I am using only the cases without NAs, so that I don't have to do any imputation.
#Update colnames
colnames(expr_df) = gsub("BmA", "Bmo", gsub("Hs2", "Hsa", gsub("Mm2", "Mmu", gsub("Bt2", "Bta", gsub("Sp2", "Spu", colnames(expr_df))))))

##############################################
##### PLOT ###################################
##############################################


orthogroups_number=nrow(expr_df)
res_PCA = PCA_basic_plot(expr_df, pca_metadata_df, clade, orthogroups_number)
res_anova = anova_plot(expr_df, pca_metadata_df, clade)
#plots
first_row = annotate_figure(ggarrange(res_PCA$results[[1]] + ggtitle(""), 
                                      res_PCA$results[[2]] + ggtitle(""), 
                                      res_PCA$results[[3]] + ggtitle(""), 
                                      common.legend=TRUE, legend="right", 
                                      ncol=3), 
                        fig.lab=paste0(str_to_title(my_clade), " PC analysis ~ ", orthogroups_number, " orthogroups"), fig.lab.pos = "top.left", fig.lab.size=18)
second_row = plot_grid(res_anova$results[[1]] + ggtitle("Anova on PCs"), 
                       res_anova$results[[2]] + ggtitle("Anova ~ species separation"), 
                       res_anova$results[[3]] + ggtitle("Anova  ~ tissues separation"), 
                       res_anova$results[[4]] + ggtitle("Variance explained"), 
                       rel_widths=c(1.25,2,2,1), align="h", ncol=4)

final_plot_1 = ggarrange(first_row, second_row, nrow=2)
print(final_plot_1)

##############################################
##### PLOT ###################################
##############################################
orthogroups_number=nrow(expr_df)
res_PCA = PCA_basic_by_species_plot(expr_df, pca_metadata_df, clade, orthogroups_number)
res_anova = anova_plot(expr_df, pca_metadata_df, clade)
#plots
first_row = annotate_figure(ggarrange(res_PCA$results[[1]] + ggtitle(""), 
                                      res_PCA$results[[2]] + ggtitle(""), 
                                      res_PCA$results[[3]] + ggtitle(""), 
                                      common.legend=TRUE, legend="right", 
                                      ncol=3), 
                        fig.lab=paste0(str_to_title(my_clade), " PC analysis ~ ", orthogroups_number, " orthogroups"), fig.lab.pos = "top.left", fig.lab.size=18)
second_row = plot_grid(res_anova$results[[1]] + ggtitle("Anova on PCs"), 
                       res_anova$results[[2]] + ggtitle("Anova ~ species separation"), 
                       res_anova$results[[3]] + ggtitle("Anova  ~ tissues separation"), 
                       res_anova$results[[4]] + ggtitle("Variance explained"), 
                       rel_widths=c(1.25,2,2,1), align="h", ncol=4)

final_plot_2 = ggarrange(first_row, second_row, nrow=2)
print(final_plot_2)

pdf(paste0("/Users/federica/Documents/CRG_PhD/bilaterian_GE/NatGen-EcoEvo_submission/Revision/Images/Figures_from_R/PCA-", my_clade, "-paralogs_average.pdf"), height = 13, width=16)
print(final_plot_1)
print(final_plot_2)
dev.off()
```

```{r PCA_bilateria_average_expr_orthologs_zscore, warning=FALSE, message=FALSE, echo=FALSE, dependson=c("setup", "functions"), fig.width=16, fig.height=13}
##############################################
##### GENERATE INPUT TABLE ###################
##############################################
my_clade = "bilateria"
######### Upload table
expr_file = paste0(broccoli_revision_dir, my_clade, "/expression_tables/", my_clade, "_conserved-metasamples_median_expr-NOSVA-log2-TPMs-paralogs_average-NORM-species_zscore.tab")
expr_df = read.delim(expr_file, sep="\t", header=TRUE, row=1)
expr_df = expr_df[complete.cases(expr_df),] #I am using only the cases without NAs, so that I don't have to do any imputation.
#Update colnames
colnames(expr_df) = gsub("BmA", "Bmo", gsub("Hs2", "Hsa", gsub("Mm2", "Mmu", gsub("Bt2", "Bta", gsub("Sp2", "Spu", colnames(expr_df))))))

##############################################
##### PLOT ###################################
##############################################


orthogroups_number=nrow(expr_df)
res_PCA = PCA_basic_plot(expr_df, pca_metadata_df, clade, orthogroups_number)
res_anova = anova_plot(expr_df, pca_metadata_df, clade)
#plots
first_row = annotate_figure(ggarrange(res_PCA$results[[1]] + ggtitle(""), 
                                      res_PCA$results[[2]] + ggtitle(""), 
                                      res_PCA$results[[3]] + ggtitle(""), 
                                      common.legend=TRUE, legend="right", 
                                      ncol=3), 
                        fig.lab=paste0(str_to_title(my_clade), " PC analysis ~ ", orthogroups_number, " orthogroups"), fig.lab.pos = "top.left", fig.lab.size=18)
second_row = plot_grid(res_anova$results[[1]] + ggtitle("Anova on PCs"), 
                       res_anova$results[[2]] + ggtitle("Anova ~ species separation"), 
                       res_anova$results[[3]] + ggtitle("Anova  ~ tissues separation"), 
                       res_anova$results[[4]] + ggtitle("Variance explained"), 
                       rel_widths=c(1.25,2,2,1), align="h", ncol=4)

final_plot_1 = ggarrange(first_row, second_row, nrow=2)
print(final_plot_1)

##############################################
##### PLOT ###################################
##############################################
orthogroups_number=nrow(expr_df)
res_PCA = PCA_basic_by_species_plot(expr_df, pca_metadata_df, clade, orthogroups_number)
res_anova = anova_plot(expr_df, pca_metadata_df, clade)
#plots
first_row = annotate_figure(ggarrange(res_PCA$results[[1]] + ggtitle(""), 
                                      res_PCA$results[[2]] + ggtitle(""), 
                                      res_PCA$results[[3]] + ggtitle(""), 
                                      common.legend=TRUE, legend="right", 
                                      ncol=3), 
                        fig.lab=paste0(str_to_title(my_clade), " PC analysis ~ ", orthogroups_number, " orthogroups"), fig.lab.pos = "top.left", fig.lab.size=18)
second_row = plot_grid(res_anova$results[[1]] + ggtitle("Anova on PCs"), 
                       res_anova$results[[2]] + ggtitle("Anova ~ species separation"), 
                       res_anova$results[[3]] + ggtitle("Anova  ~ tissues separation"), 
                       res_anova$results[[4]] + ggtitle("Variance explained"), 
                       rel_widths=c(1.25,2,2,1), align="h", ncol=4)

final_plot_2 = ggarrange(first_row, second_row, nrow=2)
print(final_plot_2)

pdf(paste0("/Users/federica/Documents/CRG_PhD/bilaterian_GE/NatGen-EcoEvo_submission/Revision/Images/Figures_from_R/PCA-", my_clade, "-paralogs_average-zscores.pdf"), height = 13, width=16)
print(final_plot_1)
print(final_plot_2)
dev.off()
```

<!-- ########################################################### -->
<!-- ############## Batch effects on PC1 ####################### -->
<!-- ########################################################### -->

```{r PC1_batch_correlation_custom, warning=FALSE, echo=FALSE, message=FALSE, dependson=c("setup", "extra_functions"), fig.width=12, fig.height=14}
######################################################
####### RUN PCA AND ISOLATE FIRST PC #################
######################################################
### Define path to expression file
expr_file = paste0(pca_dir, "/", version, "/", category, "/", clade, "/", evo_type, "/Bilateria_conserved-metasamples_median_expr-NOSVA-log2-TPMs-NORM-BH_genes.tab")
expr_df = read.delim(expr_file, sep="\t", header=TRUE, row=1)
expr_df = expr_df[complete.cases(expr_df),] #I am using only the cases without NAs, so that I don't have to do any imputation.
### Update colnames
colnames(expr_df) = gsub("BmA", "Bmo", gsub("Hs2", "Hsa", gsub("Mm2", "Mmu", gsub("Bt2", "Bta", gsub("Sp2", "Spu", colnames(expr_df))))))
### Run PCA
PCA_res = prcomp(as.matrix(t(expr_df)), center = TRUE, scale = TRUE) #center and scale the PCs.
ordered_PCA_res = as.data.frame(PCA_res$x[order(rownames(PCA_res$x)),])
PC1_values_df = ordered_PCA_res %>%
  select(PC1) %>%
  rownames_to_column("Species_metasample")

######################################################
####### GENERATE EXTRA METADATA ######################
######################################################
##### This will include both continuous and categorical variables for each metasample
#Categorical: Public vs in_house, SE vs PS, sequencing technology
#Continuos: Tot number of initial samples, Tot number of initial reads, Read length
in_house_sequencing = "Illumina_HiSeq_2500"
in_house_sequencing_Obi = "Illumina_NovaSeq"
in_house_bioproject = "PRJNA846095"

#### Build the metadata table
#Original metadata
all_species_original_metadata_df = vector()
for (species in all_species) {
  file_species = species
  if (species == "BmA") {file_species = "Bmo"}
  if (species == "Hsa") {file_species = "Hs2"}
  if (species == "Mmu") {file_species = "Mm2"}
  if (species == "Bta") {file_species = "Bt2"}
  if (species == "Spu") {file_species = "Sp2"}
  species_original_metadata_df = read.delim(paste0(metadata_dir, file_species, "_samples_info.tab"), header=TRUE)
  species_original_metadata_df$Species = rep(species, nrow(species_original_metadata_df))
  ### Modify colnames as needed
  colnames(species_original_metadata_df)[colnames(species_original_metadata_df) == "Group"] = "Metasample"
  ### Join to final dataframe
  all_species_original_metadata_df = rbind(all_species_original_metadata_df, species_original_metadata_df[,1:8])
}

#Extra metadata
all_species_extra_metadata_df = vector()
for (species in all_species) {
  file_species = species
  if (species == "BmA") {file_species = "Bmo"}
  if (species == "Hsa") {file_species = "Hs2"}
  if (species == "Mmu") {file_species = "Mm2"}
  if (species == "Bta") {file_species = "Bt2"}
  if (species == "Spu") {file_species = "Sp2"}
  species_extra_metadata_df = read.delim(paste0(metadata_dir, "extra_metadata/", file_species, "-extra_samples_info.tab"), header=TRUE, stringsAsFactors = FALSE)
  #species_extra_metadata_df$Species = rep(species, nrow(species_extra_metadata_df))
  ### Modify colnames as needed
  colnames(species_extra_metadata_df)[colnames(species_extra_metadata_df)=="Run"] = "SRA"
  ### Join to final dataframe
  all_species_extra_metadata_df = rbind(all_species_extra_metadata_df, species_extra_metadata_df)
}

all_species_all_metadata_df = all_species_original_metadata_df %>%
  left_join(all_species_extra_metadata_df, by="SRA") %>%
  select(-ProjectID) %>%
  ### Add information for in_house samples
  mutate(Platform = case_when(SRA=="in_house" ~ "ILLUMINA",
                              TRUE ~ Platform),
         Model = case_when(SRA=="in_house" & Species != "Obi" ~ in_house_sequencing,
                           SRA=="in_house" & Species == "Obi" ~ in_house_sequencing_Obi,
                           TRUE ~ Model),
         BioProject = case_when(SRA=="in_house" ~ in_house_bioproject,
                                TRUE ~ BioProject),
         Origin = case_when(SRA=="in_house" ~ "in_house", 
                            TRUE ~ "public")) %>%
  ### Change read numbers to numeric
  mutate(Read_number = as.numeric(gsub(",", "", Read_number)))

#### Collapse info at the metasample level
all_species_collapse_metadata_df = all_species_all_metadata_df %>%
  group_by(Species, Tissue, Metasample, SE_PE, Platform, Model, Origin) %>%
  summarize(Mean_read_len = round(mean(Read_len)),
            Tot_read_num = sum(Read_number),
            Tot_original_samples = n()) %>%
  ungroup() %>%
  unite(Species_metasample, c("Species", "Metasample"), remove=FALSE, sep="_") %>%
  ## Select only metasamples without ambiguities (no duplications)
  group_by(Species_metasample) %>%
  filter(n()==1) %>%
  ## Add PC1 values
  left_join(PC1_values_df, by="Species_metasample") %>%
  mutate(Origin = factor(Origin, levels=c("public", "in_house")),
         SE_PE = factor(SE_PE, levels=c("SE", "PE")))

###################################################################
################ PLOT ORIGINAL PC1 DISTRIBUTIONS ##################
###################################################################
tot_metasamples = nrow(all_species_collapse_metadata_df)
discrete_variables = c("SE_PE", "Origin", "Model")
continuos_variables = c("Tot_original_samples", "Mean_read_len", "Tot_read_num")
all_variables = c(discrete_variables, continuos_variables)
selected_discrete_modes = c("SE_PE"="PE", "Origin"="public", "Model"="Illumina_HiSeq_2500")

original_PC_distribs_list = list()
i = 1

for (variable in discrete_variables) {
  if (variable=="Model") {axis_text_angle_x = 90; axis_text_hjust_x=1; axis_text_vjust_x=1} else {axis_text_angle_x = 0; axis_text_hjust_x=0.5; axis_text_vjust_x=1}
  
  subsetted_metadata_df = all_species_collapse_metadata_df %>%
    ungroup() %>%
    filter(!!as.symbol(variable) == unname(selected_discrete_modes[variable])) %>%
    summarise(xmax = unname(selected_discrete_modes[variable]),
              xmin = unname(selected_discrete_modes[variable]),
              ymax = max(PC1),
              ymin = min(PC1))
  
  discrete_var_PC1_plot = ggplot(data=all_species_collapse_metadata_df, aes_string(x=variable, y="PC1")) +
    geom_jitter(color="dimgray", alpha=0.6) +
    geom_boxplot(alpha=0.3) +
    geom_tile(data=subsetted_metadata_df, aes(x=xmax, y=1, height = Inf, width = 1), fill="white", alpha=0, color="red", size=0.7) +
    stat_compare_means() +
    ggtitle(gsub("_", " ", variable)) +
    theme_bw() +
    theme(plot.title = element_text(color="black", hjust=0.5, face="bold"),
          axis.title = element_text(color="black", size=12),
          axis.text.x = element_text(color="black", angle=axis_text_angle_x, vjust=axis_text_vjust_x, hjust=axis_text_hjust_x, size=12),
          axis.text.y = element_text(color="black", size=12)) +
    xlab(gsub("_", " ", variable))
  
  #Add to list and print
  original_PC_distribs_list[[i]] = discrete_var_PC1_plot
  i = i+1
  #print(discrete_var_PC1_plot)
  
}

for (variable in continuos_variables) {
  ## Subset for the plot post correction
  if (variable == "Mean_read_len") {subsetted_metadata_df = all_species_collapse_metadata_df %>% filter(Mean_read_len %in% seq(140,160))}
  if (variable == "Tot_original_samples") {subsetted_metadata_df = all_species_collapse_metadata_df %>% filter(Tot_original_samples == 1)}
  if (variable == "Tot_read_num") {subsetted_metadata_df = all_species_collapse_metadata_df %>% filter(Tot_read_num <= 100000000, Tot_read_num >= 80000000)}
  
  ## Generate code to add rectangles on the plots
  subsetted_metadata_df = subsetted_metadata_df %>%
    dplyr::select(Species_metasample, as.name(variable), PC1) %>%
    ungroup() %>%
    summarise(xmax = max(!!as.symbol(variable)),
              xmin = min(!!as.symbol(variable)), 
              ymax = max(PC1),
              ymin = min(PC1))
  
  ## Compute pearson correlation between the variable and the coordinates
  original_cor_res = cor.test(as.numeric(unlist(all_species_collapse_metadata_df[,variable])), 
                     all_species_collapse_metadata_df$PC1)
  
  continuos_var_PC1_plot = ggplot(data=all_species_collapse_metadata_df, aes_string(x=variable, y="PC1")) +
    geom_jitter(color="dimgray", alpha=0.6) +
    geom_smooth(method="lm", color="black") +
    geom_rect(data=subsetted_metadata_df, aes(xmin=xmin, xmax=xmax, ymax=ymax, ymin=ymin), color="red", fill=NA, inherit.aes = FALSE) +
    annotation_custom(grobTree(textGrob(paste0("Pearson's cor: ", round(unname(original_cor_res$estimate),3), 
                                               "\nPvalue: ", formatC(original_cor_res$p.value, format="e", digits=2)),
                                        x=0.6, y=0.95, hjust=0.5))) +
    ggtitle(gsub("_", " ", variable)) +
    theme_bw() +
    ggtitle(gsub("_", " ", variable)) +
        theme(plot.title = element_text(color="black", hjust=0.5, face="bold"),
          axis.title = element_text(color="black", size=12),
          axis.text.x = element_text(color="black", size=12),
          axis.text.y = element_text(color="black", size=12)) +
    xlab(gsub("_", " ", variable))
  
  #Add to list and print
  original_PC_distribs_list[[i]] = continuos_var_PC1_plot
  i = i+1
  #print(continuos_var_PC1_plot)
}

###################################################################
################ PLOT PCA FROM SELECTION ##########################
###################################################################

new_PCA_list = list()
j = 1
for (variable in all_variables) {
  ###### Subset the metadata table
  if (variable == "Mean_read_len") {subsetted_metadata_df = all_species_collapse_metadata_df %>% filter(Mean_read_len %in% seq(140,160))}
  if (variable == "Tot_original_samples") {subsetted_metadata_df = all_species_collapse_metadata_df %>% filter(Tot_original_samples == 1)}
  if (variable == "Tot_read_num") {subsetted_metadata_df = all_species_collapse_metadata_df %>% filter(Tot_read_num <= 100000000, Tot_read_num >= 80000000)}
  if (variable %in% discrete_variables) {subsetted_metadata_df = all_species_collapse_metadata_df %>% ungroup() %>%
    filter(!!as.symbol(variable) == unname(selected_discrete_modes[variable]))}

  ###### Subset expression table to selected samples
  selected_samples = as.vector(subsetted_metadata_df$Species_metasample)
  subsetted_df = expr_df[,selected_samples]

  ###### PCA plots
  res_PCA = PCA_basic_by_species_plot(subsetted_df, pca_metadata_df, clade, orthogroups_number)
  #### PC1 vs PC2 plot
  res_PCA = PCA_basic_by_species_plot(subsetted_df, pca_metadata_df, clade, orthogroups_number)
  PCA_plot = res_PCA$results[[1]] + ggtitle("") + theme(legend.position = "none") + guides(fill=FALSE, color=FALSE, fill=FALSE, shape=FALSE)

  #### lateral distribs
  raw_PCA = prcomp(as.matrix(t(subsetted_df)), center = TRUE, scale = TRUE) #center and scale the PCs.
  raw_PCA_df = as.data.frame(raw_PCA$x[order(rownames(raw_PCA$x)),]) %>%
    rownames_to_column("Metasample") %>%
    left_join(pca_metadata_df %>% rownames_to_column("Metasample"), by="Metasample") %>%
    mutate(clade = case_when(species %in% Vertebrata ~ "Vertebrate",
                             TRUE ~ "Not_vertebrate"))
  
  #Run wilcoxon test for annotation
  wilcoxon_res = wilcox.test(as.vector(subset(raw_PCA_df, clade=="Vertebrate")$PC1), 
                             as.vector(subset(raw_PCA_df, clade=="Not_vertebrate")$PC1))
  
  lateral_distrib_plot = ggplot(data=raw_PCA_df, aes(x=PC1, color=clade, fill=clade, group=clade)) +
    geom_density(alpha=0.6) +
    annotation_custom(grobTree(textGrob(paste0("Wilcoxon, p: ", formatC(wilcoxon_res$p.value, format="e", digits=2)),
                                        x=0.5, y=0.9, hjust=0.5))) +
    scale_color_manual(values=c("Vertebrate"="#82359D", "Not_vertebrate"="goldenrod")) +
    scale_fill_manual(values=c("Vertebrate"="#82359D", "Not_vertebrate"="goldenrod")) +
    theme_bw()

  #### combine the two plots
  PCA_combined_plots =
    plot_grid(lateral_distrib_plot + theme(axis.text.x = element_blank(),
                                           axis.text.y = element_text(color="black", size=12),
                                           axis.title.x = element_blank(),
                                           axis.title.y = element_text(color="black", size=12),
                                           axis.ticks.x = element_blank(),
                                           legend.position = "none",
                                           plot.margin = unit(c(0, 0.1, 0, 0), "cm")) +
                ylab("Density"),
              PCA_plot + theme(legend.position = "none",
                               plot.margin = unit(c(0, 0.1, 0, 0), "cm")),
              nrow=2, align="v", rel_heights = c(0.2,1))

  #Add to list and print
  new_PCA_list[[j]] = PCA_combined_plots
  j = j+1
  #print(PCA_combined_plots)
}

###################################################################
################ COMBINE TWO TYPES OF PLOT ########################
###################################################################

all_combined_plot_list = list()
for (index in seq(1:length(new_PCA_list))) {
  final_plot = plot_grid(original_PC_distribs_list[[index]], new_PCA_list[[index]], ncol=2, align="h", axis="t", rel_widths = c(01,1))
  all_combined_plot_list[[index]] = final_plot
  #print(final_plot)
}

final_batch_effects_panel = plot_grid(plotlist = all_combined_plot_list, ncol=2, align="hv")
print(final_batch_effects_panel)
```

```{r PCA_extra_bilateria, warning=FALSE, echo=FALSE, message=FALSE, dependson=c("setup", "functions"), fig.width=12, fig.height=4}
#### Initialize final plot list
extra_bilateria_PCAs = list()
i = 1

######################################################
############## SVA CORRECTION ########################
######################################################
######### Upload table
expr_file = paste0(pca_dir, "/", version, "/", category, "/", clade, "/", evo_type, "/Bilateria_conserved-metasamples_median_expr-SVA-log2-TPMs-NORM-BH_genes.tab")
expr_df = read.delim(expr_file, sep="\t", header=TRUE, row=1)
expr_df = expr_df[complete.cases(expr_df),] #I am using only the cases without NAs, so that I don't have to do any imputation.
#Update colnames
colnames(expr_df) = gsub("BmA", "Bmo", gsub("Hs2", "Hsa", gsub("Mm2", "Mmu", gsub("Bt2", "Bta", gsub("Sp2", "Spu", colnames(expr_df))))))

#Run PCA
orthogroups_number=nrow(expr_df)
res_PCA = plot_PCA(expr_df, pca_metadata_df, clade, orthogroups_number)
PC_plot = res_PCA$results[[1]] +
  theme(legend.position = "bottom",
        plot.title = element_text(color="black", hjust=0.5, face="bold")) +
  guides(color=guide_legend(nrow=2), shape=guide_legend(nrow=2)) +
  ggtitle("SVA normalization")

print(PC_plot)
extra_bilateria_PCAs[[i]] = PC_plot
i = i+1

######################################################
############## AVERAGE PARALOG EXPR ##################
######################################################
my_clade = "bilateria"
######### Upload table
expr_file = paste0(broccoli_revision_dir, my_clade, "/expression_tables/", my_clade, "_conserved-metasamples_median_expr-NOSVA-log2-TPMs-paralogs_average-NORM.tab")
expr_df = read.delim(expr_file, sep="\t", header=TRUE, row=1)
expr_df = expr_df[complete.cases(expr_df),] #I am using only the cases without NAs, so that I don't have to do any imputation.
#Update colnames
colnames(expr_df) = gsub("BmA", "Bmo", gsub("Hs2", "Hsa", gsub("Mm2", "Mmu", gsub("Bt2", "Bta", gsub("Sp2", "Spu", colnames(expr_df))))))

orthogroups_number=nrow(expr_df)
res_PCA = plot_PCA(expr_df, pca_metadata_df, clade, orthogroups_number)
PC_plot = res_PCA$results[[1]] +
  theme(legend.position = "bottom",
        plot.title = element_text(color="black", hjust=0.5, face="bold")) +
  guides(color=guide_legend(nrow=2), shape=guide_legend(nrow=2)) +
  ggtitle("Average paralog expression")

print(PC_plot)
extra_bilateria_PCAs[[i]] = PC_plot
i = i+1

######################################################
############## SUMMED PARALOG EXPR ###################
######################################################
my_clade = "bilateria"
######### Upload table
expr_file = paste0(broccoli_revision_dir, my_clade, "/expression_tables/", my_clade, "_conserved-metasamples_median_expr-NOSVA-log2-TPMs-paralogs_summed-NORM.tab")
expr_df = read.delim(expr_file, sep="\t", header=TRUE, row=1)
expr_df = expr_df[complete.cases(expr_df),] #I am using only the cases without NAs, so that I don't have to do any imputation.
#Update colnames
colnames(expr_df) = gsub("BmA", "Bmo", gsub("Hs2", "Hsa", gsub("Mm2", "Mmu", gsub("Bt2", "Bta", gsub("Sp2", "Spu", colnames(expr_df))))))

orthogroups_number=nrow(expr_df)
res_PCA = plot_PCA(expr_df, pca_metadata_df, clade, orthogroups_number)
PC_plot = res_PCA$results[[1]] +
  theme(legend.position = "bottom",
        plot.title = element_text(color="black", hjust=0.5, face="bold")) +
  guides(color=guide_legend(nrow=2), shape=guide_legend(nrow=2)) +
  ggtitle("Summed paralog expression")

print(PC_plot)
extra_bilateria_PCAs[[i]] = PC_plot
i = i+1

######################################################
############## SINGLE_COPY ORTHOLOGS #################
######################################################
my_clade = "bilateria"
######### Upload table
expr_file = paste0(broccoli_revision_dir, my_clade, "/expression_tables/", my_clade, "_conserved-metasamples_median_expr-NOSVA-log2-TPMs-single_copies-NORM.tab")
expr_df = read.delim(expr_file, sep="\t", header=TRUE, row=1)
expr_df = expr_df[complete.cases(expr_df),] #I am using only the cases without NAs, so that I don't have to do any imputation.
#Update colnames
colnames(expr_df) = gsub("BmA", "Bmo", gsub("Hs2", "Hsa", gsub("Mm2", "Mmu", gsub("Bt2", "Bta", gsub("Sp2", "Spu", colnames(expr_df))))))

orthogroups_number=nrow(expr_df)
res_PCA = plot_PCA(expr_df, pca_metadata_df, clade, orthogroups_number)
PC_plot = res_PCA$results[[1]] +
  theme(legend.position = "bottom",
        plot.title = element_text(color="black", hjust=0.5, face="bold")) +
  guides(color=guide_legend(nrow=2), shape=guide_legend(nrow=2)) +
  ggtitle(paste0("Single copy orthologs (", nrow(expr_df), " OGs)"))

print(PC_plot)
extra_bilateria_PCAs[[i]] = PC_plot
i = i+1

######################################################
########## JOINED PLOTS ##############################
######################################################

joined_PCA_plots = annotate_figure(ggarrange(plotlist = extra_bilateria_PCAs,
                                          common.legend=TRUE, 
                                          legend="bottom", 
                                          ncol=4))

print(joined_PCA_plots)
```

```{r combined_figure, warning=FALSE, echo=FALSE, messsage=FALES, dependson=c("setup", "functions", "PC1_batch_correlation_custom", "PCA_extra_bilateria"), fig.width=12, fig.height=18}
plot_grid(final_batch_effects_panel, joined_PCA_plots, rel_heights = c(1, 0.3), nrow=2)

pdf("/Users/federica/Documents/CRG_PhD/bilaterian_GE/NatGen-EcoEvo_submission/Revision/Images/Figures_from_R/Batch_effects_on_PC1-v1.pdf", width=12, height=18)
plot_grid(final_batch_effects_panel, joined_PCA_plots, rel_heights = c(1, 0.3), nrow=2)
dev.off()
```

<!-- ########################################################### -->
<!-- ############## Vertebrates and Insects PCAs ############### -->
<!-- ########################################################### -->

```{r vertebrates_insects_single_copy_orthologs, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions"), fig.width=12, fig.height=5}
for (my_clade in c("vertebrata", "insecta")) { 
  ##############################################
  ##### GENERATE INPUT TABLE ###################
  ##############################################
  
  ######### Upload table
  expr_file = paste0(broccoli_revision_dir, my_clade, "/expression_tables/", my_clade, "_conserved-metasamples_median_expr-NOSVA-log2-TPMs-single_copies-NORM.tab")
  expr_df = read.delim(expr_file, sep="\t", header=TRUE, row=1)
  expr_df = expr_df[complete.cases(expr_df),] #I am using only the cases without NAs, so that I don't have to do any imputation.
  #Update colnames
  colnames(expr_df) = gsub("BmA", "Bmo", gsub("Hs2", "Hsa", gsub("Mm2", "Mmu", gsub("Bt2", "Bta", gsub("Sp2", "Spu", colnames(expr_df))))))
  
  ##############################################
  ##### PLOT ###################################
  ##############################################
  orthogroups_number=nrow(expr_df)
  res_PCA = PCA_basic_plot(expr_df, pca_metadata_df, my_clade, orthogroups_number)
  res_anova = anova_plot(expr_df, pca_metadata_df, my_clade)
  
  if (my_clade == "vertebrata") {
    #### Generate lateral distribs
    raw_PCA = prcomp(as.matrix(t(expr_df)), center = TRUE, scale = TRUE) #center and scale the PCs.
    raw_PCA_df = as.data.frame(raw_PCA$x[order(rownames(raw_PCA$x)),]) %>%
      rownames_to_column("Metasample") %>%
      left_join(pca_metadata_df %>% rownames_to_column("Metasample"), by="Metasample") %>%
      mutate(clade = case_when(species %in% Vertebrata ~ "Vertebrate",
                               TRUE ~ "Not_vertebrate"))
    #### Plot lateral distribs
    tissue_lateral_distrib_plot = ggplot(data=raw_PCA_df, aes(x=PC2, group=tissues, fill=tissues, color=tissues)) +
      geom_density(alpha=0.6) +
      scale_color_manual(values=tissue_colors) +
      scale_fill_manual(values=tissue_colors) +
      theme_bw() +
      theme(axis.title.y=element_blank(),
            axis.text.y=element_blank(),
            axis.ticks.y = element_blank(),
            axis.title.x=element_text(color="black", size=12),
            axis.text.x=element_text(color="black", size=12),
            legend.position = "none") +
      ylab("Density") +
      coord_flip()
    
    vertebrata_plot = plot_grid(res_PCA$results[[1]] + theme(legend.position = "bottom",
                                                             plot.title=element_text(color="black", face="bold")) + 
                                  ggtitle(paste0("Vertebrate single copy orthologs (n=", orthogroups_number, ")")), 
                                tissue_lateral_distrib_plot, ncol=2, rel_widths = c(1,0.4), align="h", axis="tb")
  }
  
  if (my_clade == "insecta")
    insecta_plot = plot_grid(res_PCA$results[[1]] + theme(legend.position = "bottom",
                                                          plot.title=element_text(color="black", face="bold")) +
                                ggtitle(paste0("Insect single copy orthologs (n=", orthogroups_number, ")")), 
                             res_anova$results[[2]] + ggtitle(""), rel_widths = c(1,0.7), align="h", axis="bt")
}

vertebrate_insect_pcas = plot_grid(vertebrata_plot, insecta_plot, ncol=2, align="h", axis="bt", rel_widths = c(0.8,1))

pdf("/Users/federica/Documents/CRG_PhD/bilaterian_GE/NatGen-EcoEvo_submission/Revision/Images/Figures_from_R/vertebrate_insect_pcas.pdf", width=12, height=5)
print(vertebrate_insect_pcas)
dev.off()
```

```{r PCA_clades_single_copy_orthologs, warning=FALSE, message=FALSE, echo=FALSE, dependson=c("setup", "functions"), fig.width=16, fig.height=13}
for (my_clade in "bilateria") {
  ##############################################
  ##### GENERATE INPUT TABLE ###################
  ##############################################
  
  ######### Upload table
  expr_file = paste0(broccoli_revision_dir, my_clade, "/expression_tables/", my_clade, "_conserved-metasamples_median_expr-NOSVA-log2-TPMs-single_copies-NORM.tab")
  expr_df = read.delim(expr_file, sep="\t", header=TRUE, row=1)
  expr_df = expr_df[complete.cases(expr_df),] #I am using only the cases without NAs, so that I don't have to do any imputation.
  #Update colnames
  colnames(expr_df) = gsub("BmA", "Bmo", gsub("Hs2", "Hsa", gsub("Mm2", "Mmu", gsub("Bt2", "Bta", gsub("Sp2", "Spu", colnames(expr_df))))))
  
  ##############################################
  ##### PLOT ###################################
  ##############################################
  
  
  orthogroups_number=nrow(expr_df)
  res_PCA = PCA_basic_plot(expr_df, pca_metadata_df, clade, orthogroups_number)
  res_anova = anova_plot(expr_df, pca_metadata_df, clade)
  #plots
  first_row = annotate_figure(ggarrange(res_PCA$results[[1]] + ggtitle(""), 
                                        res_PCA$results[[2]] + ggtitle(""), 
                                        res_PCA$results[[3]] + ggtitle(""), 
                                        common.legend=TRUE, legend="right", 
                                        ncol=3), 
                          fig.lab=paste0(str_to_title(my_clade), " PC analysis ~ ", orthogroups_number, " orthogroups"), fig.lab.pos = "top.left", fig.lab.size=18)
  second_row = plot_grid(res_anova$results[[1]] + ggtitle("Anova on PCs"), 
                         res_anova$results[[2]] + ggtitle("Anova ~ species separation"), 
                         res_anova$results[[3]] + ggtitle("Anova  ~ tissues separation"), 
                         res_anova$results[[4]] + ggtitle("Variance explained"), 
                         rel_widths=c(1.25,2,2,1), align="h", ncol=4)
  
  final_plot_1 = ggarrange(first_row, second_row, nrow=2)
  print(final_plot_1)
  
  ##############################################
  ##### PLOT ###################################
  ##############################################
  orthogroups_number=nrow(expr_df)
  res_PCA = PCA_basic_by_species_plot(expr_df, pca_metadata_df, clade, orthogroups_number)
  res_anova = anova_plot(expr_df, pca_metadata_df, clade)
  #plots
  first_row = annotate_figure(ggarrange(res_PCA$results[[1]] + ggtitle(""), 
                                        res_PCA$results[[2]] + ggtitle(""), 
                                        res_PCA$results[[3]] + ggtitle(""), 
                                        common.legend=TRUE, legend="right", 
                                        ncol=3), 
                          fig.lab=paste0(str_to_title(my_clade), " PC analysis ~ ", orthogroups_number, " orthogroups"), fig.lab.pos = "top.left", fig.lab.size=18)
  second_row = plot_grid(res_anova$results[[1]] + ggtitle("Anova on PCs"), 
                         res_anova$results[[2]] + ggtitle("Anova ~ species separation"), 
                         res_anova$results[[3]] + ggtitle("Anova  ~ tissues separation"), 
                         res_anova$results[[4]] + ggtitle("Variance explained"), 
                         rel_widths=c(1.25,2,2,1), align="h", ncol=4)
  
  final_plot_2 = ggarrange(first_row, second_row, nrow=2)
  print(final_plot_2)
  
  pdf(paste0("/Users/federica/Documents/CRG_PhD/bilaterian_GE/NatGen-EcoEvo_submission/Revision/Images/Figures_from_R/PCA-", my_clade, "-single_copies.pdf"), height = 13, width=16)
  print(final_plot_1)
  print(final_plot_2)
  dev.off()
}
```

<!-- ########################################################### -->
<!-- ############## ancestral TS ############################### -->
<!-- ########################################################### -->
```{r PCA_ancestral_TS, warning=FALSE, message=FALSE, echo=FALSE, dependson=c("setup", "functions"), fig.width=16, fig.height=13}
##############################################
##### GENERATE INPUT TABLE ###################
##############################################

######### Upload table
expr_file = paste0(pca_dir, "/", version, "/", category, "/", clade, "/", evo_type, "/Bilateria_conserved-metasamples_median_expr-NOSVA-log2-TPMs-NORM-BH_genes.tab")
expr_df = read.delim(expr_file, sep="\t", header=TRUE, row=1)
expr_df = expr_df[complete.cases(expr_df),] #I am using only the cases without NAs, so that I don't have to do any imputation.
#Update colnames
colnames(expr_df) = gsub("BmA", "Bmo", gsub("Hs2", "Hsa", gsub("Mm2", "Mmu", gsub("Bt2", "Bta", gsub("Sp2", "Spu", colnames(expr_df))))))

######### Filter only for orthogroups that are in the ancestral tissue-specific module
splsda_complete_dir = paste0(splsda_dir, version, "/", category, "/", clade, "/", evo_type, "/loadings/metasamples_median_expr/splsda_zscore_comb/")
all_ancestral_ts_OGs = vector()

#Cycle on the single files and upload the orthogroup IDs
for (my_tissue in ordered_tissues) {
  tissue_file = paste0(splsda_complete_dir, my_tissue, "_splsda-zscore_loadings_orthogroups-GO_input.txt")
  tissue_df = read.delim(tissue_file, header=FALSE, col.names = c("OG_ID"))
  tissue_ancestral_ts_OG = as.vector(tissue_df$OG_ID)
  #Join to orthogroups from all tissues
  all_ancestral_ts_OGs = c(all_ancestral_ts_OGs, tissue_ancestral_ts_OG)
}

#Filter expression table
all_ancestral_ts_OGs = all_ancestral_ts_OGs[all_ancestral_ts_OGs %in% rownames(expr_df)] #select only those that are conserved in all species
expr_df = expr_df[all_ancestral_ts_OGs,]

##############################################
##### PLOT ###################################
##############################################
orthogroups_number=nrow(expr_df)
res_PCA = PCA_basic_plot(expr_df, pca_metadata_df, clade, orthogroups_number)
res_anova = anova_plot(expr_df, pca_metadata_df, clade)
#plots
first_row = annotate_figure(ggarrange(res_PCA$results[[1]] + ggtitle(""), 
                                      res_PCA$results[[2]] + ggtitle(""), 
                                      res_PCA$results[[3]] + ggtitle(""), 
                                      common.legend=TRUE, legend="right", 
                                      ncol=3), 
                        fig.lab=paste0(clade, " PC analysis ~ ", orthogroups_number, " orthogroups"), fig.lab.pos = "top.left", fig.lab.size=18)
second_row = plot_grid(res_anova$results[[1]] + ggtitle("Anova on PCs"), 
                       res_anova$results[[2]] + ggtitle("Anova ~ species separation"), 
                       res_anova$results[[3]] + ggtitle("Anova  ~ tissues separation"), 
                       res_anova$results[[4]] + ggtitle("Variance explained"), 
                       rel_widths=c(1.25,2,2,1), align="h", ncol=4)

final_plot = ggarrange(first_row, second_row, nrow=2)
print(final_plot)

pdf("/Users/federica/Documents/CRG_PhD/bilaterian_GE/NatGen-EcoEvo_submission/Revision/Images/Figures_from_R/PCA-ancestral_TS_OGs-ellipses.pdf", height = 13, width=16)
print(final_plot)
dev.off()
```

```{r PCA_ancestral_TS_matched_numbers, warning=FALSE, message=FALSE, echo=FALSE, dependson=c("setup", "functions"), fig.width=16, fig.height=13}
##############################################
##### GENERATE INPUT TABLE ###################
##############################################
OG_number_by_tissue = 20

######### Upload table
expr_file = paste0(pca_dir, "/", version, "/", category, "/", clade, "/", evo_type, "/Bilateria_conserved-metasamples_median_expr-NOSVA-log2-TPMs-NORM-BH_genes.tab")
expr_df = read.delim(expr_file, sep="\t", header=TRUE, row=1)
expr_df = expr_df[complete.cases(expr_df),] #I am using only the cases without NAs, so that I don't have to do any imputation.
#Update colnames
colnames(expr_df) = gsub("BmA", "Bmo", gsub("Hs2", "Hsa", gsub("Mm2", "Mmu", gsub("Bt2", "Bta", gsub("Sp2", "Spu", colnames(expr_df))))))

######### Filter only for orthogroups that are in the ancestral tissue-specific module
splsda_complete_dir = paste0(splsda_dir, version, "/", category, "/", clade, "/", evo_type, "/loadings/metasamples_median_expr/splsda_zscore_comb/")
all_ancestral_ts_OGs = vector()

#Cycle on the single files and upload the orthogroup IDs
for (my_tissue in ordered_tissues) {
  tissue_file = paste0(splsda_complete_dir, my_tissue, "_splsda-zscore_loadings_orthogroups-GO_input.txt")
  tissue_df = read.delim(tissue_file, header=FALSE, col.names = c("OG_ID"))
  tissue_ancestral_ts_OG = as.vector(tissue_df$OG_ID)
  tissue_ancestral_ts_OG = tissue_ancestral_ts_OG[tissue_ancestral_ts_OG %in% rownames(expr_df)] #select only those that are conserved in all species VERY IMPORTANT TO AVOID NAs
  ####### select similar numbers of orthogroups across tissues
  if (length(tissue_ancestral_ts_OG) >= OG_number_by_tissue) {
    selected_tissue_ancestral_ts_OG = tissue_ancestral_ts_OG[1:OG_number_by_tissue]
  } else {
    selected_tissue_ancestral_ts_OG = tissue_ancestral_ts_OG
  }
  ########
  #Join to orthogroups from all tissues
  all_ancestral_ts_OGs = c(all_ancestral_ts_OGs, selected_tissue_ancestral_ts_OG)
}

#Filter expression table
expr_df = expr_df[all_ancestral_ts_OGs,]

##############################################
##### PLOT ###################################
##############################################
orthogroups_number=nrow(expr_df)
res_PCA = PCA_basic_plot(expr_df, pca_metadata_df, clade, orthogroups_number)
res_anova = anova_plot(expr_df, pca_metadata_df, clade)
#plots
first_row = annotate_figure(ggarrange(res_PCA$results[[1]] + ggtitle(""), 
                                      res_PCA$results[[2]] + ggtitle(""), 
                                      res_PCA$results[[3]] + ggtitle(""), 
                                      common.legend=TRUE, legend="right", 
                                      ncol=3), 
                        fig.lab=paste0(clade, " PC analysis ~ ", orthogroups_number, " orthogroups"), fig.lab.pos = "top.left", fig.lab.size=18)
second_row = plot_grid(res_anova$results[[1]] + ggtitle("Anova on PCs"), 
                       res_anova$results[[2]] + ggtitle("Anova ~ species separation"), 
                       res_anova$results[[3]] + ggtitle("Anova  ~ tissues separation"), 
                       res_anova$results[[4]] + ggtitle("Variance explained"), 
                       rel_widths=c(1.25,2,2,1), align="h", ncol=4)

final_plot = ggarrange(first_row, second_row, nrow=2)
print(final_plot)

pdf("/Users/federica/Documents/CRG_PhD/bilaterian_GE/NatGen-EcoEvo_submission/Revision/Images/Figures_from_R/PCA-ancestral_TS_OGs-matched-tissues-ellipses.pdf", height = 13, width=16)
print(final_plot)
dev.off()

##############################################
##### PLOT ###################################
##############################################
orthogroups_number=nrow(expr_df)
res_PCA = PCA_basic_by_species_plot(expr_df, pca_metadata_df, clade, orthogroups_number)
res_anova = anova_plot(expr_df, pca_metadata_df, clade)
#plots
first_row = annotate_figure(ggarrange(res_PCA$results[[1]] + ggtitle(""), 
                                      res_PCA$results[[2]] + ggtitle(""), 
                                      res_PCA$results[[3]] + ggtitle(""), 
                                      common.legend=TRUE, legend="right", 
                                      ncol=3), 
                        fig.lab=paste0(clade, " PC analysis ~ ", orthogroups_number, " orthogroups"), fig.lab.pos = "top.left", fig.lab.size=18)
second_row = plot_grid(res_anova$results[[1]] + ggtitle("Anova on PCs"), 
                       res_anova$results[[2]] + ggtitle("Anova ~ species separation"), 
                       res_anova$results[[3]] + ggtitle("Anova  ~ tissues separation"), 
                       res_anova$results[[4]] + ggtitle("Variance explained"), 
                       rel_widths=c(1.25,2,2,1), align="h", ncol=4)

final_plot = ggarrange(first_row, second_row, nrow=2)
print(final_plot)

pdf("/Users/federica/Documents/CRG_PhD/bilaterian_GE/NatGen-EcoEvo_submission/Revision/Images/Figures_from_R/PCA-ancestral_TS_OGs-matched-species.pdf", height = 13, width=16)
print(final_plot)
dev.off()

##############################################
##### PLOT ###################################
##############################################
orthogroups_number=nrow(expr_df)
res_PCA = PCA_basic_by_clade_plot(expr_df, pca_metadata_df, clade, orthogroups_number)
res_anova = anova_plot(expr_df, pca_metadata_df, clade)
#plots
first_row = annotate_figure(ggarrange(res_PCA$results[[1]] + ggtitle(""), 
                                      res_PCA$results[[2]] + ggtitle(""), 
                                      res_PCA$results[[3]] + ggtitle(""), 
                                      common.legend=TRUE, legend="right", 
                                      ncol=3), 
                        fig.lab=paste0(clade, " PC analysis ~ ", orthogroups_number, " orthogroups"), fig.lab.pos = "top.left", fig.lab.size=18)
second_row = plot_grid(res_anova$results[[1]] + ggtitle("Anova on PCs"), 
                       res_anova$results[[2]] + ggtitle("Anova ~ species separation"), 
                       res_anova$results[[3]] + ggtitle("Anova  ~ tissues separation"), 
                       res_anova$results[[4]] + ggtitle("Variance explained"), 
                       rel_widths=c(1.25,2,2,1), align="h", ncol=4)

final_plot = ggarrange(first_row, second_row, nrow=2)
print(final_plot)

pdf("/Users/federica/Documents/CRG_PhD/bilaterian_GE/NatGen-EcoEvo_submission/Revision/Images/Figures_from_R/PCA-ancestral_TS_OGs-matched-clades.pdf", height = 13, width=16)
print(final_plot)
dev.off()
```

```{r PCA_ancestral_TS_zscored, warning=FALSE, message=FALSE, echo=FALSE, dependson=c("setup", "functions"), fig.width=16, fig.height=13}
##############################################
##### GENERATE INPUT TABLE ###################
##############################################

######### Upload table
expr_file = paste0(pca_dir, "/", version, "/", category, "/", clade, "/", evo_type, "/Bilateria_conserved-metasamples_median_expr-NOSVA-log2-TPMs-NORM-species_zscore-BH_genes.tab")
expr_df = read.delim(expr_file, sep="\t", header=TRUE, row=1)
expr_df = expr_df[complete.cases(expr_df),] #I am using only the cases without NAs, so that I don't have to do any imputation.
#Update colnames
colnames(expr_df) = gsub("BmA", "Bmo", gsub("Hs2", "Hsa", gsub("Mm2", "Mmu", gsub("Bt2", "Bta", gsub("Sp2", "Spu", colnames(expr_df))))))

######### Filter only for orthogroups that are in the ancestral tissue-specific module
splsda_complete_dir = paste0(splsda_dir, version, "/", category, "/", clade, "/", evo_type, "/loadings/metasamples_median_expr/splsda_zscore_comb/")
all_ancestral_ts_OGs = vector()

#Cycle on the single files and upload the orthogroup IDs
for (my_tissue in ordered_tissues) {
  tissue_file = paste0(splsda_complete_dir, my_tissue, "_splsda-zscore_loadings_orthogroups-GO_input.txt")
  tissue_df = read.delim(tissue_file, header=FALSE, col.names = c("OG_ID"))
  tissue_ancestral_ts_OG = as.vector(tissue_df$OG_ID)
  #Join to orthogroups from all tissues
  all_ancestral_ts_OGs = c(all_ancestral_ts_OGs, tissue_ancestral_ts_OG)
}

#Filter expression table
all_ancestral_ts_OGs = all_ancestral_ts_OGs[all_ancestral_ts_OGs %in% rownames(expr_df)] #select only those that are conserved in all species
expr_df = expr_df[all_ancestral_ts_OGs,]

##############################################
##### PLOT ###################################
##############################################
orthogroups_number=nrow(expr_df)
res_PCA = PCA_basic_plot(expr_df, pca_metadata_df, clade, orthogroups_number)
res_anova = anova_plot(expr_df, pca_metadata_df, clade)
#plots
first_row = annotate_figure(ggarrange(res_PCA$results[[1]] + ggtitle(""), 
                                      res_PCA$results[[2]] + ggtitle(""), 
                                      res_PCA$results[[3]] + ggtitle(""), 
                                      common.legend=TRUE, legend="right", 
                                      ncol=3), 
                        fig.lab=paste0(clade, " PC analysis ~ ", orthogroups_number, " orthogroups"), fig.lab.pos = "top.left", fig.lab.size=18)
second_row = plot_grid(res_anova$results[[1]] + ggtitle("Anova on PCs"), 
                       res_anova$results[[2]] + ggtitle("Anova ~ species separation"), 
                       res_anova$results[[3]] + ggtitle("Anova  ~ tissues separation"), 
                       res_anova$results[[4]] + ggtitle("Variance explained"), 
                       rel_widths=c(1.25,2,2,1), align="h", ncol=4)

final_plot = ggarrange(first_row, second_row, nrow=2)
print(final_plot)

pdf("/Users/federica/Documents/CRG_PhD/bilaterian_GE/NatGen-EcoEvo_submission/Revision/Images/Figures_from_R/PCA-ancestral_TS_OGs_zscores.pdf", height = 13, width=16)
print(final_plot)
dev.off()
```

<!-- ########################################################### -->
<!-- ############## ancestral and zscores ##################### -->
<!-- ########################################################### -->

```{r PCA_ancestral_TS_matched_numbers_zscored, warning=FALSE, message=FALSE, echo=FALSE, dependson=c("setup", "functions"), fig.width=16, fig.height=13}
##############################################
##### GENERATE INPUT TABLE ###################
##############################################
OG_number_by_tissue = 20

######### Upload table
expr_file = paste0(pca_dir, "/", version, "/", category, "/", clade, "/", evo_type, "/Bilateria_conserved-metasamples_median_expr-NOSVA-log2-TPMs-NORM-species_zscore-BH_genes.tab")
expr_df = read.delim(expr_file, sep="\t", header=TRUE, row=1)
expr_df = expr_df[complete.cases(expr_df),] #I am using only the cases without NAs, so that I don't have to do any imputation.
#Update colnames
colnames(expr_df) = gsub("BmA", "Bmo", gsub("Hs2", "Hsa", gsub("Mm2", "Mmu", gsub("Bt2", "Bta", gsub("Sp2", "Spu", colnames(expr_df))))))

######### Filter only for orthogroups that are in the ancestral tissue-specific module
splsda_complete_dir = paste0(splsda_dir, version, "/", category, "/", clade, "/", evo_type, "/loadings/metasamples_median_expr/splsda_zscore_comb/")
all_ancestral_ts_OGs = vector()

#Cycle on the single files and upload the orthogroup IDs
for (my_tissue in ordered_tissues) {
  tissue_file = paste0(splsda_complete_dir, my_tissue, "_splsda-zscore_loadings_orthogroups-GO_input.txt")
  tissue_df = read.delim(tissue_file, header=FALSE, col.names = c("OG_ID"))
  tissue_ancestral_ts_OG = as.vector(tissue_df$OG_ID)
  ####### select similar numbers of orthogroups across tissues
  if (length(tissue_ancestral_ts_OG) >= OG_number_by_tissue) {
    selected_tissue_ancestral_ts_OG = tissue_ancestral_ts_OG[1:OG_number_by_tissue]
  } else {
    selected_tissue_ancestral_ts_OG = tissue_ancestral_ts_OG
  }
  ########
  #Join to orthogroups from all tissues
  all_ancestral_ts_OGs = c(all_ancestral_ts_OGs, selected_tissue_ancestral_ts_OG)
}

#Filter expression table
all_ancestral_ts_OGs = all_ancestral_ts_OGs[all_ancestral_ts_OGs %in% rownames(expr_df)] #select only those that are conserved in all species
expr_df = expr_df[all_ancestral_ts_OGs,]

##############################################
##### PLOT ###################################
##############################################
orthogroups_number=nrow(expr_df)
res_PCA = PCA_basic_plot(expr_df, pca_metadata_df, clade, orthogroups_number)
res_anova = anova_plot(expr_df, pca_metadata_df, clade)
#plots
first_row = annotate_figure(ggarrange(res_PCA$results[[1]] + ggtitle(""), 
                                      res_PCA$results[[2]] + ggtitle(""), 
                                      res_PCA$results[[3]] + ggtitle(""), 
                                      common.legend=TRUE, legend="right", 
                                      ncol=3), 
                        fig.lab=paste0(clade, " PC analysis ~ ", orthogroups_number, " orthogroups"), fig.lab.pos = "top.left", fig.lab.size=18)
second_row = plot_grid(res_anova$results[[1]] + ggtitle("Anova on PCs"), 
                       res_anova$results[[2]] + ggtitle("Anova ~ species separation"), 
                       res_anova$results[[3]] + ggtitle("Anova  ~ tissues separation"), 
                       res_anova$results[[4]] + ggtitle("Variance explained"), 
                       rel_widths=c(1.25,2,2,1), align="h", ncol=4)

final_plot = ggarrange(first_row, second_row, nrow=2)
print(final_plot)

pdf("/Users/federica/Documents/CRG_PhD/bilaterian_GE/NatGen-EcoEvo_submission/Revision/Images/Figures_from_R/PCA-ancestral_TS_OGs_zscores-matched_tissues.pdf", height = 13, width=16)
print(final_plot)
dev.off()

##############################################
##### PLOT ###################################
##############################################
orthogroups_number=nrow(expr_df)
res_PCA = PCA_basic_by_species_plot(expr_df, pca_metadata_df, clade, orthogroups_number)
res_anova = anova_plot(expr_df, pca_metadata_df, clade)
#plots
first_row = annotate_figure(ggarrange(res_PCA$results[[1]] + ggtitle(""), 
                                      res_PCA$results[[2]] + ggtitle(""), 
                                      res_PCA$results[[3]] + ggtitle(""), 
                                      common.legend=TRUE, legend="right", 
                                      ncol=3), 
                        fig.lab=paste0(clade, " PC analysis ~ ", orthogroups_number, " orthogroups"), fig.lab.pos = "top.left", fig.lab.size=18)
second_row = plot_grid(res_anova$results[[1]] + ggtitle("Anova on PCs"), 
                       res_anova$results[[2]] + ggtitle("Anova ~ species separation"), 
                       res_anova$results[[3]] + ggtitle("Anova  ~ tissues separation"), 
                       res_anova$results[[4]] + ggtitle("Variance explained"), 
                       rel_widths=c(1.25,2,2,1), align="h", ncol=4)

final_plot = ggarrange(first_row, second_row, nrow=2)
print(final_plot)

# pdf("/Users/federica/Documents/CRG_PhD/bilaterian_GE/NatGen-EcoEvo_submission/Revision/Images/Figures_from_R/PCA-ancestral_TS_OGs-matched-species.pdf", height = 13, width=16)
# print(final_plot)
# dev.off()

##############################################
##### PLOT ###################################
##############################################
orthogroups_number=nrow(expr_df)
res_PCA = PCA_basic_by_clade_plot(expr_df, pca_metadata_df, clade, orthogroups_number)
res_anova = anova_plot(expr_df, pca_metadata_df, clade)
#plots
first_row = annotate_figure(ggarrange(res_PCA$results[[1]] + ggtitle(""), 
                                      res_PCA$results[[2]] + ggtitle(""), 
                                      res_PCA$results[[3]] + ggtitle(""), 
                                      common.legend=TRUE, legend="right", 
                                      ncol=3), 
                        fig.lab=paste0(clade, " PC analysis ~ ", orthogroups_number, " orthogroups"), fig.lab.pos = "top.left", fig.lab.size=18)
second_row = plot_grid(res_anova$results[[1]] + ggtitle("Anova on PCs"), 
                       res_anova$results[[2]] + ggtitle("Anova ~ species separation"), 
                       res_anova$results[[3]] + ggtitle("Anova  ~ tissues separation"), 
                       res_anova$results[[4]] + ggtitle("Variance explained"), 
                       rel_widths=c(1.25,2,2,1), align="h", ncol=4)

final_plot = ggarrange(first_row, second_row, nrow=2)
print(final_plot)

# pdf("/Users/federica/Documents/CRG_PhD/bilaterian_GE/NatGen-EcoEvo_submission/Revision/Images/Figures_from_R/PCA-ancestral_TS_OGs-matched-clades.pdf", height = 13, width=16)
# print(final_plot)
# dev.off()
```

```{r PCA_ancestral_TS_zscored_by_tissue, warning=FALSE, message=FALSE, echo=FALSE, dependson=c("setup", "functions"), fig.width=16, fig.height=13}
##############################################
##### GENERATE INPUT TABLE ###################
##############################################
my_tissue = "Neural"

######### Upload table
expr_file = paste0(pca_dir, "/", version, "/", category, "/", clade, "/", evo_type, "/Bilateria_conserved-metasamples_median_expr-NOSVA-log2-TPMs-NORM-species_zscore-BH_genes.tab")
expr_df = read.delim(expr_file, sep="\t", header=TRUE, row=1)
expr_df = expr_df[complete.cases(expr_df),] #I am using only the cases without NAs, so that I don't have to do any imputation.
#Update colnames
colnames(expr_df) = gsub("BmA", "Bmo", gsub("Hs2", "Hsa", gsub("Mm2", "Mmu", gsub("Bt2", "Bta", gsub("Sp2", "Spu", colnames(expr_df))))))

#Filter expression table
tissue_samples = rownames(subset(pca_metadata_df, tissues==my_tissue))
#expr_df = expr_df[all_ancestral_ts_OGs,tissue_samples]
expr_df = expr_df[,tissue_samples]

##############################################
##### PLOT ###################################
##############################################
orthogroups_number=nrow(expr_df)
res_PCA = PCA_basic_plot(expr_df, pca_metadata_df, clade, orthogroups_number)
res_anova = anova_plot(expr_df, pca_metadata_df, clade)
#plots
first_row = annotate_figure(ggarrange(res_PCA$results[[1]] + ggtitle(""), 
                                      res_PCA$results[[2]] + ggtitle(""), 
                                      res_PCA$results[[3]] + ggtitle(""), 
                                      common.legend=TRUE, legend="right", 
                                      ncol=3), 
                        fig.lab=paste0(clade, " PC analysis ~ ", orthogroups_number, " orthogroups"), fig.lab.pos = "top.left", fig.lab.size=18)
second_row = plot_grid(res_anova$results[[1]] + ggtitle("Anova on PCs"), 
                       res_anova$results[[2]] + ggtitle("Anova ~ species separation"), 
                       res_anova$results[[3]] + ggtitle("Anova  ~ tissues separation"), 
                       res_anova$results[[4]] + ggtitle("Variance explained"), 
                       rel_widths=c(1.25,2,2,1), align="h", ncol=4)

final_plot = ggarrange(first_row, second_row, nrow=2)
print(final_plot)

# pdf("/Users/federica/Documents/CRG_PhD/bilaterian_GE/NatGen-EcoEvo_submission/Revision/Images/Figures_from_R/PCA-ancestral_TS_OGs-matched-tissues-ellipses.pdf", height = 13, width=16)
# print(final_plot)
# dev.off()

##############################################
##### PLOT ###################################
##############################################
orthogroups_number=nrow(expr_df)
res_PCA = PCA_basic_by_species_plot(expr_df, pca_metadata_df, clade, orthogroups_number)
res_anova = anova_plot(expr_df, pca_metadata_df, clade)
#plots
first_row = annotate_figure(ggarrange(res_PCA$results[[1]] + ggtitle(""), 
                                      res_PCA$results[[2]] + ggtitle(""), 
                                      res_PCA$results[[3]] + ggtitle(""), 
                                      common.legend=TRUE, legend="right", 
                                      ncol=3), 
                        fig.lab=paste0(clade, " PC analysis ~ ", orthogroups_number, " orthogroups"), fig.lab.pos = "top.left", fig.lab.size=18)
second_row = plot_grid(res_anova$results[[1]] + ggtitle("Anova on PCs"), 
                       res_anova$results[[2]] + ggtitle("Anova ~ species separation"), 
                       res_anova$results[[3]] + ggtitle("Anova  ~ tissues separation"), 
                       res_anova$results[[4]] + ggtitle("Variance explained"), 
                       rel_widths=c(1.25,2,2,1), align="h", ncol=4)

final_plot = ggarrange(first_row, second_row, nrow=2)
print(final_plot)

# pdf("/Users/federica/Documents/CRG_PhD/bilaterian_GE/NatGen-EcoEvo_submission/Revision/Images/Figures_from_R/PCA-ancestral_TS_OGs-matched-species.pdf", height = 13, width=16)
# print(final_plot)
# dev.off()

##############################################
##### PLOT ###################################
##############################################
orthogroups_number=nrow(expr_df)
res_PCA = PCA_basic_by_clade_plot(expr_df, pca_metadata_df, clade, orthogroups_number)
res_anova = anova_plot(expr_df, pca_metadata_df, clade)
#plots
first_row = annotate_figure(ggarrange(res_PCA$results[[1]] + ggtitle(""), 
                                      res_PCA$results[[2]] + ggtitle(""), 
                                      res_PCA$results[[3]] + ggtitle(""), 
                                      common.legend=TRUE, legend="right", 
                                      ncol=3), 
                        fig.lab=paste0(clade, " PC analysis ~ ", orthogroups_number, " orthogroups"), fig.lab.pos = "top.left", fig.lab.size=18)
second_row = plot_grid(res_anova$results[[1]] + ggtitle("Anova on PCs"), 
                       res_anova$results[[2]] + ggtitle("Anova ~ species separation"), 
                       res_anova$results[[3]] + ggtitle("Anova  ~ tissues separation"), 
                       res_anova$results[[4]] + ggtitle("Variance explained"), 
                       rel_widths=c(1.25,2,2,1), align="h", ncol=4)

final_plot = ggarrange(first_row, second_row, nrow=2)
print(final_plot)

# pdf("/Users/federica/Documents/CRG_PhD/bilaterian_GE/NatGen-EcoEvo_submission/Revision/Images/Figures_from_R/PCA-ancestral_TS_OGs-matched-clades.pdf", height = 13, width=16)
# print(final_plot)
# dev.off()
```

```{r check_zscore_correlation_all_OGs, warning=FALSE, message=FALSE, echo=FALSE, dependson=c("setup", "functions"), fig.width=16, fig.height=10}
all_cor_plots = list()
i = 1

for (my_tissue in ordered_tissues) {
  ##############################################
  ##### GENERATE INPUT TABLE ###################
  ##############################################
  
  ######### Upload table
  expr_file = paste0(pca_dir, "/", version, "/", category, "/", clade, "/", evo_type, "/Bilateria_conserved-metasamples_median_expr-NOSVA-log2-TPMs-NORM-species_zscore-BH_genes.tab")
  expr_df = read.delim(expr_file, sep="\t", header=TRUE, row=1)
  expr_df = expr_df[complete.cases(expr_df),] #I am using only the cases without NAs, so that I don't have to do any imputation.
  #Update colnames
  colnames(expr_df) = gsub("BmA", "Bmo", gsub("Hs2", "Hsa", gsub("Mm2", "Mmu", gsub("Bt2", "Bta", gsub("Sp2", "Spu", colnames(expr_df))))))
  
  ######## Filter expression table for only one tissue
  tissue_samples = rownames(subset(pca_metadata_df, tissues==my_tissue))
  expr_df = expr_df[,tissue_samples]
  
  ##############################################
  ##### COMPUTE CORRELATIONS ###################
  ##############################################
  
  ###### Generate correlation matrix
  all_cor_matrix = cor(expr_df, method="pearson")
  
  all_cor_long_df = all_cor_matrix %>%
    as.data.frame() %>%
    rownames_to_column("Sample1") %>%
    pivot_longer(cols=colnames(expr_df), names_to = "Sample2", values_to="Cor_values") %>%
    mutate(Species1 = gsub("_.*", "", Sample1)) %>%
    mutate(Species2 = gsub("_.*", "", Sample2)) %>%
    unite("Species_combo", Species1:Species2, sep="_", remove=FALSE) %>%
    left_join(all_species_distances_df, by="Species_combo") %>%
    mutate(all_species_distances = factor(all_species_distances, levels=seq(0,10)))
  
  ##############################################
  ##### PLOT CORRELATIONS ######################
  ##############################################
  tot_OGs = nrow(expr_df)
  
  zscores_correlation_plots = ggplot(data=all_cor_long_df, aes(x=all_species_distances, y=Cor_values, group=all_species_distances, color=Species1)) +
    geom_jitter(alpha=0.8) +
    geom_boxplot(color="black", alpha=0.2) +
    ggtitle(paste0(my_tissue, " ~ OG num: ", tot_OGs)) +
    scale_color_manual(values=species_colors) +
    theme_bw() +
    theme(plot.title = element_text(color="black", hjust=0.5, face="bold"),
          axis.text = element_text(color="black"),
          axis.title = element_text(color="black"),
          legend.position = "bottom") +
    guides(color=guide_legend(nrow =2, title="One species"))+
    xlab("Species distances") +
    ylab("Pearson correlation") +
    coord_cartesian(ylim=c(-0.25,1))
  
  print(zscores_correlation_plots)
  all_cor_plots[[i]] = zscores_correlation_plots
  i = i+1
}

all_panels = annotate_figure(ggarrange(plotlist = all_cor_plots,
                                       common.legend=TRUE, legend="bottom",
                                       ncol=4, nrow=2))
print(all_panels)
```

```{r check_zscore_correlation_tissue_OGs, warning=FALSE, message=FALSE, echo=FALSE, dependson=c("setup", "functions")}
##############################################
##### GENERATE INPUT TABLE ###################
##############################################
my_clade = "bilateria"
my_tissue = "Neural"

######### Upload table
expr_file = paste0(pca_dir, "/", version, "/", category, "/", clade, "/", evo_type, "/Bilateria_conserved-metasamples_median_expr-NOSVA-log2-TPMs-NORM-species_zscore-BH_genes.tab")
expr_df = read.delim(expr_file, sep="\t", header=TRUE, row=1)
expr_df = expr_df[complete.cases(expr_df),] #I am using only the cases without NAs, so that I don't have to do any imputation.
#Update colnames
colnames(expr_df) = gsub("BmA", "Bmo", gsub("Hs2", "Hsa", gsub("Mm2", "Mmu", gsub("Bt2", "Bta", gsub("Sp2", "Spu", colnames(expr_df))))))

######### Filter only for orthogroups that are in the ancestral tissue-specific module
splsda_complete_dir = paste0(splsda_dir, version, "/", category, "/", clade, "/", evo_type, "/loadings/metasamples_median_expr/splsda_zscore_comb/")

#Cycle on the single files and upload the orthogroup IDs
tissue_file = paste0(splsda_complete_dir, my_tissue, "_splsda-zscore_loadings_orthogroups-GO_input.txt")
tissue_df = read.delim(tissue_file, header=FALSE, col.names = c("OG_ID"))
tissue_ancestral_ts_OG = unique(as.vector(tissue_df$OG_ID))
#Select only those that are conserved in all 20 species
tissue_ancestral_ts_OG = tissue_ancestral_ts_OG[tissue_ancestral_ts_OG %in% rownames(expr_df)]

#Filter expression table
tissue_samples = rownames(subset(pca_metadata_df, tissues==my_tissue))
expr_df = expr_df[tissue_ancestral_ts_OG,tissue_samples]

##############################################
##### COMPUTE CORRELATIONS ###################
##############################################

###### Generate correlation matrix
all_cor_matrix = cor(expr_df, method="pearson")

all_cor_long_df = all_cor_matrix %>%
  as.data.frame() %>%
  rownames_to_column("Sample1") %>%
  pivot_longer(cols=colnames(expr_df), names_to = "Sample2", values_to="Cor_values") %>%
  mutate(Species1 = gsub("_.*", "", Sample1)) %>%
  mutate(Species2 = gsub("_.*", "", Sample2)) %>%
  unite("Species_combo", Species1:Species2, sep="_", remove=FALSE) %>%
  left_join(all_species_distances_df, by="Species_combo") %>%
  mutate(all_species_distances = factor(all_species_distances, levels=seq(0,10)))

##############################################
##### PLOT CORRELATIONS ######################
##############################################
tot_OGs = length(tissue_ancestral_ts_OG)

zscores_correlation_plots = ggplot(data=all_cor_long_df, aes(x=all_species_distances, y=Cor_values, group=all_species_distances, color=Species1)) +
  geom_jitter(alpha=0.8) +
  geom_boxplot(color="black", alpha=0.2) +
  ggtitle(paste0(my_tissue, " ~ OG num: ", tot_OGs)) +
  scale_color_manual(values=species_colors) +
  theme_bw() +
  theme(plot.title = element_text(color="black", hjust=0.5, face="bold"),
        axis.text = element_text(color="black"),
        axis.title = element_text(color="black")) +
  guides(color=guide_legend(ncol=2))+
  xlab("Species distances") +
  ylab("Pearson correlation")

print(zscores_correlation_plots)
```

<!-- This is because the z-scores look weird according to Manu -->
```{r plot_zscores, warning=FALSE, message=FALSE, echo=FALSE, dependson=c("setup", "functions")}
##############################################
##### GENERATE INPUT TABLE ###################
##############################################
OG_number_by_tissue = 20

######### Upload table
expr_file = paste0(pca_dir, "/", version, "/", category, "/", clade, "/", evo_type, "/Bilateria_conserved-metasamples_median_expr-NOSVA-log2-TPMs-NORM-species_zscore-BH_genes.tab")
expr_df = read.delim(expr_file, sep="\t", header=TRUE, row=1)
expr_df = expr_df[complete.cases(expr_df),] #I am using only the cases without NAs, so that I don't have to do any imputation.
#Update colnames
colnames(expr_df) = gsub("BmA", "Bmo", gsub("Hs2", "Hsa", gsub("Mm2", "Mmu", gsub("Bt2", "Bta", gsub("Sp2", "Spu", colnames(expr_df))))))

######### Filter only for orthogroups that are in the ancestral tissue-specific module
splsda_complete_dir = paste0(splsda_dir, version, "/", category, "/", clade, "/", evo_type, "/loadings/metasamples_median_expr/splsda_zscore_comb/")
all_ancestral_ts_OGs = vector()

#Cycle on the single files and upload the orthogroup IDs
for (my_tissue in ordered_tissues) {
  tissue_file = paste0(splsda_complete_dir, my_tissue, "_splsda-zscore_loadings_orthogroups-GO_input.txt")
  tissue_df = read.delim(tissue_file, header=FALSE, col.names = c("OG_ID"))
  tissue_ancestral_ts_OG = as.vector(tissue_df$OG_ID)
  ####### select similar numbers of orthogroups across tissues
  if (length(tissue_ancestral_ts_OG) >= OG_number_by_tissue) {
    selected_tissue_ancestral_ts_OG = tissue_ancestral_ts_OG[1:OG_number_by_tissue]
  } else {
    selected_tissue_ancestral_ts_OG = tissue_ancestral_ts_OG
  }
  ########
  #Join to orthogroups from all tissues
  all_ancestral_ts_OGs = c(all_ancestral_ts_OGs, selected_tissue_ancestral_ts_OG)
}

#Filter expression table
expr_df = expr_df[all_ancestral_ts_OGs,]

expr_long_df = expr_df %>%
  rownames_to_column("OG_ID") %>%
  pivot_longer(cols=colnames(expr_df), names_to="Sample", values_to="Zscore") %>%
  separate(Sample, c("Species", "Sample"), remove=TRUE, sep="_", extra="merge") %>%
  unite("OG_ID_Species", OG_ID:Species, , sep="_", remove=FALSE) %>%
  group_by(OG_ID_Species, Species) %>%
  summarize(mean_zscore=mean(Zscore, na.rm=TRUE), std_zscore=sd(Zscore, na.rm=TRUE))

mean_zscores_plot = ggplot(data=expr_long_df, aes(x=mean_zscore)) +
  geom_density(color="dodgerblue") +
  ggtitle("Mean zscore") +
  theme_bw()

std_zscores_plot = ggplot(data=expr_long_df, aes(x=std_zscore)) +
  geom_density(color="dodgerblue") +
  ggtitle("Std zscore") +
  theme_bw()

print(mean_zscores_plot)
print(std_zscores_plot)
```

<!-- Plot z-score expression across tissues of ancestral tissue-specific orthogroups using the average paralog expression
I would expect to see a diminished tissue-specific expression profile -->
```{r plot_expr_across_tissues_paralogs_average, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions")}
##### Load the expr df at the tissue level.
expr_file = paste0(broccoli_revision_dir, my_clade, "/expression_tables/", my_clade, "_conserved-tissue_average_expr-NOSVA-log2-TPMs-paralogs_average-NORM-species_zscore.tab")
expr_df = read.table(expr_file, header=TRUE, row=1)

#### Load the interesting OG_IDs
ancestral_OG_IDs = as.vector(read.delim(paste0(splsda_dir, my_version, "/", my_category, "/Bilateria/", my_evo_type, "/loadings", "/metasamples_median_expr/splsda_zscore_comb/Neural_splsda-zscore_loadings_orthogroups-GO_input.txt"), header=FALSE)$V1)

ancestral_expr_df = expr_df[rownames(expr_df) %in% ancestral_OG_IDs,] #Subset the expression dataframe only to the selected loadings
ancestral_expr_long_df = melt(as.matrix(ancestral_expr_df))
colnames(ancestral_expr_long_df) = c("OG_ID", "Metasample", "Expr")
ancestral_expr_long_df$Species = sub("_.*", "", ancestral_expr_long_df$Metasample)
ancestral_expr_long_df$Tissue = sub(".*_", "", ancestral_expr_long_df$Metasample)
#Add clade label based on the species
clade_expr_df = ancestral_expr_long_df %>% 
mutate(Clade = case_when(Species %in% Vertebrata ~ "Vertebrata",
                         Species %in% Insecta ~ "Insecta",
                         Species %in% Outgroups ~ "Outgroups")
           )

#Compute the median and the zscore
OG_zscore_df = clade_expr_df %>% group_by(OG_ID, Tissue, Clade) %>% #Group dataframe by orthogroup, tissue and clade
  summarize(zscore=median(Expr, na.rm=TRUE)) %>% #Compute the median value for each tissue in each clade
  ungroup()
  
OG_zscore_df$ID = paste0(OG_zscore_df$Clade, "_", OG_zscore_df$Tissue) #Combine clade and tissue in a unique ID

#Order factors for plotting
OG_zscore_df$Tissue = factor(OG_zscore_df$Tissue, levels=ordered_tissues)

ancestral_neural_genes_plot = ggplot(data=OG_zscore_df, aes(x=Tissue, y=zscore, fill=Tissue)) +
  geom_jitter(alpha=0.1, size=0.5) +
  geom_boxplot(color="black", alpha=0.8, outlier.shape = NA) +
  theme_bw() +
  theme(text=element_text(family="sans"),
        axis.text.y = element_text(color="black", size=10),
        axis.title = element_text(color="black", size=12),
        plot.margin = unit(c(0,0,0,0), "cm"),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.title.x = element_blank()
        ) +
  scale_fill_manual(values=my_color_vector) +
  guides(fill=FALSE)

print(ancestral_neural_genes_plot)
```

<!-- ############################################################################################# -->
<!-- ############## TS GAINS AND LOSSES ACROSS GENE FAMILIES ##################################### -->
<!-- ############################################################################################# -->

<!-- ##################################################################################### -->
<!-- ############## Plot numbers of OGs with gains/losses across gene families ########### -->
<!-- ##################################################################################### -->
<!-- Also compute enrichments for OGs with multi TS gains across each class -->

```{r plot_OGs_gains_losses_by_gene_family_Hs2, echo=FALSE, messsage=FALSE, warning=FALSE, dependson=c("setup", "functions"), fig.width=9, fig.height=10.5}
##### Work either with gains or with losses (i.e., inferences)
inference_type = "losses"

#######################################
###### GENERATE INPUT #################
#######################################

##### Upload table with gene families information
gene_families_info_file = paste0(pfam_transfers_dir, "PFAM_annot/orthogroups_to_PFAMClan-transferred_from_Hs2-filtered_min20OGs.txt")
gene_families_df = read.delim(gene_families_info_file, header=FALSE, col.names=c("OG_ID", "Clan_ID", "Clan_name"))

##### Generate vector of orthogroups with at least one inference
all_OG_with_inference_vector = vector()
for (tissue in ordered_tissues) {
  inference_file = paste0(ts_gains_losses_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_", inference_type, ".tab")
  OG_with_inference_vector = as.vector(read.delim(inference_file, header=TRUE)$"OG_ID")
  all_OG_with_inference_vector = c(all_OG_with_inference_vector, OG_with_inference_vector)
}
all_OG_with_inference_vector = unique(as.vector(all_OG_with_inference_vector)) #Remove duplicates

##### Generate vector of orthogroups with multi-TS inferences
multiTS_classification_file = paste0(pfam_transfers_dir, "Bilateria_conserved_orthogroups-multiTS_classification-", inference_type, ".tab")
multiTS_df = read.delim(multiTS_classification_file, header=FALSE, col.names=c("OG_ID", "Evo_class"))
multiTS_OGs_vector = unique(as.vector(multiTS_df$OG_ID))

####### Add information about inferences to gene families df
gene_families_inference_info_df = gene_families_df %>%
  mutate(Inference = case_when(OG_ID %in% multiTS_OGs_vector ~ "multiTS",
                                        OG_ID %in% all_OG_with_inference_vector ~ "singleTS",
                                        TRUE ~ "None")) %>%
  group_by(Clan_ID, Clan_name) %>%
  mutate(Clan_size = n()) %>%
  ungroup()

####### Order the gene families based on the number of orthogroups they contain
clan_order_df = gene_families_inference_info_df %>%
  filter(Inference %in% c("multiTS", "singleTS")) %>%
  group_by(Clan_ID, Clan_name, Clan_size) %>%
  summarize(Tot_inferences_prop = n()/Clan_size) %>%
  ungroup() %>%
  arrange(desc(Tot_inferences_prop)) %>%
  distinct()

#Consider clans that do not have any inferences
no_inference_clans_df = gene_families_inference_info_df %>%
  group_by(Clan_ID, Clan_name) %>%
  filter(all(Inference=="None")) %>%
  dplyr::select(Clan_ID, Clan_name) %>%
  summarize(Clan_size=n(),
            Tot_inferences_prop = 0)

#Add at the end of the complete dataframe
clan_order_df = bind_rows(clan_order_df, no_inference_clans_df)

gene_families_inference_info_df$Clan_ID = factor(gene_families_inference_info_df$Clan_ID, levels=rev(as.vector(clan_order_df$Clan_ID)))
gene_families_inference_info_df$Clan_name = factor(gene_families_inference_info_df$Clan_name, levels=rev(as.vector(clan_order_df$Clan_name)))
gene_families_inference_info_df$Inference = factor(gene_families_inference_info_df$Inference, 
                                                            levels=rev(c("multiTS", "singleTS", "None")))

####### Generate a dataframe with the number of OGs with multiTS inferences by gene family
#This will be usesd to compute the expected probablity
multiTS_count_by_gene_family_df = gene_families_df %>%
  filter(OG_ID %in% multiTS_OGs_vector) %>%
  group_by(Clan_ID, Clan_name) %>%
  summarize(Tot_multiTS = n()) %>%
  ungroup() %>%
  right_join(gene_families_inference_info_df %>% 
               dplyr::select(Clan_ID, Clan_name, Clan_size) %>%
               distinct(), by=c("Clan_ID", "Clan_name")) %>%
  #### If there are NA (meaning no inferences for that gene fam, replace Tot_multiTS with 0)
  mutate(Tot_multiTS=case_when(is.na(Tot_multiTS) ~ 0L, ## the L tells dplyr to interpret the 0 as an integer, not as a double
                                 TRUE ~ Tot_multiTS)) %>%
  ###
  mutate(multiTS_prop = Tot_multiTS / Clan_size) %>%
  left_join(gene_families_inference_info_df %>%
              filter(Inference %in% c("multiTS", "singleTS")) %>%
              group_by(Clan_ID) %>%
              summarize(Tot_inference = n()) %>%
              ungroup(), by="Clan_ID") %>%
  mutate(Tot_inference = case_when(is.na(Tot_inference) ~ 0L,
                                   TRUE ~ Tot_inference)) ## consider clans with no inferences at all


#######################################
###### EXPECTED PROPORTIONS ###########
#######################################

#Compute expected proportions of orthogroups with (multiTS) gains in each family to add to the stacked barplot
#Expected probability is defined by as e = (b/u)*a
#b = total number of OGs in a gene family
#u = total number of OGs with at least one TS gain
#a = total number of OGs with (multiTS) gains

### Define universes and initial numbers
#The universe all for gains is all orthogroups, for losses is all orthogroups with gains
if (inference_type == "gains") {universe_all = 7178}
if (inference_type == "losses") {universe_all = 3283}
universe_TS = length(all_OG_with_inference_vector)
Tot_TS_across_clans = universe_TS
Tot_multiTS_across_clans = length(multiTS_OGs_vector)

TS_prop_over_all = Tot_TS_across_clans / universe_all
multiTS_prop_over_TS = Tot_multiTS_across_clans / universe_TS

expected_df = multiTS_count_by_gene_family_df %>%
  mutate(Expected_TS = TS_prop_over_all*Clan_size,
         Expected_multiTS = multiTS_prop_over_TS*Tot_inference,
         Expected_multiTS_prop = multiTS_prop_over_TS,
         Expected_TS_prop_to_plot = Expected_TS/Clan_size,
         Expected_multiTS_prop_to_plot = Expected_multiTS/Clan_size) %>%
  dplyr::select(Clan_ID, Clan_name, Clan_size, Tot_inference, Tot_multiTS, Expected_TS, Expected_multiTS,  Expected_multiTS_prop, Expected_TS_prop_to_plot,  Expected_multiTS_prop_to_plot)


#######################################
###### BINOMIAL TEST ##################
#######################################

binomial_test_input_df = expected_df %>%
  dplyr::select(Clan_ID, Clan_name, Tot_inference, Tot_multiTS, Clan_size, Expected_multiTS_prop, Expected_TS_prop_to_plot)

### Testing if there is an enrichment in TS gains in a specific family
#I am doing the binom test with alternative=two.sided, it should be okay to test both enrichment and depletion
pvalue_cutoff = 0.05
binomial_test_output_df = binomial_test_input_df %>%
  dplyr::rowwise() %>%
  mutate(binom_TS_pvalue = binom.test(x=Tot_inference, 
                                      n=Clan_size, 
                                      p=Expected_TS_prop_to_plot, 
                                      alternative="two.sided")$p.value,
         binom_multiTS_pvalue = ifelse(Tot_inference > 0, ### Consider when the number of trials (e.g., total TS) is also zero
                                            binom.test(x=Tot_multiTS, 
                                                       n=Tot_inference, 
                                                       p=Expected_multiTS_prop, 
                                                       alternative="two.sided")$p.value,
                                       NA)) %>% 
  mutate(binom_TS_signif = ifelse(binom_TS_pvalue <= pvalue_cutoff, "*", ""),
         binom_multiTS_signif = ifelse(binom_multiTS_pvalue <= pvalue_cutoff, "*", ""))

#######################################
###### MULTI-TS BY EVO ################
#######################################

multiTS_evo_df = multiTS_df %>%
  left_join(gene_families_df, by="OG_ID") %>%
  mutate(Clan_name = factor(Clan_name, levels=rev(as.vector(clan_order_df$Clan_name))),
         Evo_class = factor(Evo_class, levels=rev(c("Vertebrata_only", "Insecta_only", "Others")))) %>%
  filter(!is.na(Clan_name))

#######################################
###### PLOT ###########################
#######################################

#########################
#### Stacked bar ########
#########################
inferences_by_gene_family_stacked_plot = ggplot() +
  geom_bar(data=gene_families_inference_info_df, 
           aes(x=Clan_name, fill=Inference, color=Inference), 
           stat="count", alpha=0.5, position = position_fill()) +
  geom_hline(aes(yintercept=universe_TS/universe_all), color="limegreen", size=1) + #The intercept is the proportion of OGs with inferences over the background
  geom_errorbar(data=expected_df, aes(x=Clan_name, xmin=Clan_name, xmax=Clan_name,
                                      y=Expected_multiTS_prop_to_plot, ymin=Expected_multiTS_prop_to_plot, ymax=Expected_multiTS_prop_to_plot,
                                      group=Clan_name), color="dodgerblue2", size=0.7) +
  geom_text(data=binomial_test_output_df, aes(x=Clan_name, y=1, label=binom_TS_signif), color="limegreen", size=6, position=position_nudge(x=-0.5)) +
  geom_text(data=binomial_test_output_df, aes(x=Clan_name, y=0.97, label=binom_multiTS_signif), color="dodgerblue2", size=6, position=position_nudge(x=-0.5)) +
  scale_fill_manual(values=c("singleTS"="limegreen", "multiTS"="dodgerblue2", "None"="white")) +
  scale_color_manual(values=c("singleTS"="white", "multiTS"="white", "None"="white")) +
  theme_bw() +
  theme(axis.text.y = element_text(color="black"),
        axis.text.x = element_text(color="black", angle=45, hjust=1, vjust=1),
        axis.title = element_text(color="black"),
        legend.position = "bottom") +
  xlab("") +
  ylab("Proportion") +
  coord_flip() +
  ggtitle(paste0("OGs with TS ", inference_type, " across clans"))

#########################
#### Total number #######
#########################
inferences_by_gene_family_count_plot = ggplot(data=gene_families_inference_info_df, aes(x=Clan_name)) +
  geom_bar(stat="count", color="dimgray", fill="dimgray", alpha=0.6) +
  #scale_fill_manual(values=c("Inference"="dimgray", "None"="white")) +
  theme_bw() +
  theme(axis.text.y = element_text(color="black"),
        axis.text.x = element_text(color="black", angle=45, hjust=1, vjust=1),
        axis.title = element_text(color="black"),
        plot.background = element_rect(fill = "transparent", colour = NA)) +
  coord_flip() +
  ggtitle("OGs in clan")

#########################################
#### Proportion of multiTS by Evo #######
#########################################
multiTS_evo_prop_plot = ggplot(data=multiTS_evo_df, aes(x=Clan_name, fill=Evo_class)) +
  geom_bar(stat="count", position=position_fill(), color="white", alpha=0.7, size=0.1) +
  scale_fill_manual(values=c("Vertebrata_only"="#82359D", "Insecta_only"="goldenrod", "Others"="coral3")) +
  theme_bw() +
  theme(axis.text.y = element_text(color="black"),
        axis.text.x = element_text(color="black", angle=45, hjust=1, vjust=1),
        legend.position="bottom") +
  scale_x_discrete(position="top", drop=FALSE) +
  xlab("") +
  ylab("Proportion") +
  coord_flip() +
  ggtitle("MultiTS OGs ~ Evo distribs")

#########################################
#### Combine plots ######################
#########################################
plot_grid(inferences_by_gene_family_stacked_plot + theme(plot.title=element_text(color="black", face="bold", size=10, hjust=0.5)), 
          inferences_by_gene_family_count_plot + theme(axis.title.y=element_blank(),
                                                       axis.text.y=element_blank(),
                                                       axis.ticks.y=element_blank(),
                                                       plot.title=element_text(color="black", face="bold", size=10, hjust=0.5)), 
          multiTS_evo_prop_plot + theme(plot.title=element_text(color="black", face="bold", size=10, hjust=0.5)),
          align="h", axis="bt", rel_widths = c(1,0.4, 1), ncol=3)
```

```{r plot_OGs_gains_losses_by_gene_family_Dme, echo=FALSE, messsage=FALSE, warning=FALSE, dependson=c("setup", "functions"), fig.width=9, fig.height=10.5}
##### Work either with gains or with losses (i.e., inferences)
inference_type = "gains"

#######################################
###### GENERATE INPUT #################
#######################################

##### Upload table with gene families information
gene_families_info_file = paste0(pfam_transfers_dir, "PFAM_annot/orthogroups_to_PFAMClan-transferred_from_Dme-filtered_min20OGs.txt")
gene_families_df = read.delim(gene_families_info_file, header=FALSE, col.names=c("OG_ID", "Clan_ID", "Clan_name"))

##### Generate vector of orthogroups with at least one inference
all_OG_with_inference_vector = vector()
for (tissue in ordered_tissues) {
  inference_file = paste0(ts_gains_losses_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_", inference_type, ".tab")
  OG_with_inference_vector = as.vector(read.delim(inference_file, header=TRUE)$"OG_ID")
  all_OG_with_inference_vector = c(all_OG_with_inference_vector, OG_with_inference_vector)
}
all_OG_with_inference_vector = unique(as.vector(all_OG_with_inference_vector)) #Remove duplicates

##### Generate vector of orthogroups with multi-TS inferences
multiTS_classification_file = paste0(pfam_transfers_dir, "Bilateria_conserved_orthogroups-multiTS_classification-", inference_type, ".tab")
multiTS_df = read.delim(multiTS_classification_file, header=FALSE, col.names=c("OG_ID", "Evo_class"))
multiTS_OGs_vector = unique(as.vector(multiTS_df$OG_ID))

####### Add information about inferences to gene families df
gene_families_inference_info_df = gene_families_df %>%
  mutate(Inference = case_when(OG_ID %in% multiTS_OGs_vector ~ "multiTS",
                                        OG_ID %in% all_OG_with_inference_vector ~ "singleTS",
                                        TRUE ~ "None")) %>%
  group_by(Clan_ID, Clan_name) %>%
  mutate(Clan_size = n()) %>%
  ungroup()

####### Order the gene families based on the number of orthogroups they contain
clan_order_df = gene_families_inference_info_df %>%
  filter(Inference %in% c("multiTS", "singleTS")) %>%
  group_by(Clan_ID, Clan_name, Clan_size) %>%
  summarize(Tot_inferences_prop = n()/Clan_size) %>%
  ungroup() %>%
  arrange(desc(Tot_inferences_prop)) %>%
  distinct()

#Consider clans that do not have any inferences
no_inference_clans_df = gene_families_inference_info_df %>%
  group_by(Clan_ID, Clan_name) %>%
  filter(all(Inference=="None")) %>%
  dplyr::select(Clan_ID, Clan_name) %>%
  summarize(Clan_size=n(),
            Tot_inferences_prop = 0)

#Add at the end of the complete dataframe
clan_order_df = bind_rows(clan_order_df, no_inference_clans_df)

gene_families_inference_info_df$Clan_ID = factor(gene_families_inference_info_df$Clan_ID, levels=rev(as.vector(clan_order_df$Clan_ID)))
gene_families_inference_info_df$Clan_name = factor(gene_families_inference_info_df$Clan_name, levels=rev(as.vector(clan_order_df$Clan_name)))
gene_families_inference_info_df$Inference = factor(gene_families_inference_info_df$Inference, 
                                                            levels=rev(c("multiTS", "singleTS", "None")))

####### Generate a dataframe with the number of OGs with multiTS inferences by gene family
#This will be usesd to compute the expected probablity
multiTS_count_by_gene_family_df = gene_families_df %>%
  filter(OG_ID %in% multiTS_OGs_vector) %>%
  group_by(Clan_ID, Clan_name) %>%
  summarize(Tot_multiTS = n()) %>%
  ungroup() %>%
  right_join(gene_families_inference_info_df %>% 
               dplyr::select(Clan_ID, Clan_name, Clan_size) %>%
               distinct(), by=c("Clan_ID", "Clan_name")) %>%
  #### If there are NA (meaning no inferences for that gene fam, replace Tot_multiTS with 0)
  mutate(Tot_multiTS=case_when(is.na(Tot_multiTS) ~ 0L, ## the L tells dplyr to interpret the 0 as an integer, not as a double
                                 TRUE ~ Tot_multiTS)) %>%
  ###
  mutate(multiTS_prop = Tot_multiTS / Clan_size) %>%
  left_join(gene_families_inference_info_df %>%
              filter(Inference %in% c("multiTS", "singleTS")) %>%
              group_by(Clan_ID) %>%
              summarize(Tot_inference = n()) %>%
              ungroup(), by="Clan_ID") %>%
  mutate(Tot_inference = case_when(is.na(Tot_inference) ~ 0L,
                                   TRUE ~ Tot_inference)) ## consider clans with no inferences at all


#######################################
###### EXPECTED PROPORTIONS ###########
#######################################

#Compute expected proportions of orthogroups with (multiTS) gains in each family to add to the stacked barplot
#Expected probability is defined by as e = (b/u)*a
#b = total number of OGs in a gene family
#u = total number of OGs with at least one TS gain
#a = total number of OGs with (multiTS) gains

### Define universes and initial numbers
universe_all = 7178
universe_TS = length(all_OG_with_inference_vector)
Tot_TS_across_clans = universe_TS
Tot_multiTS_across_clans = length(multiTS_OGs_vector)

TS_prop_over_all = Tot_TS_across_clans / universe_all
multiTS_prop_over_TS = Tot_multiTS_across_clans / universe_TS

expected_df = multiTS_count_by_gene_family_df %>%
  mutate(Expected_TS = TS_prop_over_all*Clan_size,
         Expected_multiTS = multiTS_prop_over_TS*Tot_inference,
         Expected_multiTS_prop = multiTS_prop_over_TS,
         Expected_TS_prop_to_plot = Expected_TS/Clan_size,
         Expected_multiTS_prop_to_plot = Expected_multiTS/Clan_size) %>%
  dplyr::select(Clan_ID, Clan_name, Clan_size, Tot_inference, Tot_multiTS, Expected_TS, Expected_multiTS,  Expected_multiTS_prop, Expected_TS_prop_to_plot,  Expected_multiTS_prop_to_plot)


#######################################
###### BINOMIAL TEST ##################
#######################################

binomial_test_input_df = expected_df %>%
  dplyr::select(Clan_ID, Clan_name, Tot_inference, Tot_multiTS, Clan_size, Expected_multiTS_prop, Expected_TS_prop_to_plot)

### Testing if there is an enrichment in TS gains in a specific family
#I am doing the binom test with alternative=two.sided, it should be okay to test both enrichment and depletion
pvalue_cutoff = 0.05
binomial_test_output_df = binomial_test_input_df %>%
  dplyr::rowwise() %>%
  mutate(binom_TS_pvalue = binom.test(x=Tot_inference, 
                                      n=Clan_size, 
                                      p=Expected_TS_prop_to_plot, 
                                      alternative="two.sided")$p.value,
         binom_multiTS_pvalue = ifelse(Tot_inference > 0, ### Consider when the number of trials (e.g., total TS) is also zero
                                            binom.test(x=Tot_multiTS, 
                                                       n=Tot_inference, 
                                                       p=Expected_multiTS_prop, 
                                                       alternative="two.sided")$p.value,
                                       NA)) %>% 
  mutate(binom_TS_signif = ifelse(binom_TS_pvalue <= pvalue_cutoff, "*", ""),
         binom_multiTS_signif = ifelse(binom_multiTS_pvalue <= pvalue_cutoff, "*", ""))

#######################################
###### MULTI-TS BY EVO ################
#######################################

multiTS_evo_df = multiTS_df %>%
  left_join(gene_families_df, by="OG_ID") %>%
  mutate(Clan_name = factor(Clan_name, levels=rev(as.vector(clan_order_df$Clan_name))),
         Evo_class = factor(Evo_class, levels=rev(c("Vertebrata_only", "Insecta_only", "Others")))) %>%
  filter(!is.na(Clan_name))

#######################################
###### PLOT ###########################
#######################################

#########################
#### Stacked bar ########
#########################
inferences_by_gene_family_stacked_plot = ggplot() +
  geom_bar(data=gene_families_inference_info_df, 
           aes(x=Clan_name, fill=Inference, color=Inference), 
           stat="count", alpha=0.5, position = position_fill()) +
  geom_hline(aes(yintercept=0.457), color="limegreen", size=1) +
  geom_errorbar(data=expected_df, aes(x=Clan_name, xmin=Clan_name, xmax=Clan_name,
                                      y=Expected_multiTS_prop_to_plot, ymin=Expected_multiTS_prop_to_plot, ymax=Expected_multiTS_prop_to_plot,
                                      group=Clan_name), color="dodgerblue2", size=0.7) +
  geom_text(data=binomial_test_output_df, aes(x=Clan_name, y=1, label=binom_TS_signif), color="limegreen", size=6, position=position_nudge(x=-0.5)) +
  geom_text(data=binomial_test_output_df, aes(x=Clan_name, y=0.97, label=binom_multiTS_signif), color="dodgerblue2", size=6, position=position_nudge(x=-0.5)) +
  scale_fill_manual(values=c("singleTS"="limegreen", "multiTS"="dodgerblue2", "None"="white")) +
  scale_color_manual(values=c("singleTS"="white", "multiTS"="white", "None"="white")) +
  theme_bw() +
  theme(axis.text.y = element_text(color="black"),
        axis.text.x = element_text(color="black", angle=45, hjust=1, vjust=1),
        axis.title = element_text(color="black"),
        legend.position = "bottom") +
  xlab("") +
  ylab("Proportion") +
  coord_flip() +
  ggtitle("OGs with TS gains across clans")

#########################
#### Total number #######
#########################
inferences_by_gene_family_count_plot = ggplot(data=gene_families_inference_info_df, aes(x=Clan_name)) +
  geom_bar(stat="count", color="dimgray", fill="dimgray", alpha=0.6) +
  #scale_fill_manual(values=c("Inference"="dimgray", "None"="white")) +
  theme_bw() +
  theme(axis.text.y = element_text(color="black"),
        axis.text.x = element_text(color="black", angle=45, hjust=1, vjust=1),
        axis.title = element_text(color="black"),
        plot.background = element_rect(fill = "transparent", colour = NA)) +
  coord_flip() +
  ggtitle("OGs in clan")

#########################################
#### Proportion of multiTS by Evo #######
#########################################
multiTS_evo_prop_plot = ggplot(data=multiTS_evo_df, aes(x=Clan_name, fill=Evo_class)) +
  geom_bar(stat="count", position=position_fill(), color="white", alpha=0.7, size=0.1) +
  scale_fill_manual(values=c("Vertebrata_only"="#82359D", "Insecta_only"="goldenrod", "Others"="coral3")) +
  theme_bw() +
  theme(axis.text.y = element_text(color="black"),
        axis.text.x = element_text(color="black", angle=45, hjust=1, vjust=1),
        legend.position="bottom") +
  xlab("") +
  ylab("Proportion") +
  coord_flip() +
  ggtitle("MultiTS OGs ~ Evo distribs")

#########################################
#### Combine plots ######################
#########################################
plot_grid(inferences_by_gene_family_stacked_plot + theme(plot.title=element_text(color="black", face="bold", size=10, hjust=0.5)), 
          inferences_by_gene_family_count_plot + theme(axis.title.y=element_blank(),
                                                       axis.text.y=element_blank(),
                                                       axis.ticks.y=element_blank(),
                                                       plot.title=element_text(color="black", face="bold", size=10, hjust=0.5)), 
          multiTS_evo_prop_plot + scale_x_discrete(position="top") + theme(plot.title=element_text(color="black", face="bold", size=10, hjust=0.5)),
          align="h", axis="bt", rel_widths = c(1,0.4, 1), ncol=3)
```


<!-- ############################################################################################# -->
<!-- ############## TAU CONTROLS ################################################################# -->
<!-- ############################################################################################# -->

```{r taus_general_controls, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions"), fig.width=10, fig.height=12}
control_types = c("all_samples_average", "random_metasamples_average")

for (control_type in control_types) {
  #Initialize final dataframes
  all_species_combined_taus_df = data.frame()
  all_species_cor_df = data.frame()
  
  for (species in all_species) {
    ##########################################
    ###### BUILD INPUT #######################
    ##########################################
    file_species = species
    if (species == "BmA") {file_species = "Bmo"}
    if (species == "Hsa") {file_species = "Hs2"}
    if (species == "Mmu") {file_species = "Mm2"}
    if (species == "Bta") {file_species = "Bt2"}
    if (species == "Spu") {file_species = "Sp2"}
    
    ### NB: this contains taus only for genes that pass the expression cutoff
    original_taus_file = paste0(tau_original_dir, file_species, "-protein_coding_taus.tab")
    original_taus_df = read.delim(original_taus_file, header=FALSE, col.names=c("GeneID", "Tau"))
    control_taus_file = paste0(tau_revision_dir, control_type, "/taus/", file_species, "-protein_coding_taus.tab")
    control_taus_df = read.delim(control_taus_file, header=FALSE, col.names=c("GeneID", "Tau"))
    combined_taus_df = original_taus_df %>%
      inner_join(control_taus_df, by="GeneID", suffix=c("_original", "_control")) %>%
      mutate(Species = species)
    
    all_species_combined_taus_df = rbind(all_species_combined_taus_df, combined_taus_df)
    
    ##########################################
    ###### COMPUTE CORRELATION ###############
    ##########################################
    cor_value = cor.test(combined_taus_df$Tau_original, combined_taus_df$Tau_control, method="pearson")$estimate
    all_species_cor_df = rbind(all_species_cor_df, data.frame(Species=species,
                                                              Cor=paste0("Pearson: ", round(cor_value,3)),
                                                              x_cord = 0.3,
                                                              y_cord = 1))
  }
  
  all_species_combined_taus_df$Species = factor(all_species_combined_taus_df$Species, levels=ordered_species)
  all_species_cor_df$Species = factor(all_species_cor_df$Species, levels=ordered_species)

    
  ##########################################
  ###### PLOT CORRELATION ##################
  ##########################################
  
  taus_correlation_plot = ggplot() +
    #geom_point(alpha=0.6, color=control_type_color) +
    geom_hex(data=all_species_combined_taus_df, aes(x=Tau_original, y=Tau_control), bins=100) +
    scale_fill_viridis_c(limits=c(1,50), oob = scales::squish) + #Everything 50 or higher will be represented in the same color
    geom_text(data=all_species_cor_df, aes(x=x_cord, y=y_cord, label=Cor), color="black") +
    #geom_smooth(method="lm", color="black") +
    #annotation_custom(grobTree(textGrob(paste0("Pearson: ", round(cor_value,3)),
    #                                    x=0.4, y=0.95, hjust=0.5))) +
    theme_bw() +
    theme(axis.title = element_text(color="black"),
          axis.text = element_text(color="black"),
          plot.title = element_text(color="black", face="bold", hjust=0.5),
          legend.position = "bottom") +
    coord_equal() +
    facet_wrap(~Species) +
    ggtitle(paste0("Original vs ", gsub("_", " ", control_type), " taus"))
    
    print(taus_correlation_plot)
  }
```

```{r taus_across_datasets, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions"), fig.width=6, fig.height=6}
###### Comparing the tau computed across datasets
###### Filter only for the genes that originally pass the expression cutoff
dataset_types = c("dataset_1", "dataset_2")
multiple_dataset_species = c("Bta", "Mdo", "Dre", "Cmi")

all_species_combined_taus_df = data.frame()
all_species_cor_df = data.frame()

for (species in multiple_dataset_species) {
  ##########################################
  ###### BUILD INPUT #######################
  ##########################################
  file_species = species
  if (species == "BmA") {file_species = "Bmo"}
  if (species == "Hsa") {file_species = "Hs2"}
  if (species == "Mmu") {file_species = "Mm2"}
  if (species == "Bta") {file_species = "Bt2"}
  if (species == "Spu") {file_species = "Sp2"}
  
  ### NB: this contains taus only for genes that pass the expression cutoff
  original_taus_file = paste0(tau_original_dir, file_species, "-protein_coding_taus.tab")
  original_taus_df = read.delim(original_taus_file, header=FALSE, col.names=c("GeneID", "Tau"))
  dataset1_file = paste0(tau_revision_dir, "dataset_1/taus/", file_species, "-protein_coding_taus.tab")
  dataset2_file = paste0(tau_revision_dir, "dataset_2/taus/", file_species, "-protein_coding_taus.tab")
  dataset1_taus_df = read.delim(dataset1_file, header=FALSE, col.names=c("GeneID", "Tau"))
  dataset2_taus_df = read.delim(dataset2_file, header=FALSE, col.names=c("GeneID", "Tau"))
  
  genes_to_filter = original_taus_df %>% filter(!(is.na(Tau))) %>% .$GeneID %>% as.vector()
  
  combined_taus_df = dataset1_taus_df %>%
    inner_join(dataset2_taus_df, by="GeneID", suffix=c("_dataset1", "_dataset2")) %>%
    mutate(Species = species) %>%
    ##### Filter only for the genes that are not NA in the original
    filter(GeneID %in% genes_to_filter)
  
  all_species_combined_taus_df = rbind(all_species_combined_taus_df, combined_taus_df)
  
  ##########################################
  ###### COMPUTE CORRELATION ###############
  ##########################################
  cor_value = cor.test(combined_taus_df$Tau_original, combined_taus_df$Tau_control, method="pearson")$estimate
  all_species_cor_df = rbind(all_species_cor_df, data.frame(Species=species,
                                                            Cor=paste0("Pearson: ", round(cor_value,3)),
                                                            x_cord = 0.3,
                                                            y_cord = 1))
}

all_species_combined_taus_df$Species = factor(all_species_combined_taus_df$Species, levels=ordered_species)
all_species_cor_df$Species = factor(all_species_cor_df$Species, levels=ordered_species)

  
##########################################
###### PLOT CORRELATION ##################
##########################################
  
taus_correlation_plot = ggplot() +
  #geom_point(alpha=0.6, color=control_type_color) +
  geom_hex(data=all_species_combined_taus_df, aes(x=Tau_original, y=Tau_control), bins=100) +
  scale_fill_viridis_c(limits=c(1,50), oob = scales::squish) + #Everything 50 or higher will be represented in the same color
  geom_text(data=all_species_cor_df, aes(x=x_cord, y=y_cord, label=Cor), color="black") +
  theme_bw() +
  theme(axis.title = element_text(color="black"),
        axis.text = element_text(color="black"),
        plot.title = element_text(color="black", face="bold", hjust=0.5),
        legend.position = "bottom") +
  coord_equal() +
  facet_wrap(~Species, ncol=2) +
  ggtitle("Dataset1 vs Dataset2")
  
print(taus_correlation_plot)
```

```{r taus_across_datasets_vs_original, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions"), fig.width=12, fig.height=6}
###### Comparing the tau computed across datasets
###### Filter only for the genes that originally pass the expression cutoff
dataset_types = c("dataset_1", "dataset_2")
multiple_dataset_species = c("Bta", "Mdo", "Dre", "Cmi")

for (dataset_type in dataset_types) {
  all_species_combined_taus_df = data.frame()
  all_species_cor_df = data.frame()
  for (species in multiple_dataset_species) {
    ##########################################
    ###### BUILD INPUT #######################
    ##########################################
    file_species = species
    if (species == "BmA") {file_species = "Bmo"}
    if (species == "Hsa") {file_species = "Hs2"}
    if (species == "Mmu") {file_species = "Mm2"}
    if (species == "Bta") {file_species = "Bt2"}
    if (species == "Spu") {file_species = "Sp2"}
    
    ### NB: this contains taus only for genes that pass the expression cutoff
    original_taus_file = paste0(tau_original_dir, file_species, "-protein_coding_taus.tab")
    original_taus_df = read.delim(original_taus_file, header=FALSE, col.names=c("GeneID", "Tau"))
    control_file = paste0(tau_revision_dir, dataset_type, "/taus/", file_species, "-protein_coding_taus.tab")
    control_taus_df = read.delim(control_file, header=FALSE, col.names=c("GeneID", "Tau"))

    combined_taus_df = original_taus_df %>%
      inner_join(control_taus_df, by="GeneID", suffix=c("_original", "_control")) %>%
      mutate(Species = species)
    
    all_species_combined_taus_df = rbind(all_species_combined_taus_df, combined_taus_df)
    
    ##########################################
    ###### COMPUTE CORRELATION ###############
    ##########################################
    cor_value = cor.test(combined_taus_df$Tau_original, combined_taus_df$Tau_control, method="pearson")$estimate
    all_species_cor_df = rbind(all_species_cor_df, data.frame(Species=species,
                                                              Cor=paste0("Pearson: ", round(cor_value,3)),
                                                              x_cord = 0.3,
                                                              y_cord = 1))
  }
  
  all_species_combined_taus_df$Species = factor(all_species_combined_taus_df$Species, levels=ordered_species)
  all_species_cor_df$Species = factor(all_species_cor_df$Species, levels=ordered_species)
  
    
  ##########################################
  ###### PLOT CORRELATION ##################
  ##########################################
    
  taus_correlation_plot = ggplot() +
    #geom_point(alpha=0.6, color=control_type_color) +
    geom_hex(data=all_species_combined_taus_df, aes(x=Tau_original, y=Tau_control), bins=100) +
    scale_fill_viridis_c(limits=c(1,50), oob = scales::squish) + #Everything 50 or higher will be represented in the same color
    geom_text(data=all_species_cor_df, aes(x=x_cord, y=y_cord, label=Cor), color="black") +
    theme_bw() +
    theme(axis.title = element_text(color="black"),
          axis.text = element_text(color="black"),
          plot.title = element_text(color="black", face="bold", hjust=0.5),
          legend.position = "bottom") +
    coord_equal() +
    facet_wrap(~Species, ncol=4) +
    ggtitle(paste0("Original vs ", dataset_type))
    
  print(taus_correlation_plot)
}

##########################################
###### MDO DATASET 2 #####################
##########################################

#### Check Mdo single tissue correlations
dataset2_expr_file = paste0(tau_revision_dir, "dataset_2/Mdo-protein_coding-tissue_average_expr-NOSVA-log2-TPMs-NORM.tab")
dataset2_expr_df = read.delim(dataset2_expr_file, header=TRUE, row=1)
  
dataset2_expr_long_df = dataset2_expr_df %>%
  rownames_to_column("GeneID") %>%
  pivot_longer(cols=colnames(dataset2_expr_df), names_to="Tissue", values_to="Expr")

original_expr_file = paste0(tau_original_dir, "Mdo-protein_coding-tissue_average_expr-NOSVA-log2-TPMs-NORM.tab")
original_expr_df = read.delim(original_expr_file, header=TRUE, row=1)

original_expr_long_df = original_expr_df %>%
  rownames_to_column("GeneID") %>%
  pivot_longer(cols=colnames(original_expr_df), names_to="Tissue", values_to="Expr")

######### Get genes that pass the expression cutoff ###########
original_taus_file = paste0(tau_original_dir, "Mdo-protein_coding_taus.tab")
original_taus_df = read.delim(original_taus_file, header=FALSE, col.names=c("GeneID", "Tau"))
genes_to_filter = original_taus_df %>% filter(!(is.na(Tau))) %>% .$GeneID %>% as.vector()
###############################################################

#### Combine dataframes
combined_df = original_expr_long_df %>%
  inner_join(dataset2_expr_long_df, by=c("GeneID", "Tissue"), suffix=c("_original", "_dataset2")) %>%
  filter(GeneID %in% genes_to_filter)

correlation_df = combined_df %>%
  group_by(Tissue) %>%
  summarize(Cor = round(cor.test(Expr_original, Expr_dataset2, method="pearson")$estimate, 3))

#### Plot
mdo_tissues_cor_plot = ggplot() +
  geom_hex(data=combined_df, aes(x=Expr_original, y=Expr_dataset2), bins=100) +
  geom_abline(slope=1, intercept=0) +
  geom_text(data=correlation_df, aes(x=5, y=15, label=Cor), color="black") +
  #Everything 50 or higher will be represented in the same color
  scale_fill_viridis_c(limits=c(1,50), oob = scales::squish) +
  theme_bw() +
  theme(axis.title = element_text(color="black"),
        axis.text = element_text(color="black"),
        plot.title = element_text(color="black", face="bold", hjust=0.5),
        legend.position = "bottom") +
  coord_equal() +
  facet_wrap(~Tissue, ncol=4) +
  ggtitle("Dataset2")

print(mdo_tissues_cor_plot)

##########################################
###### MDO DATASET 1 #####################
##########################################

#### Check Mdo single tissue correlations
dataset1_expr_file = paste0(tau_revision_dir, "dataset_1/Mdo-protein_coding-tissue_average_expr-NOSVA-log2-TPMs-NORM.tab")
dataset1_expr_df = read.delim(dataset1_expr_file, header=TRUE, row=1)
  
dataset1_expr_long_df = dataset1_expr_df %>%
  rownames_to_column("GeneID") %>%
  pivot_longer(cols=colnames(dataset2_expr_df), names_to="Tissue", values_to="Expr")

original_expr_file = paste0(tau_original_dir, "Mdo-protein_coding-tissue_average_expr-NOSVA-log2-TPMs-NORM.tab")
original_expr_df = read.delim(original_expr_file, header=TRUE, row=1)

original_expr_long_df = original_expr_df %>%
  rownames_to_column("GeneID") %>%
  pivot_longer(cols=colnames(original_expr_df), names_to="Tissue", values_to="Expr")

######### Get genes that pass the expression cutoff ###########
original_taus_file = paste0(tau_original_dir, "Mdo-protein_coding_taus.tab")
original_taus_df = read.delim(original_taus_file, header=FALSE, col.names=c("GeneID", "Tau"))
genes_to_filter = original_taus_df %>% filter(!(is.na(Tau))) %>% .$GeneID %>% as.vector()
###############################################################

#### Combine dataframes
combined_df = original_expr_long_df %>%
  inner_join(dataset1_expr_long_df, by=c("GeneID", "Tissue"), suffix=c("_original", "_dataset1")) %>%
  filter(GeneID %in% genes_to_filter)

correlation_df = combined_df %>%
  group_by(Tissue) %>%
  summarize(Cor = round(cor.test(Expr_original, Expr_dataset1, method="pearson")$estimate, 3))

#### Plot
mdo_tissues_cor_plot = ggplot() +
  geom_hex(data=combined_df, aes(x=Expr_original, y=Expr_dataset1), bins=100) +
  geom_abline(slope=1, intercept=0) +
  geom_text(data=correlation_df, aes(x=5, y=15, label=Cor), color="black") +
  #Everything 50 or higher will be represented in the same color
  scale_fill_viridis_c(limits=c(1,50), oob = scales::squish) +
  theme_bw() +
  theme(axis.title = element_text(color="black"),
        axis.text = element_text(color="black"),
        plot.title = element_text(color="black", face="bold", hjust=0.5),
        legend.position = "bottom") +
  coord_equal() +
  facet_wrap(~Tissue, ncol=4) +
  ggtitle("Dataset1")

print(mdo_tissues_cor_plot)
```

<!-- ############################################################################################# -->
<!-- ############## RANDOMIZED TS GAINS ########################################################## -->
<!-- ############################################################################################# -->

<!-- ANCESTORS GAINS -->
```{r duplication_levels_randomized_call_ancestors, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions")}
#Upload orthogroups and get duplication info about each gene
orthogroups_df = read.delim(paste0(ts_gains_losses_raw_dir, "Bilateria_conserved_orthogroups-All_genes-associated_tissues.tab"),
                            header=FALSE, col.names=c("OG_ID", "Species", "GeneID", "Tau", "Tissue", "Paralog_tissue", "Expr_cutoff"), stringsAsFactors = FALSE)
#### Add column with OG_ID_Species and count number of paralogs per gene
orthogroups_df$OG_ID_Species = paste0(orthogroups_df$OG_ID, "_", orthogroups_df$Species)
species_paralogs_count_df = as.data.frame(table(orthogroups_df$OG_ID_Species))
OG_ID_Species_paralogs_counts_dict = hashmap(as.vector(species_paralogs_count_df$Var1), as.vector(species_paralogs_count_df$Freq))
orthogroups_df$Paralogs_count = OG_ID_Species_paralogs_counts_dict$find(orthogroups_df$OG_ID_Species)

#########################################################
############## GENERATE BACKGROUND ######################
#########################################################
#### For each node (group of species) compute the proportion in each orthogroup that is duplicated. Get the difference for non-duplicated.
#Sum all and get proportions.
background_df = vector()
for (ancestor in all_ancestors) {
  my_species = get(ancestor)
  #Subset all orthogroups to only these species
  my_species_orthogroups_df = subset(orthogroups_df, Species %in% my_species)
  #Compute for each orthogroup the proportion of duplicated species (among the ones with conserved genes)
   final_ancestor_df = my_species_orthogroups_df %>%
      dplyr::select(OG_ID, Species, Paralogs_count) %>%
      distinct() %>%
      mutate(Dup_status=case_when(Paralogs_count > 1 ~ "Dup",
                                  Paralogs_count <= 1 ~ "NO_Dup")) %>%
      group_by(OG_ID, Dup_status) %>%
      summarise(total = n()) %>%
      ungroup() %>%
      pivot_wider(id_cols="OG_ID", names_from="Dup_status", values_from="total", values_fill=0) %>%
      pivot_longer(cols=c("Dup", "NO_Dup"), names_to = "Dup_status") %>%
      group_by(OG_ID) %>%
      mutate(freq=value/sum(value)) %>%
      mutate(freq = replace_na(freq, 0))
  
  collapsed_final_df = final_ancestor_df %>%
    group_by(Dup_status) %>%
    summarise(freq_proportion = mean(freq)) %>%
    mutate(Node = ancestor) %>%
    dplyr::select(Node, Dup_status, freq_proportion)
  
  background_df = rbind(background_df, collapsed_final_df)
}
#### Order levels
background_df$Node = factor(background_df$Node, levels=c(deuterostoma_y_labels[deuterostoma_y_labels!="Bilateria"], "Bilateria",
                                                             rev(protostoma_y_labels[protostoma_y_labels!="Bilateria"])))
background_df$Dup_status = factor(background_df$Dup_status, levels=rev(c("Dup", "NO_Dup")))
#### Subset only to duplicates
dup_background_df = subset(background_df, Dup_status=="Dup")

#########################################################
############## GENERATE FOREGROUND ######################
#########################################################

#############################
#### Upload TS gains ########
#############################

tissue_gains_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  tissue_gains_df = rbind(tissue_gains_df, input_df)
}

#############################
#### Upload TS losses #######
#############################
tissue_losses_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_losses.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_losses", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  input_df$OG_ID_loss = paste0(input_df$OG_ID, "_", input_df$TS_losses)
  tissue_losses_df = rbind(tissue_losses_df, input_df)
}

#############################
#### Select relative OGs ####
#############################
foreground_df = vector()
for (ancestor in all_ancestors) {
  #### Select ancestor species
  my_species = get(ancestor)
  my_species_orthogroups_df = vector()
  #### Cycle on each tissue
  for (tissue in ordered_tissues) {
    ancestor_gains_OGs = unique(as.vector(subset(tissue_gains_df, TS_gains==ancestor & Tissue==tissue)$OG_ID))
    #### Subset all orthogroups to only these species in the orthogroups with gains
    my_species_tissue_orthogroups_df = subset(orthogroups_df, Species %in% my_species & OG_ID %in% ancestor_gains_OGs)
    ancestor_losses_OGs = unique(as.vector(subset(tissue_losses_df, Tissue==tissue)$OG_ID_loss))
    #### Filter out the losses
    my_species_tissue_orthogroups_df = my_species_tissue_orthogroups_df %>%
      filter(!(OG_ID_Species %in% ancestor_losses_OGs))
    #### Join to final dataframe
    my_species_orthogroups_df = rbind(my_species_orthogroups_df, my_species_tissue_orthogroups_df)
  }
  #Compute the same statistics as above
    final_ancestor_df = my_species_orthogroups_df %>%
      dplyr::select(OG_ID, Species, Paralogs_count) %>%
      distinct() %>%
      mutate(Dup_status=case_when(Paralogs_count > 1 ~ "Dup",
                                  Paralogs_count <= 1 ~ "NO_Dup")) %>%
      group_by(OG_ID, Dup_status) %>%
      summarise(total = n()) %>%
      ungroup() %>%
      pivot_wider(id_cols="OG_ID", names_from="Dup_status", values_from="total", values_fill=0) %>%
      pivot_longer(cols=c("Dup", "NO_Dup"), names_to = "Dup_status") %>%
      group_by(OG_ID) %>%
      mutate(freq=value/sum(value)) %>%
      mutate(freq = replace_na(freq, 0))
  
  collapsed_final_df = final_ancestor_df %>%
    group_by(Dup_status) %>%
    summarise(freq_proportion = mean(freq)) %>%
    mutate(Node = ancestor) %>%
    dplyr::select(Node, Dup_status, freq_proportion)
  
  foreground_df = rbind(foreground_df, collapsed_final_df)
}

#Select only the TS species (species without losses) and compute the proportion of those species that are duplicated.
#Compute the difference as not-duplicated (it has to sum to 1).

####### Order:
foreground_df$Node = factor(foreground_df$Node, levels=c(deuterostoma_y_labels[deuterostoma_y_labels!="Bilateria"], "Bilateria",
                                                             rev(protostoma_y_labels[protostoma_y_labels!="Bilateria"])))
foreground_df$Dup_status = factor(foreground_df$Dup_status, levels=rev(c("Dup", "NO_Dup")))

###### Add branch info
foreground_df$Branch = rep("Deuterostoma", nrow(foreground_df))
foreground_df$Branch[foreground_df$Node %in% protostoma_y_labels] = "Protostoma"
foreground_df$Branch[foreground_df$Node=="Bilateria"] = "Bilateria"
foreground_df = foreground_df %>%
  unite("Branch_Dup_status", c(Branch,Dup_status), remove=FALSE)
foreground_df$Branch_Dup_status = factor(foreground_df$Branch_Dup_status, levels=rev(c("Deuterostoma_Dup", "Deuterostoma_NO_Dup", "Protostoma_Dup", "Protostoma_NO_Dup", "Bilateria_Dup", "Bilateria_NO_Dup")))


#########################################################
##### GENERATE FOREGROUND RANDOMIZATIONS ################
#########################################################
all_rand_foregrounds_df = vector()

for (randomization_round in seq(1,10)) {
  ######### Set directory
  randomization_dir = paste0(home, "projects/bilaterian_GE/data/ts_gains_losses_revision/randomization_", randomization_round, "/")
  
  ######### Upload TS gains
  tissue_gains_df = vector()
  for (tissue in ordered_tissues) {
    input_df = read.delim(paste0(randomization_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
    input_df$Tissue = rep(tissue, nrow(input_df))
    tissue_gains_df = rbind(tissue_gains_df, input_df)
  }
  ######### Upload TS losses
  tissue_losses_df = vector()
  for (tissue in ordered_tissues) {
    input_df = read.delim(paste0(randomization_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_losses.tab"), header=TRUE,
                          stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_losses", "Category"))
    input_df$Tissue = rep(tissue, nrow(input_df))
    input_df$OG_ID_loss = paste0(input_df$OG_ID, "_", input_df$TS_losses)
    tissue_losses_df = rbind(tissue_losses_df, input_df)
  }
  
  ######### Build foreground
  rand_foreground_df = vector()
  for (ancestor in all_ancestors) {
    #### Select ancestor species
    my_species = get(ancestor)
    my_species_orthogroups_df = vector()
    #### Cycle on each tissue
    for (tissue in ordered_tissues) {
      ancestor_gains_OGs = unique(as.vector(subset(tissue_gains_df, TS_gains==ancestor & Tissue==tissue)$OG_ID))
      #### Subset all orthogroups to only these species in the orthogroups with gains
      my_species_tissue_orthogroups_df = subset(orthogroups_df, Species %in% my_species & OG_ID %in% ancestor_gains_OGs)
      ancestor_losses_OGs = unique(as.vector(subset(tissue_losses_df, Tissue==tissue)$OG_ID_loss))
      #### Filter out the losses
      my_species_tissue_orthogroups_df = my_species_tissue_orthogroups_df %>%
        filter(!(OG_ID_Species %in% ancestor_losses_OGs))
      #### Join to final dataframe
      my_species_orthogroups_df = rbind(my_species_orthogroups_df, my_species_tissue_orthogroups_df)
    }
    ##############
    #If there are no gains at all respecting the conditions, go to the next iteration
    if (nrow(my_species_orthogroups_df)==0) {next}
    ##############
    #Compute the same statistics as above
    final_ancestor_df = my_species_orthogroups_df %>%
      dplyr::select(OG_ID, Species, Paralogs_count) %>%
      distinct() %>%
      mutate(Dup_status=case_when(Paralogs_count > 1 ~ "Dup",
                                  Paralogs_count <= 1 ~ "NO_Dup")) %>%
      group_by(OG_ID, Dup_status) %>%
      summarise(total = n()) %>%
      ungroup() %>%
      pivot_wider(id_cols="OG_ID", names_from="Dup_status", values_from="total", values_fill=0) %>%
      #### Add column if missing
      add_cols(c("Dup", "NO_Dup")) %>%
      ####
      pivot_longer(cols=c("Dup", "NO_Dup"), names_to = "Dup_status") %>%
      group_by(OG_ID) %>%
      mutate(freq=value/sum(value)) %>%
      mutate(freq = replace_na(freq, 0))
  
    collapsed_final_df = final_ancestor_df %>%
      group_by(Dup_status) %>%
      summarise(freq_proportion = mean(freq)) %>%
      mutate(Node = ancestor) %>%
      dplyr::select(Node, Dup_status, freq_proportion)
    
    rand_foreground_df = rbind(rand_foreground_df, collapsed_final_df)
  }
  
  ####### Order:
  rand_foreground_df$Node = factor(rand_foreground_df$Node, levels=c(deuterostoma_y_labels[deuterostoma_y_labels!="Bilateria"], "Bilateria",
                                                               rev(protostoma_y_labels[protostoma_y_labels!="Bilateria"])))
  rand_foreground_df$Dup_status = factor(rand_foreground_df$Dup_status, levels=rev(c("Dup", "NO_Dup")))
  
  ###### Add branch info
  rand_foreground_df$Branch = rep("Deuterostoma", nrow(rand_foreground_df))
  rand_foreground_df$Branch[rand_foreground_df$Node %in% protostoma_y_labels] = "Protostoma"
  rand_foreground_df$Branch[rand_foreground_df$Node=="Bilateria"] = "Bilateria"
  rand_foreground_df = rand_foreground_df %>%
    unite("Branch_Dup_status", c(Branch,Dup_status), remove=FALSE)
  rand_foreground_df$Branch_Dup_status = factor(rand_foreground_df$Branch_Dup_status, levels=rev(c("Deuterostoma_Dup", "Deuterostoma_NO_Dup", "Protostoma_Dup", "Protostoma_NO_Dup", "Bilateria_Dup", "Bilateria_NO_Dup")))
  
  ###### Add randomization round
  rand_foreground_df$Randomization_round = rep(randomization_round, nrow(rand_foreground_df))
  
  ###### join to final df
  all_rand_foregrounds_df = rbind(all_rand_foregrounds_df, as.data.frame(rand_foreground_df))
}

all_rand_foregrounds_dup_df = all_rand_foregrounds_df %>%
  filter(Dup_status=="Dup")

#########################################################
############## PLOT #####################################
#########################################################

ancestral_gains_duplication_status = ggplot() +
  geom_bar(data=subset(foreground_df, Node!="Bilateria"), 
           aes(x=Node, y=freq_proportion, fill=Branch_Dup_status), stat="identity", position="fill", color="black", alpha=0.75) +
  geom_errorbar(data=subset(dup_background_df, Node!="Bilateria"), 
                            aes(x=Node, ymin=freq_proportion, ymax=freq_proportion), color="black", linetype="dashed") +
  ###############
  geom_boxplot(data=all_rand_foregrounds_dup_df, aes(x=Node, y=freq_proportion), color="black", fill="black", size=0.5, alpha=0.4, outlier.size = 0.5) +
  ###############
  scale_fill_manual(values=c("Deuterostoma_Dup"="darkolivegreen4", "Protostoma_Dup"="tan3", "Bilateria_Dup"="firebrick1", "Deuterostoma_NO_Dup"="white", "Protostoma_NO_Dup"="white", "Bilateria_NO_Dup"="white")) +
  ggtitle("Proportion of TS gains orthogroups with duplications") +
  theme_bw() +
  theme(axis.text.y = element_text(color="black"),
        axis.title = element_blank(),
        axis.text.x = element_text(color="black", angle=45, hjust=1, vjust=1),
        axis.ticks.y = element_line(),
        plot.title = element_text(color="black", hjust=0.5, face="bold"))

print(ancestral_gains_duplication_status)
```

```{r duplication_levels_randomized_call_ancestors_OTHER_SPECIES, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions")}
#Upload orthogroups and get duplication info about each gene
orthogroups_df = read.delim(paste0(ts_gains_losses_raw_dir, "Bilateria_conserved_orthogroups-All_genes-associated_tissues.tab"),
                            header=FALSE, col.names=c("OG_ID", "Species", "GeneID", "Tau", "Tissue", "Paralog_tissue", "Expr_cutoff"), stringsAsFactors = FALSE)
#### Add column with OG_ID_Species and count number of paralogs per gene
orthogroups_df$OG_ID_Species = paste0(orthogroups_df$OG_ID, "_", orthogroups_df$Species)
species_paralogs_count_df = as.data.frame(table(orthogroups_df$OG_ID_Species))
OG_ID_Species_paralogs_counts_dict = hashmap(as.vector(species_paralogs_count_df$Var1), as.vector(species_paralogs_count_df$Freq))
orthogroups_df$Paralogs_count = OG_ID_Species_paralogs_counts_dict$find(orthogroups_df$OG_ID_Species)

#########################################################
############## GENERATE BACKGROUND ######################
#########################################################
#### For each node (group of species) compute the proportion in each orthogroup that is duplicated. Get the difference for non-duplicated.
#Sum all and get proportions.
background_df = vector()
for (ancestor in all_ancestors[!(all_ancestors %in% c("Deuterostoma", "Protostoma"))]) {
  ###### Select species on the same branch WITHOUT tissue-specificity
  if (ancestor %in% deuterostoma_y_labels) {my_species = Deuterostoma[!(Deuterostoma %in% get(ancestor))]}
  if (ancestor %in% protostoma_y_labels) {my_species = Protostoma[!(Protostoma %in% get(ancestor))]}
  #Subset all orthogroups to only these species
  my_species_orthogroups_df = subset(orthogroups_df, Species %in% my_species)
  #Compute for each orthogroup the proportion of duplicated species (among the ones with conserved genes)
   final_ancestor_df = my_species_orthogroups_df %>%
      dplyr::select(OG_ID, Species, Paralogs_count) %>%
      distinct() %>%
      mutate(Dup_status=case_when(Paralogs_count > 1 ~ "Dup",
                                  Paralogs_count <= 1 ~ "NO_Dup")) %>%
      group_by(OG_ID, Dup_status) %>%
      summarise(total = n()) %>%
      ungroup() %>%
      pivot_wider(id_cols="OG_ID", names_from="Dup_status", values_from="total", values_fill=0) %>%
      pivot_longer(cols=c("Dup", "NO_Dup"), names_to = "Dup_status") %>%
      group_by(OG_ID) %>%
      mutate(freq=value/sum(value)) %>%
      mutate(freq = replace_na(freq, 0))
  
  collapsed_final_df = final_ancestor_df %>%
    group_by(Dup_status) %>%
    summarise(freq_proportion = mean(freq)) %>%
    mutate(Node = ancestor) %>%
    dplyr::select(Node, Dup_status, freq_proportion)
  
  background_df = rbind(background_df, collapsed_final_df)
}
#### Order levels
background_df$Node = factor(background_df$Node, levels=c(deuterostoma_y_labels[deuterostoma_y_labels!="Bilateria"], "Bilateria",
                                                             rev(protostoma_y_labels[protostoma_y_labels!="Bilateria"])))
background_df$Dup_status = factor(background_df$Dup_status, levels=rev(c("Dup", "NO_Dup")))
#### Subset only to duplicates
dup_background_df = subset(background_df, Dup_status=="Dup")

#########################################################
############## GENERATE FOREGROUND ######################
#########################################################

#############################
#### Upload TS gains ########
#############################

tissue_gains_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  tissue_gains_df = rbind(tissue_gains_df, input_df)
}

#############################
#### Upload TS losses #######
#############################
tissue_losses_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_losses.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_losses", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  input_df$OG_ID_loss = paste0(input_df$OG_ID, "_", input_df$TS_losses)
  tissue_losses_df = rbind(tissue_losses_df, input_df)
}

#############################
#### Select relative OGs ####
#############################
foreground_df = vector()
for (ancestor in all_ancestors[!(all_ancestors %in% c("Deuterostoma", "Protostoma"))]) {
  #### Select species on the same branch WITHOUT tissue-specificity
  if (ancestor %in% deuterostoma_y_labels) {my_species = Deuterostoma[!(Deuterostoma %in% get(ancestor))]}
  if (ancestor %in% protostoma_y_labels) {my_species = Protostoma[!(Protostoma %in% get(ancestor))]}
  my_species_orthogroups_df = vector()
  #### Cycle on each tissue
  for (tissue in ordered_tissues) {
    ancestor_gains_OGs = unique(as.vector(subset(tissue_gains_df, TS_gains==ancestor & Tissue==tissue)$OG_ID))
    #### Subset all orthogroups to only these species in the orthogroups with gains
    my_species_tissue_orthogroups_df = subset(orthogroups_df, Species %in% my_species & OG_ID %in% ancestor_gains_OGs)
    ancestor_losses_OGs = unique(as.vector(subset(tissue_losses_df, Tissue==tissue & TS_losses %in% my_species)$OG_ID_loss))
    #### ADD SPECIES WITH LOSSES
    my_species_tissue_orthogroups_df = my_species_tissue_orthogroups_df %>%
      dplyr::bind_rows(orthogroups_df %>% filter(OG_ID_Species %in% ancestor_losses_OGs))
    #### Join to final dataframe
    my_species_orthogroups_df = rbind(my_species_orthogroups_df, my_species_tissue_orthogroups_df)
  }
  #Compute the same statistics as above
    final_ancestor_df = my_species_orthogroups_df %>%
      dplyr::select(OG_ID, Species, Paralogs_count) %>%
      distinct() %>%
      mutate(Dup_status=case_when(Paralogs_count > 1 ~ "Dup",
                                  Paralogs_count <= 1 ~ "NO_Dup")) %>%
      group_by(OG_ID, Dup_status) %>%
      summarise(total = n()) %>%
      ungroup() %>%
      pivot_wider(id_cols="OG_ID", names_from="Dup_status", values_from="total", values_fill=0) %>%
      pivot_longer(cols=c("Dup", "NO_Dup"), names_to = "Dup_status") %>%
      group_by(OG_ID) %>%
      mutate(freq=value/sum(value)) %>%
      mutate(freq = replace_na(freq, 0))
  
  collapsed_final_df = final_ancestor_df %>%
    group_by(Dup_status) %>%
    summarise(freq_proportion = mean(freq)) %>%
    mutate(Node = ancestor) %>%
    dplyr::select(Node, Dup_status, freq_proportion)
  
  foreground_df = rbind(foreground_df, collapsed_final_df)
}

#Select only the TS species (species without losses) and compute the proportion of those species that are duplicated.
#Compute the difference as not-duplicated (it has to sum to 1).

####### Order:
foreground_df$Node = factor(foreground_df$Node, levels=c(deuterostoma_y_labels[deuterostoma_y_labels!="Bilateria"], "Bilateria",
                                                             rev(protostoma_y_labels[protostoma_y_labels!="Bilateria"])))
foreground_df$Dup_status = factor(foreground_df$Dup_status, levels=rev(c("Dup", "NO_Dup")))

###### Add branch info
foreground_df$Branch = rep("Deuterostoma", nrow(foreground_df))
foreground_df$Branch[foreground_df$Node %in% protostoma_y_labels] = "Protostoma"
foreground_df$Branch[foreground_df$Node=="Bilateria"] = "Bilateria"
foreground_df = foreground_df %>%
  unite("Branch_Dup_status", c(Branch,Dup_status), remove=FALSE)
foreground_df$Branch_Dup_status = factor(foreground_df$Branch_Dup_status, levels=rev(c("Deuterostoma_Dup", "Deuterostoma_NO_Dup", "Protostoma_Dup", "Protostoma_NO_Dup", "Bilateria_Dup", "Bilateria_NO_Dup")))


#########################################################
##### GENERATE FOREGROUND RANDOMIZATIONS ################
#########################################################
all_rand_foregrounds_df = vector()

for (randomization_round in seq(1,10)) {
  ######### Set directory
  randomization_dir = paste0(home, "projects/bilaterian_GE/data/ts_gains_losses_revision/randomization_", randomization_round, "/")
  
  ######### Upload TS gains
  tissue_gains_df = vector()
  for (tissue in ordered_tissues) {
    input_df = read.delim(paste0(randomization_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
    input_df$Tissue = rep(tissue, nrow(input_df))
    tissue_gains_df = rbind(tissue_gains_df, input_df)
  }
  ######### Upload TS losses
  tissue_losses_df = vector()
  for (tissue in ordered_tissues) {
    input_df = read.delim(paste0(randomization_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_losses.tab"), header=TRUE,
                          stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_losses", "Category"))
    input_df$Tissue = rep(tissue, nrow(input_df))
    input_df$OG_ID_loss = paste0(input_df$OG_ID, "_", input_df$TS_losses)
    tissue_losses_df = rbind(tissue_losses_df, input_df)
  }
  
  ######### Build foreground
  rand_foreground_df = vector()
  for (ancestor in all_ancestors[!(all_ancestors %in% c("Deuterostoma", "Protostoma"))]) {
    #### Select species on the same branch WITHOUT tissue-specificity
    if (ancestor %in% deuterostoma_y_labels) {my_species = Deuterostoma[!(Deuterostoma %in% get(ancestor))]}
    if (ancestor %in% protostoma_y_labels) {my_species = Protostoma[!(Protostoma %in% get(ancestor))]}
    my_species_orthogroups_df = vector()
    #### Cycle on each tissue
    for (tissue in ordered_tissues) {
      ancestor_gains_OGs = unique(as.vector(subset(tissue_gains_df, TS_gains==ancestor & Tissue==tissue)$OG_ID))
      #### Subset all orthogroups to only these species in the orthogroups with gains
      my_species_tissue_orthogroups_df = subset(orthogroups_df, Species %in% my_species & OG_ID %in% ancestor_gains_OGs)
      ancestor_losses_OGs = unique(as.vector(subset(tissue_losses_df, Tissue==tissue & TS_losses %in% my_species)$OG_ID_loss))
      #### ADD SPECIES WITH LOSSES ON THAT SIDE
      my_species_tissue_orthogroups_df = my_species_tissue_orthogroups_df %>%
        dplyr::bind_rows(orthogroups_df %>% filter(OG_ID_Species %in% ancestor_losses_OGs))
      #### Join to final dataframe
      my_species_orthogroups_df = rbind(my_species_orthogroups_df, my_species_tissue_orthogroups_df)
    }
    ##############
    #If there are no gains at all respecting the conditions, go to the next iteration
    if (nrow(my_species_orthogroups_df)==0) {next}
    ##############
    #Compute the same statistics as above
    final_ancestor_df = my_species_orthogroups_df %>%
      dplyr::select(OG_ID, Species, Paralogs_count) %>%
      distinct() %>%
      mutate(Dup_status=case_when(Paralogs_count > 1 ~ "Dup",
                                  Paralogs_count <= 1 ~ "NO_Dup")) %>%
      group_by(OG_ID, Dup_status) %>%
      summarise(total = n()) %>%
      ungroup() %>%
      pivot_wider(id_cols="OG_ID", names_from="Dup_status", values_from="total", values_fill=0) %>%
      #### Add column if missing
      add_cols(c("Dup", "NO_Dup")) %>%
      ####
      pivot_longer(cols=c("Dup", "NO_Dup"), names_to = "Dup_status") %>%
      group_by(OG_ID) %>%
      mutate(freq=value/sum(value)) %>%
      mutate(freq = replace_na(freq, 0))
  
    collapsed_final_df = final_ancestor_df %>%
      group_by(Dup_status) %>%
      summarise(freq_proportion = mean(freq)) %>%
      mutate(Node = ancestor) %>%
      dplyr::select(Node, Dup_status, freq_proportion)
    
    rand_foreground_df = rbind(rand_foreground_df, collapsed_final_df)
  }
  
  ####### Order:
  rand_foreground_df$Node = factor(rand_foreground_df$Node, levels=c(deuterostoma_y_labels[deuterostoma_y_labels!="Bilateria"], "Bilateria",
                                                               rev(protostoma_y_labels[protostoma_y_labels!="Bilateria"])))
  rand_foreground_df$Dup_status = factor(rand_foreground_df$Dup_status, levels=rev(c("Dup", "NO_Dup")))
  
  ###### Add branch info
  rand_foreground_df$Branch = rep("Deuterostoma", nrow(rand_foreground_df))
  rand_foreground_df$Branch[rand_foreground_df$Node %in% protostoma_y_labels] = "Protostoma"
  rand_foreground_df$Branch[rand_foreground_df$Node=="Bilateria"] = "Bilateria"
  rand_foreground_df = rand_foreground_df %>%
    unite("Branch_Dup_status", c(Branch,Dup_status), remove=FALSE)
  rand_foreground_df$Branch_Dup_status = factor(rand_foreground_df$Branch_Dup_status, levels=rev(c("Deuterostoma_Dup", "Deuterostoma_NO_Dup", "Protostoma_Dup", "Protostoma_NO_Dup", "Bilateria_Dup", "Bilateria_NO_Dup")))
  
  ###### Add randomization round
  rand_foreground_df$Randomization_round = rep(randomization_round, nrow(rand_foreground_df))
  
  ###### join to final df
  all_rand_foregrounds_df = rbind(all_rand_foregrounds_df, as.data.frame(rand_foreground_df))
}

all_rand_foregrounds_dup_df = all_rand_foregrounds_df %>%
  filter(Dup_status=="Dup")

#########################################################
############## PLOT #####################################
#########################################################

ancestral_gains_duplication_status = ggplot() +
  geom_bar(data=subset(foreground_df, Node!="Bilateria"), 
           aes(x=Node, y=freq_proportion, fill=Branch_Dup_status), stat="identity", position="fill", color="black", alpha=0.75) +
  geom_errorbar(data=subset(dup_background_df, Node!="Bilateria"), 
                            aes(x=Node, ymin=freq_proportion, ymax=freq_proportion), color="black", linetype="dashed") +
  ###############
  geom_boxplot(data=all_rand_foregrounds_dup_df, aes(x=Node, y=freq_proportion), color="black", fill="black", size=0.5, alpha=0.4, outlier.size = 0.5) +
  ###############
  scale_fill_manual(values=c("Deuterostoma_Dup"="darkolivegreen4", "Protostoma_Dup"="tan3", "Bilateria_Dup"="firebrick1", "Deuterostoma_NO_Dup"="white", "Protostoma_NO_Dup"="white", "Bilateria_NO_Dup"="white")) +
  ggtitle("Proportion of TS gains orthogroups with duplications") +
  theme_bw() +
  theme(axis.text.y = element_text(color="black"),
        axis.title = element_blank(),
        axis.text.x = element_text(color="black", angle=45, hjust=1, vjust=1),
        axis.ticks.y = element_line(),
        plot.title = element_text(color="black", hjust=0.5, face="bold"))

print(ancestral_gains_duplication_status)
```

```{r duplication_levels_REAL_vs_RANDOMIZED_proportions, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions"), fig.width=7, fig.height=9}
#Upload orthogroups and get duplication info about each gene
orthogroups_df = read.delim(paste0(ts_gains_losses_raw_dir, "Bilateria_conserved_orthogroups-All_genes-associated_tissues.tab"),
                            header=FALSE, col.names=c("OG_ID", "Species", "GeneID", "Tau", "Tissue", "Paralog_tissue", "Expr_cutoff"), stringsAsFactors = FALSE)
#### Add column with OG_ID_Species and count number of paralogs per gene
orthogroups_df$OG_ID_Species = paste0(orthogroups_df$OG_ID, "_", orthogroups_df$Species)
species_paralogs_count_df = as.data.frame(table(orthogroups_df$OG_ID_Species))
OG_ID_Species_paralogs_counts_dict = hashmap(as.vector(species_paralogs_count_df$Var1), as.vector(species_paralogs_count_df$Freq))
orthogroups_df$Paralogs_count = OG_ID_Species_paralogs_counts_dict$find(orthogroups_df$OG_ID_Species)

##################################################################################################################
############## GENERATE NON-SPECIES ##############################################################################
##################################################################################################################

#############################
#### Upload TS gains ########
#############################

tissue_gains_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  tissue_gains_df = rbind(tissue_gains_df, input_df)
}

#############################
#### Upload TS losses #######
#############################
tissue_losses_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_losses.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_losses", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  input_df$OG_ID_loss = paste0(input_df$OG_ID, "_", input_df$TS_losses)
  tissue_losses_df = rbind(tissue_losses_df, input_df)
}

#############################
#### Select relative OGs ####
#############################
foreground_df = vector()
for (ancestor in all_ancestors[!(all_ancestors %in% c("Deuterostoma", "Protostoma"))]) {
  #### Select species on the same branch WITHOUT tissue-specificity
  if (ancestor %in% deuterostoma_y_labels) {my_species = Deuterostoma[!(Deuterostoma %in% get(ancestor))]}
  if (ancestor %in% protostoma_y_labels) {my_species = Protostoma[!(Protostoma %in% get(ancestor))]}
  my_species_orthogroups_df = vector()
  #### Cycle on each tissue
  for (tissue in ordered_tissues) {
    ancestor_gains_OGs = unique(as.vector(subset(tissue_gains_df, TS_gains==ancestor & Tissue==tissue)$OG_ID))
    #### Subset all orthogroups to only these species in the orthogroups with gains
    my_species_tissue_orthogroups_df = subset(orthogroups_df, Species %in% my_species & OG_ID %in% ancestor_gains_OGs)
    ancestor_losses_OGs = unique(as.vector(subset(tissue_losses_df, Tissue==tissue & TS_losses %in% my_species)$OG_ID_loss))
    #### ADD SPECIES WITH LOSSES
    my_species_tissue_orthogroups_df = my_species_tissue_orthogroups_df %>%
      dplyr::bind_rows(orthogroups_df %>% filter(OG_ID_Species %in% ancestor_losses_OGs))
    #### Join to final dataframe
    my_species_orthogroups_df = rbind(my_species_orthogroups_df, my_species_tissue_orthogroups_df)
  }
  #Compute the same statistics as above
    final_ancestor_df = my_species_orthogroups_df %>%
      dplyr::select(OG_ID, Species, Paralogs_count) %>%
      distinct() %>%
      mutate(Dup_status=case_when(Paralogs_count > 1 ~ "Dup",
                                  Paralogs_count <= 1 ~ "NO_Dup")) %>%
      group_by(OG_ID, Dup_status) %>%
      summarise(total = n()) %>%
      ungroup() %>%
      pivot_wider(id_cols="OG_ID", names_from="Dup_status", values_from="total", values_fill=0) %>%
      pivot_longer(cols=c("Dup", "NO_Dup"), names_to = "Dup_status") %>%
      group_by(OG_ID) %>%
      mutate(freq=value/sum(value)) %>%
      mutate(freq = replace_na(freq, 0)) %>%
      mutate(Node = ancestor)
  
  foreground_df = rbind(foreground_df, as.data.frame(final_ancestor_df))
}

#Select only the TS species (species without losses) and compute the proportion of those species that are duplicated.
#Compute the difference as not-duplicated (it has to sum to 1).

####### Order:
foreground_df$Node = factor(foreground_df$Node, levels=c(deuterostoma_y_labels[deuterostoma_y_labels!="Bilateria"], "Bilateria",
                                                             rev(protostoma_y_labels[protostoma_y_labels!="Bilateria"])))
foreground_df$Dup_status = factor(foreground_df$Dup_status, levels=rev(c("Dup", "NO_Dup")))

###### Add branch info
foreground_df$Branch = rep("Deuterostoma", nrow(foreground_df))
foreground_df$Branch[foreground_df$Node %in% protostoma_y_labels] = "Protostoma"
foreground_df$Branch[foreground_df$Node=="Bilateria"] = "Bilateria"
foreground_df = foreground_df %>%
  unite("Branch_Dup_status", c(Branch,Dup_status), remove=FALSE)
foreground_df$Branch_Dup_status = factor(foreground_df$Branch_Dup_status, levels=rev(c("Deuterostoma_Dup", "Deuterostoma_NO_Dup", "Protostoma_Dup", "Protostoma_NO_Dup", "Bilateria_Dup", "Bilateria_NO_Dup")))

NON_SPECIES_foreground_dup_df = foreground_df

#########################################################
##### GENERATE FOREGROUND RANDOMIZATIONS ################
#########################################################
all_rand_foregrounds_df = vector()

for (randomization_round in seq(1,10)) {
  ######### Set directory
  randomization_dir = paste0(home, "projects/bilaterian_GE/data/ts_gains_losses_revision/randomization_", randomization_round, "/")
  
  ######### Upload TS gains
  tissue_gains_df = vector()
  for (tissue in ordered_tissues) {
    input_df = read.delim(paste0(randomization_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
    input_df$Tissue = rep(tissue, nrow(input_df))
    tissue_gains_df = rbind(tissue_gains_df, input_df)
  }
  ######### Upload TS losses
  tissue_losses_df = vector()
  for (tissue in ordered_tissues) {
    input_df = read.delim(paste0(randomization_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_losses.tab"), header=TRUE,
                          stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_losses", "Category"))
    input_df$Tissue = rep(tissue, nrow(input_df))
    input_df$OG_ID_loss = paste0(input_df$OG_ID, "_", input_df$TS_losses)
    tissue_losses_df = rbind(tissue_losses_df, input_df)
  }
  
  ######### Build foreground
  rand_foreground_df = vector()
  for (ancestor in all_ancestors[!(all_ancestors %in% c("Deuterostoma", "Protostoma"))]) {
    #### Select species on the same branch WITHOUT tissue-specificity
    if (ancestor %in% deuterostoma_y_labels) {my_species = Deuterostoma[!(Deuterostoma %in% get(ancestor))]}
    if (ancestor %in% protostoma_y_labels) {my_species = Protostoma[!(Protostoma %in% get(ancestor))]}
    my_species_orthogroups_df = vector()
    #### Cycle on each tissue
    for (tissue in ordered_tissues) {
      ancestor_gains_OGs = unique(as.vector(subset(tissue_gains_df, TS_gains==ancestor & Tissue==tissue)$OG_ID))
      #### Subset all orthogroups to only these species in the orthogroups with gains
      my_species_tissue_orthogroups_df = subset(orthogroups_df, Species %in% my_species & OG_ID %in% ancestor_gains_OGs)
      ancestor_losses_OGs = unique(as.vector(subset(tissue_losses_df, Tissue==tissue & TS_losses %in% my_species)$OG_ID_loss))
      #### ADD SPECIES WITH LOSSES ON THAT SIDE
      my_species_tissue_orthogroups_df = my_species_tissue_orthogroups_df %>%
        dplyr::bind_rows(orthogroups_df %>% filter(OG_ID_Species %in% ancestor_losses_OGs))
      #### Join to final dataframe
      my_species_orthogroups_df = rbind(my_species_orthogroups_df, my_species_tissue_orthogroups_df)
    }
    ##############
    #If there are no gains at all respecting the conditions, go to the next iteration
    if (nrow(my_species_orthogroups_df)==0) {next}
    ##############
    #Compute the same statistics as above
    final_ancestor_df = my_species_orthogroups_df %>%
      dplyr::select(OG_ID, Species, Paralogs_count) %>%
      distinct() %>%
      mutate(Dup_status=case_when(Paralogs_count > 1 ~ "Dup",
                                  Paralogs_count <= 1 ~ "NO_Dup")) %>%
      group_by(OG_ID, Dup_status) %>%
      summarise(total = n()) %>%
      ungroup() %>%
      pivot_wider(id_cols="OG_ID", names_from="Dup_status", values_from="total", values_fill=0) %>%
      #### Add column if missing
      add_cols(c("Dup", "NO_Dup")) %>%
      ####
      pivot_longer(cols=c("Dup", "NO_Dup"), names_to = "Dup_status") %>%
      group_by(OG_ID) %>%
      mutate(freq=value/sum(value)) %>%
      mutate(freq = replace_na(freq, 0)) %>%
      mutate(Node = ancestor)
    
    rand_foreground_df = rbind(rand_foreground_df, as.data.frame(final_ancestor_df))
  }
  
  ####### Order:
  rand_foreground_df$Node = factor(rand_foreground_df$Node, levels=c(deuterostoma_y_labels[deuterostoma_y_labels!="Bilateria"], "Bilateria",
                                                               rev(protostoma_y_labels[protostoma_y_labels!="Bilateria"])))
  rand_foreground_df$Dup_status = factor(rand_foreground_df$Dup_status, levels=rev(c("Dup", "NO_Dup")))
  
  ###### Add branch info
  rand_foreground_df$Branch = rep("Deuterostoma", nrow(rand_foreground_df))
  rand_foreground_df$Branch[rand_foreground_df$Node %in% protostoma_y_labels] = "Protostoma"
  rand_foreground_df$Branch[rand_foreground_df$Node=="Bilateria"] = "Bilateria"
  rand_foreground_df = rand_foreground_df %>%
    unite("Branch_Dup_status", c(Branch,Dup_status), remove=FALSE)
  rand_foreground_df$Branch_Dup_status = factor(rand_foreground_df$Branch_Dup_status, levels=rev(c("Deuterostoma_Dup", "Deuterostoma_NO_Dup", "Protostoma_Dup", "Protostoma_NO_Dup", "Bilateria_Dup", "Bilateria_NO_Dup")))
  
  ###### Add randomization round
  rand_foreground_df$Randomization_round = rep(randomization_round, nrow(rand_foreground_df))
  
  ###### join to final df
  all_rand_foregrounds_df = rbind(all_rand_foregrounds_df, as.data.frame(rand_foreground_df))
}

all_rand_foregrounds_dup_df = all_rand_foregrounds_df

NON_SPECIES_all_rand_foregrounds_dup_df = all_rand_foregrounds_dup_df
##################################################################################################################
############## GENERATE SPECIES ##################################################################################
##################################################################################################################

#############################
#### Upload TS gains ########
#############################

tissue_gains_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  tissue_gains_df = rbind(tissue_gains_df, input_df)
}

#############################
#### Upload TS losses #######
#############################
tissue_losses_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_losses.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_losses", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  input_df$OG_ID_loss = paste0(input_df$OG_ID, "_", input_df$TS_losses)
  tissue_losses_df = rbind(tissue_losses_df, input_df)
}

#############################
#### Select relative OGs ####
#############################
foreground_df = vector()
for (ancestor in all_ancestors) {
  #### Select ancestor species
  my_species = get(ancestor)
  my_species_orthogroups_df = vector()
  #### Cycle on each tissue
  for (tissue in ordered_tissues) {
    ancestor_gains_OGs = unique(as.vector(subset(tissue_gains_df, TS_gains==ancestor & Tissue==tissue)$OG_ID))
    #### Subset all orthogroups to only these species in the orthogroups with gains
    my_species_tissue_orthogroups_df = subset(orthogroups_df, Species %in% my_species & OG_ID %in% ancestor_gains_OGs)
    ancestor_losses_OGs = unique(as.vector(subset(tissue_losses_df, Tissue==tissue)$OG_ID_loss))
    #### Filter out the losses
    my_species_tissue_orthogroups_df = my_species_tissue_orthogroups_df %>%
      filter(!(OG_ID_Species %in% ancestor_losses_OGs))
    #### Join to final dataframe
    my_species_orthogroups_df = rbind(my_species_orthogroups_df, my_species_tissue_orthogroups_df)
  }
  #Compute the same statistics as above
    final_ancestor_df = my_species_orthogroups_df %>%
      dplyr::select(OG_ID, Species, Paralogs_count) %>%
      distinct() %>%
      mutate(Dup_status=case_when(Paralogs_count > 1 ~ "Dup",
                                  Paralogs_count <= 1 ~ "NO_Dup")) %>%
      group_by(OG_ID, Dup_status) %>%
      summarise(total = n()) %>%
      ungroup() %>%
      pivot_wider(id_cols="OG_ID", names_from="Dup_status", values_from="total", values_fill=0) %>%
      pivot_longer(cols=c("Dup", "NO_Dup"), names_to = "Dup_status") %>%
      group_by(OG_ID) %>%
      mutate(freq=value/sum(value)) %>%
      mutate(freq = replace_na(freq, 0)) %>%
      mutate(Node = ancestor)
    
  foreground_df = rbind(foreground_df, as.data.frame(final_ancestor_df))
}

#Select only the TS species (species without losses) and compute the proportion of those species that are duplicated.
#Compute the difference as not-duplicated (it has to sum to 1).

####### Order:
foreground_df$Node = factor(foreground_df$Node, levels=c(deuterostoma_y_labels[deuterostoma_y_labels!="Bilateria"], "Bilateria",
                                                             rev(protostoma_y_labels[protostoma_y_labels!="Bilateria"])))
foreground_df$Dup_status = factor(foreground_df$Dup_status, levels=rev(c("Dup", "NO_Dup")))

###### Add branch info
foreground_df$Branch = rep("Deuterostoma", nrow(foreground_df))
foreground_df$Branch[foreground_df$Node %in% protostoma_y_labels] = "Protostoma"
foreground_df$Branch[foreground_df$Node=="Bilateria"] = "Bilateria"
foreground_df = foreground_df %>%
  unite("Branch_Dup_status", c(Branch,Dup_status), remove=FALSE)
foreground_df$Branch_Dup_status = factor(foreground_df$Branch_Dup_status, levels=rev(c("Deuterostoma_Dup", "Deuterostoma_NO_Dup", "Protostoma_Dup", "Protostoma_NO_Dup", "Bilateria_Dup", "Bilateria_NO_Dup")))

SPECIES_foreground_dup_df = foreground_df

#########################################################
##### GENERATE FOREGROUND RANDOMIZATIONS ################
#########################################################
all_rand_foregrounds_df = vector()

for (randomization_round in seq(1,10)) {
  ######### Set directory
  randomization_dir = paste0(home, "projects/bilaterian_GE/data/ts_gains_losses_revision/randomization_", randomization_round, "/")
  
  ######### Upload TS gains
  tissue_gains_df = vector()
  for (tissue in ordered_tissues) {
    input_df = read.delim(paste0(randomization_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
    input_df$Tissue = rep(tissue, nrow(input_df))
    tissue_gains_df = rbind(tissue_gains_df, input_df)
  }
  ######### Upload TS losses
  tissue_losses_df = vector()
  for (tissue in ordered_tissues) {
    input_df = read.delim(paste0(randomization_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_losses.tab"), header=TRUE,
                          stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_losses", "Category"))
    input_df$Tissue = rep(tissue, nrow(input_df))
    input_df$OG_ID_loss = paste0(input_df$OG_ID, "_", input_df$TS_losses)
    tissue_losses_df = rbind(tissue_losses_df, input_df)
  }
  
  ######### Build foreground
  rand_foreground_df = vector()
  for (ancestor in all_ancestors) {
    #### Select ancestor species
    my_species = get(ancestor)
    my_species_orthogroups_df = vector()
    #### Cycle on each tissue
    for (tissue in ordered_tissues) {
      ancestor_gains_OGs = unique(as.vector(subset(tissue_gains_df, TS_gains==ancestor & Tissue==tissue)$OG_ID))
      #### Subset all orthogroups to only these species in the orthogroups with gains
      my_species_tissue_orthogroups_df = subset(orthogroups_df, Species %in% my_species & OG_ID %in% ancestor_gains_OGs)
      ancestor_losses_OGs = unique(as.vector(subset(tissue_losses_df, Tissue==tissue)$OG_ID_loss))
      #### Filter out the losses
      my_species_tissue_orthogroups_df = my_species_tissue_orthogroups_df %>%
        filter(!(OG_ID_Species %in% ancestor_losses_OGs))
      #### Join to final dataframe
      my_species_orthogroups_df = rbind(my_species_orthogroups_df, my_species_tissue_orthogroups_df)
    }
    ##############
    #If there are no gains at all respecting the conditions, go to the next iteration
    if (nrow(my_species_orthogroups_df)==0) {next}
    ##############
    #Compute the same statistics as above
    final_ancestor_df = my_species_orthogroups_df %>%
      dplyr::select(OG_ID, Species, Paralogs_count) %>%
      distinct() %>%
      mutate(Dup_status=case_when(Paralogs_count > 1 ~ "Dup",
                                  Paralogs_count <= 1 ~ "NO_Dup")) %>%
      group_by(OG_ID, Dup_status) %>%
      summarise(total = n()) %>%
      ungroup() %>%
      pivot_wider(id_cols="OG_ID", names_from="Dup_status", values_from="total", values_fill=0) %>%
      #### Add column if missing
      add_cols(c("Dup", "NO_Dup")) %>%
      ####
      pivot_longer(cols=c("Dup", "NO_Dup"), names_to = "Dup_status") %>%
      group_by(OG_ID) %>%
      mutate(freq=value/sum(value)) %>%
      mutate(freq = replace_na(freq, 0)) %>%
      mutate(Node = ancestor)
    
    rand_foreground_df = rbind(rand_foreground_df, as.data.frame(final_ancestor_df))
  }
  
  ####### Order:
  rand_foreground_df$Node = factor(rand_foreground_df$Node, levels=c(deuterostoma_y_labels[deuterostoma_y_labels!="Bilateria"], "Bilateria",
                                                               rev(protostoma_y_labels[protostoma_y_labels!="Bilateria"])))
  rand_foreground_df$Dup_status = factor(rand_foreground_df$Dup_status, levels=rev(c("Dup", "NO_Dup")))
  
  ###### Add branch info
  rand_foreground_df$Branch = rep("Deuterostoma", nrow(rand_foreground_df))
  rand_foreground_df$Branch[rand_foreground_df$Node %in% protostoma_y_labels] = "Protostoma"
  rand_foreground_df$Branch[rand_foreground_df$Node=="Bilateria"] = "Bilateria"
  rand_foreground_df = rand_foreground_df %>%
    unite("Branch_Dup_status", c(Branch,Dup_status), remove=FALSE)
  rand_foreground_df$Branch_Dup_status = factor(rand_foreground_df$Branch_Dup_status, levels=rev(c("Deuterostoma_Dup", "Deuterostoma_NO_Dup", "Protostoma_Dup", "Protostoma_NO_Dup", "Bilateria_Dup", "Bilateria_NO_Dup")))
  
  ###### Add randomization round
  rand_foreground_df$Randomization_round = rep(randomization_round, nrow(rand_foreground_df))
  
  ###### join to final df
  all_rand_foregrounds_df = rbind(all_rand_foregrounds_df, as.data.frame(rand_foreground_df))
}

all_rand_foregrounds_dup_df = all_rand_foregrounds_df

SPECIES_all_rand_foregrounds_dup_df = all_rand_foregrounds_dup_df

#########################################################
###### COMBINED ALL DATAFRAMES ##########################
#########################################################

final_df = dplyr::bind_rows(SPECIES_foreground_dup_df %>%
                              mutate(General_category="Observed",
                                     Species_category="TS"),
                            NON_SPECIES_foreground_dup_df %>%
                               mutate(General_category="Observed",
                                     Species_category="NO_TS"),
                            SPECIES_all_rand_foregrounds_dup_df %>%
                              mutate(General_category="Random",
                                     Species_category="TS"),
                            NON_SPECIES_all_rand_foregrounds_dup_df %>%
                              mutate(General_category="Random",
                                     Species_category="NO_TS")) %>%
  mutate(Node = factor(Node, levels=all_ancestors),
         Species_category = factor(Species_category, levels=c("TS", "NO_TS"))) %>%
  filter(!(Node %in% c("Deuterostoma", "Protostoma")),
         Dup_status=="Dup")

#########################################################
############## PLOT #####################################
#########################################################

TS_vs_NO_TS_plot = ggplot(data=final_df, aes(x=Species_category, y=freq, fill=Node, alpha=General_category)) +
  geom_boxplot(outlier.size = 0.5) +
  scale_fill_manual(values=all_ancestors_colors) +
  scale_alpha_manual(values=c("Observed"=0.9, "Random"=0.4)) +
  theme_bw() +
  theme(axis.title = element_text(color="black"),
        axis.text = element_text(color="black")) +
  guides(fill="none", alpha="none") +
  facet_nested_wrap(~Node+General_category)

print(TS_vs_NO_TS_plot)
```

```{r duplication_levels_REAL_vs_RANDOMIZED_numbers, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions"), fig.width=7, fig.height=12}
#Upload orthogroups and get duplication info about each gene
orthogroups_df = read.delim(paste0(ts_gains_losses_raw_dir, "Bilateria_conserved_orthogroups-All_genes-associated_tissues.tab"),
                            header=FALSE, col.names=c("OG_ID", "Species", "GeneID", "Tau", "Tissue", "Paralog_tissue", "Expr_cutoff"), stringsAsFactors = FALSE)
#### Add column with OG_ID_Species and count number of paralogs per gene
orthogroups_df$OG_ID_Species = paste0(orthogroups_df$OG_ID, "_", orthogroups_df$Species)
species_paralogs_count_df = as.data.frame(table(orthogroups_df$OG_ID_Species))
OG_ID_Species_paralogs_counts_dict = hashmap(as.vector(species_paralogs_count_df$Var1), as.vector(species_paralogs_count_df$Freq))
orthogroups_df$Paralogs_count = OG_ID_Species_paralogs_counts_dict$find(orthogroups_df$OG_ID_Species)

##################################################################################################################
############## GENERATE NON-SPECIES ##############################################################################
##################################################################################################################

#############################
#### Upload TS gains ########
#############################

tissue_gains_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  tissue_gains_df = rbind(tissue_gains_df, input_df)
}

#############################
#### Upload TS losses #######
#############################
tissue_losses_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_losses.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_losses", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  input_df$OG_ID_loss = paste0(input_df$OG_ID, "_", input_df$TS_losses)
  tissue_losses_df = rbind(tissue_losses_df, input_df)
}

#############################
#### Select relative OGs ####
#############################
foreground_df = vector()
for (ancestor in all_ancestors[!(all_ancestors %in% c("Deuterostoma", "Protostoma"))]) {
  #### Select species on the same branch WITHOUT tissue-specificity
  if (ancestor %in% deuterostoma_y_labels) {my_species = Deuterostoma[!(Deuterostoma %in% get(ancestor))]}
  if (ancestor %in% protostoma_y_labels) {my_species = Protostoma[!(Protostoma %in% get(ancestor))]}
  my_species_orthogroups_df = vector()
  #### Cycle on each tissue
  for (tissue in ordered_tissues) {
    ancestor_gains_OGs = unique(as.vector(subset(tissue_gains_df, TS_gains==ancestor & Tissue==tissue)$OG_ID))
    #### Subset all orthogroups to only these species in the orthogroups with gains
    my_species_tissue_orthogroups_df = subset(orthogroups_df, Species %in% my_species & OG_ID %in% ancestor_gains_OGs)
    ancestor_losses_OGs = unique(as.vector(subset(tissue_losses_df, Tissue==tissue & TS_losses %in% my_species)$OG_ID_loss))
    #### ADD SPECIES WITH LOSSES
    my_species_tissue_orthogroups_df = my_species_tissue_orthogroups_df %>%
      dplyr::bind_rows(orthogroups_df %>% filter(OG_ID_Species %in% ancestor_losses_OGs))
    #### Join to final dataframe
    my_species_orthogroups_df = rbind(my_species_orthogroups_df, my_species_tissue_orthogroups_df)
  }
  #Compute the same statistics as above
    final_ancestor_df = my_species_orthogroups_df %>%
      dplyr::select(OG_ID, Species, Paralogs_count) %>%
      distinct() %>%
      mutate(Dup_status=case_when(Paralogs_count > 1 ~ "Dup",
                                  Paralogs_count <= 1 ~ "NO_Dup")) %>%
      mutate(Node = ancestor)
  
  foreground_df = rbind(foreground_df, as.data.frame(final_ancestor_df))
}

#Select only the TS species (species without losses) and compute the proportion of those species that are duplicated.
#Compute the difference as not-duplicated (it has to sum to 1).

####### Order:
foreground_df$Node = factor(foreground_df$Node, levels=c(deuterostoma_y_labels[deuterostoma_y_labels!="Bilateria"], "Bilateria",
                                                             rev(protostoma_y_labels[protostoma_y_labels!="Bilateria"])))
foreground_df$Dup_status = factor(foreground_df$Dup_status, levels=rev(c("Dup", "NO_Dup")))

###### Add branch info
foreground_df$Branch = rep("Deuterostoma", nrow(foreground_df))
foreground_df$Branch[foreground_df$Node %in% protostoma_y_labels] = "Protostoma"
foreground_df$Branch[foreground_df$Node=="Bilateria"] = "Bilateria"
foreground_df = foreground_df %>%
  unite("Branch_Dup_status", c(Branch,Dup_status), remove=FALSE)
foreground_df$Branch_Dup_status = factor(foreground_df$Branch_Dup_status, levels=rev(c("Deuterostoma_Dup", "Deuterostoma_NO_Dup", "Protostoma_Dup", "Protostoma_NO_Dup", "Bilateria_Dup", "Bilateria_NO_Dup")))

NON_SPECIES_foreground_dup_df = foreground_df

#########################################################
##### GENERATE FOREGROUND RANDOMIZATIONS ################
#########################################################
all_rand_foregrounds_df = vector()

for (randomization_round in seq(1,10)) {
  ######### Set directory
  randomization_dir = paste0(home, "projects/bilaterian_GE/data/ts_gains_losses_revision/randomization_", randomization_round, "/")
  
  ######### Upload TS gains
  tissue_gains_df = vector()
  for (tissue in ordered_tissues) {
    input_df = read.delim(paste0(randomization_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
    input_df$Tissue = rep(tissue, nrow(input_df))
    tissue_gains_df = rbind(tissue_gains_df, input_df)
  }
  ######### Upload TS losses
  tissue_losses_df = vector()
  for (tissue in ordered_tissues) {
    input_df = read.delim(paste0(randomization_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_losses.tab"), header=TRUE,
                          stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_losses", "Category"))
    input_df$Tissue = rep(tissue, nrow(input_df))
    input_df$OG_ID_loss = paste0(input_df$OG_ID, "_", input_df$TS_losses)
    tissue_losses_df = rbind(tissue_losses_df, input_df)
  }
  
  ######### Build foreground
  rand_foreground_df = vector()
  for (ancestor in all_ancestors[!(all_ancestors %in% c("Deuterostoma", "Protostoma"))]) {
    #### Select species on the same branch WITHOUT tissue-specificity
    if (ancestor %in% deuterostoma_y_labels) {my_species = Deuterostoma[!(Deuterostoma %in% get(ancestor))]}
    if (ancestor %in% protostoma_y_labels) {my_species = Protostoma[!(Protostoma %in% get(ancestor))]}
    my_species_orthogroups_df = vector()
    #### Cycle on each tissue
    for (tissue in ordered_tissues) {
      ancestor_gains_OGs = unique(as.vector(subset(tissue_gains_df, TS_gains==ancestor & Tissue==tissue)$OG_ID))
      #### Subset all orthogroups to only these species in the orthogroups with gains
      my_species_tissue_orthogroups_df = subset(orthogroups_df, Species %in% my_species & OG_ID %in% ancestor_gains_OGs)
      ancestor_losses_OGs = unique(as.vector(subset(tissue_losses_df, Tissue==tissue & TS_losses %in% my_species)$OG_ID_loss))
      #### ADD SPECIES WITH LOSSES ON THAT SIDE
      my_species_tissue_orthogroups_df = my_species_tissue_orthogroups_df %>%
        dplyr::bind_rows(orthogroups_df %>% filter(OG_ID_Species %in% ancestor_losses_OGs))
      #### Join to final dataframe
      my_species_orthogroups_df = rbind(my_species_orthogroups_df, my_species_tissue_orthogroups_df)
    }
    ##############
    #If there are no gains at all respecting the conditions, go to the next iteration
    if (nrow(my_species_orthogroups_df)==0) {next}
    ##############
    #Compute the same statistics as above
    final_ancestor_df = my_species_orthogroups_df %>%
      dplyr::select(OG_ID, Species, Paralogs_count) %>%
      distinct() %>%
      mutate(Dup_status=case_when(Paralogs_count > 1 ~ "Dup",
                                  Paralogs_count <= 1 ~ "NO_Dup")) %>%
      mutate(Node = ancestor)
    
    rand_foreground_df = rbind(rand_foreground_df, as.data.frame(final_ancestor_df))
  }
  
  ####### Order:
  rand_foreground_df$Node = factor(rand_foreground_df$Node, levels=c(deuterostoma_y_labels[deuterostoma_y_labels!="Bilateria"], "Bilateria",
                                                               rev(protostoma_y_labels[protostoma_y_labels!="Bilateria"])))
  rand_foreground_df$Dup_status = factor(rand_foreground_df$Dup_status, levels=rev(c("Dup", "NO_Dup")))
  
  ###### Add branch info
  rand_foreground_df$Branch = rep("Deuterostoma", nrow(rand_foreground_df))
  rand_foreground_df$Branch[rand_foreground_df$Node %in% protostoma_y_labels] = "Protostoma"
  rand_foreground_df$Branch[rand_foreground_df$Node=="Bilateria"] = "Bilateria"
  rand_foreground_df = rand_foreground_df %>%
    unite("Branch_Dup_status", c(Branch,Dup_status), remove=FALSE)
  rand_foreground_df$Branch_Dup_status = factor(rand_foreground_df$Branch_Dup_status, levels=rev(c("Deuterostoma_Dup", "Deuterostoma_NO_Dup", "Protostoma_Dup", "Protostoma_NO_Dup", "Bilateria_Dup", "Bilateria_NO_Dup")))
  
  ###### Add randomization round
  rand_foreground_df$Randomization_round = rep(randomization_round, nrow(rand_foreground_df))
  
  ###### join to final df
  all_rand_foregrounds_df = rbind(all_rand_foregrounds_df, as.data.frame(rand_foreground_df))
}

all_rand_foregrounds_dup_df = all_rand_foregrounds_df

NON_SPECIES_all_rand_foregrounds_dup_df = all_rand_foregrounds_dup_df
##################################################################################################################
############## GENERATE SPECIES ##################################################################################
##################################################################################################################

#############################
#### Upload TS gains ########
#############################

tissue_gains_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  tissue_gains_df = rbind(tissue_gains_df, input_df)
}

#############################
#### Upload TS losses #######
#############################
tissue_losses_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_losses.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_losses", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  input_df$OG_ID_loss = paste0(input_df$OG_ID, "_", input_df$TS_losses)
  tissue_losses_df = rbind(tissue_losses_df, input_df)
}

#############################
#### Select relative OGs ####
#############################
foreground_df = vector()
for (ancestor in all_ancestors) {
  #### Select ancestor species
  my_species = get(ancestor)
  my_species_orthogroups_df = vector()
  #### Cycle on each tissue
  for (tissue in ordered_tissues) {
    ancestor_gains_OGs = unique(as.vector(subset(tissue_gains_df, TS_gains==ancestor & Tissue==tissue)$OG_ID))
    #### Subset all orthogroups to only these species in the orthogroups with gains
    my_species_tissue_orthogroups_df = subset(orthogroups_df, Species %in% my_species & OG_ID %in% ancestor_gains_OGs)
    ancestor_losses_OGs = unique(as.vector(subset(tissue_losses_df, Tissue==tissue)$OG_ID_loss))
    #### Filter out the losses
    my_species_tissue_orthogroups_df = my_species_tissue_orthogroups_df %>%
      filter(!(OG_ID_Species %in% ancestor_losses_OGs))
    #### Join to final dataframe
    my_species_orthogroups_df = rbind(my_species_orthogroups_df, my_species_tissue_orthogroups_df)
  }
  #Compute the same statistics as above
    final_ancestor_df = my_species_orthogroups_df %>%
      dplyr::select(OG_ID, Species, Paralogs_count) %>%
      distinct() %>%
      mutate(Dup_status=case_when(Paralogs_count > 1 ~ "Dup",
                                  Paralogs_count <= 1 ~ "NO_Dup")) %>%
      mutate(Node = ancestor)
    
  foreground_df = rbind(foreground_df, as.data.frame(final_ancestor_df))
}

#Select only the TS species (species without losses) and compute the proportion of those species that are duplicated.
#Compute the difference as not-duplicated (it has to sum to 1).

####### Order:
foreground_df$Node = factor(foreground_df$Node, levels=c(deuterostoma_y_labels[deuterostoma_y_labels!="Bilateria"], "Bilateria",
                                                             rev(protostoma_y_labels[protostoma_y_labels!="Bilateria"])))
foreground_df$Dup_status = factor(foreground_df$Dup_status, levels=rev(c("Dup", "NO_Dup")))

###### Add branch info
foreground_df$Branch = rep("Deuterostoma", nrow(foreground_df))
foreground_df$Branch[foreground_df$Node %in% protostoma_y_labels] = "Protostoma"
foreground_df$Branch[foreground_df$Node=="Bilateria"] = "Bilateria"
foreground_df = foreground_df %>%
  unite("Branch_Dup_status", c(Branch,Dup_status), remove=FALSE)
foreground_df$Branch_Dup_status = factor(foreground_df$Branch_Dup_status, levels=rev(c("Deuterostoma_Dup", "Deuterostoma_NO_Dup", "Protostoma_Dup", "Protostoma_NO_Dup", "Bilateria_Dup", "Bilateria_NO_Dup")))

SPECIES_foreground_dup_df = foreground_df

#########################################################
##### GENERATE FOREGROUND RANDOMIZATIONS ################
#########################################################
all_rand_foregrounds_df = vector()

for (randomization_round in seq(1,10)) {
  ######### Set directory
  randomization_dir = paste0(home, "projects/bilaterian_GE/data/ts_gains_losses_revision/randomization_", randomization_round, "/")
  
  ######### Upload TS gains
  tissue_gains_df = vector()
  for (tissue in ordered_tissues) {
    input_df = read.delim(paste0(randomization_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
    input_df$Tissue = rep(tissue, nrow(input_df))
    tissue_gains_df = rbind(tissue_gains_df, input_df)
  }
  ######### Upload TS losses
  tissue_losses_df = vector()
  for (tissue in ordered_tissues) {
    input_df = read.delim(paste0(randomization_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_losses.tab"), header=TRUE,
                          stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_losses", "Category"))
    input_df$Tissue = rep(tissue, nrow(input_df))
    input_df$OG_ID_loss = paste0(input_df$OG_ID, "_", input_df$TS_losses)
    tissue_losses_df = rbind(tissue_losses_df, input_df)
  }
  
  ######### Build foreground
  rand_foreground_df = vector()
  for (ancestor in all_ancestors) {
    #### Select ancestor species
    my_species = get(ancestor)
    my_species_orthogroups_df = vector()
    #### Cycle on each tissue
    for (tissue in ordered_tissues) {
      ancestor_gains_OGs = unique(as.vector(subset(tissue_gains_df, TS_gains==ancestor & Tissue==tissue)$OG_ID))
      #### Subset all orthogroups to only these species in the orthogroups with gains
      my_species_tissue_orthogroups_df = subset(orthogroups_df, Species %in% my_species & OG_ID %in% ancestor_gains_OGs)
      ancestor_losses_OGs = unique(as.vector(subset(tissue_losses_df, Tissue==tissue)$OG_ID_loss))
      #### Filter out the losses
      my_species_tissue_orthogroups_df = my_species_tissue_orthogroups_df %>%
        filter(!(OG_ID_Species %in% ancestor_losses_OGs))
      #### Join to final dataframe
      my_species_orthogroups_df = rbind(my_species_orthogroups_df, my_species_tissue_orthogroups_df)
    }
    ##############
    #If there are no gains at all respecting the conditions, go to the next iteration
    if (nrow(my_species_orthogroups_df)==0) {next}
    ##############
    #Compute the same statistics as above
    final_ancestor_df = my_species_orthogroups_df %>%
      dplyr::select(OG_ID, Species, Paralogs_count) %>%
      distinct() %>%
      mutate(Dup_status=case_when(Paralogs_count > 1 ~ "Dup",
                                  Paralogs_count <= 1 ~ "NO_Dup")) %>%
      mutate(Node = ancestor)
    
    rand_foreground_df = rbind(rand_foreground_df, as.data.frame(final_ancestor_df))
  }
  
  ####### Order:
  rand_foreground_df$Node = factor(rand_foreground_df$Node, levels=c(deuterostoma_y_labels[deuterostoma_y_labels!="Bilateria"], "Bilateria",
                                                               rev(protostoma_y_labels[protostoma_y_labels!="Bilateria"])))
  rand_foreground_df$Dup_status = factor(rand_foreground_df$Dup_status, levels=rev(c("Dup", "NO_Dup")))
  
  ###### Add branch info
  rand_foreground_df$Branch = rep("Deuterostoma", nrow(rand_foreground_df))
  rand_foreground_df$Branch[rand_foreground_df$Node %in% protostoma_y_labels] = "Protostoma"
  rand_foreground_df$Branch[rand_foreground_df$Node=="Bilateria"] = "Bilateria"
  rand_foreground_df = rand_foreground_df %>%
    unite("Branch_Dup_status", c(Branch,Dup_status), remove=FALSE)
  rand_foreground_df$Branch_Dup_status = factor(rand_foreground_df$Branch_Dup_status, levels=rev(c("Deuterostoma_Dup", "Deuterostoma_NO_Dup", "Protostoma_Dup", "Protostoma_NO_Dup", "Bilateria_Dup", "Bilateria_NO_Dup")))
  
  ###### Add randomization round
  rand_foreground_df$Randomization_round = rep(randomization_round, nrow(rand_foreground_df))
  
  ###### join to final df
  all_rand_foregrounds_df = rbind(all_rand_foregrounds_df, as.data.frame(rand_foreground_df))
}

all_rand_foregrounds_dup_df = all_rand_foregrounds_df

SPECIES_all_rand_foregrounds_dup_df = all_rand_foregrounds_dup_df

#########################################################
###### COMBINED ALL DATAFRAMES ##########################
#########################################################

final_df = dplyr::bind_rows(SPECIES_foreground_dup_df %>%
                              mutate(General_category="Observed",
                                     Species_category="TS"),
                            NON_SPECIES_foreground_dup_df %>%
                               mutate(General_category="Observed",
                                     Species_category="NO_TS"),
                            SPECIES_all_rand_foregrounds_dup_df %>%
                              mutate(General_category="Random",
                                     Species_category="TS"),
                            NON_SPECIES_all_rand_foregrounds_dup_df %>%
                              mutate(General_category="Random",
                                     Species_category="NO_TS")) %>%
  mutate(Node = factor(Node, levels=all_ancestors),
         Species_category = factor(Species_category, levels=c("TS", "NO_TS"))) %>%
  filter(!(Node %in% c("Deuterostoma", "Protostoma")),
         Dup_status=="Dup")

#########################################################
############## PLOT #####################################
#########################################################

TS_vs_NO_TS_plot = ggplot(data=final_df, aes(x=Species_category, y=Paralogs_count, fill=Node, alpha=General_category)) +
  geom_boxplot(outlier.size = 0.5) +
  scale_fill_manual(values=all_ancestors_colors) +
  scale_alpha_manual(values=c("Observed"=0.9, "Random"=0.4)) +
  theme_bw() +
  theme(axis.title = element_text(color="black"),
        axis.text = element_text(color="black")) +
  guides(fill="none", alpha="none") +
  coord_cartesian(ylim=c(0,10)) +
  facet_nested_wrap(~Node+General_category)

print(TS_vs_NO_TS_plot)
```


<!-- #################################################### -->

```{r duplication_levels_randomized_call_ancestors_binom_test, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions")}
#Upload orthogroups and get duplication info about each gene
orthogroups_df = read.delim(paste0(ts_gains_losses_raw_dir, "Bilateria_conserved_orthogroups-All_genes-associated_tissues.tab"),
                            header=FALSE, col.names=c("OG_ID", "Species", "GeneID", "Tau", "Tissue", "Paralog_tissue", "Expr_cutoff"), stringsAsFactors = FALSE)
#### Add column with OG_ID_Species and count number of paralogs per gene
orthogroups_df$OG_ID_Species = paste0(orthogroups_df$OG_ID, "_", orthogroups_df$Species)
species_paralogs_count_df = as.data.frame(table(orthogroups_df$OG_ID_Species))
OG_ID_Species_paralogs_counts_dict = hashmap(as.vector(species_paralogs_count_df$Var1), as.vector(species_paralogs_count_df$Freq))
orthogroups_df$Paralogs_count = OG_ID_Species_paralogs_counts_dict$find(orthogroups_df$OG_ID_Species)

#########################################################
############## GENERATE BACKGROUND ######################
#########################################################
#### For each node (group of species) compute the proportion in each orthogroup that is duplicated. Get the difference for non-duplicated.
#Sum all and get proportions.
background_df = vector()
for (ancestor in all_ancestors) {
  my_species = get(ancestor)
  #Subset all orthogroups to only these species
  my_species_orthogroups_df = subset(orthogroups_df, Species %in% my_species)
  #Compute for each orthogroup the proportion of duplicated species (among the ones with conserved genes)
   final_ancestor_df = my_species_orthogroups_df %>%
      dplyr::select(OG_ID, Species, Paralogs_count) %>%
      distinct() %>%
      mutate(Dup_status=case_when(Paralogs_count > 1 ~ "Dup",
                                  Paralogs_count <= 1 ~ "NO_Dup")) %>%
      group_by(OG_ID, Dup_status) %>%
      summarise(total = n()) %>%
      ungroup() %>%
      pivot_wider(id_cols="OG_ID", names_from="Dup_status", values_from="total", values_fill=0) %>%
      pivot_longer(cols=c("Dup", "NO_Dup"), names_to = "Dup_status") %>%
      group_by(OG_ID) %>%
      mutate(freq=value/sum(value)) %>%
      mutate(freq = replace_na(freq, 0))
  
  collapsed_final_df = final_ancestor_df %>%
    group_by(Dup_status) %>%
    summarise(freq_proportion = mean(freq)) %>%
    mutate(Node = ancestor) %>%
    dplyr::select(Node, Dup_status, freq_proportion)
  
  background_df = rbind(background_df, collapsed_final_df)
}
#### Order levels
background_df$Node = factor(background_df$Node, levels=c(deuterostoma_y_labels[deuterostoma_y_labels!="Bilateria"], "Bilateria",
                                                             rev(protostoma_y_labels[protostoma_y_labels!="Bilateria"])))
background_df$Dup_status = factor(background_df$Dup_status, levels=rev(c("Dup", "NO_Dup")))
#### Subset only to duplicates
dup_background_df = subset(background_df, Dup_status=="Dup")

#########################################################
############## GENERATE FOREGROUND ######################
#########################################################

#############################
#### Upload TS gains ########
#############################

tissue_gains_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  tissue_gains_df = rbind(tissue_gains_df, input_df)
}

#############################
#### Upload TS losses #######
#############################
tissue_losses_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_losses.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_losses", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  input_df$OG_ID_loss = paste0(input_df$OG_ID, "_", input_df$TS_losses)
  tissue_losses_df = rbind(tissue_losses_df, input_df)
}

#############################
#### Select relative OGs ####
#############################
foreground_df = vector()
for (ancestor in all_ancestors) {
  #### Select ancestor species
  my_species = get(ancestor)
  my_species_orthogroups_df = vector()
  #### Cycle on each tissue
  for (tissue in ordered_tissues) {
    ancestor_gains_OGs = unique(as.vector(subset(tissue_gains_df, TS_gains==ancestor & Tissue==tissue)$OG_ID))
    #### Subset all orthogroups to only these species in the orthogroups with gains
    my_species_tissue_orthogroups_df = subset(orthogroups_df, Species %in% my_species & OG_ID %in% ancestor_gains_OGs)
    ancestor_losses_OGs = unique(as.vector(subset(tissue_losses_df, Tissue==tissue)$OG_ID_loss))
    #### Filter out the losses
    my_species_tissue_orthogroups_df = my_species_tissue_orthogroups_df %>%
      filter(!(OG_ID_Species %in% ancestor_losses_OGs))
    #### Join to final dataframe
    my_species_orthogroups_df = rbind(my_species_orthogroups_df, my_species_tissue_orthogroups_df)
  }
  #Compute the same statistics as above
    final_ancestor_df = my_species_orthogroups_df %>%
      dplyr::select(OG_ID, Species, Paralogs_count) %>%
      distinct() %>%
      mutate(Dup_status=case_when(Paralogs_count > 1 ~ "Dup",
                                  Paralogs_count <= 1 ~ "NO_Dup")) %>%
      group_by(OG_ID, Dup_status) %>%
      summarise(total = n()) %>%
      ungroup() %>%
      pivot_wider(id_cols="OG_ID", names_from="Dup_status", values_from="total", values_fill=0) %>%
      pivot_longer(cols=c("Dup", "NO_Dup"), names_to = "Dup_status") %>%
      group_by(OG_ID) %>%
      mutate(freq=value/sum(value)) %>%
      mutate(freq = replace_na(freq, 0))
  
  collapsed_final_df = final_ancestor_df %>%
    group_by(Dup_status) %>%
    summarise(freq_proportion = mean(freq)) %>%
    mutate(Node = ancestor) %>%
    dplyr::select(Node, Dup_status, freq_proportion)
  
  foreground_df = rbind(foreground_df, collapsed_final_df)
}

#Select only the TS species (species without losses) and compute the proportion of those species that are duplicated.
#Compute the difference as not-duplicated (it has to sum to 1).

####### Order:
foreground_df$Node = factor(foreground_df$Node, levels=c(deuterostoma_y_labels[deuterostoma_y_labels!="Bilateria"], "Bilateria",
                                                             rev(protostoma_y_labels[protostoma_y_labels!="Bilateria"])))
foreground_df$Dup_status = factor(foreground_df$Dup_status, levels=rev(c("Dup", "NO_Dup")))

###### Add branch info
foreground_df$Branch = rep("Deuterostoma", nrow(foreground_df))
foreground_df$Branch[foreground_df$Node %in% protostoma_y_labels] = "Protostoma"
foreground_df$Branch[foreground_df$Node=="Bilateria"] = "Bilateria"
foreground_df = foreground_df %>%
  unite("Branch_Dup_status", c(Branch,Dup_status), remove=FALSE)
foreground_df$Branch_Dup_status = factor(foreground_df$Branch_Dup_status, levels=rev(c("Deuterostoma_Dup", "Deuterostoma_NO_Dup", "Protostoma_Dup", "Protostoma_NO_Dup", "Bilateria_Dup", "Bilateria_NO_Dup")))


#########################################################
##### GENERATE FOREGROUND RANDOMIZATIONS ################
#########################################################
all_rand_foregrounds_df = vector()

for (randomization_round in seq(1,10)) {
  ######### Set directory
  randomization_dir = paste0(home, "projects/bilaterian_GE/data/ts_gains_losses_revision/randomization_", randomization_round, "/")
  
  ######### Upload TS gains
  tissue_gains_df = vector()
  for (tissue in ordered_tissues) {
    input_df = read.delim(paste0(randomization_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
    input_df$Tissue = rep(tissue, nrow(input_df))
    tissue_gains_df = rbind(tissue_gains_df, input_df)
  }
  ######### Upload TS losses
  tissue_losses_df = vector()
  for (tissue in ordered_tissues) {
    input_df = read.delim(paste0(randomization_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_losses.tab"), header=TRUE,
                          stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_losses", "Category"))
    input_df$Tissue = rep(tissue, nrow(input_df))
    input_df$OG_ID_loss = paste0(input_df$OG_ID, "_", input_df$TS_losses)
    tissue_losses_df = rbind(tissue_losses_df, input_df)
  }
  
  ######### Build foreground
  rand_foreground_df = vector()
  for (ancestor in all_ancestors) {
    #### Select ancestor species
    my_species = get(ancestor)
    my_species_orthogroups_df = vector()
    #### Cycle on each tissue
    for (tissue in ordered_tissues) {
      ancestor_gains_OGs = unique(as.vector(subset(tissue_gains_df, TS_gains==ancestor & Tissue==tissue)$OG_ID))
      #### Subset all orthogroups to only these species in the orthogroups with gains
      my_species_tissue_orthogroups_df = subset(orthogroups_df, Species %in% my_species & OG_ID %in% ancestor_gains_OGs)
      ancestor_losses_OGs = unique(as.vector(subset(tissue_losses_df, Tissue==tissue)$OG_ID_loss))
      #### Filter out the losses
      my_species_tissue_orthogroups_df = my_species_tissue_orthogroups_df %>%
        filter(!(OG_ID_Species %in% ancestor_losses_OGs))
      #### Join to final dataframe
      my_species_orthogroups_df = rbind(my_species_orthogroups_df, my_species_tissue_orthogroups_df)
    }
    ##############
    #If there are no gains at all respecting the conditions, go to the next iteration
    if (nrow(my_species_orthogroups_df)==0) {next}
    ##############
    #Compute the same statistics as above
    final_ancestor_df = my_species_orthogroups_df %>%
      dplyr::select(OG_ID, Species, Paralogs_count) %>%
      distinct() %>%
      mutate(Dup_status=case_when(Paralogs_count > 1 ~ "Dup",
                                  Paralogs_count <= 1 ~ "NO_Dup")) %>%
      group_by(OG_ID, Dup_status) %>%
      summarise(total = n()) %>%
      ungroup() %>%
      pivot_wider(id_cols="OG_ID", names_from="Dup_status", values_from="total", values_fill=0) %>%
      #### Add column if missing
      add_cols(c("Dup", "NO_Dup")) %>%
      ####
      pivot_longer(cols=c("Dup", "NO_Dup"), names_to = "Dup_status") %>%
      group_by(OG_ID) %>%
      mutate(freq=value/sum(value)) %>%
      mutate(freq = replace_na(freq, 0))
  
    collapsed_final_df = final_ancestor_df %>%
      group_by(Dup_status) %>%
      summarise(freq_proportion = mean(freq)) %>%
      mutate(Node = ancestor) %>%
      dplyr::select(Node, Dup_status, freq_proportion)
    
    rand_foreground_df = rbind(rand_foreground_df, collapsed_final_df)
  }
  
  ####### Order:
  rand_foreground_df$Node = factor(rand_foreground_df$Node, levels=c(deuterostoma_y_labels[deuterostoma_y_labels!="Bilateria"], "Bilateria",
                                                               rev(protostoma_y_labels[protostoma_y_labels!="Bilateria"])))
  rand_foreground_df$Dup_status = factor(rand_foreground_df$Dup_status, levels=rev(c("Dup", "NO_Dup")))
  
  ###### Add branch info
  rand_foreground_df$Branch = rep("Deuterostoma", nrow(rand_foreground_df))
  rand_foreground_df$Branch[rand_foreground_df$Node %in% protostoma_y_labels] = "Protostoma"
  rand_foreground_df$Branch[rand_foreground_df$Node=="Bilateria"] = "Bilateria"
  rand_foreground_df = rand_foreground_df %>%
    unite("Branch_Dup_status", c(Branch,Dup_status), remove=FALSE)
  rand_foreground_df$Branch_Dup_status = factor(rand_foreground_df$Branch_Dup_status, levels=rev(c("Deuterostoma_Dup", "Deuterostoma_NO_Dup", "Protostoma_Dup", "Protostoma_NO_Dup", "Bilateria_Dup", "Bilateria_NO_Dup")))
  
  ###### Add randomization round
  rand_foreground_df$Randomization_round = rep(randomization_round, nrow(rand_foreground_df))
  
  ###### join to final df
  all_rand_foregrounds_df = rbind(all_rand_foregrounds_df, as.data.frame(rand_foreground_df))
}

all_rand_foregrounds_dup_df = all_rand_foregrounds_df %>%
  filter(Dup_status=="Dup")

#########################################################
######## COMPUTE BINOMIAL TEST ##########################
#########################################################
#Here the observed is the random and the expected is the proportion from the observed
binomial_res_df = all_rand_foregrounds_dup_df %>%
  group_by(Node, Branch) %>%
  summarize(Median_freq_proportion = median(freq_proportion)) %>%
  left_join(foreground_df %>% 
              filter(Dup_status=="Dup") %>%
              dplyr::select(Node, Branch, freq_proportion) %>%
              rename(Expected = freq_proportion),
            by=c("Node", "Branch")) %>%
  ungroup() %>%
  mutate(X=round(Median_freq_proportion*100, 0)) %>%
  dplyr::rowwise() %>%
  mutate(binom_pvalue = binom.test(x=X, n=100, p=Expected, alternative="less")$p.value) %>%
  mutate(binom_signif = ifelse(binom_pvalue <= 0.05, "*", ""))
  

#########################################################
############## PLOT #####################################
#########################################################

ancestral_gains_duplication_status = ggplot() +
  geom_bar(data=subset(foreground_df, Node!="Bilateria"), 
           aes(x=Node, y=freq_proportion, fill=Branch_Dup_status), stat="identity", position="fill", color="black", alpha=0.75) +
  geom_errorbar(data=subset(dup_background_df, Node!="Bilateria"), 
                            aes(x=Node, ymin=freq_proportion, ymax=freq_proportion), color="black", linetype="dashed") +
  geom_boxplot(data=all_rand_foregrounds_dup_df, aes(x=Node, y=freq_proportion), color="black", fill="black", size=0.5, alpha=0.4, outlier.size = 0.5) +
  ###############
  geom_text(data=binomial_res_df, aes(x=Node, y=0.9, label=binom_signif), color="black", size=6) +
  ###############
  scale_fill_manual(values=c("Deuterostoma_Dup"="darkolivegreen4", "Protostoma_Dup"="tan3", "Bilateria_Dup"="firebrick1", "Deuterostoma_NO_Dup"="white", "Protostoma_NO_Dup"="white", "Bilateria_NO_Dup"="white")) +
  ggtitle("Proportion of TS gains orthogroups with duplications") +
  theme_bw() +
  theme(axis.text.y = element_text(color="black"),
        axis.title = element_blank(),
        axis.text.x = element_text(color="black", angle=45, hjust=1, vjust=1),
        axis.ticks.y = element_line(),
        plot.title = element_text(color="black", hjust=0.5, face="bold"))

print(ancestral_gains_duplication_status)
```

```{r duplication_levels_randomized_call_ancestors_STRICT_ONLY, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions")}
#Upload orthogroups and get duplication info about each gene
orthogroups_df = read.delim(paste0(ts_gains_losses_raw_dir, "Bilateria_conserved_orthogroups-All_genes-associated_tissues.tab"),
                            header=FALSE, col.names=c("OG_ID", "Species", "GeneID", "Tau", "Tissue", "Paralog_tissue", "Expr_cutoff"), stringsAsFactors = FALSE)
#### Add column with OG_ID_Species and count number of paralogs per gene
orthogroups_df$OG_ID_Species = paste0(orthogroups_df$OG_ID, "_", orthogroups_df$Species)
species_paralogs_count_df = as.data.frame(table(orthogroups_df$OG_ID_Species))
OG_ID_Species_paralogs_counts_dict = hashmap(as.vector(species_paralogs_count_df$Var1), as.vector(species_paralogs_count_df$Freq))
orthogroups_df$Paralogs_count = OG_ID_Species_paralogs_counts_dict$find(orthogroups_df$OG_ID_Species)

#########################################################
############## GENERATE BACKGROUND ######################
#########################################################
#### For each node (group of species) compute the proportion in each orthogroup that is duplicated. Get the difference for non-duplicated.
#Sum all and get proportions.
background_df = vector()
for (ancestor in all_ancestors) {
  my_species = get(ancestor)
  #Subset all orthogroups to only these species
  my_species_orthogroups_df = subset(orthogroups_df, Species %in% my_species)
  #Compute for each orthogroup the proportion of duplicated species (among the ones with conserved genes)
   final_ancestor_df = my_species_orthogroups_df %>%
      dplyr::select(OG_ID, Species, Paralogs_count) %>%
      distinct() %>%
      mutate(Dup_status=case_when(Paralogs_count > 1 ~ "Dup",
                                  Paralogs_count <= 1 ~ "NO_Dup")) %>%
      group_by(OG_ID, Dup_status) %>%
      summarise(total = n()) %>%
      ungroup() %>%
      pivot_wider(id_cols="OG_ID", names_from="Dup_status", values_from="total", values_fill=0) %>%
      pivot_longer(cols=c("Dup", "NO_Dup"), names_to = "Dup_status") %>%
      group_by(OG_ID) %>%
      mutate(freq=value/sum(value)) %>%
      mutate(freq = replace_na(freq, 0))
  
  collapsed_final_df = final_ancestor_df %>%
    group_by(Dup_status) %>%
    summarise(freq_proportion = mean(freq)) %>%
    mutate(Node = ancestor) %>%
    dplyr::select(Node, Dup_status, freq_proportion)
  
  background_df = rbind(background_df, collapsed_final_df)
}
#### Order levels
background_df$Node = factor(background_df$Node, levels=c(deuterostoma_y_labels[deuterostoma_y_labels!="Bilateria"], "Bilateria",
                                                             rev(protostoma_y_labels[protostoma_y_labels!="Bilateria"])))
background_df$Dup_status = factor(background_df$Dup_status, levels=rev(c("Dup", "NO_Dup")))
#### Subset only to duplicates
dup_background_df = subset(background_df, Dup_status=="Dup")

#########################################################
############## GENERATE FOREGROUND ######################
#########################################################

#############################
#### Upload TS gains ########
#############################

tissue_gains_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
  input_df = subset(input_df, Category=="STRICT")
  input_df$Tissue = rep(tissue, nrow(input_df))
  tissue_gains_df = rbind(tissue_gains_df, input_df)
}

#############################
#### Upload TS losses #######
#############################
tissue_losses_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_losses.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_losses", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  input_df$OG_ID_loss = paste0(input_df$OG_ID, "_", input_df$TS_losses)
  tissue_losses_df = rbind(tissue_losses_df, input_df)
}

#############################
#### Select relative OGs ####
#############################
foreground_df = vector()
for (ancestor in all_ancestors) {
  #### Select ancestor species
  my_species = get(ancestor)
  my_species_orthogroups_df = vector()
  #### Cycle on each tissue
  for (tissue in ordered_tissues) {
    ancestor_gains_OGs = unique(as.vector(subset(tissue_gains_df, TS_gains==ancestor & Tissue==tissue)$OG_ID))
    #### Subset all orthogroups to only these species in the orthogroups with gains
    my_species_tissue_orthogroups_df = subset(orthogroups_df, Species %in% my_species & OG_ID %in% ancestor_gains_OGs)
    ancestor_losses_OGs = unique(as.vector(subset(tissue_losses_df, Tissue==tissue)$OG_ID_loss))
    #### Filter out the losses
    my_species_tissue_orthogroups_df = my_species_tissue_orthogroups_df %>%
      filter(!(OG_ID_Species %in% ancestor_losses_OGs))
    #### Join to final dataframe
    my_species_orthogroups_df = rbind(my_species_orthogroups_df, my_species_tissue_orthogroups_df)
  }
  #Compute the same statistics as above
    final_ancestor_df = my_species_orthogroups_df %>%
      dplyr::select(OG_ID, Species, Paralogs_count) %>%
      distinct() %>%
      mutate(Dup_status=case_when(Paralogs_count > 1 ~ "Dup",
                                  Paralogs_count <= 1 ~ "NO_Dup")) %>%
      group_by(OG_ID, Dup_status) %>%
      summarise(total = n()) %>%
      ungroup() %>%
      pivot_wider(id_cols="OG_ID", names_from="Dup_status", values_from="total", values_fill=0) %>%
      pivot_longer(cols=c("Dup", "NO_Dup"), names_to = "Dup_status") %>%
      group_by(OG_ID) %>%
      mutate(freq=value/sum(value)) %>%
      mutate(freq = replace_na(freq, 0))
  
  collapsed_final_df = final_ancestor_df %>%
    group_by(Dup_status) %>%
    summarise(freq_proportion = mean(freq)) %>%
    mutate(Node = ancestor) %>%
    dplyr::select(Node, Dup_status, freq_proportion)
  
  foreground_df = rbind(foreground_df, collapsed_final_df)
}

#Select only the TS species (species without losses) and compute the proportion of those species that are duplicated.
#Compute the difference as not-duplicated (it has to sum to 1).

####### Order:
foreground_df$Node = factor(foreground_df$Node, levels=c(deuterostoma_y_labels[deuterostoma_y_labels!="Bilateria"], "Bilateria",
                                                             rev(protostoma_y_labels[protostoma_y_labels!="Bilateria"])))
foreground_df$Dup_status = factor(foreground_df$Dup_status, levels=rev(c("Dup", "NO_Dup")))

###### Add branch info
foreground_df$Branch = rep("Deuterostoma", nrow(foreground_df))
foreground_df$Branch[foreground_df$Node %in% protostoma_y_labels] = "Protostoma"
foreground_df$Branch[foreground_df$Node=="Bilateria"] = "Bilateria"
foreground_df = foreground_df %>%
  unite("Branch_Dup_status", c(Branch,Dup_status), remove=FALSE)
foreground_df$Branch_Dup_status = factor(foreground_df$Branch_Dup_status, levels=rev(c("Deuterostoma_Dup", "Deuterostoma_NO_Dup", "Protostoma_Dup", "Protostoma_NO_Dup", "Bilateria_Dup", "Bilateria_NO_Dup")))


#########################################################
##### GENERATE FOREGROUND RANDOMIZATIONS ################
#########################################################
all_rand_foregrounds_df = vector()

for (randomization_round in seq(1,10)) {
  ######### Set directory
  randomization_dir = paste0(home, "projects/bilaterian_GE/data/ts_gains_losses_revision/randomization_", randomization_round, "/")
  
  ######### Upload TS gains
  tissue_gains_df = vector()
  for (tissue in ordered_tissues) {
    input_df = read.delim(paste0(randomization_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
    input_df = subset(input_df, Category=="STRICT")
    input_df$Tissue = rep(tissue, nrow(input_df))
    tissue_gains_df = rbind(tissue_gains_df, input_df)
  }
  ######### Upload TS losses
  tissue_losses_df = vector()
  for (tissue in ordered_tissues) {
    input_df = read.delim(paste0(randomization_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_losses.tab"), header=TRUE,
                          stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_losses", "Category"))
    input_df$Tissue = rep(tissue, nrow(input_df))
    input_df$OG_ID_loss = paste0(input_df$OG_ID, "_", input_df$TS_losses)
    tissue_losses_df = rbind(tissue_losses_df, input_df)
  }
  
  ######### Build foreground
  rand_foreground_df = vector()
  for (ancestor in all_ancestors) {
    #### Select ancestor species
    my_species = get(ancestor)
    my_species_orthogroups_df = vector()
    #### Cycle on each tissue
    for (tissue in ordered_tissues) {
      ancestor_gains_OGs = unique(as.vector(subset(tissue_gains_df, TS_gains==ancestor & Tissue==tissue)$OG_ID))
      #### Subset all orthogroups to only these species in the orthogroups with gains
      my_species_tissue_orthogroups_df = subset(orthogroups_df, Species %in% my_species & OG_ID %in% ancestor_gains_OGs)
      ancestor_losses_OGs = unique(as.vector(subset(tissue_losses_df, Tissue==tissue)$OG_ID_loss))
      #### Filter out the losses
      my_species_tissue_orthogroups_df = my_species_tissue_orthogroups_df %>%
        filter(!(OG_ID_Species %in% ancestor_losses_OGs))
      #### Join to final dataframe
      my_species_orthogroups_df = rbind(my_species_orthogroups_df, my_species_tissue_orthogroups_df)
    }
    ##############
    #If there are no gains at all respecting the conditions, go to the next iteration
    if (nrow(my_species_orthogroups_df)==0) {next}
    ##############
    #Compute the same statistics as above
    final_ancestor_df = my_species_orthogroups_df %>%
      dplyr::select(OG_ID, Species, Paralogs_count) %>%
      distinct() %>%
      mutate(Dup_status=case_when(Paralogs_count > 1 ~ "Dup",
                                  Paralogs_count <= 1 ~ "NO_Dup")) %>%
      group_by(OG_ID, Dup_status) %>%
      summarise(total = n()) %>%
      ungroup() %>%
      pivot_wider(id_cols="OG_ID", names_from="Dup_status", values_from="total", values_fill=0) %>%
      #### Add column if missing
      add_cols(c("Dup", "NO_Dup")) %>%
      ####
      pivot_longer(cols=c("Dup", "NO_Dup"), names_to = "Dup_status") %>%
      group_by(OG_ID) %>%
      mutate(freq=value/sum(value)) %>%
      mutate(freq = replace_na(freq, 0))
  
    collapsed_final_df = final_ancestor_df %>%
      group_by(Dup_status) %>%
      summarise(freq_proportion = mean(freq)) %>%
      mutate(Node = ancestor) %>%
      dplyr::select(Node, Dup_status, freq_proportion)
    
    rand_foreground_df = rbind(rand_foreground_df, collapsed_final_df)
  }
  
  ####### Order:
  rand_foreground_df$Node = factor(rand_foreground_df$Node, levels=c(deuterostoma_y_labels[deuterostoma_y_labels!="Bilateria"], "Bilateria",
                                                               rev(protostoma_y_labels[protostoma_y_labels!="Bilateria"])))
  rand_foreground_df$Dup_status = factor(rand_foreground_df$Dup_status, levels=rev(c("Dup", "NO_Dup")))
  
  ###### Add branch info
  rand_foreground_df$Branch = rep("Deuterostoma", nrow(rand_foreground_df))
  rand_foreground_df$Branch[rand_foreground_df$Node %in% protostoma_y_labels] = "Protostoma"
  rand_foreground_df$Branch[rand_foreground_df$Node=="Bilateria"] = "Bilateria"
  rand_foreground_df = rand_foreground_df %>%
    unite("Branch_Dup_status", c(Branch,Dup_status), remove=FALSE)
  rand_foreground_df$Branch_Dup_status = factor(rand_foreground_df$Branch_Dup_status, levels=rev(c("Deuterostoma_Dup", "Deuterostoma_NO_Dup", "Protostoma_Dup", "Protostoma_NO_Dup", "Bilateria_Dup", "Bilateria_NO_Dup")))
  
  ###### Add randomization round
  rand_foreground_df$Randomization_round = rep(randomization_round, nrow(rand_foreground_df))
  
  ###### join to final df
  all_rand_foregrounds_df = rbind(all_rand_foregrounds_df, as.data.frame(rand_foreground_df))
}

all_rand_foregrounds_dup_df = all_rand_foregrounds_df %>%
  filter(Dup_status=="Dup")

#########################################################
############## PLOT #####################################
#########################################################

ancestral_gains_duplication_status = ggplot() +
  geom_bar(data=subset(foreground_df, Node!="Bilateria"), 
           aes(x=Node, y=freq_proportion, fill=Branch_Dup_status), stat="identity", position="fill", color="black", alpha=0.75) +
  geom_errorbar(data=subset(dup_background_df, Node!="Bilateria"), 
                            aes(x=Node, ymin=freq_proportion, ymax=freq_proportion), color="black", linetype="dashed") +
  ###############
  geom_boxplot(data=all_rand_foregrounds_dup_df, aes(x=Node, y=freq_proportion), color="black", fill="black", size=0.5, alpha=0.4, outlier.size = 0.5) +
  ###############
  scale_fill_manual(values=c("Deuterostoma_Dup"="darkolivegreen4", "Protostoma_Dup"="tan3", "Bilateria_Dup"="firebrick1", "Deuterostoma_NO_Dup"="white", "Protostoma_NO_Dup"="white", "Bilateria_NO_Dup"="white")) +
  ggtitle("Proportion of TS gains orthogroups with duplications") +
  theme_bw() +
  theme(axis.text.y = element_text(color="black"),
        axis.title = element_blank(),
        axis.text.x = element_text(color="black", angle=45, hjust=1, vjust=1),
        axis.ticks.y = element_line(),
        plot.title = element_text(color="black", hjust=0.5, face="bold"))

print(ancestral_gains_duplication_status)
```

<!-- #################################################### -->

<!-- SPECIES GAINS -->
```{r duplication_levels_species_gains, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions")}
#######################################################
######### UPLOAD ORTHOGROUPS ##########################
#######################################################
orthogroups_df = read.delim(paste0(ts_gains_losses_raw_dir, "Bilateria_conserved_orthogroups-All_genes-associated_tissues.tab"),
                            header=FALSE, col.names=c("OG_ID", "Species", "GeneID", "Tau", "Tissue", "Paralog_tissue", "Expr_cutoff"), stringsAsFactors = FALSE)

####### Add OGID_Species column
orthogroups_df$OG_ID_Species = paste0(orthogroups_df$OG_ID, "_", orthogroups_df$Species)
####### Count number of paralogs per species in each orthogroup
species_paralogs_count_df = as.data.frame(table(orthogroups_df$OG_ID_Species))
OG_ID_Species_paralogs_counts_dict = hashmap(as.vector(species_paralogs_count_df$Var1), as.vector(species_paralogs_count_df$Freq))
####### Add paralog count to dataframe
orthogroups_df$Paralogs_count = OG_ID_Species_paralogs_counts_dict$find(orthogroups_df$OG_ID_Species)

#######################################################
######### UPLOAD TS GAINS #############################
#######################################################
tissue_gains_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_raw_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  tissue_gains_df = rbind(tissue_gains_df, input_df)
}

#######################################################
######### UPLOAD TS LOSSES ############################
#######################################################
tissue_losses_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_raw_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_losses.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_losses", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  tissue_losses_df = rbind(tissue_losses_df, input_df)
}

###### Upload the best-TS orthogroups for which the inference was performed
all_tissue_tau_df = vector()
for (tissue in ordered_tissues) {
  tissue_tau_df = read.delim(paste0(ts_gains_losses_raw_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_specific_OGs.tab"),
                           header=FALSE, col.names=c("OG_ID", "Species", "GeneID", "Tau", "Tissue", "Expr_cutoff"))
  tissue_tau_df$Query_tissue = rep(tissue, nrow(tissue_tau_df))
  all_tissue_tau_df = rbind(all_tissue_tau_df, tissue_tau_df)
}

###### Build GeneID - OG_ID dictionary
unique_tissue_tau_df = unique(all_tissue_tau_df[,c("OG_ID", "Species", "GeneID")])
GeneID_OG_ID_dict = hashmap(as.vector(unique_tissue_tau_df$GeneID), as.vector(unique_tissue_tau_df$OG_ID))


#######################################################
######### GET DUPLICATION STATUS TS GAINS #############
#######################################################

#### Filter on species and tissue gains, then combine in a unique dataframe
all_selected_paralogs_df = vector()
for (species in original_species) {
  for (tissue in ordered_tissues) {
    #### For each tissue, get the OG_ID of the OG_ID with only one gain
    species_tissue_gains_OG_IDs = as.vector(subset(tissue_gains_df, TS_gains==species & Tissue==tissue)$OG_ID)
    #### Select all the geneIDs from the orthogroups with gains
    species_tissue_gains_genes = as.vector(subset(all_tissue_tau_df, OG_ID %in% species_tissue_gains_OG_IDs & Query_tissue==tissue)$GeneID)
    #### Subset to only the genes belonging to the orthogroup with gains in the queried species
    selected_paralogs_df = subset(orthogroups_df, GeneID %in% species_tissue_gains_genes)
    #### Add Category
    selected_paralogs_df$Category = rep("NO_TS", nrow(selected_paralogs_df))
    selected_paralogs_df$Category[selected_paralogs_df$Species==species] = "TS" #Distinguish the tissue-specific genes
    #### Add Species
    selected_paralogs_df$Node = rep(species, nrow(selected_paralogs_df))
    all_selected_paralogs_df = rbind(all_selected_paralogs_df, selected_paralogs_df)
  }
}

all_selected_paralogs_df$Category = factor(all_selected_paralogs_df$Category, levels=c("TS", "NO_TS"))
all_selected_paralogs_df$Node = factor(all_selected_paralogs_df$Node, levels = ordered_species_tree)

##### Remove all the genes used as NO_TS if they appear as TS in some other Tissues
TS_genes = as.vector(subset(all_selected_paralogs_df, Category=="TS")$GeneID)
final_all_selected_paralogs_df = unique(subset(all_selected_paralogs_df, !(GeneID %in% TS_genes & Category=="NO_TS")))

##### For each OG_ID, compute the mean paralog number bewtween TS and non-TS species
mean_paralogs_df = final_all_selected_paralogs_df %>%
  group_by(OG_ID, Node, Category) %>%
  summarize(Mean_paralog_count = mean(Paralogs_count))

#Subset only for TS species
final_mean_paralogs_df = subset(mean_paralogs_df, Category=="TS")

#Compute proportions of each species-specific gains where there is at least one duplicate
duplicated_status_df = final_mean_paralogs_df
duplicated_status_df$Dup_status = "Dup"
duplicated_status_df$Dup_status[duplicated_status_df$Mean_paralog_count==1] = "Non_Dup"
duplicated_status_df$Dup_status = factor(duplicated_status_df$Dup_status, levels=rev(c("Dup", "Non_Dup")))

#### Add branch info for plotting
duplicated_status_df$Branch = rep("Deuterostoma", nrow(duplicated_status_df))
duplicated_status_df$Branch[duplicated_status_df$Node %in% Protostoma] = "Protostoma"
duplicated_status_df$Branch_Dup_status = paste0(duplicated_status_df$Branch, "_", duplicated_status_df$Dup_status)

#### Order factor levels
duplicated_status_df$Branch_Dup_status = factor(duplicated_status_df$Branch_Dup_status, 
                                                 levels=rev(c("Deuterostoma_Dup", "Deuterostoma_Non_Dup", "Protostoma_Dup", "Protostoma_Non_Dup")))
duplicated_status_df$Node = factor(duplicated_status_df$Node, levels=ordered_species_tree)

#######################################################
######### GET DUPLICATION STATUS BACKGROUND ###########
#######################################################

background_proportions_df = orthogroups_df %>%
  dplyr::select(OG_ID, Species, Paralogs_count) %>%
  distinct() %>%
  mutate(Dup_status = case_when(Paralogs_count == 1 ~ "Non_Dup",
                                 Paralogs_count > 1 ~ "Dup"))

final_background_proportions_df = background_proportions_df %>%
  group_by(Species, Dup_status) %>%
  summarise(n = n()) %>%
  mutate(freq = n / sum(n)) %>%
  filter(Dup_status=="Dup")

#### Order factor levels
final_background_proportions_df$Species = factor(final_background_proportions_df$Species, levels=original_species)

#######################################################
#################### PLOT #############################
#######################################################

# species_specific_gains_duplication_status = ggplot() +
#   geom_bar(data=duplicated_status_df, aes(x=Node, fill=Branch_Dup_status), stat="count", position="fill", color="black", alpha=0.75) +
#   geom_errorbar(data=final_background_proportions_df, aes(x=Species, ymin=freq, ymax=freq), color="black", linetype="dashed") +
#   scale_fill_manual(values=c("Deuterostoma_Dup"="darkolivegreen4", "Protostoma_Dup"="tan3", "Deuterostoma_Non_Dup"="white", "Protostoma_Non_Dup"="white")) +
#   theme_bw() +
#   theme(axis.text.y = element_text(color="black"),
#         axis.title = element_blank(),
#         axis.text.x = element_text(color="black", angle=45, hjust=1, vjust=1),
#         axis.ticks.y = element_line()) +
#   ylab("Proportion of OGs with gains") +
#   guides(fill="none")
#   
# print(species_specific_gains_duplication_status)
```

```{r duplication_levels_randomized_call_species, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions")}
all_randomizations_duplication_status_df = vector()

for (randomization_round in seq(1,10)) {
  ####### Set randomization directory
  randomization_dir = paste0(home, "projects/bilaterian_GE/data/ts_gains_losses_revision/randomization_", randomization_round, "/")
  
  #######################################################
  ######### UPLOAD ORTHOGROUPS ##########################
  #######################################################
  orthogroups_df = read.delim(paste0(randomization_dir, "Bilateria_conserved_orthogroups-All_genes-associated_tissues.tab"),
                              header=FALSE, col.names=c("OG_ID", "Species", "GeneID", "Tau", "Tissue", "Paralog_tissue", "Expr_cutoff"), stringsAsFactors = FALSE)
  
  ####### Add OGID_Species column
  orthogroups_df$OG_ID_Species = paste0(orthogroups_df$OG_ID, "_", orthogroups_df$Species)
  ####### Count number of paralogs per species in each orthogroup
  species_paralogs_count_df = as.data.frame(table(orthogroups_df$OG_ID_Species))
  OG_ID_Species_paralogs_counts_dict = hashmap(as.vector(species_paralogs_count_df$Var1), as.vector(species_paralogs_count_df$Freq))
  ####### Add paralog count to dataframe
  orthogroups_df$Paralogs_count = OG_ID_Species_paralogs_counts_dict$find(orthogroups_df$OG_ID_Species)
  
  #######################################################
  ######### UPLOAD TS GAINS #############################
  #######################################################
  tissue_gains_df = vector()
  for (tissue in ordered_tissues) {
    input_df = read.delim(paste0(randomization_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
    input_df$Tissue = rep(tissue, nrow(input_df))
    tissue_gains_df = rbind(tissue_gains_df, input_df)
  }
  
  #######################################################
  ######### UPLOAD TS LOSSES ############################
  #######################################################
  tissue_losses_df = vector()
  for (tissue in ordered_tissues) {
    input_df = read.delim(paste0(randomization_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_losses.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_losses", "Category"))
    input_df$Tissue = rep(tissue, nrow(input_df))
    tissue_losses_df = rbind(tissue_losses_df, input_df)
  }
  
  ###### Upload the best-TS orthogroups for which the inference was performed
  all_tissue_tau_df = vector()
  for (tissue in ordered_tissues) {
    tissue_tau_df = read.delim(paste0(randomization_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_specific_OGs.tab"),
                             header=FALSE, col.names=c("OG_ID", "Species", "GeneID", "Tau", "Tissue", "Expr_cutoff"))
    tissue_tau_df$Query_tissue = rep(tissue, nrow(tissue_tau_df))
    all_tissue_tau_df = rbind(all_tissue_tau_df, tissue_tau_df)
  }
  
  ###### Build GeneID - OG_ID dictionary
  unique_tissue_tau_df = unique(all_tissue_tau_df[,c("OG_ID", "Species", "GeneID")])
  GeneID_OG_ID_dict = hashmap(as.vector(unique_tissue_tau_df$GeneID), as.vector(unique_tissue_tau_df$OG_ID))
  
  #######################################################
  ######### GET DUPLICATION STATUS TS GAINS #############
  #######################################################
  
  #### Filter on species and tissue gains, then combine in a unique dataframe
  all_selected_paralogs_df = vector()
  for (species in original_species) {
    for (tissue in ordered_tissues) {
      #### For each tissue, get the OG_ID of the OG_ID with only one gain
      species_tissue_gains_OG_IDs = as.vector(subset(tissue_gains_df, TS_gains==species & Tissue==tissue)$OG_ID)
      #### Select all the geneIDs from the orthogroups with gains
      species_tissue_gains_genes = as.vector(subset(all_tissue_tau_df, OG_ID %in% species_tissue_gains_OG_IDs & Query_tissue==tissue)$GeneID)
      #### Subset to only the genes belonging to the orthogroup with gains in the queried species
      selected_paralogs_df = subset(orthogroups_df, GeneID %in% species_tissue_gains_genes)
      #### Add Category
      selected_paralogs_df$Category = rep("NO_TS", nrow(selected_paralogs_df))
      selected_paralogs_df$Category[selected_paralogs_df$Species==species] = "TS" #Distinguish the tissue-specific genes
      #### Add Species
      selected_paralogs_df$Node = rep(species, nrow(selected_paralogs_df))
      all_selected_paralogs_df = rbind(all_selected_paralogs_df, selected_paralogs_df)
    }
  }
  
  all_selected_paralogs_df$Category = factor(all_selected_paralogs_df$Category, levels=c("TS", "NO_TS"))
  all_selected_paralogs_df$Node = factor(all_selected_paralogs_df$Node, levels = ordered_species_tree)
  
  ##### Remove all the genes used as NO_TS if they appear as TS in some other Tissues
  TS_genes = as.vector(subset(all_selected_paralogs_df, Category=="TS")$GeneID)
  final_all_selected_paralogs_df = unique(subset(all_selected_paralogs_df, !(GeneID %in% TS_genes & Category=="NO_TS")))
  
  ##### For each OG_ID, compute the mean paralog number bewtween TS and non-TS species
  mean_paralogs_df = final_all_selected_paralogs_df %>%
    group_by(OG_ID, Node, Category) %>%
    summarize(Mean_paralog_count = mean(Paralogs_count))
  
  #Subset only for TS species
  final_mean_paralogs_df = subset(mean_paralogs_df, Category=="TS")
  
  #Compute proportions of each species-specific gains where there is at least one duplicate
  rand_duplicated_status_df = final_mean_paralogs_df
  rand_duplicated_status_df$Dup_status = "Dup"
  rand_duplicated_status_df$Dup_status[rand_duplicated_status_df$Mean_paralog_count==1] = "Non_Dup"
  rand_duplicated_status_df$Dup_status = factor(rand_duplicated_status_df$Dup_status, levels=rev(c("Dup", "Non_Dup")))
  
  #### Add branch info for plotting
  rand_duplicated_status_df$Branch = rep("Deuterostoma", nrow(rand_duplicated_status_df))
  rand_duplicated_status_df$Branch[rand_duplicated_status_df$Node %in% Protostoma] = "Protostoma"
  rand_duplicated_status_df$Branch_Dup_status = paste0(rand_duplicated_status_df$Branch, "_", rand_duplicated_status_df$Dup_status)
  
  #### Add randomization round
  rand_duplicated_status_df$Randomization_round = rep(randomization_round, nrow(rand_duplicated_status_df))
  
  #### Order factor levels
  rand_duplicated_status_df$Branch_Dup_status = factor(rand_duplicated_status_df$Branch_Dup_status, 
                                                   levels=rev(c("Deuterostoma_Dup", "Deuterostoma_Non_Dup", "Protostoma_Dup", "Protostoma_Non_Dup")))
  rand_duplicated_status_df$Node = factor(rand_duplicated_status_df$Node, levels=ordered_species_tree)
  all_randomizations_duplication_status_df = rbind(all_randomizations_duplication_status_df, as.data.frame(rand_duplicated_status_df))
}

#### Count proportions of TS gains for each species that occur for duplicated genes
all_randomizations_stats_df = all_randomizations_duplication_status_df %>%
  group_by(Randomization_round, Node, Dup_status) %>% 
  summarize(Tot=n()) %>% 
  mutate(Proportion = Tot / sum(Tot)) %>%
  filter(Dup_status=="Dup") %>%
  ungroup()
```

```{r plot_species_specific_duplication_status, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions")}
#######################################################
#################### PLOT #############################
#######################################################
species_specific_gains_duplication_status = ggplot() +
  geom_bar(data=duplicated_status_df, aes(x=Node, fill=Branch_Dup_status), stat="count", position="fill", color="black", alpha=0.75) +
  #####
  geom_boxplot(data=all_randomizations_stats_df, aes(x=Node, y=Proportion), color="black", fill="black", size=0.5, alpha=0.4, outlier.size = 0.5) +
  #####
  geom_errorbar(data=final_background_proportions_df, aes(x=Species, ymin=freq, ymax=freq), color="black", linetype="dashed") +
  scale_fill_manual(values=c("Deuterostoma_Dup"="darkolivegreen4", "Protostoma_Dup"="tan3", "Deuterostoma_Non_Dup"="white", "Protostoma_Non_Dup"="white")) +
  theme_bw() +
  theme(axis.text.y = element_text(color="black"),
        axis.title = element_blank(),
        axis.text.x = element_text(color="black", angle=45, hjust=1, vjust=1),
        axis.ticks.y = element_line()) +
  ylab("Proportion of OGs with gains") +
  guides(fill="none")
  
print(species_specific_gains_duplication_status)
```

```{r randomization_gains_stats, echo=FALSE, warning=FALSE, messsage=FALSE, dependson=c("setup", "functions"), fig.width=10, fig.height=14}
#Plot total numbers of TS gains across ancestors and species coming from the randomizations

#######################################################
#################### UPLOAD INPUTS ####################
#######################################################
all_rand_tissue_gains_df = vector()
for (randomization_round in seq(1,10)) {
  ### Set randomization dir
  randomization_dir = paste0(home, "projects/bilaterian_GE/data/ts_gains_losses_revision/randomization_", randomization_round, "/")
  
  for (tissue in ordered_tissues) {
    input_df = read.delim(paste0(randomization_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
    ### Add tissue and randomization round info
    input_df$Tissue = rep(tissue, nrow(input_df))
    input_df$Randomization_round = rep(randomization_round, nrow(input_df))
    ### Join to final dataframe
    all_rand_tissue_gains_df = rbind(all_rand_tissue_gains_df, input_df)
  }
}

### Add branch info
all_rand_tissue_gains_df$Branch = rep("Deuterostoma", nrow(all_rand_tissue_gains_df))
all_rand_tissue_gains_df$Branch[all_rand_tissue_gains_df$TS_gains %in% c(protostoma_y_labels, Protostoma)] = "Protostoma"
all_rand_tissue_gains_df$Branch[all_rand_tissue_gains_df$TS_gains=="Bilateria"] = "Bilateria"

### Count total ts gains per rand round, tissue and node (both STRICT and LASSO separately and ALL)
all_rand_tissue_gains_stats_df = all_rand_tissue_gains_df %>%
  filter(Category!="MERGED", Category!="MERGED_Testis") %>%
  group_by(Randomization_round, Branch, TS_gains, Category) %>%
  summarize(Tot_gains = n()) %>%
  ungroup() %>%
  bind_rows(all_rand_tissue_gains_df %>%
              mutate(Category = "ALL") %>%
              group_by(Randomization_round, Branch, TS_gains, Category) %>%
              summarize(Tot_gains = n()) %>%
              ungroup()) %>%
  filter(Branch != "Bilateria")

### Add node level
all_rand_tissue_gains_stats_df$Node_level = ancestor_node_level_dict$find(all_rand_tissue_gains_stats_df$TS_gains)
all_rand_tissue_gains_stats_df$Node_level = factor(as.vector(all_rand_tissue_gains_stats_df$Node_level), levels=seq(0,10))
all_rand_tissue_gains_stats_df$Randomization_round = factor(all_rand_tissue_gains_stats_df$Randomization_round, levels=seq(1,10))
all_rand_tissue_gains_stats_df$TS_gains = factor(all_rand_tissue_gains_stats_df$TS_gains, levels=c(all_ancestors, original_species))
all_rand_tissue_gains_stats_df$Category = factor(all_rand_tissue_gains_stats_df$Category, levels=c("ALL", "STRICT", "LASSO"))

#######################################################
####### ORIGINAL TS GAINS #############################
#######################################################
original_ts_gains_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_raw_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), 
                        header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
  ### Add tissue and randomization round info
  input_df$Tissue = rep(tissue, nrow(input_df))
  ### Join to final dataframe
  original_ts_gains_df = rbind(original_ts_gains_df, input_df)
}

### Add branch info
original_ts_gains_df$Branch = rep("Deuterostoma", nrow(original_ts_gains_df))
original_ts_gains_df$Branch[original_ts_gains_df$TS_gains %in% c(protostoma_y_labels, Protostoma)] = "Protostoma"
original_ts_gains_df$Branch[original_ts_gains_df$TS_gains=="Bilateria"] = "Bilateria"

original_ts_gains_stats_df = original_ts_gains_df %>%
    filter(Category!="MERGED", Category!="MERGED_Testis") %>%
    group_by(Branch, TS_gains, Category) %>%
    summarize(Tot_gains = n()) %>%
    ungroup() %>%
    bind_rows(original_ts_gains_df %>%
              mutate(Category = "ALL") %>%
              group_by(Branch, TS_gains, Category) %>%
              summarize(Tot_gains = n()) %>%
              ungroup()) %>%
  filter(Branch != "Bilateria")

### Add node level and order factors
original_ts_gains_stats_df$Node_level = ancestor_node_level_dict$find(original_ts_gains_stats_df$TS_gains)
original_ts_gains_stats_df$Node_level = factor(as.vector(original_ts_gains_stats_df$Node_level), levels=seq(0,10))
original_ts_gains_stats_df$TS_gains = factor(original_ts_gains_stats_df$TS_gains, levels=c(all_ancestors, original_species))
original_ts_gains_stats_df$Category = factor(original_ts_gains_stats_df$Category, levels=c("ALL", "STRICT", "LASSO"))

#######################################################
#################### PLOT ANCESTORS ###################
#######################################################
random_ancestors_plot = ggplot() +
  geom_jitter(data=subset(all_rand_tissue_gains_stats_df, Node_level!=10), 
                                aes(x=TS_gains, y=Tot_gains, group=Category, color=Randomization_round)) +
  geom_boxplot(data=subset(all_rand_tissue_gains_stats_df, Node_level!=10),
               aes(x=TS_gains, y=Tot_gains, group=TS_gains), alpha=0.7) +
  geom_line(data=subset(original_ts_gains_stats_df, Node_level!=10),
            aes(x=TS_gains, y=Tot_gains, group=Category, shape=Category), color="purple4") +
  theme_bw() +
  theme(axis.text.x = element_text(color="black", angle=45, hjust=1, vjust=1),
        axis.text.y = element_text(color="black"),
        axis.title = element_text(color="black")) +
  theme(legend.position = "bottom") +
  facet_wrap(~Category+Branch, scales="free_x", ncol=2)

print(random_ancestors_plot)

#######################################################
#################### PLOT SPECIES #####################
#######################################################
random_species_plot = ggplot() +
  geom_jitter(data=subset(all_rand_tissue_gains_stats_df, Node_level==10), 
                                aes(x=TS_gains, y=Tot_gains, group=Category, color=Randomization_round)) +
  geom_boxplot(data=subset(all_rand_tissue_gains_stats_df, Node_level==10),
               aes(x=TS_gains, y=Tot_gains, group=TS_gains), alpha=0.7) +
  geom_line(data=subset(original_ts_gains_stats_df, Node_level==10),
            aes(x=TS_gains, y=Tot_gains, group=Category, shape=Category), color="purple4") +
  theme_bw() +
  theme(axis.text.x = element_text(color="black", angle=45, hjust=1, vjust=1),
        axis.text.y = element_text(color="black"),
        axis.title = element_text(color="black")) +
  theme(legend.position = "bottom") +
  facet_wrap(~Category+Branch, scales="free_x", ncol=2)

print(random_species_plot)
```


<!-- ALL SPECIES TS -->
```{r tissue_specific_OGs_dup_stats, echo=FALSE, warning=FALSE, messsage=FALSE, dependson=c("setup", "functions")}
#######################################################
######### UPLOAD ORTHOGROUPS ##########################
#######################################################
orthogroups_df = read.delim(paste0(ts_gains_losses_raw_dir, "Bilateria_conserved_orthogroups-All_genes-associated_tissues.tab"),
                            header=FALSE, col.names=c("OG_ID", "Species", "GeneID", "Tau", "Tissue", "Paralog_tissue", "Expr_cutoff"), stringsAsFactors = FALSE)

####### Add OGID_Species column
orthogroups_df$OG_ID_Species = paste0(orthogroups_df$OG_ID, "_", orthogroups_df$Species)
####### Count number of paralogs per species in each orthogroup
species_paralogs_count_df = as.data.frame(table(orthogroups_df$OG_ID_Species))
OG_ID_Species_paralogs_counts_dict = hashmap(as.vector(species_paralogs_count_df$Var1), as.vector(species_paralogs_count_df$Freq))
####### Add paralog count to dataframe
orthogroups_df$Paralogs_count = OG_ID_Species_paralogs_counts_dict$find(orthogroups_df$OG_ID_Species)

#######################################################
######### BACKGROUND ##################################
#######################################################

background_genes_orthogroups_df = orthogroups_df %>%
  mutate(Dup_status = case_when(Paralogs_count > 1 ~ "Dup",
                                Paralogs_count <= 1 ~ "Non_Dup")) %>%
  dplyr::select(OG_ID, Species, Dup_status) %>%
  distinct() %>%
  group_by(Species, Dup_status) %>%
  summarize(Tot_dup_status = n()) %>%
  ungroup() %>%
  group_by(Species) %>%
  mutate(Prop_dup_status = Tot_dup_status/sum(Tot_dup_status)) %>%
  filter(Dup_status=="Dup") %>%
  mutate(Species=factor(Species, ordered_species_tree))

#######################################################
######### UPLOAD TS ORTHOGROUPS #######################
#######################################################

ts_genes_orthogroups_df = orthogroups_df %>%
  filter(Tissue!="none", Tau>=tau_cutoff, Expr_cutoff=="YES_EXPR") %>%
  mutate(Dup_status = case_when(Paralogs_count > 1 ~ "Dup",
                                TRUE ~ "Non_Dup")) %>%
  group_by(Species, Dup_status) %>%
  summarize(Tot_dup_status = n()) %>%
  ungroup() %>%
  group_by(Species) %>%
  mutate(Prop_dup_status = Tot_dup_status/sum(Tot_dup_status)) %>%
  mutate(Branch = ifelse(Species %in% Deuterostoma, "Deuterostoma", "Protostoma")) %>%
  unite(Branch_Dup_status, c("Branch", "Dup_status"), remove=FALSE, sep="_") %>%
  mutate(Branch_Dup_status = factor(Branch_Dup_status, levels=rev(c("Deuterostoma_Dup", "Deuterostoma_Non_Dup", 
                                                                    "Protostoma_Dup", "Protostoma_Non_Dup")))) %>%
  mutate(Species=factor(Species, ordered_species_tree))

#######################################################
######### UPLOAD RANDOM ORTHOGROUPS ###################
#######################################################

all_random_orthogroups_df = vector()
for (randomization_round in seq(1,10)) {
  ######### Set directory
  randomization_dir = paste0(home, "projects/bilaterian_GE/data/ts_gains_losses_revision/randomization_", randomization_round, "/")
  ######### Upload input
  random_orthogroups_df = read.delim(paste0(randomization_dir, "Bilateria_conserved_orthogroups-All_genes-associated_tissues.tab"),
                                        header=FALSE, col.names=c("OG_ID", "Species", "GeneID", "Tau", "Tissue", "Paralog_tissue", "Expr_cutoff"), 
                                        stringsAsFactors = FALSE)
  ####### Add OGID_Species column
  random_orthogroups_df$OG_ID_Species = paste0(random_orthogroups_df$OG_ID, "_", random_orthogroups_df$Species)
  ####### Count number of paralogs per species in each orthogroup
  species_paralogs_count_df = as.data.frame(table(random_orthogroups_df$OG_ID_Species))
  OG_ID_Species_paralogs_counts_dict = hashmap(as.vector(species_paralogs_count_df$Var1), as.vector(species_paralogs_count_df$Freq))
  ####### Add paralog count to dataframe
  random_orthogroups_df$Paralogs_count = OG_ID_Species_paralogs_counts_dict$find(random_orthogroups_df$OG_ID_Species)
  ####### Add randomization round
  random_orthogroups_df$Randomization_round = rep(randomization_round, nrow(random_orthogroups_df))
  ####### Join to final dataframe
  all_random_orthogroups_df = rbind(all_random_orthogroups_df, random_orthogroups_df)
}

ts_genes_all_random_orthogroups_df = all_random_orthogroups_df %>%
  filter(Tissue!="none", Tau>=tau_cutoff, Expr_cutoff=="YES_EXPR") %>%
  mutate(Dup_status = case_when(Paralogs_count > 1 ~ "Dup",
                                TRUE ~ "Non_Dup")) %>%
  group_by(Species, Randomization_round, Dup_status) %>%
  summarize(Tot_dup_status = n()) %>%
  ungroup() %>%
  group_by(Species, Randomization_round) %>%
  mutate(Prop_dup_status = Tot_dup_status/sum(Tot_dup_status)) %>%
  ungroup() %>%
  filter(Dup_status=="Dup") %>%
  mutate(Species=factor(Species, ordered_species_tree))

#########################################################
######## COMPUTE BINOMIAL TEST ##########################
#########################################################
#Here the observed is the random and the expected is the proportion from the observed
binomial_res_df = ts_genes_all_random_orthogroups_df %>%
  group_by(Species) %>%
  summarize(Median_freq_proportion = median(Prop_dup_status)) %>%
  ungroup() %>%
  left_join(ts_genes_orthogroups_df %>% 
              filter(Dup_status=="Dup") %>%
              dplyr::select(Species, Branch, Prop_dup_status) %>%
              rename(Expected = Prop_dup_status),
            by="Species") %>%
  ungroup() %>%
  mutate(X=round(Median_freq_proportion*100, 0)) %>%
  dplyr::rowwise() %>%
  mutate(binom_pvalue = binom.test(x=X, n=100, p=Expected, alternative="less")$p.value) %>%
  mutate(binom_signif = ifelse(binom_pvalue <= 0.05, "*", "")) %>%
  mutate(Species=factor(Species, levels=ordered_species_tree))

#######################################################
################# PLOT ################################
#######################################################

all_ts_genes_duplication_status = ggplot() +
  geom_bar(data=ts_genes_orthogroups_df, 
           aes(x=Species, y=Prop_dup_status, fill=Branch_Dup_status), 
           stat="identity", position="fill", color="black", alpha=0.75) +
  geom_boxplot(data=ts_genes_all_random_orthogroups_df, 
               aes(x=Species, y=Prop_dup_status), 
               color="black", fill="black", size=0.5, alpha=0.4, outlier.size = 0.5) +
  
  geom_errorbar(data=background_genes_orthogroups_df, aes(x=Species, ymin=Prop_dup_status, ymax=Prop_dup_status), color="black", linetype="dashed") +
  geom_text(data=binomial_res_df, aes(x=Species, y=0.95, label=binom_signif), color="black", size=6) +
  scale_fill_manual(values=c("Deuterostoma_Dup"="darkolivegreen4", "Protostoma_Dup"="tan3", "Deuterostoma_Non_Dup"="white", "Protostoma_Non_Dup"="white")) +
  theme_bw() +
  theme(axis.text.y = element_text(color="black"),
        axis.title = element_blank(),
        axis.text.x = element_text(color="black", angle=45, hjust=1, vjust=1),
        axis.ticks.y = element_line()) +
  ylab("Proportion of OGs with gains") +
  guides(fill="none")
  
print(all_ts_genes_duplication_status)
```


<!-- ########################################################### -->
<!-- ######## further exploring the connection with paralogy ### -->
<!-- ########################################################### -->

```{r plot_correlation_ancestor_number_of_paralogs, echo=FALSE, warning=FALSE, messsage=FALSE, dependson=c("setup", "functions"), fig.width=8, fig.height=9}
#############################
#### Upload orthogroups #####
#############################
#Upload orthogroups and get duplication info about each gene
orthogroups_df = read.delim(paste0(ts_gains_losses_raw_dir, "Bilateria_conserved_orthogroups-All_genes-associated_tissues.tab"),
                            header=FALSE, col.names=c("OG_ID", "Species", "GeneID", "Tau", "Tissue", "Paralog_tissue", "Expr_cutoff"), stringsAsFactors = FALSE)
#### Count number of genes per orthogroup
orthogroups_df = orthogroups_df %>%
  group_by(OG_ID) %>%
  summarize(Tot_genes = n()) %>%
  ungroup()

#############################
#### Upload TS gains ########
#############################
tissue_gains_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
  #### Add Tissue
  input_df$Tissue = rep(tissue, nrow(input_df))
  #### Add Node level
  input_df$Node_level = ancestor_node_level_dict$find(input_df$TS_gains)
  tissue_gains_df = rbind(tissue_gains_df, input_df)
}

#############################
#### Combine dataframes #####
#############################
#Get a dataframe with TS gain and number of genes in 
tissue_gains_gene_count_df_1 = tissue_gains_df %>%
  left_join(orthogroups_df, by="OG_ID") %>%
  mutate(Node_level=factor(Node_level)) %>%
  mutate(Branch = case_when(TS_gains %in% c(Deuterostoma, deuterostoma_y_labels[1:(length(deuterostoma_y_labels)-1)]) ~ "Deuterostoma", 
                            TS_gains %in% c(Protostoma, protostoma_y_labels[1:(length(protostoma_y_labels)-1)]) ~ "Protostoma",
                            TRUE ~ "Bilateria"))
##Duplicated the Bilateria
tissue_gains_gene_count_df = tissue_gains_gene_count_df_1 %>%
  filter(Branch=="Bilateria") %>%
  mutate(Branch="Protostoma") %>%
  dplyr::bind_rows(tissue_gains_gene_count_df_1 %>%
                     mutate(Branch = case_when(Branch=="Bilateria" ~ "Deuterostoma",
                                               TRUE ~ Branch)))

#############################
#### Compute correlations ####
#############################
correlations_df = tissue_gains_gene_count_df %>%
  group_by(Branch) %>%
  summarize(cor_value = cor.test(as.numeric(Node_level), Tot_genes, method = "pearson")$estimate,
            cor_pvluae = cor.test(as.numeric(Node_level), Tot_genes, method = "pearson")$p.value)

#############################
######## PLOT ###############
#############################

OBSERVED_tissue_gains_gene_count_plot = ggplot(data=tissue_gains_gene_count_df, aes(x=Node_level, y=Tot_genes)) +
  geom_jitter(color="dodgerblue2", alpha=0.3, size=1) +
  geom_boxplot(alpha=0.7) +
  geom_text(data=correlations_df, aes(x=3, y=120, label=paste0("Pearson cor: ", round(cor_value, 2))), color="red") +
  theme_bw() +
  theme(axis.text=element_text(color="black"),
        axis.title=element_text(color="black"),
        plot.title=element_text(color="black", face="bold", hjust=0.5)) +
  facet_wrap(~Branch) +
  ggtitle("Observed")

#print(tissue_gains_gene_count_plot)

########################################
#### Upload RANDOMIZED TS gains ########
########################################

tissue_gains_df = vector()
for (randomization_round in seq(1,10)) {
  randomization_dir = paste0(home, "projects/bilaterian_GE/data/ts_gains_losses_revision/randomization_", randomization_round, "/")
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(randomization_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
  #### Add Tissue
  input_df$Tissue = rep(tissue, nrow(input_df))
  #### Add Node level
  input_df$Node_level = ancestor_node_level_dict$find(input_df$TS_gains)
  #### Add randomization round
  input_df$Randomization_round = rep(randomization_round, nrow(input_df))
  tissue_gains_df = rbind(tissue_gains_df, input_df)
 }
}

#############################
#### Combine dataframes #####
#############################
#Get a dataframe with TS gain and number of genes in 
tissue_gains_gene_count_df_1 = tissue_gains_df %>%
  left_join(orthogroups_df, by="OG_ID") %>%
  mutate(Node_level=factor(Node_level)) %>%
  mutate(Branch = case_when(TS_gains %in% c(Deuterostoma, deuterostoma_y_labels[1:(length(deuterostoma_y_labels)-1)]) ~ "Deuterostoma", 
                            TS_gains %in% c(Protostoma, protostoma_y_labels[1:(length(protostoma_y_labels)-1)]) ~ "Protostoma",
                            TRUE ~ "Bilateria"))
##Duplicated the Bilateria
tissue_gains_gene_count_df = tissue_gains_gene_count_df_1 %>%
  filter(Branch=="Bilateria") %>%
  mutate(Branch="Protostoma") %>%
  dplyr::bind_rows(tissue_gains_gene_count_df_1 %>%
                     mutate(Branch = case_when(Branch=="Bilateria" ~ "Deuterostoma",
                                               TRUE ~ Branch)))

#############################
#### Compute correlations ####
#############################
correlations_df = tissue_gains_gene_count_df %>%
  group_by(Branch) %>%
  summarize(cor_value = cor.test(as.numeric(Node_level), Tot_genes, method = "pearson")$estimate,
            cor_pvluae = cor.test(as.numeric(Node_level), Tot_genes, method = "pearson")$p.value)

#############################
######## PLOT ###############
#############################

RANDOMIZED_tissue_gains_gene_count_plot = ggplot(data=tissue_gains_gene_count_df, aes(x=Node_level, y=Tot_genes)) +
  geom_jitter(color="dodgerblue2", alpha=0.3, size=1) +
  geom_boxplot(alpha=0.7) +
  geom_text(data=correlations_df, aes(x=3, y=120, label=paste0("Pearson cor: ", round(cor_value, 2))), color="red") +
  theme_bw() +
  theme(axis.text=element_text(color="black"),
        axis.title=element_text(color="black"),
        plot.title=element_text(color="black", face="bold", hjust=0.5)) +
  facet_wrap(~Branch) +
  ggtitle("10 randomizations")

#print(tissue_gains_gene_count_plot)

plot_grid(OBSERVED_tissue_gains_gene_count_plot, RANDOMIZED_tissue_gains_gene_count_plot, ncol=1, align="hv")
```

```{r plot_number_paralogs_earliest_diverging_species, echo=FALSE, warning=FALSE, messsage=FALSE, dependson=c("setup", "functions"), fig.width=6, fig.height=8}
#############################
#### Upload orthogroups #####
#############################
#Upload orthogroups and get duplication info about each gene
orthogroups_df = read.delim(paste0(ts_gains_losses_raw_dir, "Bilateria_conserved_orthogroups-All_genes-associated_tissues.tab"),
                            header=FALSE, col.names=c("OG_ID", "Species", "GeneID", "Tau", "Tissue", "Paralog_tissue", "Expr_cutoff"), stringsAsFactors = FALSE)
#### Add column with OG_ID_Species and count number of paralogs per gene
orthogroups_df$OG_ID_Species = paste0(orthogroups_df$OG_ID, "_", orthogroups_df$Species)
species_paralogs_count_df = as.data.frame(table(orthogroups_df$OG_ID_Species))
OG_ID_Species_paralogs_counts_dict = hashmap(as.vector(species_paralogs_count_df$Var1), as.vector(species_paralogs_count_df$Freq))
orthogroups_df$Paralogs_count = OG_ID_Species_paralogs_counts_dict$find(orthogroups_df$OG_ID_Species)

#############################
#### Upload TS gains ########
#############################
tissue_gains_df = vector()

for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
  #### Add Tissue
  input_df$Tissue = rep(tissue, nrow(input_df))
  #### Add Node level
  input_df$Node_level = ancestor_node_level_dict$find(input_df$TS_gains)
  #### Add randomization category
  input_df$Randomization_round = rep("Observed", nrow(input_df))
  tissue_gains_df = rbind(tissue_gains_df, input_df)
}

########################################
#### Upload RANDOMIZED TS gains ########
########################################
#Join to the same dataframe as before
for (randomization_round in seq(1,10)) {
  randomization_dir = paste0(home, "projects/bilaterian_GE/data/ts_gains_losses_revision/randomization_", randomization_round, "/")
  for (tissue in ordered_tissues) {
    input_df = read.delim(paste0(randomization_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
    #### Add Tissue
    input_df$Tissue = rep(tissue, nrow(input_df))
    #### Add Node level
    input_df$Node_level = ancestor_node_level_dict$find(input_df$TS_gains)
    #### Add randomization category
    input_df$Randomization_round = rep("Randomized", nrow(input_df))
    tissue_gains_df = rbind(tissue_gains_df, input_df)
  }
}

###########################################################
#### Add earliest diverging species and count paralogs ####
###########################################################
tissue_gains_diverging_species_df = tissue_gains_df %>%
  inner_join(earliest_diverging_species_df, by=c("TS_gains"="Ancestor")) %>%
  unite(OG_ID_Species, c(OG_ID, Diverging_species), remove=FALSE) %>%
  left_join(orthogroups_df %>% 
              select(OG_ID_Species, Paralogs_count) %>% 
              distinct(), 
            by="OG_ID_Species") %>%
  mutate(Diverging_species = factor(Diverging_species, levels=original_species))


########################################
#### PLot ##############################
########################################
earliest_diverging_species_plot = ggplot(data=tissue_gains_diverging_species_df, 
                                         aes(x=Randomization_round, y=Paralogs_count)) +
  geom_boxplot() +
  theme_bw() +
  facet_wrap(~Diverging_species, scales = "free_y")

print(earliest_diverging_species_plot)
```



<!-- ########################################################### -->
<!-- ############## Backup ##################################### -->
<!-- ########################################################### -->

```{r check_zscore_correlation_all_OGs_v1, warning=FALSE, message=FALSE, echo=FALSE, dependson=c("setup", "functions")}
##############################################
##### GENERATE INPUT TABLE ###################
##############################################
my_tissue = "Neural"

######### Upload table
expr_file = paste0(pca_dir, "/", version, "/", category, "/", clade, "/", evo_type, "/Bilateria_conserved-metasamples_median_expr-NOSVA-log2-TPMs-NORM-species_zscore-BH_genes.tab")
expr_df = read.delim(expr_file, sep="\t", header=TRUE, row=1)
expr_df = expr_df[complete.cases(expr_df),] #I am using only the cases without NAs, so that I don't have to do any imputation.
#Update colnames
colnames(expr_df) = gsub("BmA", "Bmo", gsub("Hs2", "Hsa", gsub("Mm2", "Mmu", gsub("Bt2", "Bta", gsub("Sp2", "Spu", colnames(expr_df))))))

######## Filter expression table for only one tissue
tissue_samples = rownames(subset(pca_metadata_df, tissues==my_tissue))
expr_df = expr_df[,tissue_samples]

##############################################
##### COMPUTE CORRELATIONS ###################
##############################################

###### Generate correlation matrix
all_cor_matrix = cor(expr_df, method="pearson")

all_cor_long_df = all_cor_matrix %>%
  as.data.frame() %>%
  rownames_to_column("Sample1") %>%
  pivot_longer(cols=colnames(expr_df), names_to = "Sample2", values_to="Cor_values") %>%
  mutate(Species1 = gsub("_.*", "", Sample1)) %>%
  mutate(Species2 = gsub("_.*", "", Sample2)) %>%
  unite("Species_combo", Species1:Species2, sep="_", remove=FALSE) %>%
  left_join(all_species_distances_df, by="Species_combo") %>%
  mutate(all_species_distances = factor(all_species_distances, levels=seq(0,10)))

##############################################
##### PLOT CORRELATIONS ######################
##############################################
tot_OGs = nrow(expr_df)

zscores_correlation_plots = ggplot(data=all_cor_long_df, aes(x=all_species_distances, y=Cor_values, group=all_species_distances, color=Species1)) +
  geom_jitter(alpha=0.8) +
  geom_boxplot(color="black", alpha=0.2) +
  ggtitle(paste0(my_tissue, " ~ OG num: ", tot_OGs)) +
  scale_color_manual(values=species_colors) +
  theme_bw() +
  theme(plot.title = element_text(color="black", hjust=0.5, face="bold"),
        axis.text = element_text(color="black"),
        axis.title = element_text(color="black")) +
  guides(color=guide_legend(ncol=2, title="One species"))+
  xlab("Species distances") +
  ylab("Pearson correlation")

print(zscores_correlation_plots)
```

```{r PC_batch_correlation_custom_v1, warning=FALSE, echo=FALSE, message=FALSE, dependson=c("setup", "extra_functions")}
######################################################
####### RUN PCA AND ISOLATE FIRST PC #################
######################################################
### Define path to expression file
expr_file = paste0(pca_dir, "/", version, "/", category, "/", clade, "/", evo_type, "/Bilateria_conserved-metasamples_median_expr-NOSVA-log2-TPMs-NORM-BH_genes.tab")
expr_df = read.delim(expr_file, sep="\t", header=TRUE, row=1)
expr_df = expr_df[complete.cases(expr_df),] #I am using only the cases without NAs, so that I don't have to do any imputation.
### Update colnames
colnames(expr_df) = gsub("BmA", "Bmo", gsub("Hs2", "Hsa", gsub("Mm2", "Mmu", gsub("Bt2", "Bta", gsub("Sp2", "Spu", colnames(expr_df))))))
### Run PCA
PCA_res = prcomp(as.matrix(t(expr_df)), center = TRUE, scale = TRUE) #center and scale the PCs.
ordered_PCA_res = as.data.frame(PCA_res$x[order(rownames(PCA_res$x)),])
PC1_values_df = ordered_PCA_res %>%
  select(PC1) %>%
  rownames_to_column("Species_metasample")

######################################################
####### GENERATE EXTRA METADATA ######################
######################################################
##### This will include both continuous and categorical variables for each metasample
#Categorical: Public vs in_house, SE vs PS, sequencing technology
#Continuos: Tot number of initial samples, Tot number of initial reads, Read length
in_house_sequencing = "Illumina_HiSeq_2500"
in_house_sequencing_Obi = "Illumina_NovaSeq"
in_house_bioproject = "PRJNA846095"

#### Build the metadata table
#Original metadata
all_species_original_metadata_df = vector()
for (species in all_species) {
  file_species = species
  if (species == "BmA") {file_species = "Bmo"}
  if (species == "Hsa") {file_species = "Hs2"}
  if (species == "Mmu") {file_species = "Mm2"}
  if (species == "Bta") {file_species = "Bt2"}
  if (species == "Spu") {file_species = "Sp2"}
  species_original_metadata_df = read.delim(paste0(metadata_dir, file_species, "_samples_info.tab"), header=TRUE)
  species_original_metadata_df$Species = rep(species, nrow(species_original_metadata_df))
  ### Modify colnames as needed
  colnames(species_original_metadata_df)[colnames(species_original_metadata_df) == "Group"] = "Metasample"
  ### Join to final dataframe
  all_species_original_metadata_df = rbind(all_species_original_metadata_df, species_original_metadata_df[,1:8])
}

#Extra metadata
all_species_extra_metadata_df = vector()
for (species in all_species) {
  file_species = species
  if (species == "BmA") {file_species = "Bmo"}
  if (species == "Hsa") {file_species = "Hs2"}
  if (species == "Mmu") {file_species = "Mm2"}
  if (species == "Bta") {file_species = "Bt2"}
  if (species == "Spu") {file_species = "Sp2"}
  species_extra_metadata_df = read.delim(paste0(metadata_dir, "extra_metadata/", file_species, "-extra_samples_info.tab"), header=TRUE, stringsAsFactors = FALSE)
  #species_extra_metadata_df$Species = rep(species, nrow(species_extra_metadata_df))
  ### Modify colnames as needed
  colnames(species_extra_metadata_df)[colnames(species_extra_metadata_df)=="Run"] = "SRA"
  ### Join to final dataframe
  all_species_extra_metadata_df = rbind(all_species_extra_metadata_df, species_extra_metadata_df)
}

all_species_all_metadata_df = all_species_original_metadata_df %>%
  left_join(all_species_extra_metadata_df, by="SRA") %>%
  select(-ProjectID) %>%
  ### Add information for in_house samples
  mutate(Platform = case_when(SRA=="in_house" ~ "ILLUMINA",
                              TRUE ~ Platform),
         Model = case_when(SRA=="in_house" & Species != "Obi" ~ in_house_sequencing,
                           SRA=="in_house" & Species == "Obi" ~ in_house_sequencing_Obi,
                           TRUE ~ Model),
         BioProject = case_when(SRA=="in_house" ~ in_house_bioproject,
                                TRUE ~ BioProject),
         Origin = case_when(SRA=="in_house" ~ "in_house", 
                            TRUE ~ "public")) %>%
  ### Change read numbers to numeric
  mutate(Read_number = as.numeric(gsub(",", "", Read_number)))

#### Collapse info at the metasample level
all_species_collapse_metadata_df = all_species_all_metadata_df %>%
  group_by(Species, Tissue, Metasample, SE_PE, Platform, Model, Origin) %>%
  summarize(Mean_read_len = round(mean(Read_len)),
            Tot_read_num = sum(Read_number),
            Tot_original_samples = n()) %>%
  ungroup() %>%
  unite(Species_metasample, c("Species", "Metasample"), remove=FALSE, sep="_") %>%
  ## Select only metasamples without ambiguities (no duplications)
  group_by(Species_metasample) %>%
  filter(n()==1) %>%
  ## Add PC1 values
  left_join(PC1_values_df, by="Species_metasample")

######################################################
####### PLOT PC1 DISTRIBUTIONS #######################
######################################################
tot_metasamples = nrow(all_species_collapse_metadata_df)
discrete_variables = c("SE_PE", "Model", "Origin")
continuos_variables = c("Mean_read_len", "Tot_read_num", "Tot_original_samples")

for (variable in discrete_variables) {
  discrete_var_PC1_plot = ggplot(data=all_species_collapse_metadata_df, aes_string(x=variable, y="PC1")) +
    geom_jitter(color="dimgray", alpha=0.6) +
    geom_boxplot(alpha=0.3) +
    stat_compare_means() +
    theme_bw() +
    ggtitle(paste0(variable, " ~ Tot dots plotted: ", tot_metasamples)) +
    theme(plot.title = element_text(color="black", hjust=0.5, face="bold"))
  
  print(discrete_var_PC1_plot)
}

for (variable in continuos_variables) {
  continuos_var_PC1_plot = ggplot(data=all_species_collapse_metadata_df, aes_string(x=variable, y="PC1")) +
    geom_jitter(color="dimgray", alpha=0.6) +
    geom_smooth(method="lm") +
    theme_bw() +
    ggtitle(paste0(variable, " ~ Tot dots plotted: ", tot_metasamples)) +
    theme(plot.title = element_text(color="black", hjust=0.5, face="bold"))
  
  print(continuos_var_PC1_plot)
}

continuos_var_PC1_plot = ggplot(data=subset(all_species_collapse_metadata_df, Mean_read_len <= 125), aes_string(x="Mean_read_len", y="PC1")) +
  geom_point(color="dimgray", alpha=0.6) +
  geom_smooth(method="lm") +
  theme_bw() +
  ggtitle(paste0(variable, " ~ Tot dots plotted: ", tot_metasamples)) +
  theme(plot.title = element_text(color="black", hjust=0.5, face="bold"))

print(continuos_var_PC1_plot)

continuos_var_PC1_plot = ggplot(data=subset(all_species_collapse_metadata_df, Tot_read_num <= 100000000 & Tot_read_num >= 80000000), aes_string(x="Tot_read_num", y="PC1")) +
  geom_point(color="dimgray", alpha=0.6) +
  geom_smooth(method="lm") +
  theme_bw() +
  ggtitle(paste0(variable, " ~ Tot dots plotted: ", tot_metasamples)) +
  theme(plot.title = element_text(color="black", hjust=0.5, face="bold"))

print(continuos_var_PC1_plot)
```

```{r PC_batch_correlation_custom_v2, warning=FALSE, echo=FALSE, message=FALSE, dependson=c("setup", "extra_functions")}
######################################################
####### RUN PCA AND ISOLATE FIRST PC #################
######################################################
### Define path to expression file
expr_file = paste0(pca_dir, "/", version, "/", category, "/", clade, "/", evo_type, "/Bilateria_conserved-metasamples_median_expr-NOSVA-log2-TPMs-NORM-BH_genes.tab")
expr_df = read.delim(expr_file, sep="\t", header=TRUE, row=1)
expr_df = expr_df[complete.cases(expr_df),] #I am using only the cases without NAs, so that I don't have to do any imputation.
### Update colnames
colnames(expr_df) = gsub("BmA", "Bmo", gsub("Hs2", "Hsa", gsub("Mm2", "Mmu", gsub("Bt2", "Bta", gsub("Sp2", "Spu", colnames(expr_df))))))
### Run PCA
PCA_res = prcomp(as.matrix(t(expr_df)), center = TRUE, scale = TRUE) #center and scale the PCs.
ordered_PCA_res = as.data.frame(PCA_res$x[order(rownames(PCA_res$x)),])
PC1_values_df = ordered_PCA_res %>%
  select(PC1) %>%
  rownames_to_column("Species_metasample")

######################################################
####### GENERATE EXTRA METADATA ######################
######################################################
##### This will include both continuous and categorical variables for each metasample
#Categorical: Public vs in_house, SE vs PS, sequencing technology
#Continuos: Tot number of initial samples, Tot number of initial reads, Read length
in_house_sequencing = "Illumina_HiSeq_2500"
in_house_sequencing_Obi = "Illumina_NovaSeq"
in_house_bioproject = "PRJNA846095"

#### Build the metadata table
#Original metadata
all_species_original_metadata_df = vector()
for (species in all_species) {
  file_species = species
  if (species == "BmA") {file_species = "Bmo"}
  if (species == "Hsa") {file_species = "Hs2"}
  if (species == "Mmu") {file_species = "Mm2"}
  if (species == "Bta") {file_species = "Bt2"}
  if (species == "Spu") {file_species = "Sp2"}
  species_original_metadata_df = read.delim(paste0(metadata_dir, file_species, "_samples_info.tab"), header=TRUE)
  species_original_metadata_df$Species = rep(species, nrow(species_original_metadata_df))
  ### Modify colnames as needed
  colnames(species_original_metadata_df)[colnames(species_original_metadata_df) == "Group"] = "Metasample"
  ### Join to final dataframe
  all_species_original_metadata_df = rbind(all_species_original_metadata_df, species_original_metadata_df[,1:8])
}

#Extra metadata
all_species_extra_metadata_df = vector()
for (species in all_species) {
  file_species = species
  if (species == "BmA") {file_species = "Bmo"}
  if (species == "Hsa") {file_species = "Hs2"}
  if (species == "Mmu") {file_species = "Mm2"}
  if (species == "Bta") {file_species = "Bt2"}
  if (species == "Spu") {file_species = "Sp2"}
  species_extra_metadata_df = read.delim(paste0(metadata_dir, "extra_metadata/", file_species, "-extra_samples_info.tab"), header=TRUE, stringsAsFactors = FALSE)
  #species_extra_metadata_df$Species = rep(species, nrow(species_extra_metadata_df))
  ### Modify colnames as needed
  colnames(species_extra_metadata_df)[colnames(species_extra_metadata_df)=="Run"] = "SRA"
  ### Join to final dataframe
  all_species_extra_metadata_df = rbind(all_species_extra_metadata_df, species_extra_metadata_df)
}

all_species_all_metadata_df = all_species_original_metadata_df %>%
  left_join(all_species_extra_metadata_df, by="SRA") %>%
  select(-ProjectID) %>%
  ### Add information for in_house samples
  mutate(Platform = case_when(SRA=="in_house" ~ "ILLUMINA",
                              TRUE ~ Platform),
         Model = case_when(SRA=="in_house" & Species != "Obi" ~ in_house_sequencing,
                           SRA=="in_house" & Species == "Obi" ~ in_house_sequencing_Obi,
                           TRUE ~ Model),
         BioProject = case_when(SRA=="in_house" ~ in_house_bioproject,
                                TRUE ~ BioProject),
         Origin = case_when(SRA=="in_house" ~ "in_house", 
                            TRUE ~ "public")) %>%
  ### Change read numbers to numeric
  mutate(Read_number = as.numeric(gsub(",", "", Read_number)))

#### Collapse info at the metasample level
all_species_collapse_metadata_df = all_species_all_metadata_df %>%
  group_by(Species, Tissue, Metasample, SE_PE, Platform, Model, Origin) %>%
  summarize(Mean_read_len = round(mean(Read_len)),
            Tot_read_num = sum(Read_number),
            Tot_original_samples = n()) %>%
  ungroup() %>%
  unite(Species_metasample, c("Species", "Metasample"), remove=FALSE, sep="_") %>%
  ## Select only metasamples without ambiguities (no duplications)
  group_by(Species_metasample) %>%
  filter(n()==1) %>%
  ## Add PC1 values
  left_join(PC1_values_df, by="Species_metasample") %>%
  mutate(Origin = factor(Origin, levels=c("public", "in_house")),
         SE_PE = factor(SE_PE, levels=c("SE", "PE")))

###################################################################
################ PLOT ORIGINAL PC1 DISTRIBUTIONS ##################
###################################################################
tot_metasamples = nrow(all_species_collapse_metadata_df)
discrete_variables = c("SE_PE", "Origin", "Model")
continuos_variables = c("Mean_read_len", "Tot_read_num", "Tot_original_samples")

for (variable in discrete_variables) {
  if (variable=="Model") {axis_text_angle_x = 30; axis_text_hjust_x=1; axis_text_vjust_x=1} else {axis_text_angle_x = 0; axis_text_hjust_x=0.5; axis_text_vjust_x=0.5}
  
  discrete_var_PC1_plot = ggplot(data=all_species_collapse_metadata_df, aes_string(x=variable, y="PC1")) +
    geom_jitter(color="dimgray", alpha=0.6) +
    geom_boxplot(alpha=0.3) +
    stat_compare_means() +
    ggtitle(gsub("_", " ", variable)) +
    theme_bw() +
    #ggtitle(paste0(variable, " ~ Tot dots plotted: ", tot_metasamples)) +
    theme(plot.title = element_text(color="black", hjust=0.5, face="bold"),
          axis.title = element_text(color="black"),
          axis.text.x = element_text(color="black", angle=axis_text_angle_x, vjust=1, hjust=1),
          axis.text.y = element_text(color="black"))
  
  print(discrete_var_PC1_plot)
}

for (variable in continuos_variables) {
  ## Compute pearson correlation between the variable and the coordinates
  cor_res = cor.test(as.numeric(unlist(all_species_collapse_metadata_df[,variable])), 
                     all_species_collapse_metadata_df$PC1)
  
  continuos_var_PC1_plot = ggplot(data=all_species_collapse_metadata_df, aes_string(x=variable, y="PC1")) +
    geom_jitter(color="dimgray", alpha=0.6) +
    geom_smooth(method="lm", color="black") +
    annotation_custom(grobTree(textGrob(paste0("Pearson's cor: ", round(unname(cor_res$estimate),3), 
                                               "\nPvalue: ", cor_res$p.value),
                                        x=0.7, y=0.9, hjust=0.5))) +
    ggtitle(variable) +
    theme_bw() +
    ggtitle(gsub("_", " ", variable)) +
        theme(plot.title = element_text(color="black", hjust=0.5, face="bold"),
          axis.title = element_text(color="black"),
          axis.text.x = element_text(color="black"),
          axis.text.y = element_text(color="black"))
  
  print(continuos_var_PC1_plot)
}

###################################################################
################ PLOT SELECTED PC1 DISTRIBUTIONS ##################
###################################################################

```

```{r PC_batch_correlation_custom_V3, warning=FALSE, echo=FALSE, message=FALSE, dependson=c("setup", "extra_functions")}
######################################################
####### RUN PCA AND ISOLATE FIRST PC #################
######################################################
### Define path to expression file
expr_file = paste0(pca_dir, "/", version, "/", category, "/", clade, "/", evo_type, "/Bilateria_conserved-metasamples_median_expr-NOSVA-log2-TPMs-NORM-BH_genes.tab")
expr_df = read.delim(expr_file, sep="\t", header=TRUE, row=1)
expr_df = expr_df[complete.cases(expr_df),] #I am using only the cases without NAs, so that I don't have to do any imputation.
### Update colnames
colnames(expr_df) = gsub("BmA", "Bmo", gsub("Hs2", "Hsa", gsub("Mm2", "Mmu", gsub("Bt2", "Bta", gsub("Sp2", "Spu", colnames(expr_df))))))
### Run PCA
PCA_res = prcomp(as.matrix(t(expr_df)), center = TRUE, scale = TRUE) #center and scale the PCs.
ordered_PCA_res = as.data.frame(PCA_res$x[order(rownames(PCA_res$x)),])
PC1_values_df = ordered_PCA_res %>%
  select(PC1) %>%
  rownames_to_column("Species_metasample")

######################################################
####### GENERATE EXTRA METADATA ######################
######################################################
##### This will include both continuous and categorical variables for each metasample
#Categorical: Public vs in_house, SE vs PS, sequencing technology
#Continuos: Tot number of initial samples, Tot number of initial reads, Read length
in_house_sequencing = "Illumina_HiSeq_2500"
in_house_sequencing_Obi = "Illumina_NovaSeq"
in_house_bioproject = "PRJNA846095"

#### Build the metadata table
#Original metadata
all_species_original_metadata_df = vector()
for (species in all_species) {
  file_species = species
  if (species == "BmA") {file_species = "Bmo"}
  if (species == "Hsa") {file_species = "Hs2"}
  if (species == "Mmu") {file_species = "Mm2"}
  if (species == "Bta") {file_species = "Bt2"}
  if (species == "Spu") {file_species = "Sp2"}
  species_original_metadata_df = read.delim(paste0(metadata_dir, file_species, "_samples_info.tab"), header=TRUE)
  species_original_metadata_df$Species = rep(species, nrow(species_original_metadata_df))
  ### Modify colnames as needed
  colnames(species_original_metadata_df)[colnames(species_original_metadata_df) == "Group"] = "Metasample"
  ### Join to final dataframe
  all_species_original_metadata_df = rbind(all_species_original_metadata_df, species_original_metadata_df[,1:8])
}

#Extra metadata
all_species_extra_metadata_df = vector()
for (species in all_species) {
  file_species = species
  if (species == "BmA") {file_species = "Bmo"}
  if (species == "Hsa") {file_species = "Hs2"}
  if (species == "Mmu") {file_species = "Mm2"}
  if (species == "Bta") {file_species = "Bt2"}
  if (species == "Spu") {file_species = "Sp2"}
  species_extra_metadata_df = read.delim(paste0(metadata_dir, "extra_metadata/", file_species, "-extra_samples_info.tab"), header=TRUE, stringsAsFactors = FALSE)
  #species_extra_metadata_df$Species = rep(species, nrow(species_extra_metadata_df))
  ### Modify colnames as needed
  colnames(species_extra_metadata_df)[colnames(species_extra_metadata_df)=="Run"] = "SRA"
  ### Join to final dataframe
  all_species_extra_metadata_df = rbind(all_species_extra_metadata_df, species_extra_metadata_df)
}

all_species_all_metadata_df = all_species_original_metadata_df %>%
  left_join(all_species_extra_metadata_df, by="SRA") %>%
  select(-ProjectID) %>%
  ### Add information for in_house samples
  mutate(Platform = case_when(SRA=="in_house" ~ "ILLUMINA",
                              TRUE ~ Platform),
         Model = case_when(SRA=="in_house" & Species != "Obi" ~ in_house_sequencing,
                           SRA=="in_house" & Species == "Obi" ~ in_house_sequencing_Obi,
                           TRUE ~ Model),
         BioProject = case_when(SRA=="in_house" ~ in_house_bioproject,
                                TRUE ~ BioProject),
         Origin = case_when(SRA=="in_house" ~ "in_house", 
                            TRUE ~ "public")) %>%
  ### Change read numbers to numeric
  mutate(Read_number = as.numeric(gsub(",", "", Read_number)))

#### Collapse info at the metasample level
all_species_collapse_metadata_df = all_species_all_metadata_df %>%
  group_by(Species, Tissue, Metasample, SE_PE, Platform, Model, Origin) %>%
  summarize(Mean_read_len = round(mean(Read_len)),
            Tot_read_num = sum(Read_number),
            Tot_original_samples = n()) %>%
  ungroup() %>%
  unite(Species_metasample, c("Species", "Metasample"), remove=FALSE, sep="_") %>%
  ## Select only metasamples without ambiguities (no duplications)
  group_by(Species_metasample) %>%
  filter(n()==1) %>%
  ## Add PC1 values
  left_join(PC1_values_df, by="Species_metasample") %>%
  mutate(Origin = factor(Origin, levels=c("public", "in_house")),
         SE_PE = factor(SE_PE, levels=c("SE", "PE")))

###################################################################
################ PLOT ORIGINAL PC1 DISTRIBUTIONS ##################
###################################################################
tot_metasamples = nrow(all_species_collapse_metadata_df)
discrete_variables = c("SE_PE", "Origin", "Model")
continuos_variables = c("Tot_original_samples", "Mean_read_len", "Tot_read_num")

for (variable in discrete_variables) {
  if (variable=="Model") {axis_text_angle_x = 30; axis_text_hjust_x=1; axis_text_vjust_x=1} else {axis_text_angle_x = 0; axis_text_hjust_x=0.5; axis_text_vjust_x=0.5}
  
  discrete_var_PC1_plot = ggplot(data=all_species_collapse_metadata_df, aes_string(x=variable, y="PC1")) +
    geom_jitter(color="dimgray", alpha=0.6) +
    geom_boxplot(alpha=0.3) +
    stat_compare_means() +
    ggtitle(gsub("_", " ", variable)) +
    theme_bw() +
    #ggtitle(paste0(variable, " ~ Tot dots plotted: ", tot_metasamples)) +
    theme(plot.title = element_text(color="black", hjust=0.5, face="bold"),
          axis.title = element_text(color="black"),
          axis.text.x = element_text(color="black", angle=axis_text_angle_x, vjust=1, hjust=1),
          axis.text.y = element_text(color="black"))
  
  print(discrete_var_PC1_plot)
}

for (variable in continuos_variables) {
  ## Subset for the plot post correction
  if (variable == "Mean_read_len") {subsetted_metadata_df = all_species_collapse_metadata_df %>% filter(Mean_read_len %in% seq(150,160))}
  if (variable == "Tot_original_samples") {subsetted_metadata_df = all_species_collapse_metadata_df %>% filter(Tot_original_samples == 1)}
  if (variable == "Tot_read_num") {subsetted_metadata_df = all_species_collapse_metadata_df %>% filter(Tot_read_num <= 100000000, Tot_read_num >= 80000000)}
  
  ## Compute pearson correlation between the variable and the coordinates
  original_cor_res = cor.test(as.numeric(unlist(all_species_collapse_metadata_df[,variable])), 
                     all_species_collapse_metadata_df$PC1)
  subsetted_cor_res = cor.test(as.numeric(unlist(subsetted_metadata_df[,variable])), 
                     subsetted_metadata_df$PC1)
  
  continuos_var_PC1_plot = ggplot(data=all_species_collapse_metadata_df, aes_string(x=variable, y="PC1")) +
    geom_jitter(color="dimgray", alpha=0.6) +
    geom_smooth(method="lm", color="black") +
    geom_smooth(data=subsetted_metadata_df, aes_string(x=variable, y="PC1"), method="lm", color="red") +
    annotation_custom(grobTree(textGrob(paste0("Pearson's cor: ", round(unname(original_cor_res$estimate),3), 
                                               "\nPvalue: ", original_cor_res$p.value),
                                        x=0.7, y=0.9, hjust=0.5))) +
    annotation_custom(grobTree(textGrob(paste0("Pearson's cor: ", round(unname(subsetted_cor_res$estimate),3), 
                                               "\nPvalue: ", subsetted_cor_res$p.value),
                                        x=0.7, y=0.8, hjust=0.5, gp=gpar(col="red")))) +
    ggtitle(variable) +
    theme_bw() +
    ggtitle(gsub("_", " ", variable)) +
        theme(plot.title = element_text(color="black", hjust=0.5, face="bold"),
          axis.title = element_text(color="black"),
          axis.text.x = element_text(color="black"),
          axis.text.y = element_text(color="black"))
  
  print(continuos_var_PC1_plot)
}

###################################################################
################ PLOT SELECTED PC1 DISTRIBUTIONS ##################
###################################################################

```

```{r PC_batch_correlation_custom_v4, warning=FALSE, echo=FALSE, message=FALSE, dependson=c("setup", "extra_functions")}
######################################################
####### RUN PCA AND ISOLATE FIRST PC #################
######################################################
### Define path to expression file
expr_file = paste0(pca_dir, "/", version, "/", category, "/", clade, "/", evo_type, "/Bilateria_conserved-metasamples_median_expr-NOSVA-log2-TPMs-NORM-BH_genes.tab")
expr_df = read.delim(expr_file, sep="\t", header=TRUE, row=1)
expr_df = expr_df[complete.cases(expr_df),] #I am using only the cases without NAs, so that I don't have to do any imputation.
### Update colnames
colnames(expr_df) = gsub("BmA", "Bmo", gsub("Hs2", "Hsa", gsub("Mm2", "Mmu", gsub("Bt2", "Bta", gsub("Sp2", "Spu", colnames(expr_df))))))
### Run PCA
PCA_res = prcomp(as.matrix(t(expr_df)), center = TRUE, scale = TRUE) #center and scale the PCs.
ordered_PCA_res = as.data.frame(PCA_res$x[order(rownames(PCA_res$x)),])
PC1_values_df = ordered_PCA_res %>%
  select(PC1) %>%
  rownames_to_column("Species_metasample")

######################################################
####### GENERATE EXTRA METADATA ######################
######################################################
##### This will include both continuous and categorical variables for each metasample
#Categorical: Public vs in_house, SE vs PS, sequencing technology
#Continuos: Tot number of initial samples, Tot number of initial reads, Read length
in_house_sequencing = "Illumina_HiSeq_2500"
in_house_sequencing_Obi = "Illumina_NovaSeq"
in_house_bioproject = "PRJNA846095"

#### Build the metadata table
#Original metadata
all_species_original_metadata_df = vector()
for (species in all_species) {
  file_species = species
  if (species == "BmA") {file_species = "Bmo"}
  if (species == "Hsa") {file_species = "Hs2"}
  if (species == "Mmu") {file_species = "Mm2"}
  if (species == "Bta") {file_species = "Bt2"}
  if (species == "Spu") {file_species = "Sp2"}
  species_original_metadata_df = read.delim(paste0(metadata_dir, file_species, "_samples_info.tab"), header=TRUE)
  species_original_metadata_df$Species = rep(species, nrow(species_original_metadata_df))
  ### Modify colnames as needed
  colnames(species_original_metadata_df)[colnames(species_original_metadata_df) == "Group"] = "Metasample"
  ### Join to final dataframe
  all_species_original_metadata_df = rbind(all_species_original_metadata_df, species_original_metadata_df[,1:8])
}

#Extra metadata
all_species_extra_metadata_df = vector()
for (species in all_species) {
  file_species = species
  if (species == "BmA") {file_species = "Bmo"}
  if (species == "Hsa") {file_species = "Hs2"}
  if (species == "Mmu") {file_species = "Mm2"}
  if (species == "Bta") {file_species = "Bt2"}
  if (species == "Spu") {file_species = "Sp2"}
  species_extra_metadata_df = read.delim(paste0(metadata_dir, "extra_metadata/", file_species, "-extra_samples_info.tab"), header=TRUE, stringsAsFactors = FALSE)
  #species_extra_metadata_df$Species = rep(species, nrow(species_extra_metadata_df))
  ### Modify colnames as needed
  colnames(species_extra_metadata_df)[colnames(species_extra_metadata_df)=="Run"] = "SRA"
  ### Join to final dataframe
  all_species_extra_metadata_df = rbind(all_species_extra_metadata_df, species_extra_metadata_df)
}

all_species_all_metadata_df = all_species_original_metadata_df %>%
  left_join(all_species_extra_metadata_df, by="SRA") %>%
  select(-ProjectID) %>%
  ### Add information for in_house samples
  mutate(Platform = case_when(SRA=="in_house" ~ "ILLUMINA",
                              TRUE ~ Platform),
         Model = case_when(SRA=="in_house" & Species != "Obi" ~ in_house_sequencing,
                           SRA=="in_house" & Species == "Obi" ~ in_house_sequencing_Obi,
                           TRUE ~ Model),
         BioProject = case_when(SRA=="in_house" ~ in_house_bioproject,
                                TRUE ~ BioProject),
         Origin = case_when(SRA=="in_house" ~ "in_house", 
                            TRUE ~ "public")) %>%
  ### Change read numbers to numeric
  mutate(Read_number = as.numeric(gsub(",", "", Read_number)))

#### Collapse info at the metasample level
all_species_collapse_metadata_df = all_species_all_metadata_df %>%
  group_by(Species, Tissue, Metasample, SE_PE, Platform, Model, Origin) %>%
  summarize(Mean_read_len = round(mean(Read_len)),
            Tot_read_num = sum(Read_number),
            Tot_original_samples = n()) %>%
  ungroup() %>%
  unite(Species_metasample, c("Species", "Metasample"), remove=FALSE, sep="_") %>%
  ## Select only metasamples without ambiguities (no duplications)
  group_by(Species_metasample) %>%
  filter(n()==1) %>%
  ## Add PC1 values
  left_join(PC1_values_df, by="Species_metasample") %>%
  mutate(Origin = factor(Origin, levels=c("public", "in_house")),
         SE_PE = factor(SE_PE, levels=c("SE", "PE")))

###################################################################
################ PLOT ORIGINAL PC1 DISTRIBUTIONS ##################
###################################################################
tot_metasamples = nrow(all_species_collapse_metadata_df)
discrete_variables = c("SE_PE", "Origin", "Model")
continuos_variables = c("Tot_original_samples", "Mean_read_len", "Tot_read_num")
selected_discrete_modes = c("SE_PE"="PE", "Origin"="public", "Model"="Illumina_HiSeq_2500")

for (variable in discrete_variables) {
  if (variable=="Model") {axis_text_angle_x = 30; axis_text_hjust_x=1; axis_text_vjust_x=1} else {axis_text_angle_x = 0; axis_text_hjust_x=0.5; axis_text_vjust_x=0.5}
  
  subsetted_metadata_df = all_species_collapse_metadata_df %>%
    ungroup() %>%
    filter(!!as.symbol(variable) == unname(selected_discrete_modes[variable])) %>%
    summarise(xmax = unname(selected_discrete_modes[variable]),
              xmin = unname(selected_discrete_modes[variable]),
              ymax = max(PC1),
              ymin = min(PC1))
  
  discrete_var_PC1_plot = ggplot(data=all_species_collapse_metadata_df, aes_string(x=variable, y="PC1")) +
    geom_jitter(color="dimgray", alpha=0.6) +
    geom_boxplot(alpha=0.3) +
    geom_tile(data=subsetted_metadata_df, aes(x=xmax, y=1, height = Inf, width = 1), fill="white", alpha=0, color="red", size=0.7) +
    #geom_rect(data=subsetted_metadata_df, aes(xmin=xmin+0.5, xmax=xmax, ymax=ymax-0.5, ymin=ymin), color="red", fill=NA, inherit.aes = FALSE) +
    stat_compare_means() +
    ggtitle(gsub("_", " ", variable)) +
    theme_bw() +
    #ggtitle(paste0(variable, " ~ Tot dots plotted: ", tot_metasamples)) +
    theme(plot.title = element_text(color="black", hjust=0.5, face="bold"),
          axis.title = element_text(color="black"),
          axis.text.x = element_text(color="black", angle=axis_text_angle_x, vjust=1, hjust=1),
          axis.text.y = element_text(color="black"))
  
  print(discrete_var_PC1_plot)
}

for (variable in continuos_variables) {
  ## Subset for the plot post correction
  if (variable == "Mean_read_len") {subsetted_metadata_df = all_species_collapse_metadata_df %>% filter(Mean_read_len %in% seq(140,160))}
  if (variable == "Tot_original_samples") {subsetted_metadata_df = all_species_collapse_metadata_df %>% filter(Tot_original_samples == 1)}
  if (variable == "Tot_read_num") {subsetted_metadata_df = all_species_collapse_metadata_df %>% filter(Tot_read_num <= 100000000, Tot_read_num >= 80000000)}
  
  ## Generate code to add rectangles on the plots
  subsetted_metadata_df = subsetted_metadata_df %>%
    dplyr::select(Species_metasample, as.name(variable), PC1) %>%
    ungroup() %>%
    summarise(xmax = max(!!as.symbol(variable)),
              xmin = min(!!as.symbol(variable)), 
              ymax = max(PC1),
              ymin = min(PC1))
  
  ## Compute pearson correlation between the variable and the coordinates
  original_cor_res = cor.test(as.numeric(unlist(all_species_collapse_metadata_df[,variable])), 
                     all_species_collapse_metadata_df$PC1)
  
  continuos_var_PC1_plot = ggplot(data=all_species_collapse_metadata_df, aes_string(x=variable, y="PC1")) +
    geom_jitter(color="dimgray", alpha=0.6) +
    geom_smooth(method="lm", color="black") +
    geom_rect(data=subsetted_metadata_df, aes(xmin=xmin, xmax=xmax, ymax=ymax, ymin=ymin), color="red", fill=NA, inherit.aes = FALSE) +
    annotation_custom(grobTree(textGrob(paste0("Pearson's cor: ", round(unname(original_cor_res$estimate),3), 
                                               "\nPvalue: ", original_cor_res$p.value),
                                        x=0.7, y=0.9, hjust=0.5))) +
    ggtitle(variable) +
    theme_bw() +
    ggtitle(gsub("_", " ", variable)) +
        theme(plot.title = element_text(color="black", hjust=0.5, face="bold"),
          axis.title = element_text(color="black"),
          axis.text.x = element_text(color="black"),
          axis.text.y = element_text(color="black"))
  
  print(continuos_var_PC1_plot)
}

###################################################################
################ PLOT PCA FROM SELECTION ##########################
###################################################################


```

```{r PC1_batch_correlation_custom_v5, warning=FALSE, echo=FALSE, message=FALSE, dependson=c("setup", "extra_functions"), fig.width=12, fig.height=14}
######################################################
####### RUN PCA AND ISOLATE FIRST PC #################
######################################################
### Define path to expression file
expr_file = paste0(pca_dir, "/", version, "/", category, "/", clade, "/", evo_type, "/Bilateria_conserved-metasamples_median_expr-NOSVA-log2-TPMs-NORM-BH_genes.tab")
expr_df = read.delim(expr_file, sep="\t", header=TRUE, row=1)
expr_df = expr_df[complete.cases(expr_df),] #I am using only the cases without NAs, so that I don't have to do any imputation.
### Update colnames
colnames(expr_df) = gsub("BmA", "Bmo", gsub("Hs2", "Hsa", gsub("Mm2", "Mmu", gsub("Bt2", "Bta", gsub("Sp2", "Spu", colnames(expr_df))))))
### Run PCA
PCA_res = prcomp(as.matrix(t(expr_df)), center = TRUE, scale = TRUE) #center and scale the PCs.
ordered_PCA_res = as.data.frame(PCA_res$x[order(rownames(PCA_res$x)),])
PC1_values_df = ordered_PCA_res %>%
  select(PC1) %>%
  rownames_to_column("Species_metasample")

######################################################
####### GENERATE EXTRA METADATA ######################
######################################################
##### This will include both continuous and categorical variables for each metasample
#Categorical: Public vs in_house, SE vs PS, sequencing technology
#Continuos: Tot number of initial samples, Tot number of initial reads, Read length
in_house_sequencing = "Illumina_HiSeq_2500"
in_house_sequencing_Obi = "Illumina_NovaSeq"
in_house_bioproject = "PRJNA846095"

#### Build the metadata table
#Original metadata
all_species_original_metadata_df = vector()
for (species in all_species) {
  file_species = species
  if (species == "BmA") {file_species = "Bmo"}
  if (species == "Hsa") {file_species = "Hs2"}
  if (species == "Mmu") {file_species = "Mm2"}
  if (species == "Bta") {file_species = "Bt2"}
  if (species == "Spu") {file_species = "Sp2"}
  species_original_metadata_df = read.delim(paste0(metadata_dir, file_species, "_samples_info.tab"), header=TRUE)
  species_original_metadata_df$Species = rep(species, nrow(species_original_metadata_df))
  ### Modify colnames as needed
  colnames(species_original_metadata_df)[colnames(species_original_metadata_df) == "Group"] = "Metasample"
  ### Join to final dataframe
  all_species_original_metadata_df = rbind(all_species_original_metadata_df, species_original_metadata_df[,1:8])
}

#Extra metadata
all_species_extra_metadata_df = vector()
for (species in all_species) {
  file_species = species
  if (species == "BmA") {file_species = "Bmo"}
  if (species == "Hsa") {file_species = "Hs2"}
  if (species == "Mmu") {file_species = "Mm2"}
  if (species == "Bta") {file_species = "Bt2"}
  if (species == "Spu") {file_species = "Sp2"}
  species_extra_metadata_df = read.delim(paste0(metadata_dir, "extra_metadata/", file_species, "-extra_samples_info.tab"), header=TRUE, stringsAsFactors = FALSE)
  #species_extra_metadata_df$Species = rep(species, nrow(species_extra_metadata_df))
  ### Modify colnames as needed
  colnames(species_extra_metadata_df)[colnames(species_extra_metadata_df)=="Run"] = "SRA"
  ### Join to final dataframe
  all_species_extra_metadata_df = rbind(all_species_extra_metadata_df, species_extra_metadata_df)
}

all_species_all_metadata_df = all_species_original_metadata_df %>%
  left_join(all_species_extra_metadata_df, by="SRA") %>%
  select(-ProjectID) %>%
  ### Add information for in_house samples
  mutate(Platform = case_when(SRA=="in_house" ~ "ILLUMINA",
                              TRUE ~ Platform),
         Model = case_when(SRA=="in_house" & Species != "Obi" ~ in_house_sequencing,
                           SRA=="in_house" & Species == "Obi" ~ in_house_sequencing_Obi,
                           TRUE ~ Model),
         BioProject = case_when(SRA=="in_house" ~ in_house_bioproject,
                                TRUE ~ BioProject),
         Origin = case_when(SRA=="in_house" ~ "in_house", 
                            TRUE ~ "public")) %>%
  ### Change read numbers to numeric
  mutate(Read_number = as.numeric(gsub(",", "", Read_number)))

#### Collapse info at the metasample level
all_species_collapse_metadata_df = all_species_all_metadata_df %>%
  group_by(Species, Tissue, Metasample, SE_PE, Platform, Model, Origin) %>%
  summarize(Mean_read_len = round(mean(Read_len)),
            Tot_read_num = sum(Read_number),
            Tot_original_samples = n()) %>%
  ungroup() %>%
  unite(Species_metasample, c("Species", "Metasample"), remove=FALSE, sep="_") %>%
  ## Select only metasamples without ambiguities (no duplications)
  group_by(Species_metasample) %>%
  filter(n()==1) %>%
  ## Add PC1 values
  left_join(PC1_values_df, by="Species_metasample") %>%
  mutate(Origin = factor(Origin, levels=c("public", "in_house")),
         SE_PE = factor(SE_PE, levels=c("SE", "PE")))

###################################################################
################ PLOT ORIGINAL PC1 DISTRIBUTIONS ##################
###################################################################
tot_metasamples = nrow(all_species_collapse_metadata_df)
discrete_variables = c("SE_PE", "Origin", "Model")
continuos_variables = c("Tot_original_samples", "Mean_read_len", "Tot_read_num")
all_variables = c(discrete_variables, continuos_variables)
selected_discrete_modes = c("SE_PE"="PE", "Origin"="public", "Model"="Illumina_HiSeq_2500")

original_PC_distribs_list = list()
i = 1

for (variable in discrete_variables) {
  if (variable=="Model") {axis_text_angle_x = 90; axis_text_hjust_x=1; axis_text_vjust_x=1} else {axis_text_angle_x = 0; axis_text_hjust_x=0.5; axis_text_vjust_x=1}
  
  subsetted_metadata_df = all_species_collapse_metadata_df %>%
    ungroup() %>%
    filter(!!as.symbol(variable) == unname(selected_discrete_modes[variable])) %>%
    summarise(xmax = unname(selected_discrete_modes[variable]),
              xmin = unname(selected_discrete_modes[variable]),
              ymax = max(PC1),
              ymin = min(PC1))
  
  discrete_var_PC1_plot = ggplot(data=all_species_collapse_metadata_df, aes_string(x=variable, y="PC1")) +
    geom_jitter(color="dimgray", alpha=0.6) +
    geom_boxplot(alpha=0.3) +
    geom_tile(data=subsetted_metadata_df, aes(x=xmax, y=1, height = Inf, width = 1), fill="white", alpha=0, color="red", size=0.7) +
    stat_compare_means() +
    ggtitle(gsub("_", " ", variable)) +
    theme_bw() +
    theme(plot.title = element_text(color="black", hjust=0.5, face="bold"),
          axis.title = element_text(color="black", size=12),
          axis.text.x = element_text(color="black", angle=axis_text_angle_x, vjust=axis_text_vjust_x, hjust=axis_text_hjust_x, size=12),
          axis.text.y = element_text(color="black", size=12)) +
    xlab(gsub("_", " ", variable))
  
  #Add to list and print
  original_PC_distribs_list[[i]] = discrete_var_PC1_plot
  i = i+1
  #print(discrete_var_PC1_plot)
  
}

for (variable in continuos_variables) {
  ## Subset for the plot post correction
  if (variable == "Mean_read_len") {subsetted_metadata_df = all_species_collapse_metadata_df %>% filter(Mean_read_len %in% seq(140,160))}
  if (variable == "Tot_original_samples") {subsetted_metadata_df = all_species_collapse_metadata_df %>% filter(Tot_original_samples == 1)}
  if (variable == "Tot_read_num") {subsetted_metadata_df = all_species_collapse_metadata_df %>% filter(Tot_read_num <= 100000000, Tot_read_num >= 80000000)}
  
  ## Generate code to add rectangles on the plots
  subsetted_metadata_df = subsetted_metadata_df %>%
    dplyr::select(Species_metasample, as.name(variable), PC1) %>%
    ungroup() %>%
    summarise(xmax = max(!!as.symbol(variable)),
              xmin = min(!!as.symbol(variable)), 
              ymax = max(PC1),
              ymin = min(PC1))
  
  ## Compute pearson correlation between the variable and the coordinates
  original_cor_res = cor.test(as.numeric(unlist(all_species_collapse_metadata_df[,variable])), 
                     all_species_collapse_metadata_df$PC1)
  
  continuos_var_PC1_plot = ggplot(data=all_species_collapse_metadata_df, aes_string(x=variable, y="PC1")) +
    geom_jitter(color="dimgray", alpha=0.6) +
    geom_smooth(method="lm", color="black") +
    geom_rect(data=subsetted_metadata_df, aes(xmin=xmin, xmax=xmax, ymax=ymax, ymin=ymin), color="red", fill=NA, inherit.aes = FALSE) +
    annotation_custom(grobTree(textGrob(paste0("Pearson's cor: ", round(unname(original_cor_res$estimate),3), 
                                               "\nPvalue: ", formatC(original_cor_res$p.value, format="e", digits=2)),
                                        x=0.6, y=0.95, hjust=0.5))) +
    ggtitle(gsub("_", " ", variable)) +
    theme_bw() +
    ggtitle(gsub("_", " ", variable)) +
        theme(plot.title = element_text(color="black", hjust=0.5, face="bold"),
          axis.title = element_text(color="black", size=12),
          axis.text.x = element_text(color="black", size=12),
          axis.text.y = element_text(color="black", size=12)) +
    xlab(gsub("_", " ", variable))
  
  #Add to list and print
  original_PC_distribs_list[[i]] = continuos_var_PC1_plot
  i = i+1
  #print(continuos_var_PC1_plot)
}

###################################################################
################ PLOT PCA FROM SELECTION ##########################
###################################################################

new_PCA_list = list()
j = 1
for (variable in all_variables) {
  ###### Subset the metadata table
  if (variable == "Mean_read_len") {subsetted_metadata_df = all_species_collapse_metadata_df %>% filter(Mean_read_len %in% seq(140,160))}
  if (variable == "Tot_original_samples") {subsetted_metadata_df = all_species_collapse_metadata_df %>% filter(Tot_original_samples == 1)}
  if (variable == "Tot_read_num") {subsetted_metadata_df = all_species_collapse_metadata_df %>% filter(Tot_read_num <= 100000000, Tot_read_num >= 80000000)}
  if (variable %in% discrete_variables) {subsetted_metadata_df = all_species_collapse_metadata_df %>% ungroup() %>%
    filter(!!as.symbol(variable) == unname(selected_discrete_modes[variable]))}

  ###### Subset expression table to selected samples
  selected_samples = as.vector(subsetted_metadata_df$Species_metasample)
  subsetted_df = expr_df[,selected_samples]

  ###### PCA plots
  res_PCA = PCA_basic_by_species_plot(subsetted_df, pca_metadata_df, clade, orthogroups_number)
  #### PC1 vs PC2 plot
  res_PCA = PCA_basic_by_species_plot(subsetted_df, pca_metadata_df, clade, orthogroups_number)
  PCA_plot = res_PCA$results[[1]] + ggtitle("") + theme(legend.position = "none") + guides(fill=FALSE, color=FALSE, fill=FALSE, shape=FALSE)

  #### lateral distribs
  raw_PCA = prcomp(as.matrix(t(subsetted_df)), center = TRUE, scale = TRUE) #center and scale the PCs.
  raw_PCA_df = as.data.frame(raw_PCA$x[order(rownames(raw_PCA$x)),]) %>%
    rownames_to_column("Metasample") %>%
    left_join(pca_metadata_df %>% rownames_to_column("Metasample"), by="Metasample") %>%
    mutate(clade = case_when(species %in% Vertebrata ~ "Vertebrate",
                             TRUE ~ "Not_vertebrate"))
  
  #Run wilcoxon test for annotation
  wilcoxon_res = wilcox.test(as.vector(subset(raw_PCA_df, clade=="Vertebrate")$PC1), 
                             as.vector(subset(raw_PCA_df, clade=="Not_vertebrate")$PC1))
  
  lateral_distrib_plot = ggplot(data=raw_PCA_df, aes(x=PC1, color=clade, fill=clade, group=clade)) +
    geom_density(alpha=0.6) +
    annotation_custom(grobTree(textGrob(paste0("Wilcoxon, p: ", formatC(wilcoxon_res$p.value, format="e", digits=2)),
                                        x=0.5, y=0.9, hjust=0.5))) +
    scale_color_manual(values=c("Vertebrate"="#82359D", "Not_vertebrate"="goldenrod")) +
    scale_fill_manual(values=c("Vertebrate"="#82359D", "Not_vertebrate"="goldenrod")) +
    theme_bw()

  #### combine the two plots
  PCA_combined_plots =
    plot_grid(lateral_distrib_plot + theme(axis.text.x = element_blank(),
                                           axis.text.y = element_text(color="black", size=12),
                                           axis.title.x = element_blank(),
                                           axis.title.y = element_text(color="black", size=12),
                                           axis.ticks.x = element_blank(),
                                           legend.position = "none",
                                           plot.margin = unit(c(0, 0.1, 0, 0), "cm")) +
                ylab("Density"),
              PCA_plot + theme(legend.position = "none",
                               plot.margin = unit(c(0, 0.1, 0, 0), "cm")),
              nrow=2, align="v", rel_heights = c(0.2,1))

  #Add to list and print
  new_PCA_list[[j]] = PCA_combined_plots
  j = j+1
  #print(PCA_combined_plots)
}

###################################################################
################ COMBINE TWO TYPES OF PLOT ########################
###################################################################

all_combined_plot_list = list()
for (index in seq(1:length(new_PCA_list))) {
  final_plot = plot_grid(original_PC_distribs_list[[index]], new_PCA_list[[index]], ncol=2, align="h", axis="t", rel_widths = c(01,1))
  all_combined_plot_list[[index]] = final_plot
  #print(final_plot)
}

final_batch_effects_panel = plot_grid(plotlist = all_combined_plot_list, ncol=2, align="hv")
print(final_batch_effects_panel)
```

```{r PC_batch_correlation_custom_plot, warning=FALSE, echo=FALSE, message=FALSE, dependson=c("setup", "extra_functions"), fig.width=16, fig.height=13}
######################################################
####### EXTRA PCA TO DISPROVE BATCH EFFECTS ##########
######################################################

######################################################
####### PUBLIC SAMPLES ONLY ##########################
######################################################
selected_samples = all_species_collapse_metadata_df %>% filter(Origin=="in_house") %>% .$Species_metasample
subsetted_df = expr_df[,selected_samples]

orthogroups_number=nrow(subsetted_df)
res_PCA = PCA_basic_by_clade_plot(subsetted_df, pca_metadata_df, clade, orthogroups_number)
res_anova = anova_plot(subsetted_df, pca_metadata_df, clade)
#plots
first_row = annotate_figure(ggarrange(res_PCA$results[[1]] + ggtitle(""), 
                                      res_PCA$results[[2]] + ggtitle(""), 
                                      res_PCA$results[[3]] + ggtitle(""), 
                                      common.legend=TRUE, legend="right", 
                                      ncol=3), 
                        fig.lab=paste0(str_to_title(my_clade), " PC analysis ~ ", orthogroups_number, " orthogroups. ", "in house only"), fig.lab.pos = "top.left", fig.lab.size=18)
second_row = plot_grid(res_anova$results[[1]] + ggtitle("Anova on PCs"), 
                       res_anova$results[[2]] + ggtitle("Anova ~ species separation"), 
                       res_anova$results[[3]] + ggtitle("Anova  ~ tissues separation"), 
                       res_anova$results[[4]] + ggtitle("Variance explained"), 
                       rel_widths=c(1.25,2,2,1), align="h", ncol=4)

final_plot = ggarrange(first_row, second_row, nrow=2)
print(final_plot)

######################################################
####### PE VS SE #####################################
######################################################
selected_samples = all_species_collapse_metadata_df %>% filter(SE_PE == "SE") %>% .$Species_metasample
subsetted_df = expr_df[,selected_samples]

orthogroups_number=nrow(subsetted_df)
res_PCA = PCA_basic_by_clade_plot(subsetted_df, pca_metadata_df, clade, orthogroups_number)
res_anova = anova_plot(subsetted_df, pca_metadata_df, clade)
#plots
first_row = annotate_figure(ggarrange(res_PCA$results[[1]] + ggtitle(""), 
                                      res_PCA$results[[2]] + ggtitle(""), 
                                      res_PCA$results[[3]] + ggtitle(""), 
                                      common.legend=TRUE, legend="right", 
                                      ncol=3), 
                        fig.lab=paste0(str_to_title(my_clade), " PC analysis ~ ", orthogroups_number, " orthogroups. ", "PE only"), fig.lab.pos = "top.left", fig.lab.size=18)
second_row = plot_grid(res_anova$results[[1]] + ggtitle("Anova on PCs"), 
                       res_anova$results[[2]] + ggtitle("Anova ~ species separation"), 
                       res_anova$results[[3]] + ggtitle("Anova  ~ tissues separation"), 
                       res_anova$results[[4]] + ggtitle("Variance explained"), 
                       rel_widths=c(1.25,2,2,1), align="h", ncol=4)

final_plot = ggarrange(first_row, second_row, nrow=2)
print(final_plot)


######################################################
####### ORIGINAL METASAMPLES #########################
######################################################
average_original_samples = round(mean(all_species_collapse_metadata_df$Tot_original_samples))
selected_samples = all_species_collapse_metadata_df %>% filter(Tot_original_samples == 1) %>% .$Species_metasample
subsetted_df = expr_df[,selected_samples]

orthogroups_number=nrow(subsetted_df)
res_PCA = PCA_basic_by_clade_plot(subsetted_df, pca_metadata_df, clade, orthogroups_number)
res_anova = anova_plot(subsetted_df, pca_metadata_df, clade)
#plots
first_row = annotate_figure(ggarrange(res_PCA$results[[1]] + ggtitle(""), 
                                      res_PCA$results[[2]] + ggtitle(""), 
                                      res_PCA$results[[3]] + ggtitle(""), 
                                      common.legend=TRUE, legend="right", 
                                      ncol=3), 
                        fig.lab=paste0(str_to_title(my_clade), " PC analysis ~ ", orthogroups_number, " orthogroups. ", "Tot original samples == 1"), fig.lab.pos = "top.left", fig.lab.size=18)
second_row = plot_grid(res_anova$results[[1]] + ggtitle("Anova on PCs"), 
                       res_anova$results[[2]] + ggtitle("Anova ~ species separation"), 
                       res_anova$results[[3]] + ggtitle("Anova  ~ tissues separation"), 
                       res_anova$results[[4]] + ggtitle("Variance explained"), 
                       rel_widths=c(1.25,2,2,1), align="h", ncol=4)

final_plot = ggarrange(first_row, second_row, nrow=2)
print(final_plot)

######################################################
############ READ LENGTH #############################
######################################################
selected_samples = all_species_collapse_metadata_df %>% filter(Mean_read_len %in% c(150, 151, 152, 153, 154, 155)) %>% .$Species_metasample
subsetted_df = expr_df[,selected_samples]

orthogroups_number=nrow(subsetted_df)
res_PCA = PCA_basic_by_clade_plot(subsetted_df, pca_metadata_df, clade, orthogroups_number)
res_anova = anova_plot(subsetted_df, pca_metadata_df, clade)
#plots
first_row = annotate_figure(ggarrange(res_PCA$results[[1]] + ggtitle(""), 
                                      res_PCA$results[[2]] + ggtitle(""), 
                                      res_PCA$results[[3]] + ggtitle(""), 
                                      common.legend=TRUE, legend="right", 
                                      ncol=3), 
                        fig.lab=paste0(str_to_title(my_clade), " PC analysis ~ ", orthogroups_number, " orthogroups. ", "Mean read len >=100 & <=125"), fig.lab.pos = "top.left", fig.lab.size=18)
second_row = plot_grid(res_anova$results[[1]] + ggtitle("Anova on PCs"), 
                       res_anova$results[[2]] + ggtitle("Anova ~ species separation"), 
                       res_anova$results[[3]] + ggtitle("Anova  ~ tissues separation"), 
                       res_anova$results[[4]] + ggtitle("Variance explained"), 
                       rel_widths=c(1.25,2,2,1), align="h", ncol=4)

final_plot = ggarrange(first_row, second_row, nrow=2)
print(final_plot)


######################################################
############ TOT READ NUMBER #########################
######################################################
selected_samples = all_species_collapse_metadata_df %>% filter(Tot_read_num <= 100000000 & Tot_read_num >= 80000000) %>% .$Species_metasample
subsetted_df = expr_df[,selected_samples]

orthogroups_number=nrow(subsetted_df)
res_PCA = PCA_basic_by_clade_plot(subsetted_df, pca_metadata_df, clade, orthogroups_number)
res_anova = anova_plot(subsetted_df, pca_metadata_df, clade)
#plots
first_row = annotate_figure(ggarrange(res_PCA$results[[1]] + ggtitle(""), 
                                      res_PCA$results[[2]] + ggtitle(""), 
                                      res_PCA$results[[3]] + ggtitle(""), 
                                      common.legend=TRUE, legend="right", 
                                      ncol=3), 
                        fig.lab=paste0(str_to_title(my_clade), " PC analysis ~ ", orthogroups_number, " orthogroups. ", "Tot read num >=80*10^6 & <=100*10^6 "), fig.lab.pos = "top.left", fig.lab.size=18)
second_row = plot_grid(res_anova$results[[1]] + ggtitle("Anova on PCs"), 
                       res_anova$results[[2]] + ggtitle("Anova ~ species separation"), 
                       res_anova$results[[3]] + ggtitle("Anova  ~ tissues separation"), 
                       res_anova$results[[4]] + ggtitle("Variance explained"), 
                       rel_widths=c(1.25,2,2,1), align="h", ncol=4)

final_plot = ggarrange(first_row, second_row, nrow=2)
print(final_plot)

######################################################
############ SEQUENCING TECHNOLOGY ###################
######################################################
selected_samples = all_species_collapse_metadata_df %>% filter(Model == "Illumina_HiSeq_2500") %>% .$Species_metasample
subsetted_df = expr_df[,selected_samples]

orthogroups_number=nrow(subsetted_df)
res_PCA = PCA_basic_by_clade_plot(subsetted_df, pca_metadata_df, clade, orthogroups_number)
res_anova = anova_plot(subsetted_df, pca_metadata_df, clade)
#plots
first_row = annotate_figure(ggarrange(res_PCA$results[[1]] + ggtitle(""), 
                                      res_PCA$results[[2]] + ggtitle(""), 
                                      res_PCA$results[[3]] + ggtitle(""), 
                                      common.legend=TRUE, legend="right", 
                                      ncol=3), 
                        fig.lab=paste0(str_to_title(my_clade), " PC analysis ~ ", orthogroups_number, " orthogroups. ", "Model == Illumina_HiSeq_2500"), fig.lab.pos = "top.left", fig.lab.size=18)
second_row = plot_grid(res_anova$results[[1]] + ggtitle("Anova on PCs"), 
                       res_anova$results[[2]] + ggtitle("Anova ~ species separation"), 
                       res_anova$results[[3]] + ggtitle("Anova  ~ tissues separation"), 
                       res_anova$results[[4]] + ggtitle("Variance explained"), 
                       rel_widths=c(1.25,2,2,1), align="h", ncol=4)

final_plot = ggarrange(first_row, second_row, nrow=2)
print(final_plot)
```

```{r PC1_batch_correlation_eigencor, warning=FALSE, echo=FALSE, message=FALSE, dependson=c("setup", "extra_functions"), fig.width=16, fig.height=13}
#Define path to expression file
expr_file = paste0(pca_dir, "/", version, "/", category, "/", clade, "/", evo_type, "/Bilateria_conserved-metasamples_median_expr-NOSVA-log2-TPMs-NORM-BH_genes.tab")
expr_df = read.delim(expr_file, sep="\t", header=TRUE, row=1)
expr_df = expr_df[complete.cases(expr_df),] #I am using only the cases without NAs, so that I don't have to do any imputation.
#Update colnames
colnames(expr_df) = gsub("BmA", "Bmo", gsub("Hs2", "Hsa", gsub("Mm2", "Mmu", gsub("Bt2", "Bta", gsub("Sp2", "Spu", colnames(expr_df))))))

################################
######### PCA ##################
################################

orthogroups_number=nrow(expr_df)
res_PCA = PCA_basic_plot(expr_df, pca_metadata_df, clade, orthogroups_number)
res_anova = anova_plot(expr_df, pca_metadata_df, clade)
#plots
first_row = annotate_figure(ggarrange(res_PCA$results[[1]] + ggtitle(""), 
                                      res_PCA$results[[2]] + ggtitle(""), 
                                      res_PCA$results[[3]] + ggtitle(""), 
                                      common.legend=TRUE, legend="right", 
                                      ncol=3), 
                        fig.lab=paste0(str_to_title(my_clade), " PC analysis ~ ", orthogroups_number, " orthogroups"), fig.lab.pos = "top.left", fig.lab.size=18)
second_row = plot_grid(res_anova$results[[1]] + ggtitle("Anova on PCs"), 
                       res_anova$results[[2]] + ggtitle("Anova ~ species separation"), 
                       res_anova$results[[3]] + ggtitle("Anova  ~ tissues separation"), 
                       res_anova$results[[4]] + ggtitle("Variance explained"), 
                       rel_widths=c(1.25,2,2,1), align="h", ncol=4)

final_plot = ggarrange(first_row, second_row, nrow=2)
print(final_plot)


################################
######### BATCHES ##############
################################
test_metadata = pca_metadata_df
test_metadata$clade = "Vertebrata"
test_metadata$clade[!(test_metadata$species %in% Vertebrata)] = "Non_vertebrata"

PCA_res = prcomp(as.matrix(t(expr_df)), center = TRUE, scale = TRUE) #center and scale the PCs.
ordered_PCA_res = as.data.frame(PCA_res$x[order(rownames(PCA_res$x)),])
#test_metadata$species = factor(test_metadata$species, levels=rev(c("Hsa", "Mmu", "Bta", "Mdo", "Gga", "Xtr", "Dre", "Cmi", "Bla", "Spu", rev(c("Dme", "Eba", "Aae", "Bmo", "Tca", "Ame", "Bge", "Cdi", "Sma", "Obi")))))
eigencorplot(ordered_PCA_res, test_metadata, metavars=c("clade", "tissues", "species"))
```

```{r PCA_extra_bilateria_v1, warning=FALSE, echo=FALSE, message=FALSE, dependson=c("setup", "functions")}
#### Initialize final plot list
extra_bilateria_PCAs = list()
i = 1

######################################################
############## SVA CORRECTION ########################
######################################################
######### Upload table
expr_file = paste0(pca_dir, "/", version, "/", category, "/", clade, "/", evo_type, "/Bilateria_conserved-metasamples_median_expr-SVA-log2-TPMs-NORM-BH_genes.tab")
expr_df = read.delim(expr_file, sep="\t", header=TRUE, row=1)
expr_df = expr_df[complete.cases(expr_df),] #I am using only the cases without NAs, so that I don't have to do any imputation.
#Update colnames
colnames(expr_df) = gsub("BmA", "Bmo", gsub("Hs2", "Hsa", gsub("Mm2", "Mmu", gsub("Bt2", "Bta", gsub("Sp2", "Spu", colnames(expr_df))))))

#Run PCA
orthogroups_number=nrow(expr_df)
res_PCA = plot_PCA(expr_df, pca_metadata_df, clade, orthogroups_number)
PC_plot = res_PCA$results[[1]] +
  theme(legend.position = "right",
        plot.title = element_text(color="black", hjust=0.5, face="bold")) +
  guides(color=guide_legend(ncol=2), shape=guide_legend(ncol=2)) +
  ggtitle("SVA normalization")

print(PC_plot)
extra_bilateria_PCAs[[i]] = PC_plot
i = i+1

######################################################
############## AVERAGE PARALOG EXPR ##################
######################################################
my_clade = "bilateria"
######### Upload table
expr_file = paste0(broccoli_revision_dir, my_clade, "/expression_tables/", my_clade, "_conserved-metasamples_median_expr-NOSVA-log2-TPMs-paralogs_average-NORM.tab")
expr_df = read.delim(expr_file, sep="\t", header=TRUE, row=1)
expr_df = expr_df[complete.cases(expr_df),] #I am using only the cases without NAs, so that I don't have to do any imputation.
#Update colnames
colnames(expr_df) = gsub("BmA", "Bmo", gsub("Hs2", "Hsa", gsub("Mm2", "Mmu", gsub("Bt2", "Bta", gsub("Sp2", "Spu", colnames(expr_df))))))

orthogroups_number=nrow(expr_df)
res_PCA = plot_PCA(expr_df, pca_metadata_df, clade, orthogroups_number)
PC_plot = res_PCA$results[[1]] +
  theme(legend.position = "right",
        plot.title = element_text(color="black", hjust=0.5, face="bold")) +
  guides(color=guide_legend(ncol=2), shape=guide_legend(ncol=2)) +
  ggtitle("Average paralog expression")

print(PC_plot)
extra_bilateria_PCAs[[i]] = PC_plot
i = i+1

######################################################
############## SUMMED PARALOG EXPR ###################
######################################################
my_clade = "bilateria"
######### Upload table
expr_file = paste0(broccoli_revision_dir, my_clade, "/expression_tables/", my_clade, "_conserved-metasamples_median_expr-NOSVA-log2-TPMs-paralogs_summed-NORM.tab")
expr_df = read.delim(expr_file, sep="\t", header=TRUE, row=1)
expr_df = expr_df[complete.cases(expr_df),] #I am using only the cases without NAs, so that I don't have to do any imputation.
#Update colnames
colnames(expr_df) = gsub("BmA", "Bmo", gsub("Hs2", "Hsa", gsub("Mm2", "Mmu", gsub("Bt2", "Bta", gsub("Sp2", "Spu", colnames(expr_df))))))

orthogroups_number=nrow(expr_df)
res_PCA = plot_PCA(expr_df, pca_metadata_df, clade, orthogroups_number)
PC_plot = res_PCA$results[[1]] +
  theme(legend.position = "right",
        plot.title = element_text(color="black", hjust=0.5, face="bold")) +
  guides(color=guide_legend(ncol=2), shape=guide_legend(ncol=2)) +
  ggtitle("Summed paralog expression")

print(PC_plot)
extra_bilateria_PCAs[[i]] = PC_plot
i = i+1

######################################################
############## SINGLE_COPY ORTHOLOGS #################
######################################################
my_clade = "bilateria"
######### Upload table
expr_file = paste0(broccoli_revision_dir, my_clade, "/expression_tables/", my_clade, "_conserved-metasamples_median_expr-NOSVA-log2-TPMs-single_copies-NORM.tab")
expr_df = read.delim(expr_file, sep="\t", header=TRUE, row=1)
expr_df = expr_df[complete.cases(expr_df),] #I am using only the cases without NAs, so that I don't have to do any imputation.
#Update colnames
colnames(expr_df) = gsub("BmA", "Bmo", gsub("Hs2", "Hsa", gsub("Mm2", "Mmu", gsub("Bt2", "Bta", gsub("Sp2", "Spu", colnames(expr_df))))))

orthogroups_number=nrow(expr_df)
res_PCA = plot_PCA(expr_df, pca_metadata_df, clade, orthogroups_number)
PC_plot = res_PCA$results[[1]] +
  theme(legend.position = "right",
        plot.title = element_text(color="black", hjust=0.5, face="bold")) +
  guides(color=guide_legend(ncol=2), shape=guide_legend(ncol=2)) +
  ggtitle(paste0("Single copy orthologs (", nrow(expr_df), " OGs)"))

print(PC_plot)
extra_bilateria_PCAs[[i]] = PC_plot
i = i+1

######################################################
########## JOINED PLOTS ##############################
######################################################

joined_PCA_plots = annotate_figure(ggarrange(plotlist = extra_bilateria_PCAs,
                                          common.legend=TRUE, 
                                          legend="bottom", 
                                          ncol=4))

print(joined_PCA_plots)


# for (my_clade in c("vertebrata", "insecta")) {
#   ######### Upload table
#   expr_file = paste0(broccoli_revision_dir, my_clade, "/expression_tables/", my_clade, "_conserved-metasamples_median_expr-NOSVA-log2-TPMs-single_copies-NORM.tab")
#   expr_df = read.delim(expr_file, sep="\t", header=TRUE, row=1)
#   expr_df = expr_df[complete.cases(expr_df),] #I am using only the cases without NAs, so that I don't have to do any imputation.
#   #Update colnames
#   colnames(expr_df) = gsub("BmA", "Bmo", gsub("Hs2", "Hsa", gsub("Mm2", "Mmu", gsub("Bt2", "Bta", gsub("Sp2", "Spu", colnames(expr_df))))))
#   
#   orthogroups_number=nrow(expr_df)
#   res_PCA = plot_PCA(expr_df, pca_metadata_df, clade, orthogroups_number)
#   PC_plot = res_PCA$results[[1]] +
#     theme(legend.position = "right",
#           plot.title = element_text(color="black", hjust=0.5, face="bold")) +
#     guides(color=guide_legend(ncol=2), shape=guide_legend(ncol=2)) +
#     ggtitle(paste0(my_clade, " ~ single copy orthologs"))
#   
#   print(PC_plot)
# }
```

```{r vertebrates_insects_single_copy_orthologs_v1, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions"), fig.width=8, fig.height=9}
for (my_clade in c("vertebrata", "insecta")[1]) { 
  ##############################################
  ##### GENERATE INPUT TABLE ###################
  ##############################################
  
  ######### Upload table
  expr_file = paste0(broccoli_revision_dir, my_clade, "/expression_tables/", my_clade, "_conserved-metasamples_median_expr-NOSVA-log2-TPMs-single_copies-NORM.tab")
  expr_df = read.delim(expr_file, sep="\t", header=TRUE, row=1)
  expr_df = expr_df[complete.cases(expr_df),] #I am using only the cases without NAs, so that I don't have to do any imputation.
  #Update colnames
  colnames(expr_df) = gsub("BmA", "Bmo", gsub("Hs2", "Hsa", gsub("Mm2", "Mmu", gsub("Bt2", "Bta", gsub("Sp2", "Spu", colnames(expr_df))))))
  
  ##############################################
  ##### PLOT ###################################
  ##############################################
  
  
  orthogroups_number=nrow(expr_df)
  res_PCA = PCA_basic_plot(expr_df, pca_metadata_df, clade, orthogroups_number)
  res_anova = anova_plot(expr_df, pca_metadata_df, clade)
  #plots
  first_row = annotate_figure(ggarrange(res_PCA$results[[1]] + ggtitle(""), 
                                        res_PCA$results[[2]] + ggtitle(""), 
                                        common.legend=TRUE, legend="right", 
                                        ncol=2), 
                          fig.lab=paste0(str_to_title(my_clade), " PC analysis ~ ", orthogroups_number, " orthogroups"), fig.lab.pos = "top.left", fig.lab.size=12)
  second_row = plot_grid(res_anova$results[[1]] + ggtitle("Anova on PCs"), 
                         res_anova$results[[2]] + ggtitle("Anova ~ species separation"), 
                         res_anova$results[[3]] + ggtitle("Anova  ~ tissues separation"), 
                         res_anova$results[[4]] + ggtitle("Variance explained"), 
                         rel_widths=c(1.25,2,2,1), align="h", ncol=4)
  
  final_plot_1 = ggarrange(first_row, second_row, nrow=2)
  print(final_plot_1)
}



#### lateral distribs
raw_PCA = prcomp(as.matrix(t(expr_df)), center = TRUE, scale = TRUE) #center and scale the PCs.
raw_PCA_df = as.data.frame(raw_PCA$x[order(rownames(raw_PCA$x)),]) %>%
  rownames_to_column("Metasample") %>%
  left_join(pca_metadata_df %>% rownames_to_column("Metasample"), by="Metasample") %>%
  mutate(clade = case_when(species %in% Vertebrata ~ "Vertebrate",
                           TRUE ~ "Not_vertebrate"))

tissue_lateral_distrib_plot = ggplot(data=raw_PCA_df, aes(x=PC2, group=tissues, fill=tissues, color=tissues)) +
  geom_density(alpha=0.6) +
  scale_color_manual(values=tissue_colors) +
  scale_fill_manual(values=tissue_colors) +
  theme_bw() +
  theme(axis.title.y=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y = element_blank(),
        axis.title.x=element_text(color="black", size=12),
        axis.text.x=element_text(color="black", size=12),
        legend.position = "none") +
  ylab("Density") +
  coord_flip()

print(tissue_lateral_distrib_plot)
plot_grid(res_PCA$results[[1]] + theme(legend.position = "bottom"), tissue_lateral_distrib_plot, ncol=2, rel_widths = c(1,0.4), align="h", axis="tb")
```

```{r plot_OGs_gains_losses_by_gene_family_v1, echo=FALSE, messsage=FALSE, warning=FALSE, dependson=c("setup", "functions"), fig.width=8, fig.height=10}
##### Work either with gains or with losses (i.e., inferences)
inference_type = "gains"

#######################################
###### GENERATE INPUT #################
#######################################

##### Upload table with gene families information
gene_families_info_file = paste0(pfam_transfers_dir, "PFAM_annot/orthogroups_to_PFAMClan-transferred_from_Hs2-filtered_min20OGs.txt")
gene_families_df = read.delim(gene_families_info_file, header=FALSE, col.names=c("OG_ID", "Clan_ID", "Clan_name"))

##### Generate vector of orthogroups with at least one inference and add info to the gene families
all_OG_with_inference_vector = vector()
for (tissue in ordered_tissues) {
  inference_file = paste0(ts_gains_losses_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_", inference_type, ".tab")
  OG_with_inference_vector = as.vector(read.delim(inference_file, header=TRUE)$"OG_ID")
  all_OG_with_inference_vector = c(all_OG_with_inference_vector, OG_with_inference_vector)
}
all_OG_with_inference_vector = unique(as.vector(all_OG_with_inference_vector)) #Remove duplicates

## Add information to gene families df
gene_families_inference_info_df = gene_families_df %>%
  mutate(inference_presence = case_when(OG_ID %in% all_OG_with_inference_vector ~ "Inference", 
                                        TRUE ~ "NO_inference")) %>%
  group_by(Clan_ID, Clan_name) %>%
  mutate(Clan_size = n()) %>%
  ungroup()

#Order the gene families based on the number of orthogroups they contain
clan_order_df = gene_families_inference_info_df %>%
  arrange(desc(Clan_size)) %>%
  dplyr::select(Clan_ID, Clan_name, Clan_size) %>%
  distinct()

gene_families_inference_info_df$Clan_ID = factor(gene_families_inference_info_df$Clan_ID, levels=rev(as.vector(clan_order_df$Clan_ID)))
gene_families_inference_info_df$Clan_name = factor(gene_families_inference_info_df$Clan_name, levels=rev(as.vector(clan_order_df$Clan_name)))

##### Generate a dataframe where for each gene family we count the number of OGs with multiTS inferences
multiTS_classification_file = paste0(pfam_transfers_dir, "Bilateria_conserved_orthogroups-multiTS_classification-", inference_type, ".tab")
multiTS_df = read.delim(multiTS_classification_file, header=FALSE, col.names=c("OG_ID", "Evo_class"))
multiTS_OGs_vector = unique(as.vector(multiTS_df$OG_ID))

multiTS_count_by_gene_family_df = gene_families_df %>%
  filter(OG_ID %in% multiTS_OGs_vector) %>%
  group_by(Clan_ID, Clan_name) %>%
  summarize(Tot_multiTS = n()) %>%
  ungroup() %>%
  right_join(gene_families_inference_info_df %>% 
               dplyr::select(Clan_ID, Clan_name, Clan_size) %>%
               distinct(), by=c("Clan_ID", "Clan_name")) %>%
  mutate(multiTS_prop = Tot_multiTS / Clan_size) %>%
  left_join(gene_families_inference_info_df %>%
              filter(inference_presence=="Inference") %>%
              group_by(Clan_ID) %>%
              summarize(Tot_inference = n()) %>%
              ungroup(), by="Clan_ID")


#######################################
###### EXPECTED PROBABILITY ###########
#######################################

#Compute expected proportions of orthogroups with (multiTS) gains in each family to add to the stacked barplot
#Expected probability is defined by as e = (b/u)*a
#b = total number of OGs in a gene family
#u = total number of OGs with at least one TS gain
#a = total number of OGs with (multiTS) gains

### Define universes and initial numbers
universe_all = 7178
universe_TS = length(all_OG_with_inference_vector)
Tot_TS_across_clans = universe_TS
Tot_multiTS_across_clans = length(multiTS_OGs_vector)

TS_prop_over_all = Tot_TS_across_clans / universe_all
multiTS_prop_over_TS = Tot_multiTS_across_clans / universe_TS

expected_df = multiTS_count_by_gene_family_df %>%
  mutate(Expected_TS = TS_prop_over_all*Clan_size,
         Expected_multiTS = multiTS_prop_over_TS*Tot_inference,
         Expected_TS_prop_to_plot = Expected_TS/Clan_size,
         Expected_multiTS_prop_to_plot = Expected_multiTS/Clan_size) %>%
  dplyr::select(Clan_ID, Clan_name, Clan_size, Tot_inference, Tot_multiTS, Expected_TS, Expected_multiTS,  Expected_TS_prop_to_plot,  Expected_multiTS_prop_to_plot)

#######################################
###### PLOT ###########################
#######################################

#### Stacked bar
gene_families_inference_info_df$inference_presence = factor(gene_families_inference_info_df$inference_presence, levels=rev(c("Inference", "NO_inference")))

inferences_by_gene_family_stacked_plot = ggplot() +
  geom_bar(data=gene_families_inference_info_df, aes(x=Clan_name, fill=inference_presence, color=inference_presence), 
           stat="count", alpha=0.7, position = position_fill()) +
  
  geom_errorbar(data=multiTS_count_by_gene_family_df, aes(x=Clan_name, xmin=Clan_name, xmax=Clan_name, 
                                                          y=multiTS_prop, ymin=multiTS_prop, ymax=multiTS_prop, 
                                                          group=Clan_name), color="black") +
  
  geom_errorbar(data=expected_df, aes(x=Clan_name, xmin=Clan_name, xmax=Clan_name,
                                      y=Expected_TS_prop_to_plot, ymin=Expected_TS_prop_to_plot, ymax=Expected_TS_prop_to_plot,
                                      group=Clan_name), color="black", linetype="solid") +
  
  geom_errorbar(data=expected_df, aes(x=Clan_name, xmin=Clan_name, xmax=Clan_name,
                                      y=Expected_multiTS_prop_to_plot, ymin=Expected_multiTS_prop_to_plot, ymax=Expected_multiTS_prop_to_plot,
                                      group=Clan_name), color="white", linetype="solid") +
  
  scale_fill_manual(values=c("Inference"="coral3", "NO_inference"="white")) +
  scale_color_manual(values=c("Inference"="coral3", "NO_inference"="coral3")) +
  theme_bw() +
  theme(axis.text = element_text(color="black"),
        axis.title = element_text(color="black"),
        legend.position = "bottom") +
  ylab("") +
  coord_flip()

#### Total number
inferences_by_gene_family_count_plot = ggplot(data=gene_families_inference_info_df, aes(x=Clan_name)) +
  geom_bar(stat="count", color="black", fill="dimgray", alpha=0.6) +
  #scale_fill_manual(values=c("Inference"="dimgray", "NO_inference"="white")) +
  theme_bw() +
  theme(axis.text = element_text(color="black"),
        axis.title = element_text(color="black")) +
  coord_flip()

# print(inferences_by_gene_family_count_plot)
# print(inferences_by_gene_family_stacked_plot)

plot_grid(inferences_by_gene_family_stacked_plot, inferences_by_gene_family_count_plot + theme(axis.title.y=element_blank(),
                                                                                               axis.text.y=element_blank(),
                                                                                               axis.ticks.y=element_blank()), 
          align="h", axis="bt", rel_widths = c(1,0.6))
```

```{r plot_OGs_gains_losses_by_gene_family_v2, echo=FALSE, messsage=FALSE, warning=FALSE, dependson=c("setup", "functions"), fig.width=8, fig.height=10}
##### Work either with gains or with losses (i.e., inferences)
inference_type = "gains"

#######################################
###### GENERATE INPUT #################
#######################################

##### Upload table with gene families information
gene_families_info_file = paste0(pfam_transfers_dir, "PFAM_annot/orthogroups_to_PFAMClan-transferred_from_Hs2-filtered_min20OGs.txt")
gene_families_df = read.delim(gene_families_info_file, header=FALSE, col.names=c("OG_ID", "Clan_ID", "Clan_name"))

##### Generate vector of orthogroups with at least one inference
all_OG_with_inference_vector = vector()
for (tissue in ordered_tissues) {
  inference_file = paste0(ts_gains_losses_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_", inference_type, ".tab")
  OG_with_inference_vector = as.vector(read.delim(inference_file, header=TRUE)$"OG_ID")
  all_OG_with_inference_vector = c(all_OG_with_inference_vector, OG_with_inference_vector)
}
all_OG_with_inference_vector = unique(as.vector(all_OG_with_inference_vector)) #Remove duplicates

##### Generate vector of orthogroups with multi-TS inferences
multiTS_classification_file = paste0(pfam_transfers_dir, "Bilateria_conserved_orthogroups-multiTS_classification-", inference_type, ".tab")
multiTS_df = read.delim(multiTS_classification_file, header=FALSE, col.names=c("OG_ID", "Evo_class"))
multiTS_OGs_vector = unique(as.vector(multiTS_df$OG_ID))

## Add information to gene families df
gene_families_inference_info_df = gene_families_df %>%
  mutate(inference_presence = case_when(OG_ID %in% multiTS_OGs_vector ~ "Inference_multiTS",
                                        OG_ID %in% all_OG_with_inference_vector ~ "Inference_singleTS",
                                        TRUE ~ "NO_inference")) %>%
  group_by(Clan_ID, Clan_name) %>%
  mutate(Clan_size = n()) %>%
  ungroup()

### Order the gene families based on the number of orthogroups they contain
# clan_order_df = gene_families_inference_info_df %>%
#   arrange(desc(Clan_size)) %>%
#   dplyr::select(Clan_ID, Clan_name, Clan_size) %>%
#   distinct()

clan_order_df = gene_families_inference_info_df %>%
  filter(inference_presence %in% c("Inference_multiTS", "Inference_singleTS")) %>%
  group_by(Clan_ID, Clan_name, Clan_size) %>%
  summarize(Tot_inferences_prop = n()/Clan_size) %>%
  ungroup() %>%
  arrange(desc(Tot_inferences_prop)) %>%
  distinct()

gene_families_inference_info_df$Clan_ID = factor(gene_families_inference_info_df$Clan_ID, levels=rev(as.vector(clan_order_df$Clan_ID)))
gene_families_inference_info_df$Clan_name = factor(gene_families_inference_info_df$Clan_name, levels=rev(as.vector(clan_order_df$Clan_name)))

##### Generate a dataframe where for each gene family we count the number of OGs with multiTS inferences
multiTS_count_by_gene_family_df = gene_families_df %>%
  filter(OG_ID %in% multiTS_OGs_vector) %>%
  group_by(Clan_ID, Clan_name) %>%
  summarize(Tot_multiTS = n()) %>%
  ungroup() %>%
  right_join(gene_families_inference_info_df %>% 
               dplyr::select(Clan_ID, Clan_name, Clan_size) %>%
               distinct(), by=c("Clan_ID", "Clan_name")) %>%
  mutate(multiTS_prop = Tot_multiTS / Clan_size) %>%
  left_join(gene_families_inference_info_df %>%
              filter(inference_presence %in% c("Inference_multiTS", "Inference_singleTS")) %>%
              group_by(Clan_ID) %>%
              summarize(Tot_inference = n()) %>%
              ungroup(), by="Clan_ID")


#######################################
###### EXPECTED PROBABILITY ###########
#######################################

#Compute expected proportions of orthogroups with (multiTS) gains in each family to add to the stacked barplot
#Expected probability is defined by as e = (b/u)*a
#b = total number of OGs in a gene family
#u = total number of OGs with at least one TS gain
#a = total number of OGs with (multiTS) gains

### Define universes and initial numbers
universe_all = 7178
universe_TS = length(all_OG_with_inference_vector)
Tot_TS_across_clans = universe_TS
Tot_multiTS_across_clans = length(multiTS_OGs_vector)

TS_prop_over_all = Tot_TS_across_clans / universe_all
multiTS_prop_over_TS = Tot_multiTS_across_clans / universe_TS

expected_df = multiTS_count_by_gene_family_df %>%
  mutate(Expected_TS = TS_prop_over_all*Clan_size,
         Expected_multiTS = multiTS_prop_over_TS*Tot_inference,
         Expected_TS_prop_to_plot = Expected_TS/Clan_size,
         Expected_multiTS_prop_to_plot = Expected_multiTS/Clan_size) %>%
  dplyr::select(Clan_ID, Clan_name, Clan_size, Tot_inference, Tot_multiTS, Expected_TS, Expected_multiTS,  Expected_TS_prop_to_plot,  Expected_multiTS_prop_to_plot)

#######################################
###### PLOT ###########################
#######################################

#########################
#### Stacked bar ########
#########################
gene_families_inference_info_df$inference_presence = factor(gene_families_inference_info_df$inference_presence, levels=rev(c("Inference_multiTS", "Inference_singleTS", "NO_inference")))

inferences_by_gene_family_stacked_plot = ggplot() +
  geom_bar(data=gene_families_inference_info_df, aes(x=Clan_name, fill=inference_presence, color=inference_presence), 
           stat="count", alpha=0.5, position = position_fill()) +
  
  geom_hline(aes(yintercept=0.457), color="limegreen", size=1) +
  geom_errorbar(data=expected_df, aes(x=Clan_name, xmin=Clan_name, xmax=Clan_name,
                                      y=Expected_multiTS_prop_to_plot, ymin=Expected_multiTS_prop_to_plot, ymax=Expected_multiTS_prop_to_plot,
                                      group=Clan_name), color="dodgerblue2", size=0.7) +
  
  scale_fill_manual(values=c("Inference_singleTS"="limegreen", "Inference_multiTS"="dodgerblue2", "NO_inference"="white")) +
  scale_color_manual(values=c("Inference_singleTS"="white", "Inference_multiTS"="white", "NO_inference"="white")) +
  theme_bw() +
  theme(axis.text = element_text(color="black"),
        axis.title = element_text(color="black"),
        legend.position = "bottom") +
  ylab("") +
  coord_flip()

#########################
#### Total number #######
#########################
inferences_by_gene_family_count_plot = ggplot(data=gene_families_inference_info_df, aes(x=Clan_name)) +
  geom_bar(stat="count", color="black", fill="dimgray", alpha=0.6) +
  #scale_fill_manual(values=c("Inference"="dimgray", "NO_inference"="white")) +
  theme_bw() +
  theme(axis.text = element_text(color="black"),
        axis.title = element_text(color="black")) +
  coord_flip()


####### Combine plots
plot_grid(inferences_by_gene_family_stacked_plot, inferences_by_gene_family_count_plot + theme(axis.title.y=element_blank(),
                                                                                               axis.text.y=element_blank(),
                                                                                               axis.ticks.y=element_blank()), 
          align="h", axis="bt", rel_widths = c(1,0.6))
```

```{r plot_OGs_gains_losses_by_gene_family_v3, echo=FALSE, messsage=FALSE, warning=FALSE, dependson=c("setup", "functions"), fig.width=8, fig.height=10}
##### Work either with gains or with losses (i.e., inferences)
inference_type = "gains"

#######################################
###### GENERATE INPUT #################
#######################################

##### Upload table with gene families information
gene_families_info_file = paste0(pfam_transfers_dir, "PFAM_annot/orthogroups_to_PFAMClan-transferred_from_Hs2-filtered_min20OGs.txt")
gene_families_df = read.delim(gene_families_info_file, header=FALSE, col.names=c("OG_ID", "Clan_ID", "Clan_name"))

##### Generate vector of orthogroups with at least one inference
all_OG_with_inference_vector = vector()
for (tissue in ordered_tissues) {
  inference_file = paste0(ts_gains_losses_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_", inference_type, ".tab")
  OG_with_inference_vector = as.vector(read.delim(inference_file, header=TRUE)$"OG_ID")
  all_OG_with_inference_vector = c(all_OG_with_inference_vector, OG_with_inference_vector)
}
all_OG_with_inference_vector = unique(as.vector(all_OG_with_inference_vector)) #Remove duplicates

##### Generate vector of orthogroups with multi-TS inferences
multiTS_classification_file = paste0(pfam_transfers_dir, "Bilateria_conserved_orthogroups-multiTS_classification-", inference_type, ".tab")
multiTS_df = read.delim(multiTS_classification_file, header=FALSE, col.names=c("OG_ID", "Evo_class"))
multiTS_OGs_vector = unique(as.vector(multiTS_df$OG_ID))

####### Add information about inferences to gene families df
gene_families_inference_info_df = gene_families_df %>%
  mutate(inference_presence = case_when(OG_ID %in% multiTS_OGs_vector ~ "Inference_multiTS",
                                        OG_ID %in% all_OG_with_inference_vector ~ "Inference_singleTS",
                                        TRUE ~ "NO_inference")) %>%
  group_by(Clan_ID, Clan_name) %>%
  mutate(Clan_size = n()) %>%
  ungroup()

####### Order the gene families based on the number of orthogroups they contain
clan_order_df = gene_families_inference_info_df %>%
  filter(inference_presence %in% c("Inference_multiTS", "Inference_singleTS")) %>%
  group_by(Clan_ID, Clan_name, Clan_size) %>%
  summarize(Tot_inferences_prop = n()/Clan_size) %>%
  ungroup() %>%
  arrange(desc(Tot_inferences_prop)) %>%
  distinct()

gene_families_inference_info_df$Clan_ID = factor(gene_families_inference_info_df$Clan_ID, levels=rev(as.vector(clan_order_df$Clan_ID)))
gene_families_inference_info_df$Clan_name = factor(gene_families_inference_info_df$Clan_name, levels=rev(as.vector(clan_order_df$Clan_name)))

####### Generate a dataframe with the number of OGs with multiTS inferences by gene family
#This will be usesd to compute the expected probablity
multiTS_count_by_gene_family_df = gene_families_df %>%
  filter(OG_ID %in% multiTS_OGs_vector) %>%
  group_by(Clan_ID, Clan_name) %>%
  summarize(Tot_multiTS = n()) %>%
  ungroup() %>%
  right_join(gene_families_inference_info_df %>% 
               dplyr::select(Clan_ID, Clan_name, Clan_size) %>%
               distinct(), by=c("Clan_ID", "Clan_name")) %>%
  mutate(multiTS_prop = Tot_multiTS / Clan_size) %>%
  left_join(gene_families_inference_info_df %>%
              filter(inference_presence %in% c("Inference_multiTS", "Inference_singleTS")) %>%
              group_by(Clan_ID) %>%
              summarize(Tot_inference = n()) %>%
              ungroup(), by="Clan_ID")


#######################################
###### EXPECTED PROPORTIONS ###########
#######################################

#Compute expected proportions of orthogroups with (multiTS) gains in each family to add to the stacked barplot
#Expected probability is defined by as e = (b/u)*a
#b = total number of OGs in a gene family
#u = total number of OGs with at least one TS gain
#a = total number of OGs with (multiTS) gains

### Define universes and initial numbers
universe_all = 7178
universe_TS = length(all_OG_with_inference_vector)
Tot_TS_across_clans = universe_TS
Tot_multiTS_across_clans = length(multiTS_OGs_vector)

TS_prop_over_all = Tot_TS_across_clans / universe_all
multiTS_prop_over_TS = Tot_multiTS_across_clans / universe_TS

expected_df = multiTS_count_by_gene_family_df %>%
  mutate(Expected_TS = TS_prop_over_all*Clan_size,
         Expected_multiTS = multiTS_prop_over_TS*Tot_inference,
         Expected_multiTS_prop = multiTS_prop_over_TS,
         Expected_TS_prop_to_plot = Expected_TS/Clan_size,
         Expected_multiTS_prop_to_plot = Expected_multiTS/Clan_size) %>%
  dplyr::select(Clan_ID, Clan_name, Clan_size, Tot_inference, Tot_multiTS, Expected_TS, Expected_multiTS,  Expected_multiTS_prop, Expected_TS_prop_to_plot,  Expected_multiTS_prop_to_plot)


#######################################
###### BINOMIAL TEST ##################
#######################################

binomial_test_input_df = expected_df %>%
  dplyr::select(Clan_ID, Clan_name, Tot_inference, Tot_multiTS, Clan_size, Expected_multiTS_prop, Expected_TS_prop_to_plot)

### Testing if there is an enrichment in TS gains in a specific family
#I am doing the binom test with alternative=two.sided, it should be okay to test both enrichment and depletion
pvalue_cutoff = 0.05
binomial_test_output_df = binomial_test_input_df %>%
  dplyr::rowwise() %>%
  mutate(binom_TS_pvalue = binom.test(x=Tot_inference, n=Clan_size, p=Expected_TS_prop_to_plot, alternative="two.sided")$p.value,
         binom_multiTS_pvalue = binom.test(x=Tot_multiTS, n=Tot_inference, p=Expected_multiTS_prop, alternative="two.sided")$p.value) %>%
  mutate(binom_TS_signif = case_when(binom_TS_pvalue <= pvalue_cutoff ~ "*",
                                     TRUE ~ ""),
         binom_multiTS_signif = case_when(binom_multiTS_pvalue <= pvalue_cutoff ~ "*",
                                     TRUE ~ ""))

#######################################
###### BINOMIAL TEST ##################
#######################################

#######################################
###### PLOT ###########################
#######################################

#########################
#### Stacked bar ########
#########################
gene_families_inference_info_df$inference_presence = factor(gene_families_inference_info_df$inference_presence, levels=rev(c("Inference_multiTS", "Inference_singleTS", "NO_inference")))

inferences_by_gene_family_stacked_plot = ggplot() +
  geom_bar(data=gene_families_inference_info_df, 
           aes(x=Clan_name, fill=inference_presence, color=inference_presence), 
           stat="count", alpha=0.5, position = position_fill()) +
  geom_hline(aes(yintercept=0.457), color="limegreen", size=1) +
  geom_errorbar(data=expected_df, aes(x=Clan_name, xmin=Clan_name, xmax=Clan_name,
                                      y=Expected_multiTS_prop_to_plot, ymin=Expected_multiTS_prop_to_plot, ymax=Expected_multiTS_prop_to_plot,
                                      group=Clan_name), color="dodgerblue2", size=0.7) +
  scale_fill_manual(values=c("Inference_singleTS"="limegreen", "Inference_multiTS"="dodgerblue2", "NO_inference"="white")) +
  scale_color_manual(values=c("Inference_singleTS"="white", "Inference_multiTS"="white", "NO_inference"="white")) +
  theme_bw() +
  theme(axis.text = element_text(color="black"),
        axis.title = element_text(color="black"),
        legend.position = "bottom") +
  ylab("") +
  coord_flip()

#########################
#### Total number #######
#########################
inferences_by_gene_family_count_plot = ggplot(data=gene_families_inference_info_df, aes(x=Clan_name)) +
  geom_bar(stat="count", color="black", fill="dimgray", alpha=0.6) +
  #scale_fill_manual(values=c("Inference"="dimgray", "NO_inference"="white")) +
  theme_bw() +
  theme(axis.text = element_text(color="black"),
        axis.title = element_text(color="black")) +
  coord_flip()


####### Combine plots
plot_grid(inferences_by_gene_family_stacked_plot, inferences_by_gene_family_count_plot + theme(axis.title.y=element_blank(),
                                                                                               axis.text.y=element_blank(),
                                                                                               axis.ticks.y=element_blank()), 
          align="h", axis="bt", rel_widths = c(1,0.6))
```

```{r plot_OGs_gains_losses_by_gene_family_v4, echo=FALSE, messsage=FALSE, warning=FALSE, dependson=c("setup", "functions"), fig.width=9, fig.height=10.5}
##### Work either with gains or with losses (i.e., inferences)
inference_type = "gains"

#######################################
###### GENERATE INPUT #################
#######################################

##### Upload table with gene families information
gene_families_info_file = paste0(pfam_transfers_dir, "PFAM_annot/orthogroups_to_PFAMClan-transferred_from_Hs2-filtered_min20OGs.txt")
gene_families_df = read.delim(gene_families_info_file, header=FALSE, col.names=c("OG_ID", "Clan_ID", "Clan_name"))

##### Generate vector of orthogroups with at least one inference
all_OG_with_inference_vector = vector()
for (tissue in ordered_tissues) {
  inference_file = paste0(ts_gains_losses_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_", inference_type, ".tab")
  OG_with_inference_vector = as.vector(read.delim(inference_file, header=TRUE)$"OG_ID")
  all_OG_with_inference_vector = c(all_OG_with_inference_vector, OG_with_inference_vector)
}
all_OG_with_inference_vector = unique(as.vector(all_OG_with_inference_vector)) #Remove duplicates

##### Generate vector of orthogroups with multi-TS inferences
multiTS_classification_file = paste0(pfam_transfers_dir, "Bilateria_conserved_orthogroups-multiTS_classification-", inference_type, ".tab")
multiTS_df = read.delim(multiTS_classification_file, header=FALSE, col.names=c("OG_ID", "Evo_class"))
multiTS_OGs_vector = unique(as.vector(multiTS_df$OG_ID))

####### Add information about inferences to gene families df
gene_families_inference_info_df = gene_families_df %>%
  mutate(inference_presence = case_when(OG_ID %in% multiTS_OGs_vector ~ "Inference_multiTS",
                                        OG_ID %in% all_OG_with_inference_vector ~ "Inference_singleTS",
                                        TRUE ~ "NO_inference")) %>%
  group_by(Clan_ID, Clan_name) %>%
  mutate(Clan_size = n()) %>%
  ungroup()

####### Order the gene families based on the number of orthogroups they contain
clan_order_df = gene_families_inference_info_df %>%
  filter(inference_presence %in% c("Inference_multiTS", "Inference_singleTS")) %>%
  group_by(Clan_ID, Clan_name, Clan_size) %>%
  summarize(Tot_inferences_prop = n()/Clan_size) %>%
  ungroup() %>%
  arrange(desc(Tot_inferences_prop)) %>%
  distinct()

gene_families_inference_info_df$Clan_ID = factor(gene_families_inference_info_df$Clan_ID, levels=rev(as.vector(clan_order_df$Clan_ID)))
gene_families_inference_info_df$Clan_name = factor(gene_families_inference_info_df$Clan_name, levels=rev(as.vector(clan_order_df$Clan_name)))
gene_families_inference_info_df$inference_presence = factor(gene_families_inference_info_df$inference_presence, 
                                                            levels=rev(c("Inference_multiTS", "Inference_singleTS", "NO_inference")))

####### Generate a dataframe with the number of OGs with multiTS inferences by gene family
#This will be usesd to compute the expected probablity
multiTS_count_by_gene_family_df = gene_families_df %>%
  filter(OG_ID %in% multiTS_OGs_vector) %>%
  group_by(Clan_ID, Clan_name) %>%
  summarize(Tot_multiTS = n()) %>%
  ungroup() %>%
  right_join(gene_families_inference_info_df %>% 
               dplyr::select(Clan_ID, Clan_name, Clan_size) %>%
               distinct(), by=c("Clan_ID", "Clan_name")) %>%
  mutate(multiTS_prop = Tot_multiTS / Clan_size) %>%
  left_join(gene_families_inference_info_df %>%
              filter(inference_presence %in% c("Inference_multiTS", "Inference_singleTS")) %>%
              group_by(Clan_ID) %>%
              summarize(Tot_inference = n()) %>%
              ungroup(), by="Clan_ID")


#######################################
###### EXPECTED PROPORTIONS ###########
#######################################

#Compute expected proportions of orthogroups with (multiTS) gains in each family to add to the stacked barplot
#Expected probability is defined by as e = (b/u)*a
#b = total number of OGs in a gene family
#u = total number of OGs with at least one TS gain
#a = total number of OGs with (multiTS) gains

### Define universes and initial numbers
universe_all = 7178
universe_TS = length(all_OG_with_inference_vector)
Tot_TS_across_clans = universe_TS
Tot_multiTS_across_clans = length(multiTS_OGs_vector)

TS_prop_over_all = Tot_TS_across_clans / universe_all
multiTS_prop_over_TS = Tot_multiTS_across_clans / universe_TS

expected_df = multiTS_count_by_gene_family_df %>%
  mutate(Expected_TS = TS_prop_over_all*Clan_size,
         Expected_multiTS = multiTS_prop_over_TS*Tot_inference,
         Expected_multiTS_prop = multiTS_prop_over_TS,
         Expected_TS_prop_to_plot = Expected_TS/Clan_size,
         Expected_multiTS_prop_to_plot = Expected_multiTS/Clan_size) %>%
  dplyr::select(Clan_ID, Clan_name, Clan_size, Tot_inference, Tot_multiTS, Expected_TS, Expected_multiTS,  Expected_multiTS_prop, Expected_TS_prop_to_plot,  Expected_multiTS_prop_to_plot)


#######################################
###### BINOMIAL TEST ##################
#######################################

binomial_test_input_df = expected_df %>%
  dplyr::select(Clan_ID, Clan_name, Tot_inference, Tot_multiTS, Clan_size, Expected_multiTS_prop, Expected_TS_prop_to_plot)

### Testing if there is an enrichment in TS gains in a specific family
#I am doing the binom test with alternative=two.sided, it should be okay to test both enrichment and depletion
pvalue_cutoff = 0.05
binomial_test_output_df = binomial_test_input_df %>%
  dplyr::rowwise() %>%
  mutate(binom_TS_pvalue = binom.test(x=Tot_inference, n=Clan_size, p=Expected_TS_prop_to_plot, alternative="two.sided")$p.value,
         binom_multiTS_pvalue = binom.test(x=Tot_multiTS, n=Tot_inference, p=Expected_multiTS_prop, alternative="two.sided")$p.value) %>%
  mutate(binom_TS_signif = case_when(binom_TS_pvalue <= pvalue_cutoff ~ "*",
                                     TRUE ~ ""),
         binom_multiTS_signif = case_when(binom_multiTS_pvalue <= pvalue_cutoff ~ "*",
                                     TRUE ~ ""))

#######################################
###### MULTI-TS BY EVO ################
#######################################

multiTS_evo_df = multiTS_df %>%
  left_join(gene_families_df, by="OG_ID") %>%
  mutate(Clan_name = factor(Clan_name, levels=rev(as.vector(clan_order_df$Clan_name))),
         Evo_class = factor(Evo_class, levels=rev(c("Vertebrata_only", "Insecta_only", "Others")))) %>%
  filter(!is.na(Clan_name))

#######################################
###### PLOT ###########################
#######################################

#########################
#### Stacked bar ########
#########################
inferences_by_gene_family_stacked_plot = ggplot() +
  geom_bar(data=gene_families_inference_info_df, 
           aes(x=Clan_name, fill=inference_presence, color=inference_presence), 
           stat="count", alpha=0.5, position = position_fill()) +
  geom_hline(aes(yintercept=0.457), color="limegreen", size=1) +
  geom_errorbar(data=expected_df, aes(x=Clan_name, xmin=Clan_name, xmax=Clan_name,
                                      y=Expected_multiTS_prop_to_plot, ymin=Expected_multiTS_prop_to_plot, ymax=Expected_multiTS_prop_to_plot,
                                      group=Clan_name), color="dodgerblue2", size=0.7) +
  geom_text(data=binomial_test_output_df, aes(x=Clan_name, y=1, label=binom_TS_signif), color="limegreen", size=6, position=position_nudge(x=-0.5)) +
  geom_text(data=binomial_test_output_df, aes(x=Clan_name, y=0.97, label=binom_multiTS_signif), color="dodgerblue2", size=6, position=position_nudge(x=-0.5)) +
  scale_fill_manual(values=c("Inference_singleTS"="limegreen", "Inference_multiTS"="dodgerblue2", "NO_inference"="white")) +
  scale_color_manual(values=c("Inference_singleTS"="white", "Inference_multiTS"="white", "NO_inference"="white")) +
  theme_bw() +
  theme(axis.text.y = element_text(color="black"),
        axis.text.x = element_text(color="black", angle=45, hjust=1, vjust=1),
        axis.title = element_text(color="black"),
        legend.position = "bottom") +
  xlab("") +
  coord_flip() +
  ggtitle("OGs with TS gains across clans")

#########################
#### Total number #######
#########################
inferences_by_gene_family_count_plot = ggplot(data=gene_families_inference_info_df, aes(x=Clan_name)) +
  geom_bar(stat="count", color="dimgray", fill="dimgray", alpha=0.6) +
  #scale_fill_manual(values=c("Inference"="dimgray", "NO_inference"="white")) +
  theme_bw() +
  theme(axis.text.y = element_text(color="black"),
        axis.text.x = element_text(color="black", angle=45, hjust=1, vjust=1),
        axis.title = element_text(color="black")) +
  coord_flip() +
  ggtitle("OGs in clan")

#########################################
#### Proportion of multiTS by Evo #######
#########################################
multiTS_evo_prop_plot = ggplot(data=multiTS_evo_df, aes(x=Clan_name, fill=Evo_class)) +
  geom_bar(stat="count", position=position_fill(), color="white", alpha=0.7, size=0.1) +
  scale_fill_manual(values=c("Vertebrata_only"="#82359D", "Insecta_only"="goldenrod", "Others"="coral3")) +
  theme_bw() +
  theme(axis.text.y = element_text(color="black"),
        axis.text.x = element_text(color="black", angle=45, hjust=1, vjust=1),
        legend.position="bottom") +
  xlab("") +
  coord_flip() +
  ggtitle("MultiTS OGs ~ Evo distribs")

#########################################
#### Combine plots ######################
#########################################
plot_grid(inferences_by_gene_family_stacked_plot + theme(plot.title=element_text(color="black", face="bold", size=10, hjust=0.5)), 
          inferences_by_gene_family_count_plot + theme(axis.title.y=element_blank(),
                                                       axis.text.y=element_blank(),
                                                       axis.ticks.y=element_blank(),
                                                       plot.title=element_text(color="black", face="bold", size=10, hjust=0.5)), 
          multiTS_evo_prop_plot + scale_x_discrete(position="top") + theme(plot.title=element_text(color="black", face="bold", size=10, hjust=0.5)),
          align="h", axis="bt", rel_widths = c(1,0.4, 1), ncol=3)
```

```{r taus_from_all_samples_average_v1, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions"), fig.width=10, fig.height=13}
control_types = c("all_samples_average", "random_metasamples_average")
control_type_colors = c("all_samples_average"="dodgerblue2", "random_metasamples_average"="coral3")

for (control_type in control_types) {
  control_type_color = unname(control_type_colors[control_type])
  all_plots_list = list()
  i = 1
  for (species in all_species[all_species != "Tca"]) {
    ##########################################
    ###### BUILD INPUT #######################
    ##########################################
    file_species = species
    if (species == "BmA") {file_species = "Bmo"}
    if (species == "Hsa") {file_species = "Hs2"}
    if (species == "Mmu") {file_species = "Mm2"}
    if (species == "Bta") {file_species = "Bt2"}
    if (species == "Spu") {file_species = "Sp2"}
    
    original_taus_file = paste0(tau_original_dir, file_species, "-protein_coding_taus.tab")
    original_taus_df = read.delim(original_taus_file, header=FALSE, col.names=c("GeneID", "Tau"))
    control_taus_file = paste0(tau_revision_dir, control_type, "/taus/", file_species, "-protein_coding_taus.tab")
    control_taus_df = read.delim(control_taus_file, header=FALSE, col.names=c("GeneID", "Tau"))
    combined_taus_df = original_taus_df %>%
      inner_join(control_taus_df, by="GeneID", suffix=c("_original", "_control"))
    
    ##########################################
    ###### COMPUTE CORRELATION ###############
    ##########################################
    cor_value = cor.test(combined_taus_df$Tau_original, combined_taus_df$Tau_control, method="pearson")$estimate
      
    ##########################################
    ###### PLOT CORRELATION ##################
    ##########################################
    
    taus_correlation_plot = ggplot(data=combined_taus_df, aes(x=Tau_original, y=Tau_control)) +
      geom_point(alpha=0.6, color=control_type_color) +
      geom_smooth(method="lm", color="black") +
      annotation_custom(grobTree(textGrob(paste0("Pearson: ", round(cor_value,3)),
                                          x=0.4, y=0.95, hjust=0.5))) +
      theme_bw() +
      theme(axis.title = element_text(color="black"),
            axis.text = element_text(color="black"),
            plot.title = element_text(color="black", face="bold", hjust=0.5)) +
      ggtitle(species)
      #ggtitle(paste(species, " ~ original vs", gsub("_", " ", control_type), "taus"))
    
    #print(taus_correlation_plot)
    
    all_plots_list[[i]] = taus_correlation_plot
    i = i+1
  }
  final_plots = plot_grid(plotlist=all_plots_list, ncol=4, align="hv")
  print(final_plots)
}
```

```{r taus_controls_v1, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions"), fig.width=10, fig.height=13}
control_types = c("all_samples_average", "random_metasamples_average")
control_type_colors = c("all_samples_average"="dodgerblue2", "random_metasamples_average"="coral3")

for (control_type in control_types[1]) {
  control_type_color = unname(control_type_colors[control_type])
  all_plots_list = list()
  i = 1
  for (species in all_species[all_species != "Tca"]) {
    ##########################################
    ###### BUILD INPUT #######################
    ##########################################
    file_species = species
    if (species == "BmA") {file_species = "Bmo"}
    if (species == "Hsa") {file_species = "Hs2"}
    if (species == "Mmu") {file_species = "Mm2"}
    if (species == "Bta") {file_species = "Bt2"}
    if (species == "Spu") {file_species = "Sp2"}
    
    original_taus_file = paste0(tau_original_dir, file_species, "-protein_coding_taus.tab")
    original_taus_df = read.delim(original_taus_file, header=FALSE, col.names=c("GeneID", "Tau"))
    control_taus_file = paste0(tau_revision_dir, control_type, "/taus/", file_species, "-protein_coding_taus.tab")
    control_taus_df = read.delim(control_taus_file, header=FALSE, col.names=c("GeneID", "Tau"))
    combined_taus_df = original_taus_df %>%
      inner_join(control_taus_df, by="GeneID", suffix=c("_original", "_control"))
    
    ##########################################
    ###### COMPUTE CORRELATION ###############
    ##########################################
    cor_value = cor.test(combined_taus_df$Tau_original, combined_taus_df$Tau_control, method="pearson")$estimate
      
    ##########################################
    ###### PLOT CORRELATION ##################
    ##########################################
    
    taus_correlation_plot = ggplot(data=combined_taus_df, aes(x=Tau_original, y=Tau_control)) +
      #geom_point(alpha=0.6, color=control_type_color) +
      geom_hex(bins=100) +
      scale_fill_continuous(type="viridis") +
      #geom_smooth(method="lm", color="black") +
      annotation_custom(grobTree(textGrob(paste0("Pearson: ", round(cor_value,3)),
                                          x=0.4, y=0.95, hjust=0.5))) +
      theme_bw() +
      theme(axis.title = element_text(color="black"),
            axis.text = element_text(color="black"),
            plot.title = element_text(color="black", face="bold", hjust=0.5)) +
      ggtitle(species) +
      guides(fill=FALSE)
      #ggtitle(paste(species, " ~ original vs", gsub("_", " ", control_type), "taus"))
    
    #print(taus_correlation_plot)
    
    all_plots_list[[i]] = taus_correlation_plot
    i = i+1
  }
  final_plots = plot_grid(plotlist=all_plots_list, ncol=4, align="hv")
  print(final_plots)
}
```

```{r taus_controls_v2, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions"), fig.width=10, fig.height=13}
control_types = c("all_samples_average", "random_metasamples_average")
control_type_colors = c("all_samples_average"="dodgerblue2", "random_metasamples_average"="coral3")

for (control_type in control_types[1]) {
  control_type_color = unname(control_type_colors[control_type])
  all_plots_list = list()
  i = 1
  for (species in all_species[all_species != "Tca"]) {
    ##########################################
    ###### BUILD INPUT #######################
    ##########################################
    file_species = species
    if (species == "BmA") {file_species = "Bmo"}
    if (species == "Hsa") {file_species = "Hs2"}
    if (species == "Mmu") {file_species = "Mm2"}
    if (species == "Bta") {file_species = "Bt2"}
    if (species == "Spu") {file_species = "Sp2"}
    
    original_taus_file = paste0(tau_original_dir, file_species, "-protein_coding_taus.tab")
    original_taus_df = read.delim(original_taus_file, header=FALSE, col.names=c("GeneID", "Tau"))
    control_taus_file = paste0(tau_revision_dir, control_type, "/taus/", file_species, "-protein_coding_taus.tab")
    control_taus_df = read.delim(control_taus_file, header=FALSE, col.names=c("GeneID", "Tau"))
    combined_taus_df = original_taus_df %>%
      inner_join(control_taus_df, by="GeneID", suffix=c("_original", "_control"))
    
    ##########################################
    ###### COMPUTE CORRELATION ###############
    ##########################################
    cor_value = cor.test(combined_taus_df$Tau_original, combined_taus_df$Tau_control, method="pearson")$estimate
      
    ##########################################
    ###### PLOT CORRELATION ##################
    ##########################################
    
    taus_correlation_plot = ggplot(data=combined_taus_df, aes(x=Tau_original, y=Tau_control)) +
      #geom_point(alpha=0.6, color=control_type_color) +
      geom_hex(bins=70) +
      scale_fill_viridis_c() +
      #geom_smooth(method="lm", color="black") +
      annotation_custom(grobTree(textGrob(paste0("Pearson: ", round(cor_value,3)),
                                          x=0.4, y=0.95, hjust=0.5))) +
      theme_bw() +
      theme(axis.title = element_text(color="black"),
            axis.text = element_text(color="black"),
            plot.title = element_text(color="black", face="bold", hjust=0.5)) +
      ggtitle(species) #+
      #guides(fill=FALSE)
      #ggtitle(paste(species, " ~ original vs", gsub("_", " ", control_type), "taus"))
    
    #print(taus_correlation_plot)
    
    all_plots_list[[i]] = taus_correlation_plot
    i = i+1
  }
  final_plots = plot_grid(plotlist=all_plots_list, ncol=4, align="hv")
  print(final_plots)
}
```

```{r taus_controls_v3, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions"), fig.width=10, fig.height=12}
control_types = c("all_samples_average", "random_metasamples_average")
control_type_colors = c("all_samples_average"="dodgerblue2", "random_metasamples_average"="coral3")

for (control_type in control_types) {
  control_type_color = unname(control_type_colors[control_type])
  all_plots_list = list()
  i = 1
  all_species_combined_taus_df = data.frame()
  for (species in all_species) {
    ##########################################
    ###### BUILD INPUT #######################
    ##########################################
    file_species = species
    if (species == "BmA") {file_species = "Bmo"}
    if (species == "Hsa") {file_species = "Hs2"}
    if (species == "Mmu") {file_species = "Mm2"}
    if (species == "Bta") {file_species = "Bt2"}
    if (species == "Spu") {file_species = "Sp2"}
    
    original_taus_file = paste0(tau_original_dir, file_species, "-protein_coding_taus.tab")
    original_taus_df = read.delim(original_taus_file, header=FALSE, col.names=c("GeneID", "Tau"))
    control_taus_file = paste0(tau_revision_dir, control_type, "/taus/", file_species, "-protein_coding_taus.tab")
    control_taus_df = read.delim(control_taus_file, header=FALSE, col.names=c("GeneID", "Tau"))
    combined_taus_df = original_taus_df %>%
      inner_join(control_taus_df, by="GeneID", suffix=c("_original", "_control")) %>%
      mutate(Species = species)
    
    all_species_combined_taus_df = rbind(all_species_combined_taus_df, combined_taus_df)
  }
  
  all_species_combined_taus_df$Species = factor(all_species_combined_taus_df$Species, levels=ordered_species)
    
  ##########################################
  ###### COMPUTE CORRELATION ###############
  ##########################################
  #cor_value = cor.test(combined_taus_df$Tau_original, combined_taus_df$Tau_control, method="pearson")$estimate
    
  ##########################################
  ###### PLOT CORRELATION ##################
  ##########################################
  
  taus_correlation_plot = ggplot(data=all_species_combined_taus_df, aes(x=Tau_original, y=Tau_control)) +
    #geom_point(alpha=0.6, color=control_type_color) +
    geom_hex(bins=100) +
    scale_fill_viridis_c(limits=c(1,50), oob = scales::squish) + #Everything 50 or higher will be represented in the same color
    #geom_smooth(method="lm", color="black") +
    #annotation_custom(grobTree(textGrob(paste0("Pearson: ", round(cor_value,3)),
    #                                    x=0.4, y=0.95, hjust=0.5))) +
    theme_bw() +
    theme(axis.title = element_text(color="black"),
          axis.text = element_text(color="black"),
          plot.title = element_text(color="black", face="bold", hjust=0.5),
          legend.position = "bottom") +
    coord_equal() +
    facet_wrap(~Species) +
    ggtitle(paste0("Original vs ", gsub("_", " ", control_type), " taus"))
      #ggtitle(species) +
      #guides(fill=FALSE)
      #ggtitle(paste(species, " ~ original vs", gsub("_", " ", control_type), "taus"))
    
    print(taus_correlation_plot)
    
    #all_plots_list[[i]] = taus_correlation_plot
    #i = i+1
  }
  #final_plots = plot_grid(plotlist=all_plots_list, ncol=4, align="hv")
  #print(final_plots)
#}
```

```{r taus_controls_v4, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions"), fig.width=10, fig.height=12}
control_types = c("all_samples_average", "random_metasamples_average")
control_type_colors = c("all_samples_average"="dodgerblue2", "random_metasamples_average"="coral3")

for (control_type in control_types) {
  control_type_color = unname(control_type_colors[control_type])
  all_plots_list = list()
  i = 1
  all_species_combined_taus_df = data.frame()
  for (species in all_species) {
    ##########################################
    ###### BUILD INPUT #######################
    ##########################################
    file_species = species
    if (species == "BmA") {file_species = "Bmo"}
    if (species == "Hsa") {file_species = "Hs2"}
    if (species == "Mmu") {file_species = "Mm2"}
    if (species == "Bta") {file_species = "Bt2"}
    if (species == "Spu") {file_species = "Sp2"}
    
    ### NB: this contains taus only for genes that pass the expression cutoff
    original_taus_file = paste0(tau_original_dir, file_species, "-protein_coding_taus.tab")
    original_taus_df = read.delim(original_taus_file, header=FALSE, col.names=c("GeneID", "Tau"))
    control_taus_file = paste0(tau_revision_dir, control_type, "/taus/", file_species, "-protein_coding_taus.tab")
    control_taus_df = read.delim(control_taus_file, header=FALSE, col.names=c("GeneID", "Tau"))
    combined_taus_df = original_taus_df %>%
      inner_join(control_taus_df, by="GeneID", suffix=c("_original", "_control")) %>%
      mutate(Species = species)
    
    all_species_combined_taus_df = rbind(all_species_combined_taus_df, combined_taus_df)
  }
  
  all_species_combined_taus_df$Species = factor(all_species_combined_taus_df$Species, levels=ordered_species)
    
  ##########################################
  ###### COMPUTE CORRELATION ###############
  ##########################################
  #cor_value = cor.test(combined_taus_df$Tau_original, combined_taus_df$Tau_control, method="pearson")$estimate
    
  ##########################################
  ###### PLOT CORRELATION ##################
  ##########################################
  
  taus_correlation_plot = ggplot(data=all_species_combined_taus_df, aes(x=Tau_original, y=Tau_control)) +
    #geom_point(alpha=0.6, color=control_type_color) +
    geom_hex(bins=100) +
    scale_fill_viridis_c(limits=c(1,50), oob = scales::squish) + #Everything 50 or higher will be represented in the same color
    #geom_smooth(method="lm", color="black") +
    #annotation_custom(grobTree(textGrob(paste0("Pearson: ", round(cor_value,3)),
    #                                    x=0.4, y=0.95, hjust=0.5))) +
    theme_bw() +
    theme(axis.title = element_text(color="black"),
          axis.text = element_text(color="black"),
          plot.title = element_text(color="black", face="bold", hjust=0.5),
          legend.position = "bottom") +
    coord_equal() +
    facet_wrap(~Species) +
    ggtitle(paste0("Original vs ", gsub("_", " ", control_type), " taus"))
    
    print(taus_correlation_plot)
  }
```

```{r duplication_levels_randomized_call_species_v1, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions")}
####### Build color vector
# deuterostome_colors = rep("olivedrab3", length(deuterostoma_y_labels)-1)
# names(deuterostome_colors) = deuterostoma_y_labels[deuterostoma_y_labels!="Bilateria"]
# protostome_colors = rep("tan3", length(protostoma_y_labels)-1)
# names(protostome_colors) = protostoma_y_labels[protostoma_y_labels!="Bilateria"]
# all_nodes_species_colors = c(deuterostome_colors, protostome_colors, my_species_colors)

#######################################################
######### UPLOAD ORTHOGROUPS ##########################
#######################################################
orthogroups_df = read.delim(paste0(ts_gains_losses_raw_dir, "Bilateria_conserved_orthogroups-All_genes-associated_tissues.tab"),
                            header=FALSE, col.names=c("OG_ID", "Species", "GeneID", "Tau", "Tissue", "Paralog_tissue", "Expr_cutoff"), stringsAsFactors = FALSE)

####### Add OGID_Species column
orthogroups_df$OG_ID_Species = paste0(orthogroups_df$OG_ID, "_", orthogroups_df$Species)
####### Count number of paralogs per species in each orthogroup
species_paralogs_count_df = as.data.frame(table(orthogroups_df$OG_ID_Species))
OG_ID_Species_paralogs_counts_dict = hashmap(as.vector(species_paralogs_count_df$Var1), as.vector(species_paralogs_count_df$Freq))
####### Add paralog count to dataframe
orthogroups_df$Paralogs_count = OG_ID_Species_paralogs_counts_dict$find(orthogroups_df$OG_ID_Species)

#######################################################
######### UPLOAD TS GAINS #############################
#######################################################
tissue_gains_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_raw_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  tissue_gains_df = rbind(tissue_gains_df, input_df)
}

#######################################################
######### UPLOAD TS LOSSES ############################
#######################################################
tissue_losses_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_raw_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_losses.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_losses", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  tissue_losses_df = rbind(tissue_losses_df, input_df)
}

###### Filter only for gains that happened in the node of interest
all_OGs_tissue_gains = paste0(tissue_gains_df$OG_ID, "_", tissue_gains_df$Tissue)
single_tissue_gains_OG_IDs = all_OGs_tissue_gains[!duplicated(all_OGs_tissue_gains)]
tissue_gains_df$OG_ID_tissue = paste0(tissue_gains_df$OG_ID, "_", tissue_gains_df$Tissue)
single_tissue_gains_df = subset(tissue_gains_df, OG_ID_tissue %in% single_tissue_gains_OG_IDs)

###### Upload the best-TS orthogroups for which the inference was performed
all_tissue_tau_df = vector()
for (tissue in ordered_tissues) {
  tissue_tau_df = read.delim(paste0(ts_gains_losses_raw_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_specific_OGs.tab"),
                           header=FALSE, col.names=c("OG_ID", "Species", "GeneID", "Tau", "Tissue", "Expr_cutoff"))
  tissue_tau_df$Query_tissue = rep(tissue, nrow(tissue_tau_df))
  all_tissue_tau_df = rbind(all_tissue_tau_df, tissue_tau_df)
}

###### Build GeneID - OG_ID dictionary
unique_tissue_tau_df = unique(all_tissue_tau_df[,c("OG_ID", "Species", "GeneID")])
GeneID_OG_ID_dict = hashmap(as.vector(unique_tissue_tau_df$GeneID), as.vector(unique_tissue_tau_df$OG_ID))


#######################################################
######### UPLOAD TS LOSSES ############################
#######################################################

#### Filter on species and tissue gains, then combine in a unique dataframe
all_selected_paralogs_df = vector()
for (species in original_species) {
  for (tissue in ordered_tissues) {
    #### For each tissue, get the OG_ID of the OG_ID with only one gain
    species_tissue_gains_OG_IDs = as.vector(subset(single_tissue_gains_df, TS_gains==species & Tissue==tissue)$OG_ID)
    #### Select all the geneIDs from the orthogroups with gains
    species_tissue_gains_genes = as.vector(subset(all_tissue_tau_df, OG_ID %in% species_tissue_gains_OG_IDs & Query_tissue==tissue)$GeneID)
    #### Subset to only the genes belonging to the orthogroup with gains in the queried species
    selected_paralogs_df = subset(orthogroups_df, GeneID %in% species_tissue_gains_genes)
    #### Add Category
    #selected_paralogs_df$OG_ID = GeneID_OG_ID_dict$find(selected_paralogs_df$GeneID)
    selected_paralogs_df$Category = rep("NO_TS", nrow(selected_paralogs_df))
    selected_paralogs_df$Category[selected_paralogs_df$Species==species] = "TS" #Distinguish the tissue-specific genes
    
    #### Cycle on each OG_ID and remove the species whose tau is NA (so the gene is not expressed enough to pass the cutoffs
    # for (my_OG_ID in unique(as.vector(selected_paralogs_df$OG_ID))) {
    #   selected_paralogs_df = subset(selected_paralogs_df, !(OG_ID==my_OG_ID & Species %in% as.vector(subset(all_tissue_tau_df, 
    #                                                                                 Query_tissue==tissue & 
    #                                                                                   OG_ID==my_OG_ID & 
    #                                                                                   is.na(Tau)))))
    # }
    selected_paralogs_df$Node = rep(species, nrow(selected_paralogs_df))
    all_selected_paralogs_df = rbind(all_selected_paralogs_df, selected_paralogs_df)
  }
}

all_selected_paralogs_df$Category = factor(all_selected_paralogs_df$Category, levels=c("TS", "NO_TS"))
all_selected_paralogs_df$Node = factor(all_selected_paralogs_df$Node, levels = ordered_species_tree)

##### Remove all the genes used as NO_TS if they appear as TS in some other Tissues
TS_genes = as.vector(subset(all_selected_paralogs_df, Category=="TS")$GeneID)
final_all_selected_paralogs_df = unique(subset(all_selected_paralogs_df, !(GeneID %in% TS_genes & Category=="NO_TS")))

##### For each OG_ID, compute the mean paralog number bewtween TS and non-TS species
mean_paralogs_df = final_all_selected_paralogs_df %>%
  group_by(OG_ID, Node, Category) %>%
  summarize(Mean_paralog_count = mean(Paralogs_count))

#Subset only for TS species
final_mean_paralogs_df = subset(mean_paralogs_df, Category=="TS")

#Compute proportions of each species-specific gains where there is at least one duplicate
duplicated_status_df = final_mean_paralogs_df
duplicated_status_df$Dup_status = "Dup"
duplicated_status_df$Dup_status[duplicated_status_df$Mean_paralog_count==1] = "Non_Dup"
duplicated_status_df$Dup_status = factor(duplicated_status_df$Dup_status, levels=rev(c("Dup", "Non_Dup")))

#### Add branch info for plotting
duplicated_status_df$Branch = rep("Deuterostoma", nrow(duplicated_status_df))
duplicated_status_df$Branch[duplicated_status_df$Node %in% Protostoma] = "Protostoma"

##########
background_proportions_df = orthogroups_df %>%
  dplyr::select(OG_ID, Species, Paralogs_count) %>%
  distinct() %>%
  mutate(Dup_status = case_when(Paralogs_count == 1 ~ "Non_Dup",
                                 Paralogs_count > 1 ~ "Dup"))
background_proportions_df$Species = factor(background_proportions_df$Species, levels=original_species)
background_proportions_df$Dup_status = factor(background_proportions_df$Dup_status, levels=rev(c("Dup", "Non_Dup")))

##########
final_background_proportions_df = background_proportions_df %>%
  group_by(Species, Dup_status) %>%
  summarise(n = n()) %>%
  mutate(freq = n / sum(n)) %>%
  filter(Dup_status=="Dup")

final_background_proportions_df$Species = factor(final_background_proportions_df$Species, levels=original_species)

#### Combine branch and dup status
duplicated_status_df$Branch_Dup_status = paste0(duplicated_status_df$Branch, "_", duplicated_status_df$Dup_status)
duplicated_status_df$Branch_Dup_status = factor(duplicated_status_df$Branch_Dup_status, 
                                                 levels=rev(c("Deuterostoma_Dup", "Deuterostoma_Non_Dup", "Protostoma_Dup", "Protostoma_Non_Dup")))
duplicated_status_df$Node = factor(duplicated_status_df$Node, levels=original_species)

### Plot
species_specific_gains_duplication_status = ggplot() +
  geom_bar(data=duplicated_status_df, aes(x=Node, fill=Branch_Dup_status), stat="count", position="fill", color="black", alpha=0.75) +
  geom_errorbar(data=final_background_proportions_df, aes(x=Species, ymin=freq, ymax=freq), color="black", linetype="dashed") +
  scale_fill_manual(values=c("Deuterostoma_Dup"="darkolivegreen4", "Protostoma_Dup"="tan3", "Deuterostoma_Non_Dup"="white", "Protostoma_Non_Dup"="white")) +
  theme_bw() +
  theme(axis.text.y = element_text(color="black"),
        axis.title = element_blank(),
        axis.text.x = element_text(color="black", angle=45, hjust=1, vjust=1),
        axis.ticks.y = element_line()) +
  ylab("Proportion of OGs with gains") +
  guides(fill="none")
  
print(species_specific_gains_duplication_status)
```

```{r duplication_levels_randomized_call_species_v2, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions")}
#######################################################
######### UPLOAD ORTHOGROUPS ##########################
#######################################################
orthogroups_df = read.delim(paste0(ts_gains_losses_raw_dir, "Bilateria_conserved_orthogroups-All_genes-associated_tissues.tab"),
                            header=FALSE, col.names=c("OG_ID", "Species", "GeneID", "Tau", "Tissue", "Paralog_tissue", "Expr_cutoff"), stringsAsFactors = FALSE)

####### Add OGID_Species column
orthogroups_df$OG_ID_Species = paste0(orthogroups_df$OG_ID, "_", orthogroups_df$Species)
####### Count number of paralogs per species in each orthogroup
species_paralogs_count_df = as.data.frame(table(orthogroups_df$OG_ID_Species))
OG_ID_Species_paralogs_counts_dict = hashmap(as.vector(species_paralogs_count_df$Var1), as.vector(species_paralogs_count_df$Freq))
####### Add paralog count to dataframe
orthogroups_df$Paralogs_count = OG_ID_Species_paralogs_counts_dict$find(orthogroups_df$OG_ID_Species)

#######################################################
######### UPLOAD TS GAINS #############################
#######################################################
tissue_gains_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_raw_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  tissue_gains_df = rbind(tissue_gains_df, input_df)
}

#######################################################
######### UPLOAD TS LOSSES ############################
#######################################################
tissue_losses_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_raw_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_losses.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_losses", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  tissue_losses_df = rbind(tissue_losses_df, input_df)
}

###### Upload the best-TS orthogroups for which the inference was performed
all_tissue_tau_df = vector()
for (tissue in ordered_tissues) {
  tissue_tau_df = read.delim(paste0(ts_gains_losses_raw_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_specific_OGs.tab"),
                           header=FALSE, col.names=c("OG_ID", "Species", "GeneID", "Tau", "Tissue", "Expr_cutoff"))
  tissue_tau_df$Query_tissue = rep(tissue, nrow(tissue_tau_df))
  all_tissue_tau_df = rbind(all_tissue_tau_df, tissue_tau_df)
}

###### Build GeneID - OG_ID dictionary
unique_tissue_tau_df = unique(all_tissue_tau_df[,c("OG_ID", "Species", "GeneID")])
GeneID_OG_ID_dict = hashmap(as.vector(unique_tissue_tau_df$GeneID), as.vector(unique_tissue_tau_df$OG_ID))


#######################################################
######### GET DUPLICATION STATUS TS GAINS #############
#######################################################

#### Filter on species and tissue gains, then combine in a unique dataframe
all_selected_paralogs_df = vector()
for (species in original_species) {
  for (tissue in ordered_tissues) {
    #### For each tissue, get the OG_ID of the OG_ID with only one gain
    species_tissue_gains_OG_IDs = as.vector(subset(tissue_gains_df, TS_gains==species & Tissue==tissue)$OG_ID)
    #### Select all the geneIDs from the orthogroups with gains
    species_tissue_gains_genes = as.vector(subset(all_tissue_tau_df, OG_ID %in% species_tissue_gains_OG_IDs & Query_tissue==tissue)$GeneID)
    #### Subset to only the genes belonging to the orthogroup with gains in the queried species
    selected_paralogs_df = subset(orthogroups_df, GeneID %in% species_tissue_gains_genes)
    #### Add Category
    selected_paralogs_df$Category = rep("NO_TS", nrow(selected_paralogs_df))
    selected_paralogs_df$Category[selected_paralogs_df$Species==species] = "TS" #Distinguish the tissue-specific genes
    #### Add Species
    selected_paralogs_df$Node = rep(species, nrow(selected_paralogs_df))
    all_selected_paralogs_df = rbind(all_selected_paralogs_df, selected_paralogs_df)
  }
}

all_selected_paralogs_df$Category = factor(all_selected_paralogs_df$Category, levels=c("TS", "NO_TS"))
all_selected_paralogs_df$Node = factor(all_selected_paralogs_df$Node, levels = ordered_species_tree)

##### Remove all the genes used as NO_TS if they appear as TS in some other Tissues
TS_genes = as.vector(subset(all_selected_paralogs_df, Category=="TS")$GeneID)
final_all_selected_paralogs_df = unique(subset(all_selected_paralogs_df, !(GeneID %in% TS_genes & Category=="NO_TS")))

##### For each OG_ID, compute the mean paralog number bewtween TS and non-TS species
mean_paralogs_df = final_all_selected_paralogs_df %>%
  group_by(OG_ID, Node, Category) %>%
  summarize(Mean_paralog_count = mean(Paralogs_count))

#Subset only for TS species
final_mean_paralogs_df = subset(mean_paralogs_df, Category=="TS")

#Compute proportions of each species-specific gains where there is at least one duplicate
duplicated_status_df = final_mean_paralogs_df
duplicated_status_df$Dup_status = "Dup"
duplicated_status_df$Dup_status[duplicated_status_df$Mean_paralog_count==1] = "Non_Dup"
duplicated_status_df$Dup_status = factor(duplicated_status_df$Dup_status, levels=rev(c("Dup", "Non_Dup")))

#### Add branch info for plotting
duplicated_status_df$Branch = rep("Deuterostoma", nrow(duplicated_status_df))
duplicated_status_df$Branch[duplicated_status_df$Node %in% Protostoma] = "Protostoma"
duplicated_status_df$Branch_Dup_status = paste0(duplicated_status_df$Branch, "_", duplicated_status_df$Dup_status)

#### Order factor levels
duplicated_status_df$Branch_Dup_status = factor(duplicated_status_df$Branch_Dup_status, 
                                                 levels=rev(c("Deuterostoma_Dup", "Deuterostoma_Non_Dup", "Protostoma_Dup", "Protostoma_Non_Dup")))
duplicated_status_df$Node = factor(duplicated_status_df$Node, levels=ordered_species_tree)

#######################################################
######### GET DUPLICATION STATUS BACKGROUND ###########
#######################################################

background_proportions_df = orthogroups_df %>%
  dplyr::select(OG_ID, Species, Paralogs_count) %>%
  distinct() %>%
  mutate(Dup_status = case_when(Paralogs_count == 1 ~ "Non_Dup",
                                 Paralogs_count > 1 ~ "Dup"))

final_background_proportions_df = background_proportions_df %>%
  group_by(Species, Dup_status) %>%
  summarise(n = n()) %>%
  mutate(freq = n / sum(n)) %>%
  filter(Dup_status=="Dup")

#### Order factor levels
final_background_proportions_df$Species = factor(final_background_proportions_df$Species, levels=original_species)

#######################################################
#################### PLOT #############################
#######################################################

species_specific_gains_duplication_status = ggplot() +
  geom_bar(data=duplicated_status_df, aes(x=Node, fill=Branch_Dup_status), stat="count", position="fill", color="black", alpha=0.75) +
  geom_errorbar(data=final_background_proportions_df, aes(x=Species, ymin=freq, ymax=freq), color="black", linetype="dashed") +
  scale_fill_manual(values=c("Deuterostoma_Dup"="darkolivegreen4", "Protostoma_Dup"="tan3", "Deuterostoma_Non_Dup"="white", "Protostoma_Non_Dup"="white")) +
  theme_bw() +
  theme(axis.text.y = element_text(color="black"),
        axis.title = element_blank(),
        axis.text.x = element_text(color="black", angle=45, hjust=1, vjust=1),
        axis.ticks.y = element_line()) +
  ylab("Proportion of OGs with gains") +
  guides(fill="none")
  
print(species_specific_gains_duplication_status)
```

```{r randomization_gains_stats_v1, echo=FALSE, warning=FALSE, messsage=FALSE, dependson=c("setup", "functions"), fig.width=10, fig.height=6}
#Plot total numbers of TS gains across ancestors and species coming from the randomizations

#######################################################
#################### UPLOAD INPUTS ####################
#######################################################
all_rand_tissue_gains_df = vector()
for (randomization_round in seq(1,10)) {
  ### Set randomization dir
  randomization_dir = paste0(home, "projects/bilaterian_GE/data/ts_gains_losses_revision/randomization_", randomization_round, "/")
  
  for (tissue in ordered_tissues) {
    input_df = read.delim(paste0(randomization_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
    ### Add tissue and randomization round info
    input_df$Tissue = rep(tissue, nrow(input_df))
    input_df$Randomization_round = rep(randomization_round, nrow(input_df))
    ### Join to final dataframe
    all_rand_tissue_gains_df = rbind(all_rand_tissue_gains_df, input_df)
  }
}

### Add branch info
all_rand_tissue_gains_df$Branch = rep("Deuterostoma", nrow(all_rand_tissue_gains_df))
all_rand_tissue_gains_df$Branch[all_rand_tissue_gains_df$TS_gains %in% c(protostoma_y_labels, Protostoma)] = "Protostoma"
all_rand_tissue_gains_df$Branch[all_rand_tissue_gains_df$TS_gains=="Bilateria"] = "Bilateria"

### Count total ts gains per rand round, tissue and node (both STRICT and LASSO separately and ALL)
all_rand_tissue_gains_stats_df = all_rand_tissue_gains_df %>%
  group_by(Randomization_round, Branch, TS_gains, Category) %>%
  summarize(Tot_gains = n()) %>%
  ungroup() %>%
  bind_rows(all_rand_tissue_gains_df %>%
              mutate(Category = "ALL") %>%
              group_by(Randomization_round, Branch, TS_gains, Category) %>%
              summarize(Tot_gains = n()) %>%
              ungroup()) %>%
  filter(Branch != "Bilateria")

### Add node level
all_rand_tissue_gains_stats_df$Node_level = ancestor_node_level_dict$find(all_rand_tissue_gains_stats_df$TS_gains)
all_rand_tissue_gains_stats_df$Node_level = factor(as.vector(all_rand_tissue_gains_stats_df$Node_level), levels=seq(0,10))
all_rand_tissue_gains_stats_df$Randomization_round = factor(all_rand_tissue_gains_stats_df$Randomization_round, levels=seq(1,10))
all_rand_tissue_gains_stats_df$TS_gains = factor(all_rand_tissue_gains_stats_df$TS_gains, levels=c(all_ancestors, all_species))

#######################################################
####### ORIGINAL TS GAINS #############################
#######################################################
original_ts_gains_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_raw_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), 
                        header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
  ### Add tissue and randomization round info
  input_df$Tissue = rep(tissue, nrow(input_df))
  ### Join to final dataframe
  original_ts_gains_df = rbind(original_ts_gains_df, input_df)
}

### Add branch info
original_ts_gains_df$Branch = rep("Deuterostoma", nrow(original_ts_gains_df))
original_ts_gains_df$Branch[original_ts_gains_df$TS_gains %in% c(protostoma_y_labels, Protostoma)] = "Protostoma"
original_ts_gains_df$Branch[original_ts_gains_df$TS_gains=="Bilateria"] = "Bilateria"

original_ts_gains_stats_df = original_ts_gains_df %>%
    group_by(Branch, TS_gains, Category) %>%
    summarize(Tot_gains = n()) %>%
    ungroup() %>%
    bind_rows(original_ts_gains_df %>%
              mutate(Category = "ALL") %>%
              group_by(Branch, TS_gains, Category) %>%
              summarize(Tot_gains = n()) %>%
              ungroup()) %>%
  filter(Branch != "Bilateria")

### Add node level
original_ts_gains_stats_df$Node_level = ancestor_node_level_dict$find(original_ts_gains_stats_df$TS_gains)
original_ts_gains_stats_df$Node_level = factor(as.vector(original_ts_gains_stats_df$Node_level), levels=seq(0,10))
original_ts_gains_stats_df$TS_gains = factor(original_ts_gains_stats_df$TS_gains, levels=c(all_ancestors, all_species))


#######################################################
#################### PLOT #############################
#######################################################
random_inferences_plot = ggplot(data=subset(all_rand_tissue_gains_stats_df, Category=="ALL" & Node_level!=10), 
                                aes(x=TS_gains, y=Tot_gains, group=Category, shape=Category, color=Randomization_round)) +
  geom_jitter() +
  geom_boxplot(data=subset(all_rand_tissue_gains_stats_df, Category=="ALL" & Node_level!=10),
               aes(x=TS_gains, y=Tot_gains, group=TS_gains)) +
  #######
  geom_line(data=subset(original_ts_gains_stats_df, Category=="ALL" & Node_level!=10),
            aes(x=TS_gains, y=Tot_gains, group=Category, shape=Category), color="black") +
  ######
  theme_bw() +
  theme(axis.text.x = element_text(color="black", angle=45, hjust=1, vjust=1),
        axis.text.y = element_text(color="black"),
        axis.title = element_text(color="black")) +
  theme(legend.position = "bottom") +
  facet_wrap(~Branch, scales="free_x")

print(random_inferences_plot)
  #stat_summary(fun=median, geom="path")
```

```{r duplication_levels_randomized_call_ancestors_v1, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions")}
#Upload orthogroups and get duplication info about each gene
orthogroups_df = read.delim(paste0(ts_gains_losses_raw_dir, "Bilateria_conserved_orthogroups-All_genes-associated_tissues.tab"),
                            header=FALSE, col.names=c("OG_ID", "Species", "GeneID", "Tau", "Tissue", "Paralog_tissue", "Expr_cutoff"), stringsAsFactors = FALSE)
#### Add column with OG_ID_Species and count number of paralogs per gene
orthogroups_df$OG_ID_Species = paste0(orthogroups_df$OG_ID, "_", orthogroups_df$Species)
species_paralogs_count_df = as.data.frame(table(orthogroups_df$OG_ID_Species))
OG_ID_Species_paralogs_counts_dict = hashmap(as.vector(species_paralogs_count_df$Var1), as.vector(species_paralogs_count_df$Freq))
orthogroups_df$Paralogs_count = OG_ID_Species_paralogs_counts_dict$find(orthogroups_df$OG_ID_Species)

#########################################################
############## GENERATE BACKGROUND ######################
#########################################################
#### For each node (group of species) compute the proportion in each orthogroup that is duplicated. Get the difference for non-duplicated.
#Sum all and get proportions.
background_df = vector()
for (ancestor in all_ancestors) {
  my_species = get(ancestor)
  #Subset all orthogroups to only these species
  my_species_orthogroups_df = subset(orthogroups_df, Species %in% my_species)
  #Compute for each orthogroup the proportion of duplicated species (among the ones with conserved genes)
   final_ancestor_df = my_species_orthogroups_df %>%
      dplyr::select(OG_ID, Species, Paralogs_count) %>%
      distinct() %>%
      mutate(Dup_status=case_when(Paralogs_count > 1 ~ "Dup",
                                  Paralogs_count <= 1 ~ "NO_Dup")) %>%
      group_by(OG_ID, Dup_status) %>%
      summarise(total = n()) %>%
      ungroup() %>%
      pivot_wider(id_cols="OG_ID", names_from="Dup_status", values_from="total", values_fill=0) %>%
      pivot_longer(cols=c("Dup", "NO_Dup"), names_to = "Dup_status") %>%
      group_by(OG_ID) %>%
      mutate(freq=value/sum(value)) %>%
      mutate(freq = replace_na(freq, 0))
  
  collapsed_final_df = final_ancestor_df %>%
    group_by(Dup_status) %>%
    summarise(freq_proportion = mean(freq)) %>%
    mutate(Node = ancestor) %>%
    dplyr::select(Node, Dup_status, freq_proportion)
  
  background_df = rbind(background_df, collapsed_final_df)
}
#### Order levels
background_df$Node = factor(background_df$Node, levels=c(deuterostoma_y_labels[deuterostoma_y_labels!="Bilateria"], "Bilateria",
                                                             rev(protostoma_y_labels[protostoma_y_labels!="Bilateria"])))
background_df$Dup_status = factor(background_df$Dup_status, levels=rev(c("Dup", "NO_Dup")))
#### Subset only to duplicates
dup_background_df = subset(background_df, Dup_status=="Dup")

#########################################################
############## GENERATE FOREGROUND ######################
#########################################################

#############################
#### Upload TS gains ########
#############################

tissue_gains_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  tissue_gains_df = rbind(tissue_gains_df, input_df)
}

#############################
#### Upload TS losses #######
#############################
tissue_losses_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_losses.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  tissue_losses_df = rbind(tissue_losses_df, input_df)
}

#############################
#### Select relative OGs ####
#############################
foreground_df = vector()
for (ancestor in all_ancestors) {
  #### Select orthogroups with gains
  ancestor_gains_OGs = unique(as.vector(subset(tissue_gains_df, TS_gains==ancestor)$OG_ID))
  #### Select ancestor species
  my_species = get(ancestor)
  #### Subset all orthogroups to only these species in the orthogroups with gains
  my_species_orthogroups_df = subset(orthogroups_df, Species %in% my_species & TS_gains %in% ancestor_gains_OGs &)
  ####################################################################################
  #### For each orthogroup, remove the species with losses ###########################
  ####################################################################################
  ancestor_tissue_losses_df = tissue_losses_df %>% filter(OG_ID %in% ancestor_gains_OGs, Species %in% my_species)
  ####################################################################################
  
  #Compute the same statistics as above
    final_ancestor_df = my_species_orthogroups_df %>%
      dplyr::select(OG_ID, Species, Paralogs_count) %>%
      distinct() %>%
      mutate(Dup_status=case_when(Paralogs_count > 1 ~ "Dup",
                                  Paralogs_count <= 1 ~ "NO_Dup")) %>%
      group_by(OG_ID, Dup_status) %>%
      summarise(total = n()) %>%
      ungroup() %>%
      pivot_wider(id_cols="OG_ID", names_from="Dup_status", values_from="total", values_fill=0) %>%
      pivot_longer(cols=c("Dup", "NO_Dup"), names_to = "Dup_status") %>%
      group_by(OG_ID) %>%
      mutate(freq=value/sum(value)) %>%
      mutate(freq = replace_na(freq, 0))
  
  collapsed_final_df = final_ancestor_df %>%
    group_by(Dup_status) %>%
    summarise(freq_proportion = mean(freq)) %>%
    mutate(Node = ancestor) %>%
    dplyr::select(Node, Dup_status, freq_proportion)
  
  foreground_df = rbind(foreground_df, collapsed_final_df)
}

#Select only the TS species (species without losses) and compute the proportion of those species that are duplicated.
#Compute the difference as not-duplicated (it has to sum to 1).

####### Order:
foreground_df$Node = factor(foreground_df$Node, levels=c(deuterostoma_y_labels[deuterostoma_y_labels!="Bilateria"], "Bilateria",
                                                             rev(protostoma_y_labels[protostoma_y_labels!="Bilateria"])))
foreground_df$Dup_status = factor(foreground_df$Dup_status, levels=rev(c("Dup", "NO_Dup")))

###### Add branch info
foreground_df$Branch = rep("Deuterostoma", nrow(foreground_df))
foreground_df$Branch[foreground_df$Node %in% protostoma_y_labels] = "Protostoma"
foreground_df$Branch[foreground_df$Node=="Bilateria"] = "Bilateria"
foreground_df = foreground_df %>%
  unite("Branch_Dup_status", c(Branch,Dup_status), remove=FALSE)
foreground_df$Branch_Dup_status = factor(foreground_df$Branch_Dup_status, levels=rev(c("Deuterostoma_Dup", "Deuterostoma_NO_Dup", "Protostoma_Dup", "Protostoma_NO_Dup", "Bilateria_Dup", "Bilateria_NO_Dup")))


#########################################################
##### GENERATE FOREGROUND RANDOMIZATIONS ################
#########################################################
all_rand_foregrounds_df = vector()

for (randomization_round in seq(1,10)) {
  ######### Set directory
  randomization_dir = paste0(home, "projects/bilaterian_GE/data/ts_gains_losses_revision/randomization_", randomization_round, "/")
  
  ######### Upload TS gains
  tissue_gains_df = vector()
  for (tissue in ordered_tissues) {
    input_df = read.delim(paste0(randomization_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
    input_df$Tissue = rep(tissue, nrow(input_df))
    tissue_gains_df = rbind(tissue_gains_df, input_df)
  }
  
  ######### Build foreground
  rand_foreground_df = vector()
  for (ancestor in all_ancestors) {
    #### Select orthogroups with gains
    ancestor_gains_OGs = unique(as.vector(subset(tissue_gains_df, TS_gains==ancestor)$OG_ID))
    #### Select ancestor species
    my_species = get(ancestor)
    #### Subset all orthogroups to only these species in the orthogroups with gains
    my_species_orthogroups_df = subset(orthogroups_df, Species %in% my_species & OG_ID %in% ancestor_gains_OGs)
    #### For each orthogroup, remove the species with losses
    #Leave this pending for one second
    #Compute the same statistics as above
      final_ancestor_df = my_species_orthogroups_df %>%
        dplyr::select(OG_ID, Species, Paralogs_count) %>%
        distinct() %>%
        mutate(Dup_status=case_when(Paralogs_count > 1 ~ "Dup",
                                    Paralogs_count <= 1 ~ "NO_Dup")) %>%
        group_by(OG_ID, Dup_status) %>%
        summarise(total = n()) %>%
        ungroup() %>%
        pivot_wider(id_cols="OG_ID", names_from="Dup_status", values_from="total", values_fill=0) %>%
        pivot_longer(cols=c("Dup", "NO_Dup"), names_to = "Dup_status") %>%
        group_by(OG_ID) %>%
        mutate(freq=value/sum(value)) %>%
        mutate(freq = replace_na(freq, 0))
    
    collapsed_final_df = final_ancestor_df %>%
      group_by(Dup_status) %>%
      summarise(freq_proportion = mean(freq)) %>%
      mutate(Node = ancestor) %>%
      dplyr::select(Node, Dup_status, freq_proportion)
    
    rand_foreground_df = rbind(rand_foreground_df, collapsed_final_df)
  }
  
  ####### Order:
  rand_foreground_df$Node = factor(rand_foreground_df$Node, levels=c(deuterostoma_y_labels[deuterostoma_y_labels!="Bilateria"], "Bilateria",
                                                               rev(protostoma_y_labels[protostoma_y_labels!="Bilateria"])))
  rand_foreground_df$Dup_status = factor(rand_foreground_df$Dup_status, levels=rev(c("Dup", "NO_Dup")))
  
  ###### Add branch info
  rand_foreground_df$Branch = rep("Deuterostoma", nrow(rand_foreground_df))
  rand_foreground_df$Branch[rand_foreground_df$Node %in% protostoma_y_labels] = "Protostoma"
  rand_foreground_df$Branch[rand_foreground_df$Node=="Bilateria"] = "Bilateria"
  rand_foreground_df = rand_foreground_df %>%
    unite("Branch_Dup_status", c(Branch,Dup_status), remove=FALSE)
  rand_foreground_df$Branch_Dup_status = factor(rand_foreground_df$Branch_Dup_status, levels=rev(c("Deuterostoma_Dup", "Deuterostoma_NO_Dup", "Protostoma_Dup", "Protostoma_NO_Dup", "Bilateria_Dup", "Bilateria_NO_Dup")))
  
  ###### Add randomization round
  rand_foreground_df$Randomization_round = rep(randomization_round, nrow(rand_foreground_df))
  
  ###### join to final df
  all_rand_foregrounds_df = rbind(all_rand_foregrounds_df, as.data.frame(rand_foreground_df))
}

all_rand_foregrounds_dup_df = all_rand_foregrounds_df %>%
  filter(Dup_status=="Dup")

#########################################################
############## PLOT #####################################
#########################################################

ancestral_gains_duplication_status = ggplot() +
  geom_bar(data=subset(foreground_df, Node!="Bilateria"), 
           aes(x=Node, y=freq_proportion, fill=Branch_Dup_status), stat="identity", position="fill", color="black", alpha=0.75) +
  geom_errorbar(data=subset(dup_background_df, Node!="Bilateria"), 
                            aes(x=Node, ymin=freq_proportion, ymax=freq_proportion), color="black", linetype="dashed") +
  ###############
  geom_boxplot(data=all_rand_foregrounds_dup_df, aes(x=Node, y=freq_proportion), color="black", fill="black", size=0.5, alpha=0.4, outlier.size = 0.5) +
  ###############
  scale_fill_manual(values=c("Deuterostoma_Dup"="darkolivegreen4", "Protostoma_Dup"="tan3", "Bilateria_Dup"="firebrick1", "Deuterostoma_NO_Dup"="white", "Protostoma_NO_Dup"="white", "Bilateria_NO_Dup"="white")) +
  ggtitle("Proportion of TS gains orthogroups with duplications") +
  theme_bw() +
  theme(axis.text.y = element_text(color="black"),
        axis.title = element_blank(),
        axis.text.x = element_text(color="black", angle=45, hjust=1, vjust=1),
        axis.ticks.y = element_line(),
        plot.title = element_text(color="black", hjust=0.5, face="bold"))

print(ancestral_gains_duplication_status)
```

```{r duplication_levels_randomized_call_ancestors_v2, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions")}
#Upload orthogroups and get duplication info about each gene
orthogroups_df = read.delim(paste0(ts_gains_losses_raw_dir, "Bilateria_conserved_orthogroups-All_genes-associated_tissues.tab"),
                            header=FALSE, col.names=c("OG_ID", "Species", "GeneID", "Tau", "Tissue", "Paralog_tissue", "Expr_cutoff"), stringsAsFactors = FALSE)
#### Add column with OG_ID_Species and count number of paralogs per gene
orthogroups_df$OG_ID_Species = paste0(orthogroups_df$OG_ID, "_", orthogroups_df$Species)
species_paralogs_count_df = as.data.frame(table(orthogroups_df$OG_ID_Species))
OG_ID_Species_paralogs_counts_dict = hashmap(as.vector(species_paralogs_count_df$Var1), as.vector(species_paralogs_count_df$Freq))
orthogroups_df$Paralogs_count = OG_ID_Species_paralogs_counts_dict$find(orthogroups_df$OG_ID_Species)

#########################################################
############## GENERATE BACKGROUND ######################
#########################################################
#### For each node (group of species) compute the proportion in each orthogroup that is duplicated. Get the difference for non-duplicated.
#Sum all and get proportions.
background_df = vector()
for (ancestor in all_ancestors) {
  my_species = get(ancestor)
  #Subset all orthogroups to only these species
  my_species_orthogroups_df = subset(orthogroups_df, Species %in% my_species)
  #Compute for each orthogroup the proportion of duplicated species (among the ones with conserved genes)
   final_ancestor_df = my_species_orthogroups_df %>%
      dplyr::select(OG_ID, Species, Paralogs_count) %>%
      distinct() %>%
      mutate(Dup_status=case_when(Paralogs_count > 1 ~ "Dup",
                                  Paralogs_count <= 1 ~ "NO_Dup")) %>%
      group_by(OG_ID, Dup_status) %>%
      summarise(total = n()) %>%
      ungroup() %>%
      pivot_wider(id_cols="OG_ID", names_from="Dup_status", values_from="total", values_fill=0) %>%
      pivot_longer(cols=c("Dup", "NO_Dup"), names_to = "Dup_status") %>%
      group_by(OG_ID) %>%
      mutate(freq=value/sum(value)) %>%
      mutate(freq = replace_na(freq, 0))
  
  collapsed_final_df = final_ancestor_df %>%
    group_by(Dup_status) %>%
    summarise(freq_proportion = mean(freq)) %>%
    mutate(Node = ancestor) %>%
    dplyr::select(Node, Dup_status, freq_proportion)
  
  background_df = rbind(background_df, collapsed_final_df)
}
#### Order levels
background_df$Node = factor(background_df$Node, levels=c(deuterostoma_y_labels[deuterostoma_y_labels!="Bilateria"], "Bilateria",
                                                             rev(protostoma_y_labels[protostoma_y_labels!="Bilateria"])))
background_df$Dup_status = factor(background_df$Dup_status, levels=rev(c("Dup", "NO_Dup")))
#### Subset only to duplicates
dup_background_df = subset(background_df, Dup_status=="Dup")

#########################################################
############## GENERATE FOREGROUND ######################
#########################################################

#############################
#### Upload TS gains ########
#############################

tissue_gains_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  tissue_gains_df = rbind(tissue_gains_df, input_df)
}

#############################
#### Upload TS losses #######
#############################
tissue_losses_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_losses.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_losses", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  input_df$OG_ID_loss = paste0(input_df$OG_ID, "_", input_df$TS_losses)
  tissue_losses_df = rbind(tissue_losses_df, input_df)
}

#############################
#### Select relative OGs ####
#############################
foreground_df = vector()
for (ancestor in all_ancestors) {
  #### Select ancestor species
  my_species = get(ancestor)
  my_species_orthogroups_df = vector()
  #### Cycle on each tissue
  for (tissue in ordered_tissues) {
    ancestor_gains_OGs = unique(as.vector(subset(tissue_gains_df, TS_gains==ancestor & Tissue==tissue)$OG_ID))
    #### Subset all orthogroups to only these species in the orthogroups with gains
    my_species_tissue_orthogroups_df = subset(orthogroups_df, Species %in% my_species & OG_ID %in% ancestor_gains_OGs)
    ancestor_losses_OGs = unique(as.vector(subset(tissue_losses_df, Tissue==tissue)$OG_ID_loss))
    #### Filter out the losses
    my_species_tissue_orthogroups_df = my_species_tissue_orthogroups_df %>%
      filter(!(OG_ID_Species %in% ancestor_losses_OGs))
    #### Join to final dataframe
    my_species_orthogroups_df = rbind(my_species_orthogroups_df, my_species_tissue_orthogroups_df)
  }
  #Compute the same statistics as above
    final_ancestor_df = my_species_orthogroups_df %>%
      dplyr::select(OG_ID, Species, Paralogs_count) %>%
      distinct() %>%
      mutate(Dup_status=case_when(Paralogs_count > 1 ~ "Dup",
                                  Paralogs_count <= 1 ~ "NO_Dup")) %>%
      group_by(OG_ID, Dup_status) %>%
      summarise(total = n()) %>%
      ungroup() %>%
      pivot_wider(id_cols="OG_ID", names_from="Dup_status", values_from="total", values_fill=0) %>%
      pivot_longer(cols=c("Dup", "NO_Dup"), names_to = "Dup_status") %>%
      group_by(OG_ID) %>%
      mutate(freq=value/sum(value)) %>%
      mutate(freq = replace_na(freq, 0))
  
  collapsed_final_df = final_ancestor_df %>%
    group_by(Dup_status) %>%
    summarise(freq_proportion = mean(freq)) %>%
    mutate(Node = ancestor) %>%
    dplyr::select(Node, Dup_status, freq_proportion)
  
  foreground_df = rbind(foreground_df, collapsed_final_df)
}

#Select only the TS species (species without losses) and compute the proportion of those species that are duplicated.
#Compute the difference as not-duplicated (it has to sum to 1).

####### Order:
foreground_df$Node = factor(foreground_df$Node, levels=c(deuterostoma_y_labels[deuterostoma_y_labels!="Bilateria"], "Bilateria",
                                                             rev(protostoma_y_labels[protostoma_y_labels!="Bilateria"])))
foreground_df$Dup_status = factor(foreground_df$Dup_status, levels=rev(c("Dup", "NO_Dup")))

###### Add branch info
foreground_df$Branch = rep("Deuterostoma", nrow(foreground_df))
foreground_df$Branch[foreground_df$Node %in% protostoma_y_labels] = "Protostoma"
foreground_df$Branch[foreground_df$Node=="Bilateria"] = "Bilateria"
foreground_df = foreground_df %>%
  unite("Branch_Dup_status", c(Branch,Dup_status), remove=FALSE)
foreground_df$Branch_Dup_status = factor(foreground_df$Branch_Dup_status, levels=rev(c("Deuterostoma_Dup", "Deuterostoma_NO_Dup", "Protostoma_Dup", "Protostoma_NO_Dup", "Bilateria_Dup", "Bilateria_NO_Dup")))


#########################################################
##### GENERATE FOREGROUND RANDOMIZATIONS ################
#########################################################
all_rand_foregrounds_df = vector()

for (randomization_round in seq(1,10)) {
  ######### Set directory
  randomization_dir = paste0(home, "projects/bilaterian_GE/data/ts_gains_losses_revision/randomization_", randomization_round, "/")
  
  ######### Upload TS gains
  tissue_gains_df = vector()
  for (tissue in ordered_tissues) {
    input_df = read.delim(paste0(randomization_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
    input_df$Tissue = rep(tissue, nrow(input_df))
    tissue_gains_df = rbind(tissue_gains_df, input_df)
  }
  ######### Upload TS losses
  tissue_losses_df = vector()
  for (tissue in ordered_tissues) {
    input_df = read.delim(paste0(randomization_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_losses.tab"), header=TRUE,
                          stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_losses", "Category"))
    input_df$Tissue = rep(tissue, nrow(input_df))
    input_df$OG_ID_loss = paste0(input_df$OG_ID, "_", input_df$TS_losses)
    tissue_losses_df = rbind(tissue_losses_df, input_df)
  }
  
  ######### Build foreground
  rand_foreground_df = vector()
  for (ancestor in all_ancestors) {
    #### Select ancestor species
    my_species = get(ancestor)
    my_species_orthogroups_df = vector()
    #### Cycle on each tissue
    for (tissue in ordered_tissues) {
      ancestor_gains_OGs = unique(as.vector(subset(tissue_gains_df, TS_gains==ancestor & Tissue==tissue)$OG_ID))
      #### Subset all orthogroups to only these species in the orthogroups with gains
      my_species_tissue_orthogroups_df = subset(orthogroups_df, Species %in% my_species & OG_ID %in% ancestor_gains_OGs)
      ancestor_losses_OGs = unique(as.vector(subset(tissue_losses_df, Tissue==tissue)$OG_ID_loss))
      #### Filter out the losses
      my_species_tissue_orthogroups_df = my_species_tissue_orthogroups_df %>%
        filter(!(OG_ID_Species %in% ancestor_losses_OGs))
      #### Join to final dataframe
      my_species_orthogroups_df = rbind(my_species_orthogroups_df, my_species_tissue_orthogroups_df)
    }
    #Compute the same statistics as above
      final_ancestor_df = my_species_orthogroups_df %>%
        dplyr::select(OG_ID, Species, Paralogs_count) %>%
        distinct() %>%
        mutate(Dup_status=case_when(Paralogs_count > 1 ~ "Dup",
                                    Paralogs_count <= 1 ~ "NO_Dup")) %>%
        group_by(OG_ID, Dup_status) %>%
        summarise(total = n()) %>%
        ungroup() %>%
        pivot_wider(id_cols="OG_ID", names_from="Dup_status", values_from="total", values_fill=0) %>%
        pivot_longer(cols=c("Dup", "NO_Dup"), names_to = "Dup_status") %>%
        group_by(OG_ID) %>%
        mutate(freq=value/sum(value)) %>%
        mutate(freq = replace_na(freq, 0))
    
    collapsed_final_df = final_ancestor_df %>%
      group_by(Dup_status) %>%
      summarise(freq_proportion = mean(freq)) %>%
      mutate(Node = ancestor) %>%
      dplyr::select(Node, Dup_status, freq_proportion)
    
    rand_foreground_df = rbind(rand_foreground_df, collapsed_final_df)
  }
  
  ####### Order:
  rand_foreground_df$Node = factor(rand_foreground_df$Node, levels=c(deuterostoma_y_labels[deuterostoma_y_labels!="Bilateria"], "Bilateria",
                                                               rev(protostoma_y_labels[protostoma_y_labels!="Bilateria"])))
  rand_foreground_df$Dup_status = factor(rand_foreground_df$Dup_status, levels=rev(c("Dup", "NO_Dup")))
  
  ###### Add branch info
  rand_foreground_df$Branch = rep("Deuterostoma", nrow(rand_foreground_df))
  rand_foreground_df$Branch[rand_foreground_df$Node %in% protostoma_y_labels] = "Protostoma"
  rand_foreground_df$Branch[rand_foreground_df$Node=="Bilateria"] = "Bilateria"
  rand_foreground_df = rand_foreground_df %>%
    unite("Branch_Dup_status", c(Branch,Dup_status), remove=FALSE)
  rand_foreground_df$Branch_Dup_status = factor(rand_foreground_df$Branch_Dup_status, levels=rev(c("Deuterostoma_Dup", "Deuterostoma_NO_Dup", "Protostoma_Dup", "Protostoma_NO_Dup", "Bilateria_Dup", "Bilateria_NO_Dup")))
  
  ###### Add randomization round
  rand_foreground_df$Randomization_round = rep(randomization_round, nrow(rand_foreground_df))
  
  ###### join to final df
  all_rand_foregrounds_df = rbind(all_rand_foregrounds_df, as.data.frame(rand_foreground_df))
}

all_rand_foregrounds_dup_df = all_rand_foregrounds_df %>%
  filter(Dup_status=="Dup")

#########################################################
############## PLOT #####################################
#########################################################

ancestral_gains_duplication_status = ggplot() +
  geom_bar(data=subset(foreground_df, Node!="Bilateria"), 
           aes(x=Node, y=freq_proportion, fill=Branch_Dup_status), stat="identity", position="fill", color="black", alpha=0.75) +
  geom_errorbar(data=subset(dup_background_df, Node!="Bilateria"), 
                            aes(x=Node, ymin=freq_proportion, ymax=freq_proportion), color="black", linetype="dashed") +
  ###############
  geom_boxplot(data=all_rand_foregrounds_dup_df, aes(x=Node, y=freq_proportion), color="black", fill="black", size=0.5, alpha=0.4, outlier.size = 0.5) +
  ###############
  scale_fill_manual(values=c("Deuterostoma_Dup"="darkolivegreen4", "Protostoma_Dup"="tan3", "Bilateria_Dup"="firebrick1", "Deuterostoma_NO_Dup"="white", "Protostoma_NO_Dup"="white", "Bilateria_NO_Dup"="white")) +
  ggtitle("Proportion of TS gains orthogroups with duplications") +
  theme_bw() +
  theme(axis.text.y = element_text(color="black"),
        axis.title = element_blank(),
        axis.text.x = element_text(color="black", angle=45, hjust=1, vjust=1),
        axis.ticks.y = element_line(),
        plot.title = element_text(color="black", hjust=0.5, face="bold"))

print(ancestral_gains_duplication_status)
```

```{r plot_correlation_ancestor_number_of_paralogs_v1, echo=FALSE, warning=FALSE, messsage=FALSE, dependson=c("setup", "functions")}
#############################
#### Upload orthogroups #####
#############################
#Upload orthogroups and get duplication info about each gene
orthogroups_df = read.delim(paste0(ts_gains_losses_raw_dir, "Bilateria_conserved_orthogroups-All_genes-associated_tissues.tab"),
                            header=FALSE, col.names=c("OG_ID", "Species", "GeneID", "Tau", "Tissue", "Paralog_tissue", "Expr_cutoff"), stringsAsFactors = FALSE)
#### Count number of genes per orthogroup
orthogroups_df = orthogroups_df %>%
  group_by(OG_ID) %>%
  summarize(Tot_genes = n()) %>%
  ungroup()

#############################
#### Upload TS gains ########
#############################

tissue_gains_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, "Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
  #### Add Tissue
  input_df$Tissue = rep(tissue, nrow(input_df))
  #### Add Node level
  input_df$Node_level = ancestor_node_level_dict$find(input_df$TS_gains)
  tissue_gains_df = rbind(tissue_gains_df, input_df)
}

#############################
#### Combine dataframes #####
#############################
#Get a dataframe with TS gain and number of genes in 
tissue_gains_gene_count_df_1 = tissue_gains_df %>%
  left_join(orthogroups_df, by="OG_ID") %>%
  mutate(Node_level=factor(Node_level)) %>%
  mutate(Branch = case_when(TS_gains %in% c(Deuterostoma, deuterostoma_y_labels[1:(length(deuterostoma_y_labels)-1)]) ~ "Deuterostoma", 
                            TS_gains %in% c(Protostoma, protostoma_y_labels[1:(length(protostoma_y_labels)-1)]) ~ "Protostoma",
                            TRUE ~ "Bilateria"))
##Duplicated the Bilateria
tissue_gains_gene_count_df = tissue_gains_gene_count_df_1 %>%
  filter(Branch=="Bilateria") %>%
  mutate(Branch="Protostoma") %>%
  dplyr::bind_rows(tissue_gains_gene_count_df_1 %>%
                     mutate(Branch = case_when(Branch=="Bilateria" ~ "Deuterostoma",
                                               TRUE ~ Branch)))

#############################
######## PLOT ###############
#############################

tissue_gains_gene_count_plot = ggplot(data=tissue_gains_gene_count_df, aes(x=Node_level, y=Tot_genes)) +
  geom_jitter(color="dodgerblue2", alpha=0.5) +
  geom_boxplot(alpha=0.6) +
  geom_smooth(formula=y~x, method="lm", color="black") +
  theme_bw() +
  ggtitle("Observed")

print(tissue_gains_gene_count_plot)

########################################
#### Upload RANDOMIZED TS gains ########
########################################

tissue_gains_df = vector()
for (randomization_round in seq(1,10)) {
  randomization_dir = paste0(home, "projects/bilaterian_GE/data/ts_gains_losses_revision/randomization_", randomization_round, "/")
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(randomization_dir, "inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
  #### Add Tissue
  input_df$Tissue = rep(tissue, nrow(input_df))
  #### Add Node level
  input_df$Node_level = ancestor_node_level_dict$find(input_df$TS_gains)
  #### Add randomization round
  input_df$Randomization_round = rep(randomization_round, nrow(input_df))
  tissue_gains_df = rbind(tissue_gains_df, input_df)
 }
}

#############################
#### Combine dataframes #####
#############################
#Get a dataframe with TS gain and number of genes in 
tissue_gains_gene_count_df_1 = tissue_gains_df %>%
  left_join(orthogroups_df, by="OG_ID") %>%
  mutate(Node_level=factor(Node_level)) %>%
  mutate(Branch = case_when(TS_gains %in% c(Deuterostoma, deuterostoma_y_labels[1:(length(deuterostoma_y_labels)-1)]) ~ "Deuterostoma", 
                            TS_gains %in% c(Protostoma, protostoma_y_labels[1:(length(protostoma_y_labels)-1)]) ~ "Protostoma",
                            TRUE ~ "Bilateria"))
##Duplicated the Bilateria
tissue_gains_gene_count_df = tissue_gains_gene_count_df_1 %>%
  filter(Branch=="Bilateria") %>%
  mutate(Branch="Protostoma") %>%
  dplyr::bind_rows(tissue_gains_gene_count_df_1 %>%
                     mutate(Branch = case_when(Branch=="Bilateria" ~ "Deuterostoma",
                                               TRUE ~ Branch)))

#############################
######## PLOT ###############
#############################

tissue_gains_gene_count_plot = ggplot(data=tissue_gains_gene_count_df, aes(x=Node_level, y=Tot_genes)) +
  geom_jitter(color="dodgerblue2", alpha=0.4, size=1) +
  geom_boxplot(alpha=0.6) +
  geom_smooth(formula=y~x, method="lm", color="black") +
  theme_bw() +
  facet_wrap(~Branch) +
  ggtitle("Randomization")

print(tissue_gains_gene_count_plot)
```
