---
title: "Fig. 5: Tissue-specific gains are associated with the emergence of unique phenotypes."
output: html_document
---

```{r setup, include=FALSE}
#set Knitr option
knitr::opts_chunk$set(echo = TRUE)
knitr::knit_engines$set(python = reticulate::eng_python)

#upload libraries
library(ggplot2)
library(mixOmics)
library(hashmap)
library(reshape2)
library(pheatmap)
library(RColorBrewer)
library(cowplot)
library(tidyverse)
library(viridis)
library(gridExtra)
library(egg)
library(plotly)
library(ggrepel)
library(pathview)
library(gage)
library(ggtree)
library(ape)
library(purrr)
library(ggh4x)
library(ggpubr)
library(forcats)
library(ggridges)
library(ggalluvial)
library(data.table)
library(fgsea)

#set directories
home="/Users/federica/mnt/"
ts_call_dir=paste0(home, "projects/bilaterian_GE/data/ts_call/")
splsda_dir = paste0(home, "projects/bilaterian_GE/data/splsda_analysis/")
tau_dir = paste0(home, "projects/bilaterian_GE/data/ts_call/species_QN_5_TPM_taus/")
original_tau_dir = paste0(home, "projects/bilaterian_GE/data/ts_call/taus/")
gene_sets_dir=paste0(home, "projects/bilaterian_GE/data/gene_sets/")
randomization_dir=paste0(home, "projects/bilaterian_GE/data/ts_call/QN_randomizations/")
evo_distances_dir = paste0(home, "projects/bilaterian_GE/data/evo_distances/")
ts_gains_losses_dir = paste0(home, "projects/bilaterian_GE/data/ts_gains_losses/species_QN_5_TPM_v6/")
GO_transfers_dir = paste0(home, "projects/bilaterian_GE/data/GO_transfers/")
intergenic_dist_dir = paste0(home, "projects/bilaterian_GE/data/ts_intergenic_regions/")
# #tissue_expr_dir=paste0(home, "projects/bilaterian_GE/data/preprocessing/average_expr_by_tissue/")
# pca_dir=paste0(home, "projects/bilaterian_GE/data/pca_analysis_all_norms/")
# metadata_dir=paste0(home, "projects/bilaterian_GE/data/samples_metadata/")
# splsda_dir= paste0(home, "projects/bilaterian_GE/data/splsda_analysis/")
# diff_expr_dir = paste0(home, "projects/bilaterian_GE/data/differential_expression/")
# general_patterns_dir = paste0(home, "projects/bilaterian_GE/data/vertebrates_vs_insects_general_patterns/")
# branches_length_dir = paste0(home, "projects/bilaterian_GE/data/branches_length_analysis/")

#set variables
my_version = "All_version2"
my_category = "STRICT"
my_evo_type = "conserved"
clades = c("Vertebrata", "Insecta", "Bilateria")
clades_PCA = c("Vertebrata", "Insecta", "Bilateria")
options("scipen"=100, "digits"=4)

all_species = c("Hs2", "Mm2", "Bt2", "Mdo", "Gga", "Xtr", "Dre", "Cmi", "Bla", "Sp2", "Dme", "Eba", "Aae", "Bmo", "Tca", "Ame", "Bge", "Cdi", "Sma", "Obi")
ordered_species_tree = c("Hs2", "Mm2", "Bt2", "Mdo", "Gga", "Xtr", "Dre", "Cmi", "Bla", "Sp2", rev(c("Dme", "Eba", "Aae", "Bmo", "Tca", "Ame", "Bge", "Cdi", "Sma", "Obi")))
ordered_species = all_species
Bilateria = all_species
Vertebrata = c("Hs2", "Mm2", "Bt2", "Mdo", "Gga", "Xtr", "Dre", "Cmi")
Insecta = c("Dme", "Eba", "Aae", "BmA", "Tca", "Ame", "Bge", "Cdi")
Outgroups = c("Bla", "Sp2", "Sma", "Obi")
all_tissues=c("Adipose", "DigestiveTract", "Kidney", "Epithelial", "Muscle", "Neural", "Ovary", "Testis")
ordered_tissues = c("Neural", "Testis", "Ovary", "Muscle", "Kidney", "Epithelial", "DigestiveTract", "Adipose")

#set aestetics vectors
my_color_vector = c("Neural"="mediumblue", "Testis"="violet", "Ovary"="darkorchid", "Muscle"="springgreen3", 
                    "Kidney"="gold", "Epithelial"="steelblue1", "DigestiveTract"="chocolate1", "Adipose"="firebrick1")
my_shapes_vector = c("Hs2"=0, "Mm2"=1, "Bt2"=2, "Mdo"=3, "Gga"=4, "Xtr"=5, "Dre"=6, "Cmi"=7, "Bla"=9, "Sp2"=8, 
                     "Dme"=15, "Eba"=18, "Aae"= 23, "BmA"=25, "Tca"=17, "Bge"=10, "Ame"=16, "Cdi"=13, "Sma"=14, "Obi"=11)
my_alpha_vector=c("Hs2"=1, "Mm2"=1, "Bt2"=1, "Mdo"=1, "Gga"=1, "Xtr"=1, "Dre"=1, "Cmi"=1, "Bla"=0.5, "Sp2"=0.5, "Sma"=0.5, "Ame"=0.5, "Dme"=0.5, "Cdi"=0.5, "Obi"=0.5, "Bge"=0.5, "BmA"=0.5, "Tca"=0.5)
my_sets_color = c("Vertebrata"= "mediumorchid", "Deuterostoma"="olivedrab3", "Insecta"="tan3", "Protostoma"="yellow", "Bilateria"="firebrick1", "Outgroups"="dimgray")
my_species_colors=c("Hs2"="#00BFFF", "Mm2"="#159DFF", "Bt2"="#1873CD", "Mdo"="#174B8B", "Gga"="#3F3F8B", "Xtr"="#64398B", "Dre"="#82359D", "Cmi"="#9932CC", "Bla"="#2E8B57", "Sp2"="#96B12B", "Obi"="#FFFF00", "Sma"="#FFCC33", "Cdi"="goldenrod", "Bge"="burlywood2", "Ame"="#FF9966", "Tca"="#FF6600", "Bmo"="#CC6600", "Aae"="#993300", "Eba"="firebrick3", "Dme"="red")
my_copy_status_colors = c("single_copy_ortholog"="gray18", "duplicated"="gray76")
my_branch_colors = c("Deuterostoma"="olivedrab3", "Protostoma"="tan3", "Bilateria"="firebrick1")
my_ancestral_colors = c("Ancestral"="dimgray", "Species_specific"="coral3")

#Create named vector with PC and corresponding tissue
PC_tissue_named_vector = c("1"="Neural", "2"="Testis", "3"="Ovary", 
                          "4"="Muscle", "5"="Kidney", "6"="Epithelial", 
                          "7"="DigestiveTract", "8"="Adipose")
paletteLength = 75
palette_brewer = colorRampPalette(rev(brewer.pal(11, "RdBu")))(paletteLength)

multi_tissues_vector = c(ordered_tissues, "MultiTS")
multi_tissues_color_vector = c(my_color_vector, "MultiTS"="black")


####### Variables neeeded to plot the tree
Bilateria = all_species
Euarchontoglires = c("Hs2", "Mm2")
Eutheria = c(Euarchontoglires, "Bt2")
Mammalia = c(Eutheria, "Mdo")
Amniota = c(Mammalia, "Gga")
Tetrapoda = c(Amniota, "Xtr")
Euteleostomi = c(Tetrapoda, "Dre")
Vertebrata = c(Euteleostomi, "Cmi")
Chordata = c(Vertebrata, "Bla")
Deuterostoma = c(Chordata, "Sp2")
Cyclorrapha = c("Dme", "Eba")
Diptera = c(Cyclorrapha, "Aae")
Panorpidae = c(Diptera, "Bmo")
Oligoneoptera = c(Panorpidae, "Tca")
Holometabola = c(Oligoneoptera, "Ame")
Neoptera = c(Holometabola, "Bge")
Insecta = c(Neoptera, "Cdi")
Arthropoda = c(Insecta, "Sma")
Protostoma = c(Arthropoda, "Obi")

all_ancestors = c("Euarchontoglires", "Eutheria", "Mammalia", "Amniota", "Tetrapoda", "Euteleostomi", "Vertebrata", "Chordata", "Deuterostoma",
              "Bilateria", "Cyclorrapha", "Diptera", "Panorpidae", "Oligoneoptera", "Holometabola", "Neoptera", "Insecta", "Arthropoda", "Protostoma")

all_ancestors_colors = c("Euarchontoglires"="#159DFF",
                         "Eutheria" = "#1873CD",
                         "Mammalia" = "#174B8B",
                         "Amniota" = "#3F3F8B",  
                         "Tetrapoda" =  "#64398B",
                         "Euteleostomi" = "#82359D",
                         "Vertebrata" = "#9932CC",
                         "Chordata" = "#2E8B57",
                         "Deuterostoma" = "#96B12B",
                         "Cyclorrapha" = "firebrick3",
                         "Diptera" = "#993300",
                         "Panorpidae" = "#CC6600",
                         "Oligoneoptera" = "#FF6600", 
                         "Holometabola" = "#FF9966",
                         "Neoptera" = "burlywood2",
                         "Insecta" = "goldenrod",
                         "Arthropoda" = "#FFCC33",
                         "Protostoma" = "#FFFF00",
                         "species\nspecific" = "gray71")

clade_species = c("Euarchontoglires" = list(Euarchontoglires),
                  "Eutheria" = list(Eutheria),
                  "Mammalia" = list(Mammalia),
                  "Amniota" = list(Amniota),
                  "Tetrapoda" = list(Tetrapoda),
                  "Euteleostomi" = list(Euteleostomi),
                  "Vertebrata" = list(Vertebrata),
                  "Chordata" = list(Chordata),
                  "Deuterostoma" = list(Deuterostoma),
                  "Cyclorrapha" = list(Cyclorrapha),
                  "Diptera" = list(Diptera),
                  "Panorpidae" = list(Panorpidae),
                  "Oligoneoptera" = list(Oligoneoptera),
                  "Holometabola" = list(Holometabola),
                  "Neoptera" = list(Neoptera),
                  "Insecta" = list(Insecta),
                  "Arthropoda" = list(Arthropoda),
                  "Protostoma" = list(Protostoma),
                  "Bilateria" = list(Bilateria),
                  "Hs2" = "Hs2", "Mm2" = "Mm2", "Bt2" = "Bt2", "Mdo" = "Mdo", "Gga" = "Gga", "Xtr" = "Xtr", "Dre" = "Dre", "Cmi" = "Cmi",
                  "Bla" = "Bla", "Sp2" = "Sp2", "Sma" = "Sma", "Obi" = "Obi",
                  "Dme" = "Dme", "Eba" = "Eba", "Aae" = "Aae", "Bmo" = "Bmo", "Tca" = "Tca", "Ame" = "Ame", "Bge" = "Bge", "Cdi" = "Cdi")


node_x_position = c("Hs2"=1, "Euarchontoglires"=1.5, "Mm2"=2, "Eutheria"=2, "Bt2"=3, "Mammalia"=2.5,
                    "Mdo"=4, "Amniota"=3, "Gga"=5, "Tetrapoda"=3.5, "Xtr"=6, "Euteleostomi"=4,
                    "Dre"=7, "Vertebrata"=4.5, "Cmi"=8, "Chordata"=5, "Bla"=9, "Deuterostoma"=5.5, "Sp2"=10,
                    "Bilateria" = 11,
                    "Dme"=21, "Cyclorrapha"=20.5, "Eba"=20, "Diptera"=20, "Aae"=19, "Panorpidae"=19.5,
                    "Bmo"=18, "Oligoneoptera"=19, "Tca"=17, "Holometabola"=18.5, "Ame"=16, "Neoptera"=18,
                    "Bge"=15, "Insecta"=17.5, "Cdi"=14, "Arthropoda"=17, "Sma"=13, "Protostoma"=16.5, "Obi"=12)

node_y_position = c("Hs2"=1, "Euarchontoglires"=2, "Mm2"=1, "Eutheria"=3, "Bt2"=1, "Mammalia"=4,
                    "Mdo"=1, "Amniota"=5, "Gga"=1, "Tetrapoda"=6, "Xtr"=1, "Euteleostomi"=7,
                    "Dre"=1, "Vertebrata"=8, "Cmi"=1, "Chordata"=9, "Bla"=1, "Deuterostoma"=10, "Sp2"=1,
                    "Bilateria" = 11,
                    "Dme"=1, "Cyclorrapha"=2, "Eba"=1, "Diptera"=3, "Aae"=1, "Panorpidae"=4,
                    "Bmo"=1, "Oligoneoptera"=5, "Tca"=1, "Holometabola"=6, "Ame"=1, "Neoptera"=7,
                    "Bge"=1, "Insecta"=8, "Cdi"=1, "Arthropoda"=9, "Sma"=1, "Protostoma"=10, "Obi"=1)

node_x_labels = c("1.0"="Hs2", "1.5"="", "2.0"="Mm2", "2.5"="", "3.0"="Bt2", "3.5"="", "4.0"="Mdo", "4.5"="", "5.0"="Gga", "5.5"="",
                  "6.0"="Xtr", "7.0"="Dre", "8.0"="Cmi", "9.0"="Bla", "10.0"="Sp2",
                   "11.0"="",
                   "12.0"="Obi", "13.0"="Sma", "14.0"="Cdi", "15.0"="Bge", "16.0"="Ame", "16.5"="", "17.0"="Tca", "17.5"="", "18.0"="Bmo",
                  "18.5"="", "19.0"="Aae", "19.5"="", "20.0"="Eba", "20.5"="", "21.0"="Dme")

multi_tissues_vector = c(ordered_tissues, "MultiTS")
multi_tissues_color_vector = c(my_color_vector, "MultiTS"="black")

deuterostoma_y_labels = c("Euarchontoglires", "Eutheria", "Mammalia", "Amniota", "Tetrapoda", "Euteleostomi", "Vertebrata", "Chordata", "Deuterostoma", "Bilateria")
protostoma_y_labels = c("Cyclorrapha", "Diptera", "Panorpidae", "Oligoneoptera", "Holometabola", "Neoptera", "Insecta", "Arthropoda", "Protostoma", "Bilateria")


####### Get sub-ancestors
sub_Euarchontoglires = c("Hs2","Mm2")
sub_Eutheria = c("Euarchontoglires")
sub_Mammalia = c(sub_Eutheria, "Eutheria")
sub_Amniota = c(sub_Mammalia, "Mammalia")
sub_Tetrapoda = c(sub_Amniota, "Amniota")
sub_Euteleostomi = c(sub_Tetrapoda, "Tetrapoda")
sub_Vertebrata = c(sub_Euteleostomi, "Euteleostomi")
sub_Chordata = c(sub_Vertebrata, "Vertebrata")
sub_Deuterostoma = c(sub_Chordata, "Chordata")
sub_Cyclorrapha = c("Dme","Eba")
sub_Diptera = c(sub_Cyclorrapha, "Cyclorrapha")
sub_Panorpidae = c(sub_Diptera, "Diptera")
sub_Oligoneoptera = c(sub_Panorpidae, "Panorpidae")
sub_Holometabola = c(sub_Oligoneoptera, "Oligoneoptera")
sub_Neoptera = c(sub_Holometabola, "Holometabola")
sub_Insecta = c(sub_Neoptera, "Neoptera")
sub_Arthropoda = c(sub_Insecta, "Insecta")
sub_Protostoma = c(sub_Arthropoda, "Arthropoda")
sub_Bilateria = c(sub_Deuterostoma, sub_Protostoma)
```

```{r functions, warning=FALSE, message=FALSE, echo=FALSE, dependson="setup"}
number_ticks <- function(n) {function(limits) pretty(limits, n)}

####################################
####################################

firstup <- function(x) {
  substr(x, 1, 1) <- toupper(substr(x, 1, 1))
  x
}

####################################
####################################

get_density <- function(x, y, ...) {
  dens <- MASS::kde2d(x, y, ...)
  ix <- findInterval(x, dens$x)
  iy <- findInterval(y, dens$y)
  ii <- cbind(ix, iy)
  return(dens$z[ii])
}

####################################
####################################

plot_ts_genes_stats = function(annotation_row_df, ordered_tissues) {
  my_extra_color_vector = c("Adipose"="firebrick1", "Neural"="mediumblue", "Muscle"="springgreen3", "Ovary"="darkorchid", "Kidney"="gold", "Testis"="violet", "DigestiveTract"="chocolate1", "Epithelial"="steelblue1", "no_tissue"="white")
  ordered_tissue_combs = rev(names(table(annotation_row_df$Tissue)[order(table(annotation_row_df$Tissue))]))
  #Modify input
  annotation_row_df$Tissue = factor(annotation_row_df$Tissue, levels=ordered_tissue_combs)
  TS_genes_number_plot = ggplot(data=annotation_row_df) +
    geom_bar(aes(x=Tissue), stat="count", fill="darkgray", color="darkgray", alpha=0.8) +
    theme_bw() +
    geom_text(stat="count", aes(x=Tissue, label=..count..), vjust=-1, size=3) + #add count on top of barplot
    ylab("Number of TS genes") +
    ggtitle("Number of TS genes by tissue") +
    theme(axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        axis.ticks = element_blank(),
        title = element_text(size=12, color="black"),
        panel.background = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.grid.major = element_blank()) +
    scale_y_continuous(breaks=number_ticks(15), 
                       limits = c(0, max(unname(table(annotation_row_df$Tissue)))+60))
  
  #Build input for UpSet-like kind of plot
  shared_tissues_df = data.frame(tissue_comb=unique(as.vector(annotation_row_df$Tissue)))
  for (i in seq(length(ordered_tissues))) {
    shared_tissues_df$tissue = as.vector(shared_tissues_df$tissue_comb)
    shared_tissues_df$tissue[grep(ordered_tissues[i], shared_tissues_df$tissue_comb)] = ordered_tissues[i]
    shared_tissues_df$tissue[grep(ordered_tissues[i], shared_tissues_df$tissue_comb, invert=TRUE)] = "no_tissue"
    colnames(shared_tissues_df)[ncol(shared_tissues_df)] = ordered_tissues[i]
  }
  
  shared_tissues_long_df = melt(shared_tissues_df, id.vars = "tissue_comb")
  shared_tissues_long_df$tissue_comb = factor(shared_tissues_long_df$tissue_comb, levels=ordered_tissue_combs)
  colnames(shared_tissues_long_df) = c("tissue_comb", "tissue", "status")
  shared_tissues_long_df$tissue = factor(shared_tissues_long_df$tissue, levels=rev(ordered_tissues))
  shared_tissues_long_df = subset(shared_tissues_long_df, status != "no_tissue")
  
  shared_tissues_plot = ggplot(data=shared_tissues_long_df, aes(x=tissue_comb, y=tissue, group=tissue_comb)) +
    geom_line() +
    geom_point(aes(x=tissue_comb, y=tissue, fill=status), color="black", shape=21, size=3) +
    scale_fill_manual(values=my_extra_color_vector) +
    theme_bw() +
    theme(axis.title.x = element_blank(),
        axis.ticks = element_blank(),
        axis.text.y = element_text(color="black", size=10),
        axis.text.x=element_blank(),
        panel.background = element_blank(), 
        panel.grid.minor = element_blank(), 
        panel.grid.major = element_blank(),
        panel.border = element_blank()) +
    ylab("Tissues") +
    guides(fill=FALSE)
  
  final_plot = plot_grid(TS_genes_number_plot, shared_tissues_plot, align="v", rel_heights = c(1,0.6), ncol=1)
  return(final_plot)
}

####################################
get_density <- function(x, y, ...) {
  dens <- MASS::kde2d(x, y, ...)
  ix <- findInterval(x, dens$x)
  iy <- findInterval(y, dens$y)
  ii <- cbind(ix, iy)
  return(dens$z[ii])
}


####################################
see_pathview <- function(..., save_image = FALSE)
{
  msg <- capture.output(pathview::pathview(...), type = "message")
  msg <- grep("image file", msg, value = T)
  filename <- sapply(strsplit(msg, " "), function(x) x[length(x)])
  img <- png::readPNG(filename)
  grid::grid.raster(img)
  if(!save_image) invisible(file.remove(filename))
}

#######################################
########## Tree function ##############
#######################################
tree_y <-  function(ggtree, data){
  if(!inherits(ggtree, "ggtree"))
    stop("not a ggtree object")
  left_join(dplyr::select(data, label), dplyr::select(ggtree$data, label, y)) %>%
    pull(y)
}

#This is just to see if I can actually paier a ggtree with a ggplot decently.
# overwrite the default expand for continuous scales
scale_y_tree <- function(expand=expand_scale(0, 0.6), ...){
    scale_y_continuous(expand=expand, ...)
}

# get the range of the ggtree y-axis data
tree_ylim <- function(ggtree){
  if(!inherits(ggtree, "ggtree"))
    stop("not a ggtree object")
  range(ggtree$data$y)
}

# plot data next to a ggtree aligned by shared labels
ggtreeplot <- function(ggtree, data = NULL, mapping = aes(), flip=FALSE,
     expand_limits=expand_scale(0,.6), ...){
  
  if(!inherits(ggtree, "ggtree"))
    stop("not a ggtree object")

  # match the tree limits
  limits <- tree_ylim(ggtree)
  limits[1] <- limits[1] + (limits[1] * expand_limits[1]) - expand_limits[2]
  limits[2] <- limits[2] + (limits[2] * expand_limits[3]) + expand_limits[4]
  
  if(flip){
    mapping <- modifyList(aes_(x=~x), mapping)
    data <- mutate(data, x=tree_y(ggtree, data))
    gg <- ggplot(data=data, mapping = mapping, ...) +
      scale_x_continuous(limits=limits, expand=c(0,0))
  }else{
    mapping <- modifyList(aes_(y=~y), mapping)
    data <- mutate(data, y=tree_y(ggtree, data))
    gg <- ggplot(data=data, mapping = mapping, ...) +
      scale_y_continuous(limits=limits, expand=c(0,0))
  }
  gg
}

#######################################
########## Randomizations #############
#######################################

get_random_labels_stats = function(ts_long_df) {
  final_mat = vector()
  current_species = vector()
  for (index in seq(1, length(ordered_species))) {
    new_species = ordered_species[index]
    current_species = c(current_species, new_species)
    current_OGs_df = subset(ts_long_df, Species %in% current_species) #Select the orthogroups containing at least one of the species
    current_OGs_num = length(unique(as.vector(current_OGs_df$OG_ID)))
    TS_OGs_df = subset(current_OGs_df, TS_labels == "TS")
    TS_OGs_num = length(unique(as.vector(TS_OGs_df$OG_ID)))
    TS_OGs_proportion = TS_OGs_num/current_OGs_num
    final_mat = rbind(final_mat, c(length(current_species), TS_OGs_num, TS_OGs_proportion, new_species))
  }
  colnames(final_mat) = c("Species_number", "TS_OG_number", "TS_OG_proportion", "new_species")
  final_df = as.data.frame(final_mat)
  final_df$TS_OG_number = as.numeric(as.vector(final_df$TS_OG_number))
  final_df$TS_OG_proportion = as.numeric(as.vector(final_df$TS_OG_proportion))
  final_df$Species_number = as.numeric(as.vector(final_df$Species_number))
  final_df$new_species = as.character(final_df$new_species)
  return(final_df)
}

#######################################
######## TS gains and losses ##########
#######################################

plot_tree_structure = function(node_x_position, node_y_position, node_x_labels) {
  ancestors = c("Euarchontoglires", "Eutheria", "Mammalia", "Amniota", "Tetrapoda", "Euteleostomi", "Vertebrata", "Chordata", "Deuterostoma",
                "Bilateria", "Hs2", "Dme",
                "Cyclorrapha", "Diptera", "Panorpidae", "Oligoneoptera", "Holometabola", "Neoptera", "Insecta", "Arthropoda", "Protostoma")
  deuterostoma_side = c("Euarchontoglires", "Eutheria", "Mammalia", "Amniota", "Tetrapoda", "Euteleostomi", "Vertebrata", "Chordata", "Deuterostoma")
  protostoma_side = c("Cyclorrapha", "Diptera", "Panorpidae", "Oligoneoptera", "Holometabola", "Neoptera", "Insecta", "Arthropoda", "Protostoma")
  
  node_y_values = unique(as.numeric(unname(node_y_position))[as.numeric(unname(node_y_position)) != 1])
  deuterostoma_y_labels = c("Euarchontoglires", "Eutheria", "Mammalia", "Amniota", "Tetrapoda", "Euteleostomi", "Vertebrata", "Chordata", "Deuterostoma", "")
  protostoma_y_labels = c("Cyclorrapha", "Diptera", "Panorpidae", "Oligoneoptera", "Holometabola", "Neoptera", "Insecta", "Arthropoda", "Protostoma", "")
  
  node_x_position_dict = hashmap(names(node_x_position), unname(node_x_position))
  node_y_position_dict = hashmap(names(node_y_position), unname(node_y_position))
  
  tree_input_df = data.frame(Node=c(all_species, all_ancestors))
  #Add coordinates for each node
  tree_input_df$Node_x_position = node_x_position_dict$find(tree_input_df$Node)
  tree_input_df$Node_y_position = node_y_position_dict$find(tree_input_df$Node)
  #Add groups so that lines between all ancestors and all pairs will be drawn.
  tree_input_df$ancestor_state = rep("none", nrow(tree_input_df))
  tree_input_df$ancestor_state[tree_input_df$Node %in% ancestors] = "ancestor"
  #Add side of the tree (Deuterostoma or Protostoma)
  
  #This is long and manual
  tree_input_df$node_state = rep("none", nrow(tree_input_df))
  tree_input_df$node_state[tree_input_df$Node %in% c("Mm2", "Euarchontoglires")] = "Euarchontoglires"
  tree_input_df$node_state[tree_input_df$Node %in% c("Bt2", "Eutheria")] = "Eutheria"
  tree_input_df$node_state[tree_input_df$Node %in% c("Mdo", "Mammalia")] = "Mammalia"
  tree_input_df$node_state[tree_input_df$Node %in% c("Gga", "Amniota")] = "Amniota"
  tree_input_df$node_state[tree_input_df$Node %in% c("Xtr", "Tetrapoda")] = "Tetrapoda"
  tree_input_df$node_state[tree_input_df$Node %in% c("Dre", "Euteleostomi")] = "Euteleostomi"
  tree_input_df$node_state[tree_input_df$Node %in% c("Cmi", "Vertebrata")] = "Vertebrata"
  tree_input_df$node_state[tree_input_df$Node %in% c("Bla", "Chordata")] = "Chordata"
  tree_input_df$node_state[tree_input_df$Node %in% c("Sp2", "Deuterostoma")] = "Deuterostoma"
  
  tree_input_df$node_state[tree_input_df$Node %in% c("Eba", "Cyclorrapha")] = "Cyclorrapha"
  tree_input_df$node_state[tree_input_df$Node %in% c("Aae", "Diptera")] = "Diptera"
  tree_input_df$node_state[tree_input_df$Node %in% c("Bmo", "Panorpidae")] = "Panorpidae"
  tree_input_df$node_state[tree_input_df$Node %in% c("Tca", "Oligoneoptera")] = "Oligoneoptera"
  tree_input_df$node_state[tree_input_df$Node %in% c("Ame", "Holometabola")] = "Holometabola"
  tree_input_df$node_state[tree_input_df$Node %in% c("Bge", "Neoptera")] = "Neoptera"
  tree_input_df$node_state[tree_input_df$Node %in% c("Cdi", "Insecta")] = "Insecta"
  tree_input_df$node_state[tree_input_df$Node %in% c("Sma", "Arthropoda")] = "Arthropoda"
  tree_input_df$node_state[tree_input_df$Node %in% c("Obi", "Protostoma")] = "Protostoma"
  
  ts_gains_plot = ggplot(data=tree_input_df, aes(x=Node_x_position, y=Node_y_position)) +
    geom_line(data=subset(tree_input_df, ancestor_state=="ancestor"), 
              aes(group=ancestor_state),
              color="dimgray") +
    geom_line(data=subset(tree_input_df, node_state!="none"), 
              aes(group=node_state),
              color="dimgray") +
    geom_segment(data=subset(tree_input_df, ancestor_state=="ancestor" & Node %in% deuterostoma_side), 
                 aes(x=0, y=Node_y_position, xend=Node_x_position, yend=Node_y_position), 
                 linetype="dashed", stroke=0.2, color="dimgray") +
    geom_segment(data=subset(tree_input_df, ancestor_state=="ancestor" & Node %in% protostoma_side),
                 aes(x=Node_x_position, y=Node_y_position, xend=21.5, yend=Node_y_position), 
                 linetype="dashed", stroke=0.2, color="dimgray") +
    theme_void() +
    #add different labels on the two sides
    scale_x_continuous(breaks=as.numeric(names(node_x_labels)), labels=unname(node_x_labels)) +
    scale_y_continuous(breaks=node_y_values, labels=deuterostoma_y_labels,
                       sec.axis = dup_axis(labels=protostoma_y_labels)) +
    theme(axis.title=element_blank(),
          axis.text.x = element_text(angle=30, color="black"),
          axis.text.y = element_text(color="black"),
          plot.title=element_text(hjust=0.5))

  return(ts_gains_plot)
}

##########################
##########################

plot_taus_ts_gains = function(clade_tau_df, my_clade_species, nodelist) {
  category_tau_plot = ggplot(data=clade_tau_df, aes(x=Species, y=Tau, fill=Species)) +
      geom_boxplot(alpha=0) + #very ugly line of code to set the axis.
      
      geom_rect(data=subset(clade_tau_df, Species %in% c(my_clade_species[1], my_clade_species[length(my_clade_species)]))[1,],
          aes(ymin=0, ymax=1,
              xmin=which(nodelist==my_clade_species[1])-0.5,
              xmax=which(nodelist==my_clade_species[length(my_clade_species)])+0.5),
          color="white", fill="dimgray", alpha=0.3) +
      
      geom_boxplot(color="black", alpha=0.8) +
      geom_hline(yintercept = 0.75, color="red") +
      scale_fill_manual(values=my_species_colors) +
      theme_bw() +
      theme(axis.title = element_text(color="black"),
            axis.text.y = element_text(color="black"),
            axis.text.x = element_text(color="black", angle=30)) +
      scale_y_continuous(breaks=number_ticks(10), limits=c(0,1)) +
      guides(fill=FALSE)
    
    return(category_tau_plot)
}
```

```{r extra_functions, warning=FALSE, message=FALSE, echo=FALSE, dependson=c("setup", "functions")}
plot_tree_structure_v1 = function(node_x_position, node_y_position, node_x_labels) {
  ancestors = c("Euarchontoglires", "Eutheria", "Mammalia", "Amniota", "Tetrapoda", "Euteleostomi", "Vertebrata", "Chordata", "Deuterostoma",
                "Bilateria", "Hs2", "Dme",
                "Cyclorrapha", "Diptera", "Panorpidae", "Oligoneoptera", "Holometabola", "Neoptera", "Insecta", "Arthropoda", "Protostoma")
  deuterostoma_side = c("Euarchontoglires", "Eutheria", "Mammalia", "Amniota", "Tetrapoda", "Euteleostomi", "Vertebrata", "Chordata", "Deuterostoma")
  protostoma_side = c("Cyclorrapha", "Diptera", "Panorpidae", "Oligoneoptera", "Holometabola", "Neoptera", "Insecta", "Arthropoda", "Protostoma")
  
  node_y_values = unique(as.numeric(unname(node_y_position))[as.numeric(unname(node_y_position)) != 1])
  deuterostoma_y_labels = c("Euarchontoglires", "Eutheria", "Mammalia", "Amniota", "Tetrapoda", "Euteleostomi", "Vertebrata", "Chordata", "Deuterostoma", "")
  protostoma_y_labels = c("Cyclorrapha", "Diptera", "Panorpidae", "Oligoneoptera", "Holometabola", "Neoptera", "Insecta", "Arthropoda", "Protostoma", "")
  
  node_x_position_dict = hashmap(names(node_x_position), unname(node_x_position))
  node_y_position_dict = hashmap(names(node_y_position), unname(node_y_position))
  
  tree_input_df = data.frame(Node=c(all_species, all_ancestors))
  #Add coordinates for each node
  tree_input_df$Node_x_position = node_x_position_dict$find(tree_input_df$Node)
  tree_input_df$Node_y_position = node_y_position_dict$find(tree_input_df$Node)
  #Add groups so that lines between all ancestors and all pairs will be drawn.
  tree_input_df$ancestor_state = rep("none", nrow(tree_input_df))
  tree_input_df$ancestor_state[tree_input_df$Node %in% ancestors] = "ancestor"
  #Add side of the tree (Deuterostoma or Protostoma)
  
  #This is long and manual
  tree_input_df$node_state = rep("none", nrow(tree_input_df))
  tree_input_df$node_state[tree_input_df$Node %in% c("Mm2", "Euarchontoglires")] = "Euarchontoglires"
  tree_input_df$node_state[tree_input_df$Node %in% c("Bt2", "Eutheria")] = "Eutheria"
  tree_input_df$node_state[tree_input_df$Node %in% c("Mdo", "Mammalia")] = "Mammalia"
  tree_input_df$node_state[tree_input_df$Node %in% c("Gga", "Amniota")] = "Amniota"
  tree_input_df$node_state[tree_input_df$Node %in% c("Xtr", "Tetrapoda")] = "Tetrapoda"
  tree_input_df$node_state[tree_input_df$Node %in% c("Dre", "Euteleostomi")] = "Euteleostomi"
  tree_input_df$node_state[tree_input_df$Node %in% c("Cmi", "Vertebrata")] = "Vertebrata"
  tree_input_df$node_state[tree_input_df$Node %in% c("Bla", "Chordata")] = "Chordata"
  tree_input_df$node_state[tree_input_df$Node %in% c("Sp2", "Deuterostoma")] = "Deuterostoma"
  
  tree_input_df$node_state[tree_input_df$Node %in% c("Eba", "Cyclorrapha")] = "Cyclorrapha"
  tree_input_df$node_state[tree_input_df$Node %in% c("Aae", "Diptera")] = "Diptera"
  tree_input_df$node_state[tree_input_df$Node %in% c("Bmo", "Panorpidae")] = "Panorpidae"
  tree_input_df$node_state[tree_input_df$Node %in% c("Tca", "Oligoneoptera")] = "Oligoneoptera"
  tree_input_df$node_state[tree_input_df$Node %in% c("Ame", "Holometabola")] = "Holometabola"
  tree_input_df$node_state[tree_input_df$Node %in% c("Bge", "Neoptera")] = "Neoptera"
  tree_input_df$node_state[tree_input_df$Node %in% c("Cdi", "Insecta")] = "Insecta"
  tree_input_df$node_state[tree_input_df$Node %in% c("Sma", "Arthropoda")] = "Arthropoda"
  tree_input_df$node_state[tree_input_df$Node %in% c("Obi", "Protostoma")] = "Protostoma"
  
  ts_gains_plot = ggplot(data=tree_input_df, aes(x=Node_x_position, y=Node_y_position)) +
    geom_line(data=subset(tree_input_df, ancestor_state=="ancestor"), 
              aes(group=ancestor_state),
              color="dimgray") +
    geom_line(data=subset(tree_input_df, node_state!="none"), 
              aes(group=node_state),
              color="dimgray") +
    #geom_segment(data=subset(tree_input_df, ancestor_state=="ancestor" & Node %in% deuterostoma_side), 
    #             aes(x=0, y=Node_y_position, xend=Node_x_position, yend=Node_y_position), 
    #             linetype="dashed", stroke=0.2, color="dimgray") +
    #geom_segment(data=subset(tree_input_df, ancestor_state=="ancestor" & Node %in% protostoma_side),
    #             aes(x=Node_x_position, y=Node_y_position, xend=21.5, yend=Node_y_position), 
    #             linetype="dashed", stroke=0.2, color="dimgray") +
    theme_void() +
    #add different labels on the two sides
    scale_x_continuous(breaks=as.numeric(names(node_x_labels)), labels=unname(node_x_labels)) +
    scale_y_continuous(breaks=node_y_values, labels=deuterostoma_y_labels,
                       sec.axis = dup_axis(labels=protostoma_y_labels)) +
    theme(axis.title=element_blank(),
          axis.text.x = element_text(angle=30, color="black", vjust=2, hjust=1),
          axis.text.y.right = element_text(color="black", hjust=1),
          axis.text.y.left = element_text(color="black", hjust=-0.1),
          plot.title=element_text(hjust=0.5),
          plot.margin = unit(c(1,1,2,1), "cm")
          )

  return(ts_gains_plot)
}
```

<!-- Panel A Number of convergen gains per tissue on the two side -->
```{r convergent_ts_gains_on_the_two_branches, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions"), eval=FALSE, fig.width=5, fig.height=3}
tissue_gains_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, my_version, "/inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  input_df$Node_Tissue = paste0(input_df$TS_gains, "_", input_df$Tissue)
  tissue_gains_df = rbind(tissue_gains_df, input_df)
}

##Isolate orthogroups with gains in only one tissue on each side of the tree
mod_tissue_gains_df = tissue_gains_df
mod_tissue_gains_df = subset(mod_tissue_gains_df, TS_gains!="Bilateria")
mod_tissue_gains_df$Clade = rep("Deuterostoma", nrow(mod_tissue_gains_df))
mod_tissue_gains_df$Clade[mod_tissue_gains_df$TS_gains %in% c(Protostoma, protostoma_y_labels)] = "Protostoma"
#Deuterostome
unique_deuterostome_df = unique(subset(mod_tissue_gains_df, Clade=="Deuterostoma")[,c("OG_ID", "Tissue")])
unique_deuterostome_counts = table(unique_deuterostome_df$OG_ID)
unique_deuterostome_OG_IDs = names(unique_deuterostome_counts[unique_deuterostome_counts==1])
deuterostome_OG_ID_tissue_dict = hashmap(as.vector(unique_deuterostome_df$OG_ID), as.vector(unique_deuterostome_df$Tissue))
#Protostome
unique_protostome_df = unique(subset(mod_tissue_gains_df, Clade=="Protostoma")[,c("OG_ID", "Tissue")])
unique_protostome_counts = table(unique_protostome_df$OG_ID)
unique_protostome_OG_IDs = names(unique_protostome_counts[unique_protostome_counts==1])
protostome_OG_ID_tissue_dict = hashmap(as.vector(unique_protostome_df$OG_ID), as.vector(unique_protostome_df$Tissue))

##Select cases with only one tissue both for deuterostome and protostome
unique_both_branches_OG_IDs = unique_deuterostome_OG_IDs[unique_deuterostome_OG_IDs %in% unique_protostome_OG_IDs]
##Select the cases where the tissue is the same for for Deuterostome and Protostome
diff_tissues_unique_both_branches_OG_IDs = unique_both_branches_OG_IDs[deuterostome_OG_ID_tissue_dict$find(unique_both_branches_OG_IDs) == protostome_OG_ID_tissue_dict$find(unique_both_branches_OG_IDs)]

##Select the most ancient gain for all those cases
mod_tissue_gains_df$TS_gains[mod_tissue_gains_df$TS_gains %in% ordered_species] = "species\nspecific"
#Deuterostome
deuterostome_final_entries_df = subset(mod_tissue_gains_df, Clade=="Deuterostoma" & OG_ID %in% diff_tissues_unique_both_branches_OG_IDs)
#Protostome
protostome_final_entries_df = subset(mod_tissue_gains_df, Clade=="Protostoma" & OG_ID %in% diff_tissues_unique_both_branches_OG_IDs)
#Joint
joint_df = rbind(deuterostome_final_entries_df, protostome_final_entries_df)
joint_df$TS_gains = factor(joint_df$TS_gains, levels=c("species\nspecific", all_ancestors), ordered = TRUE)
#Operate selection
final_join_df = joint_df %>%
  dplyr::select(-c(Node_Tissue, Category)) %>%
  distinct(OG_ID, TS_gains, Tissue, Clade) %>%
  group_by(OG_ID, Clade, Tissue) %>%
  filter(TS_gains == max(TS_gains))
  
two_gains_two_branches_same_tissue_df = as.data.frame(final_join_df)
two_gains_two_branches_same_tissue_df$Tissue = factor(two_gains_two_branches_same_tissue_df$Tissue, levels=ordered_tissues)
two_gains_two_branches_same_tissue_df$TS_gains = as.character(as.vector(two_gains_two_branches_same_tissue_df$TS_gains))

convergent_cases_df = unique(two_gains_two_branches_same_tissue_df[, c("OG_ID", "Tissue")])
convergent_cases_count_df = as.data.frame(table(convergent_cases_df$Tissue))
colnames(convergent_cases_count_df) = c("Tissue", "Count")
convergent_cases_count_df$Tissue = factor(convergent_cases_count_df$Tissue, levels=ordered_tissues)

convergent_cases_plot = ggplot(data=convergent_cases_count_df, aes(x=Tissue, y=Count, fill=Tissue, size=Count)) +
  geom_point(alpha=0.65, shape=21, color="black") +
  geom_text(aes(label=Count), size=4) +
  theme_bw() +
  scale_fill_manual(values = my_color_vector) +
  #scale_color_manual(values = my_color_vector) +
  scale_size(range = c(2,15)) +
  theme(axis.text.x = element_text(color="black", angle=30, hjust=1, vjust=1),
        axis.text.y = element_text(color="black"),
        axis.title.x = element_blank()) +
  guides(fill="none", color="none", size="none") +
  scale_y_continuous(breaks = number_ticks(6), limits=c(0,165)) +
  ylab("Number of convergent TS gains")



pdf("/Users/federica/Documents/CRG_PhD/bilaterian_GE/Images/Images_from_R/new_Fig5/Fig5_panelN.pdf", width=5, height=3)
print(convergent_cases_plot)
dev.off()
```

<!-- Panel B: TESMIN example. All in illustrator in the end -->

<!-- Panel C: Repeated enriched GO heatmap results -->
```{r repeated_GO_enrichments_all_tissues, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions"), fig.width=10, fig.height=5, eval=FALSE}
####### Upload GO transfer df
GO_transfers_df = read.delim(paste0(GO_transfers_dir, my_version, "/GO_annot/Hs2_transfers/orthogroups_from_Hs2-transferred_GO-reduced.txt"),
                             header=FALSE, col.names=c("OG_ID", "GO_term", "GO_name"))
### Generate dictionary from GO term to GO name
name_id_transfers_df = unique(GO_transfers_df[,c("GO_term", "GO_name")])
GO_ID_term_dict = hashmap(as.vector(name_id_transfers_df$GO_term), as.vector(name_id_transfers_df$GO_name))

###### Upload TS-gains df
tissue_gains_df = vector() #initiate dataframe
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, my_version, "/inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category")) #read file.
  input_df$Tissue = rep(tissue, nrow(input_df)) #add tissue info.
  input_df$Node_Tissue = paste0(input_df$TS_gains, "_", input_df$Tissue) #Add node_tissue info.
  tissue_gains_df = rbind(tissue_gains_df, input_df) #join to general dataframe,
}

##### For each species and each ancestors, get the total number of gains in each tissue (to use for the proportion computation).
tissue_gains_counts = table(as.vector(tissue_gains_df$Node_Tissue)) #Total number of gains in each tissue for each node/species.
all_combinations_df = expand.grid(c(all_ancestors, ordered_species), ordered_tissues) #Create a dataframe with col1=node/species and col2=tissue (all combinations).
all_combinations = paste0(all_combinations_df$Var1, "_", all_combinations_df$Var2) #Create unique variable combining node/species and tissue: node_tissue.
missing_IDs = all_combinations[!(all_combinations %in% names(tissue_gains_counts))] #Isolate all the nodes/species that do not have gains in a given tissue.
if (length(missing_IDs) > 0) { #if there are any such nodes, add zero counts for them to the tissue_gains dataframe
  for (missing_ID in missing_IDs) {
    tissue_gains_counts = c(tissue_gains_counts, 0)
    names(tissue_gains_counts)[length(tissue_gains_counts)] = missing_ID
  }
}
tissue_gains_counts_dict = hashmap(names(tissue_gains_counts), as.vector(tissue_gains_counts))

######################################################
##### Select significant categories ##################
######################################################
final_GO_res_df = vector()
for (tissue in ordered_tissues) {
  #For each tissue, upload all GO results that respect some pre-defined cut-offs.
  for (node in c(ordered_species, all_ancestors)) {
    #Upload all GO results from the gain files.
    GO_res_file = paste0(ts_gains_losses_dir, my_version, "/inferences/GO_enrichments/Bilateria_conserved_orthogroups-All_genes-", 
                         tissue, "_inferred_gains-", node, "-orthogroups_from_Hs2-GO_res.tab")
    if (!file.exists(GO_res_file) | file.size(GO_res_file) == 1 | is.na(file.size(GO_res_file))) { #If the file does not exists
      next
    }
    else { #if file exists
      GO_res_df = read.delim(GO_res_file, header=TRUE)
      #Select entries with at least 5 genes in the set.
      GO_res_df = subset(GO_res_df, significant=="TRUE" & intersection_size>=5)
      #If there are no significant entries, continue
      if (nrow(GO_res_df)==0) {
        next
      }
      #Add tissue
      GO_res_df$node_tissue = rep(paste0(tissue, "_", node), nrow(GO_res_df))
      #Add to final df
      final_GO_res_df = rbind(final_GO_res_df, GO_res_df)
    }
  }
}
#For all tissues, select the categories that are significant in at least X nodes/species
heatmap_input_df = final_GO_res_df[,c("term_name", "term_id", "node_tissue")] #Subset to only the needed values
heatmap_input_df$significant = rep(1, nrow(heatmap_input_df)) #Add value of 1 (it indicates that the category is significant)
##### Generate wide dataframe (rows=GO_term, cols=node_tissue)
heatmap_input_wide_tibble = heatmap_input_df %>% pivot_wider(names_from="node_tissue", values_from="significant", values_fill=0)
#Transform to dataframe and modify it so that we can work on it.
heatmap_input_wide_df = as.data.frame(heatmap_input_wide_tibble)
rownames(heatmap_input_wide_df) = heatmap_input_wide_df$term_name
heatmap_input_wide_df$term_name = NULL
heatmap_input_wide_df[is.na(heatmap_input_wide_df)] = 0 #Replace NA with 0

#####Select all the GO categories significant in at least 15 nodes/species in that tissue
selected_ids = names(table(heatmap_input_df$term_id)[table(heatmap_input_df$term_id)>=15])

#######################################################################
#### Compute EXPECTED propotions of annotated gains ##################
#######################################################################
selected_GO_transfer_df = subset(GO_transfers_df, GO_term %in% selected_ids)
selected_GO_transfer_df$GO_term = as.vector(selected_GO_transfer_df$GO_term)
selected_GO_transfer_df$OG_ID = as.vector(selected_GO_transfer_df$OG_ID)

##### For each selected IDs, compute proportion of genes belonging to each category 
#over the total numbers of genes that:
#1. have a functional association.
#2. are bilaterian conserved.
#This will be used to compute the expected number of gains in each category, if these gains were random.
genes_with_annot = unique(as.vector(GO_transfers_df$OG_ID))
#Uload info about bilaterian conserved
all_orthogroups_df = read.delim(paste0(gene_sets_dir, my_version, "/STRICT/Bilateria/conserved/Bilateria_conserved_orthogroups-EXPR_genes.txt"),
                                 header=FALSE, col.names = c("OG_ID", "Species", "GeneID"))
bilaterian_conserved_OGs = unique(as.vector(all_orthogroups_df$OG_ID))
#Count how many annotated genes are also bilaterian conserved.
conserved_genes_with_annot = genes_with_annot[genes_with_annot %in% bilaterian_conserved_OGs]
tot_genes_with_annot = length(conserved_genes_with_annot)

#Initiate dataframe to compute the expected proportion (the dataframe is already subsetted to the enriched categories).
category_proportion_df = as.data.frame(table(selected_GO_transfer_df$GO_term)) 
colnames(category_proportion_df) = c("GO_term", "Freq")
category_proportion_df$proportion = category_proportion_df$Freq/Tot_genes_with_annot
#Create dictionary with GO term to proportion correspondence.
category_proportion_dict = hashmap(as.vector(category_proportion_df$GO_term), as.vector(category_proportion_df$proportion))

#######################################################################
#### Compute OBSERVED propotions of annotated gains ###################
#######################################################################
all_node_category_proportion_df = vector()
#for each node/species_tissue
for (comb in all_combinations) {
  node_gains_OGs = as.vector(subset(tissue_gains_df, Node_Tissue==comb)$OG_ID) #Select all orthogroups with gains in that node_tissue.
  #Count how many gains in the selected categories for that node/species_tissue
  node_gains_GO_transfer_count_df = as.data.frame(table(as.vector(subset(selected_GO_transfer_df, GO_term %in% selected_ids & OG_ID %in% node_gains_OGs)$GO_term)))
  #### If there is at least one gain for that node/species_tissue in any of the categories.
  if (nrow(node_gains_GO_transfer_count_df) > 0) { 
    colnames(node_gains_GO_transfer_count_df) = c("GO_term", "num") #set colnames
    # Add zero counts for any missing categories
    missing_categories = selected_ids[!(selected_ids %in% as.vector(node_gains_GO_transfer_count_df$GO_term))] #isolate missing categories
    #if there are any missing categories
    if (length(missing_categories) > 0) {
      #add one line with zero to the node/species_tissue df for each of the missing categories
      node_gains_GO_transfer_count_df = rbind(node_gains_GO_transfer_count_df, 
                                              data.frame("GO_term"=missing_categories, num=rep(0,length(missing_categories))))
      }
    } else { #If there are no gains for that node/species_tissue, add 0 for ALL categories
      node_gains_GO_transfer_count_df = data.frame("GO_term"=selected_ids, num=rep(0, length(selected_ids)))
    }
  ### Add information about the total number of gains for each node/species_tissue
  node_gains_GO_transfer_count_df$background = rep(tissue_gains_counts_dict$find(comb), nrow(node_gains_GO_transfer_count_df))
  ### Compute the EXPECTED distribution for each category (number of gains in the nodes multiplied by the total proportion of genes in that category)
  node_gains_GO_transfer_count_df$expected = node_gains_GO_transfer_count_df$background * category_proportion_dict$find(node_gains_GO_transfer_count_df$GO_term)
  node_gains_GO_transfer_count_df$GO_name = GO_ID_term_dict$find(node_gains_GO_transfer_count_df$GO_term) #Add GO name
  node_gains_GO_transfer_count_df$Node = rep(comb, nrow(node_gains_GO_transfer_count_df)) #Add node/species_tissue info
  ### Join into final dataframe
  all_node_category_proportion_df = rbind(all_node_category_proportion_df, node_gains_GO_transfer_count_df)
}

#### Compute observed vs expected ratio for all groups (use the log2 values)
#NB: num is the number of gains in each node/species_tissue belonging to each category.
all_node_category_proportion_df$proportion = log2((all_node_category_proportion_df$num / all_node_category_proportion_df$expected)+1)
#Order levels for plotting
all_node_category_proportion_df$Node = factor(all_node_category_proportion_df$Node, levels=all_combinations)
all_node_category_proportion_df$GO_term = factor(all_node_category_proportion_df$GO_term, selected_ids)

#######################################################################
####################### Generate final table ##########################
#######################################################################

wider_all_node_category_proportion_df = all_node_category_proportion_df %>%
  pivot_wider(GO_term, names_from = Node, values_from=proportion) %>%
  column_to_rownames("GO_term")

wider_all_node_category_proportion_df = wider_all_node_category_proportion_df[selected_ids,] #there should be no need to do this, in principle...

#Filter for at least one enrichment >=3 in the row
wider_all_node_category_proportion_df = wider_all_node_category_proportion_df[apply(wider_all_node_category_proportion_df, 1, function(x) {max(x, na.rm = TRUE) >= 3}),]
rownames(wider_all_node_category_proportion_df) = GO_ID_term_dict$find(rownames(wider_all_node_category_proportion_df))
#Remove all cases where there are only NAs (not sure why this should happen).
wider_all_node_category_proportion_df = wider_all_node_category_proportion_df[,colSums(is.na(wider_all_node_category_proportion_df)) < nrow(wider_all_node_category_proportion_df)]


#######################################################################
####################### Get annotation col df #########################
#######################################################################
annotation_col_df = data.frame(Tissues=sub(".*_", "", colnames(wider_all_node_category_proportion_df)))
rownames(annotation_col_df) = colnames(wider_all_node_category_proportion_df)
annotation_col_df$Ancestrality = rep("Ancestral", nrow(annotation_col_df))
annotation_col_df$Ancestrality[grep(paste(ordered_species, collapse="|"), rownames(annotation_col_df))] = "Species_specific"

#######################################################################
####################### Plot ##########################################
#######################################################################
all_tissues_heatmap_inputs_df = wider_all_node_category_proportion_df

GO_res_all_tissues_heatmap = pheatmap(wider_all_node_category_proportion_df,
                             color = c(brewer.pal(9, "PuRd")[1],brewer.pal(9, "PuRd")[3:9]),
                             breaks = c(0,1.2,2,2.5,3,4,5,6),
                             border_color = "black",
                             main = paste0("Recurrent gains in functional categories"),
                             cluster_cols = FALSE,
                             cluster_rows = TRUE,
                             annotation_col = annotation_col_df,
                             annotation_colors = list(Tissues = my_color_vector, 
                                                      #Branch = my_branch_colors, 
                                                      Ancestrality = my_ancestral_colors, 
                                                      Recurrent_tissue=c("Most_tissues"="black", "Non_germiline"="gray82", "Germline"="mediumorchid2")),
                             annotation_legend = TRUE, 
                             show_rownames = TRUE,
                             show_colnames = FALSE,
                             #silent = TRUE,
                             #legend = FALSE,
                             #angle_col = 45,
                             #fontsize_col = 8,
                             fontsize_row = 8,
                             angle_col = 0
                             )
```

```{r repeated_GO_enrichments_single_tissues, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions"), fig.width=10, fig.height=5, eval=FALSE}
####### Upload GO transfer df
GO_transfers_df = read.delim(paste0(GO_transfers_dir, my_version, "/GO_annot/Hs2_transfers/orthogroups_from_Hs2-transferred_GO-reduced.txt"),
                             header=FALSE, col.names=c("OG_ID", "GO_term", "GO_name"))
### Generate dictionary from GO term to GO name
name_id_transfers_df = unique(GO_transfers_df[,c("GO_term", "GO_name")])
GO_ID_term_dict = hashmap(as.vector(name_id_transfers_df$GO_term), as.vector(name_id_transfers_df$GO_name))

###### Upload TS-gains df
tissue_gains_df = vector() #initiate dataframe
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, my_version, "/inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category")) #read file.
  input_df$Tissue = rep(tissue, nrow(input_df)) #add tissue info.
  input_df$Node_Tissue = paste0(input_df$TS_gains, "_", input_df$Tissue) #Add node_tissue info.
  tissue_gains_df = rbind(tissue_gains_df, input_df) #join to general dataframe,
}

##### For each species and each ancestors, get the total number of gains in each tissue (to use for the proportion computation).
tissue_gains_counts = table(as.vector(tissue_gains_df$Node_Tissue)) #Total number of gains in each tissue for each node/species.
all_combinations_df = expand.grid(c(all_ancestors, ordered_species), ordered_tissues) #Create a dataframe with col1=node/species and col2=tissue (all combinations).
all_combinations = paste0(all_combinations_df$Var1, "_", all_combinations_df$Var2) #Create unique variable combining node/species and tissue: node_tissue.
missing_IDs = all_combinations[!(all_combinations %in% names(tissue_gains_counts))] #Isolate all the nodes/species that do not have gains in a given tissue.
if (length(missing_IDs) > 0) { #if there are any such nodes, add zero counts for them to the tissue_gains dataframe
  for (missing_ID in missing_IDs) {
    tissue_gains_counts = c(tissue_gains_counts, 0)
    names(tissue_gains_counts)[length(tissue_gains_counts)] = missing_ID
  }
}
tissue_gains_counts_dict = hashmap(names(tissue_gains_counts), as.vector(tissue_gains_counts))

######################################################
##### Select significant categories ##################
######################################################
all_filtered_significant_terms = vector()
all_significant_terms = vector()
annotation_row_df = vector()
for (tissue in ordered_tissues[ordered_tissues!="Ovary"]) {
  #For each tissue, upload all GO results that respect some pre-defined cut-offs.
  final_GO_res_df = vector()
  if (tissue == "Testis") {my_tissue = c("Testis", "Ovary")} else {my_tissue = tissue} #this is a trick to consider Testis and Ovary together
  for (tissue in my_tissue) {
    for (node in c(ordered_species, all_ancestors)) {
      #Upload all GO results from the gain files.
      GO_res_file = paste0(ts_gains_losses_dir, my_version, "/inferences/GO_enrichments/Bilateria_conserved_orthogroups-All_genes-", 
                           tissue, "_inferred_gains-", node, "-orthogroups_from_Hs2-GO_res.tab")
      if (!file.exists(GO_res_file) | file.size(GO_res_file) == 1 | is.na(file.size(GO_res_file))) { #If the file does not exists
        next
      }
      else { #if file exists
        GO_res_df = read.delim(GO_res_file, header=TRUE)
        #Select entries with at least 5 genes in the set.
        GO_res_df = subset(GO_res_df, significant=="TRUE" & intersection_size>=5)
        #If there are no significant entries, continue
        if (nrow(GO_res_df)==0) {
          next
        }
        #Add tissue
        GO_res_df$node_tissue = rep(paste0(tissue, "_", node), nrow(GO_res_df))
        #Add to final df
        final_GO_res_df = rbind(final_GO_res_df, GO_res_df)
      }
    }
  }
  #For each tissue, select the categories that are significant in at least X nodes/species
  heatmap_input_df = final_GO_res_df[,c("term_name", "term_id", "node_tissue")] #Subset to only the needed values
  heatmap_input_df$significant = rep(1, nrow(heatmap_input_df)) #Add value of 1 (it indicates that the category is significant)
  ##### Generate wide dataframe (rows=GO_term, cols=node_tissue)
  heatmap_input_wide_tibble = heatmap_input_df %>% pivot_wider(names_from="node_tissue", values_from="significant", values_fill=0)
  #Transform to dataframe and modify it so that we can work on it.
  heatmap_input_wide_df = as.data.frame(heatmap_input_wide_tibble)
  rownames(heatmap_input_wide_df) = heatmap_input_wide_df$term_name
  heatmap_input_wide_df$term_name = NULL
  heatmap_input_wide_df[is.na(heatmap_input_wide_df)] = 0 #Replace NA with 0
  
  #####Select all the GO categories significant in at least X nodes/species in that tissue
  filtered_significant_terms = names(table(heatmap_input_df$term_id)[table(heatmap_input_df$term_id)>=8])
  all_tissue_significant_terms = names(table(heatmap_input_df$term_id)[table(heatmap_input_df$term_id)>=1])
  #Add the filtered significant terms to the relative final vector
  all_filtered_significant_terms = c(all_filtered_significant_terms, filtered_significant_terms)
  #Add all significant terms to the relative final vector
  all_significant_terms = c(all_significant_terms, all_tissue_significant_terms)
  #annotation_row_df = rbind(annotation_row_df, data.frame(GO_term=tissue_selected_ids, GO_class=rep(tissue, length(tissue_selected_ids))))
}

### select only terms that are significant in only one tissue: if there are duplicated values, it is because they come from different tissues
non_repeated_significant_terms = all_significant_terms[!(duplicated(all_significant_terms)|duplicated(all_significant_terms, fromLast=TRUE))]
enriched_non_repeated_significant_terms = non_repeated_significant_terms[non_repeated_significant_terms %in% all_filtered_significant_terms]
selected_ids = enriched_non_repeated_significant_terms

#######################################################################
#### Compute EXPECTED propotions of annotated gains ##################
#######################################################################
selected_GO_transfer_df = subset(GO_transfers_df, GO_term %in% selected_ids)
selected_GO_transfer_df$GO_term = as.vector(selected_GO_transfer_df$GO_term)
selected_GO_transfer_df$OG_ID = as.vector(selected_GO_transfer_df$OG_ID)

##### For each selected IDs, compute proportion of genes belonging to each category 
#over the total numbers of genes that:
#1. have a functional association.
#2. are bilaterian conserved.
#This will be used to compute the expected number of gains in each category, if these gains were random.
genes_with_annot = unique(as.vector(GO_transfers_df$OG_ID))
#Uload info about bilaterian conserved
all_orthogroups_df = read.delim(paste0(gene_sets_dir, my_version, "/STRICT/Bilateria/conserved/Bilateria_conserved_orthogroups-EXPR_genes.txt"),
                                 header=FALSE, col.names = c("OG_ID", "Species", "GeneID"))
bilaterian_conserved_OGs = unique(as.vector(all_orthogroups_df$OG_ID))
#Count how many annotated genes are also bilaterian conserved.
conserved_genes_with_annot = genes_with_annot[genes_with_annot %in% bilaterian_conserved_OGs]
tot_genes_with_annot = length(conserved_genes_with_annot)

#Initiate dataframe to compute the expected proportion (the dataframe is already subsetted to the enriched categories).
category_proportion_df = as.data.frame(table(selected_GO_transfer_df$GO_term)) 
colnames(category_proportion_df) = c("GO_term", "Freq")
category_proportion_df$proportion = category_proportion_df$Freq/Tot_genes_with_annot
#Create dictionary with GO term to proportion correspondence.
category_proportion_dict = hashmap(as.vector(category_proportion_df$GO_term), as.vector(category_proportion_df$proportion))

#######################################################################
#### Compute OBSERVED propotions of annotated gains ###################
#######################################################################
all_node_category_proportion_df = vector()
#for each node/species_tissue
for (comb in all_combinations) {
  node_gains_OGs = as.vector(subset(tissue_gains_df, Node_Tissue==comb)$OG_ID) #Select all orthogroups with gains in that node_tissue.
  #Count how many gains in the selected categories for that node/species_tissue
  node_gains_GO_transfer_count_df = as.data.frame(table(as.vector(subset(selected_GO_transfer_df, GO_term %in% selected_ids & OG_ID %in% node_gains_OGs)$GO_term)))
  #### If there is at least one gain for that node/species_tissue in any of the categories.
  if (nrow(node_gains_GO_transfer_count_df) > 0) { 
    colnames(node_gains_GO_transfer_count_df) = c("GO_term", "num") #set colnames
    # Add zero counts for any missing categories
    missing_categories = selected_ids[!(selected_ids %in% as.vector(node_gains_GO_transfer_count_df$GO_term))] #isolate missing categories
    #if there are any missing categories
    if (length(missing_categories) > 0) {
      #add one line with zero to the node/species_tissue df for each of the missing categories
      node_gains_GO_transfer_count_df = rbind(node_gains_GO_transfer_count_df, 
                                              data.frame("GO_term"=missing_categories, num=rep(0,length(missing_categories))))
      }
    } else { #If there are no gains for that node/species_tissue, add 0 for ALL categories
      node_gains_GO_transfer_count_df = data.frame("GO_term"=selected_ids, num=rep(0, length(selected_ids)))
    }
  ### Add information about the total number of gains for each node/species_tissue
  node_gains_GO_transfer_count_df$background = rep(tissue_gains_counts_dict$find(comb), nrow(node_gains_GO_transfer_count_df))
  ### Compute the EXPECTED distributtion for each category (number of gains in the nodes multiplied by the total proportion of genes in that category)
  node_gains_GO_transfer_count_df$expected = node_gains_GO_transfer_count_df$background * category_proportion_dict$find(node_gains_GO_transfer_count_df$GO_term)
  node_gains_GO_transfer_count_df$GO_name = GO_ID_term_dict$find(node_gains_GO_transfer_count_df$GO_term) #Add GO name
  node_gains_GO_transfer_count_df$Node = rep(comb, nrow(node_gains_GO_transfer_count_df)) #Add node/species_tissue info
  ### Join into final dataframe
  all_node_category_proportion_df = rbind(all_node_category_proportion_df, node_gains_GO_transfer_count_df)
}

#### Compute observed vs expected ratio for all groups (use the log2 values)
#NB: num is the number of gains in each node/species_tissue belonging to each category.
all_node_category_proportion_df$proportion = log2((all_node_category_proportion_df$num / all_node_category_proportion_df$expected)+1)
#Order levels for plotting
all_node_category_proportion_df$Node = factor(all_node_category_proportion_df$Node, levels=all_combinations)
all_node_category_proportion_df$GO_term = factor(all_node_category_proportion_df$GO_term, selected_ids)

#######################################################################
####################### Generate final table ##########################
#######################################################################

wider_all_node_category_proportion_df = all_node_category_proportion_df %>%
  pivot_wider(GO_term, names_from = Node, values_from=proportion) %>%
  column_to_rownames("GO_term")

wider_all_node_category_proportion_df = wider_all_node_category_proportion_df[selected_ids,] #there should be no need to do this, in principle...

#Filter for at least one enrichment >=3 in the row
wider_all_node_category_proportion_df = wider_all_node_category_proportion_df[apply(wider_all_node_category_proportion_df, 1, function(x) {max(x, na.rm = TRUE) >= 3}),]
rownames(wider_all_node_category_proportion_df) = GO_ID_term_dict$find(rownames(wider_all_node_category_proportion_df))
#Remove all cases where there are only NAs (not sure why this should happen).
wider_all_node_category_proportion_df = wider_all_node_category_proportion_df[,colSums(is.na(wider_all_node_category_proportion_df)) < nrow(wider_all_node_category_proportion_df)]


#######################################################################
####################### Get annotation col df #########################
#######################################################################
annotation_col_df = data.frame(Tissues=sub(".*_", "", colnames(wider_all_node_category_proportion_df)))
rownames(annotation_col_df) = colnames(wider_all_node_category_proportion_df)
annotation_col_df$Ancestrality = rep("Ancestral", nrow(annotation_col_df))
annotation_col_df$Ancestrality[grep(paste(ordered_species, collapse="|"), rownames(annotation_col_df))] = "Species_specific"

#######################################################################
####################### Plot ##########################################
#######################################################################

single_tissues_heatmap_inputs_df = wider_all_node_category_proportion_df

GO_res_single_tissues_heatmap = pheatmap(wider_all_node_category_proportion_df,
                             color = c(brewer.pal(9, "PuRd")[1],brewer.pal(9, "PuRd")[3:9]),
                             breaks = c(0,1.2,2,2.5,3,4,5,6),
                             border_color = "black",
                             main = paste0("Recurrent gains in functional categories"),
                             cluster_cols = FALSE,
                             cluster_rows = TRUE,
                             annotation_col = annotation_col_df,
                             annotation_colors = list(Tissues = my_color_vector, 
                                                      #Branch = my_branch_colors, 
                                                      Ancestrality = my_ancestral_colors, 
                                                      Recurrent_tissue=c("Most_tissues"="black", "Non_germiline"="gray82", "Germline"="mediumorchid2")),
                             annotation_legend = TRUE, 
                             show_rownames = TRUE,
                             show_colnames = FALSE,
                             #silent = TRUE,
                             #legend = FALSE,
                             #angle_col = 45,
                             #fontsize_col = 8,
                             fontsize_row = 8,
                             angle_col = 0
                             )
```

```{r join_repeated_enrichments, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions"), fig.width=10, fig.height=4, eval=FALSE}
joint_heatmaps_input_df = rbind(all_tissues_heatmap_inputs_df, single_tissues_heatmap_inputs_df)

#######################################################################
####################### Get annotation col df #########################
#######################################################################
annotation_col_df = data.frame(Tissues=sub(".*_", "", colnames(joint_heatmaps_input_df)))
rownames(annotation_col_df) = colnames(joint_heatmaps_input_df)
annotation_col_df$Ancestrality = rep("Ancestral", nrow(annotation_col_df))
annotation_col_df$Ancestrality[grep(paste(ordered_species, collapse="|"), rownames(annotation_col_df))] = "Species_specific"


#######################################################################
####################### Plot ##########################################
#######################################################################

GO_res_heatmap = pheatmap(joint_heatmaps_input_df,
                             color = c(brewer.pal(9, "PuRd")[1],brewer.pal(9, "PuRd")[3:9]),
                             breaks = c(0,1.2,2,2.5,3,4,5,6),
                             border_color = "black",
                             main = paste0("Recurrent gains in functional categories"),
                             cluster_cols = FALSE,
                             cluster_rows = TRUE,
                             annotation_col = annotation_col_df,
                             annotation_colors = list(Tissues = my_color_vector, 
                                                      #Branch = my_branch_colors, 
                                                      Ancestrality = my_ancestral_colors, 
                                                      Recurrent_tissue=c("Most_tissues"="black", "Non_germiline"="gray82", "Germline"="mediumorchid2")),
                             annotation_legend = TRUE, 
                             show_rownames = TRUE,
                             show_colnames = FALSE,
                             #silent = TRUE,
                             #legend = FALSE,
                             #angle_col = 45,
                             #fontsize_col = 8,
                             fontsize_row = 8,
                             angle_col = 0
                             )


#######################################################################
############Select categories and define order ########################
#######################################################################
final_categories = c("sequence-specific_double-stranded_DNA_binding", "animal_organ_morphogenesis", "cation_transport",
                     "integral_component_of_plasma_membrane", 
                     "plasma_membrane_region",
                     "cell_projection_organization", "cell_cycle", "meiotic_nuclear_division")

final_heatmaps_df = joint_heatmaps_input_df[final_categories,]

#pdf("/Users/federica/Documents/CRG_PhD/bilaterian_GE/Images/Images_from_R/fig6/GO_heatmap-v1.pdf", width=10, height=4)
final_GO_res_heatmap = pheatmap(final_heatmaps_df,
                             color = c(brewer.pal(9, "PuRd")[1],brewer.pal(9, "PuRd")[3:9]),
                             breaks = c(0,1.2,2,2.5,3,4,5,6),
                             border_color = "black",
                             main = paste0("Recurrent gains in functional categories"),
                             cluster_cols = FALSE,
                             cluster_rows = FALSE,
                             annotation_col = annotation_col_df,
                             annotation_colors = list(Tissues = my_color_vector, 
                                                      #Branch = my_branch_colors, 
                                                      Ancestrality = my_ancestral_colors, 
                                                      Recurrent_tissue=c("Most_tissues"="black", "Non_germiline"="gray82", "Germline"="mediumorchid2")),
                             annotation_legend = TRUE, 
                             show_rownames = TRUE,
                             show_colnames = FALSE,
                             fontsize_row = 8,
                             angle_col = 0
                             )
#dev.off()
```


<!-- Panel D: check species-specific gains for enrichments in morphogen/cell signalling activity -->
```{r species_specific_gains_GOs, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions"), eval=FALSE, fig.width=5, fig.height=3}
#### Upload Go transfers dataframe
#interesting_GOs = c("GO:0016015", "GO:0007267", "GO:0008134")
GO_transfers_df = read.delim(paste0(GO_transfers_dir, my_version, "/GO_annot/Hs2_transfers/orthogroups_from_Hs2-transferred_GO-reduced.txt"),
                             header=FALSE, col.names=c("OG_ID", "GO_term", "GO_name"))

#Upload gains df
tissue_gains_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, my_version, "/inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  input_df$Node_Tissue = paste0(input_df$TS_gains, "_", input_df$Tissue)
  tissue_gains_df = rbind(tissue_gains_df, input_df)
}

#Isolate all orthogroups with at least one tissue-specific gene
all_species_specific_gains_orthogroups = unique(as.vector(subset(tissue_gains_df, TS_gains %in% ordered_species)$OG_ID))

#Upload and select all bilaterian conserved orthogroups
all_orthogroups_df = read.delim(paste0(gene_sets_dir, my_version, "/STRICT/Bilateria/conserved/Bilateria_conserved_orthogroups-EXPR_genes.txt"),
                                header=FALSE, col.names = c("OG_ID", "Species", "GeneID"))
all_orthogroups = unique(as.vector(all_orthogroups_df$OG_ID))
OG_conserved_orthogroups = all_orthogroups[all_orthogroups %in% GO_OG_IDs]

### Subset GO transfers to only bilaterian conserved orthogroups
bilaterian_conserved_GO_transfers_df = subset(GO_transfers_df, OG_ID %in% all_orthogroups)

#Filter only for those GOs with at least 10 annotated bilaterian conserved genes
bilaterian_conserved_genes_counts = table(bilaterian_conserved_GO_transfers_df$GO_term)
selected_bilaterian_conserved = names(bilaterian_conserved_genes_counts[bilaterian_conserved_genes_counts >= 10])
bilaterian_conserved_GO_transfers_df = subset(bilaterian_conserved_GO_transfers_df, GO_term %in% selected_bilaterian_conserved)

#Add tag relative to tissue specificity
bilaterian_conserved_GO_transfers_df$TS_tag = rep(FALSE, nrow(bilaterian_conserved_GO_transfers_df))
bilaterian_conserved_GO_transfers_df$TS_tag[bilaterian_conserved_GO_transfers_df$OG_ID %in% all_species_specific_gains_orthogroups] = TRUE

### Proportion of bilaterian conserved orthogroups in each GO that have at least one species-specific gain.
GO_species_specific_prop_df = bilaterian_conserved_GO_transfers_df %>%
  mutate(Total = TRUE) %>%
  group_by(GO_name) %>%
  summarize(species_specific_prop = sum(TS_tag)/sum(Total))

#Isolate the 95th percentile of the distribution
percentile_95 = quantile(GO_species_specific_prop_df$species_specific_prop, 0.95)
percentile_95_df = subset(GO_species_specific_prop_df, species_specific_prop >= percentile_95)
percentile_95_categories = as.vector(percentile_95_df$GO_name)[order(as.vector(percentile_95_df$species_specific_prop), decreasing=TRUE)]

#Select categories I want to plot (I need them to add lines)
categories_to_plot = c("cell-cell_signaling", "tissue_development", "anatomical_structure_formation_involved_in_morphogenesis")
lines_to_add = as.vector(subset(percentile_95_df, GO_name %in% categories_to_plot)$species_specific_prop)

#Plot
GO_species_specific_gains_prop_plot = ggplot(data=subset(GO_species_specific_prop_df, species_specific_prop > 0), aes(x=species_specific_prop)) +
  geom_density(color="dimgray", fill="dimgray", alpha=0.4) +
  geom_vline(xintercept = percentile_95, linetype="dashed") +
  geom_vline(xintercept = lines_to_add, color="black") +
  theme_bw() +
  theme(axis.text = element_text(color="black"),
        axis.title.y = element_text(color="black"),
        axis.title.x = element_blank()) +
  scale_y_continuous(breaks=number_ticks(5))

print(GO_species_specific_gains_prop_plot)

#Isolate developmental categories
developmental_categories_95_percentile = percentile_95_categories[grep("develop|differentiation|determination|morphogen|commitment|specification|regionalization|formation|genesis", percentile_95_categories)]
percentile_95_df$Category = rep("Other", nrow(percentile_95_df))
percentile_95_df$Category[percentile_95_df$GO_name %in% developmental_categories_95_percentile] = "Developmental"
percentile_95_df$Category = factor(percentile_95_df$Category, levels=rev(c("Developmental", "Other")))
percentile_95_df$Group = rep("95th_percentile", nrow(percentile_95_df))

developmental_proportion_barplot = ggplot(data=percentile_95_df, aes(x=Group, fill=Category, color=Category)) +
  geom_bar(stat="count", position="fill", alpha=0.4) +
  scale_fill_manual(values=c("Developmental"="firebrick1", "Other"="dimgray")) +
  scale_color_manual(values=c("Developmental"="firebrick1", "Other"="dimgray")) +
  theme_bw() +
  theme(axis.title = element_blank(),
        axis.text = element_text(color="black"),
        panel.grid = element_blank()) +
  guides(fill="none", color="none") +
  scale_y_continuous(position="right")

plot(developmental_proportion_barplot)

panel_P = plot_grid(GO_species_specific_gains_prop_plot, developmental_proportion_barplot, ncol=2, rel_widths = c(1, 0.3), align="h")

#pdf("/Users/federica/Documents/CRG_PhD/bilaterian_GE/Images/Images_from_R/new_Fig5/Fig5_panelP.pdf", width=5, height=3)
#print(panel_P)
#dev.off()


#############################
#### Test the background ####
#############################
all_categories = unique(as.vector(bilaterian_conserved_GO_transfers_df$GO_name))
all_categories_developmental = all_categories[grep("develop|differentiation|determination|morphogen|commitment|specification|regionalization|formation|genesis", all_categories)]
background_proportion = length(all_categories_developmental[!(all_categories_developmental %in% percentile_95_categories)]) / (length(all_categories)-length(percentile_95_categories))
```

<!-- Panel E: Fgf17 example -->
```{r fgf17_expr, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions"), eval=FALSE, fig.height=3, fig.width=7}
#### Normalize by reference
reference_OG = c("GF_013551", "GF_013830")
all_orthogroups_df = read.delim(paste0(gene_sets_dir, my_version, "/STRICT/Bilateria/conserved/Bilateria_conserved_orthogroups-EXPR_genes.txt"),
                                header=FALSE, col.names = c("OG_ID", "Species", "GeneID"))
reference_OGs_df = subset(all_orthogroups_df, OG_ID %in% reference_OG & 
                            !(GeneID %in% c("FBgn0004403", "FBgn0033555", "ENSBTAG00000048169", "ENSBTAG00000054615", "ENSBTAG00000037600"))) #These are the genes that I need to exclude.
reference_OGs_df = reference_OGs_df[,c("Species", "GeneID")]
reference_OGs_df$Species = as.character(as.vector(reference_OGs_df$Species))
reference_OGs_df$Species[reference_OGs_df$Species=="BmA"] = "Bmo"

#### Build expression table: all genes from all species across all tissues
all_species_expr_df = vector()
for (species in ordered_species) {
  input_file = paste0(tau_dir, my_version, "/", species, "-protein_coding-tissue_average_expr-NOSVA-log2-TPMs-NORM.tab")
  input_df = read.delim(input_file, header=TRUE, row=1)
  #Isolate reference gene expr and scale input expr
  #reference_gene = as.vector(subset(reference_OGs_df, Species==species)$GeneID)
  #reference_gene_expr = as.numeric(unname(apply(input_df[reference_gene,], 2, mean))) #Compute the mean between the selected reference genes
  #scaled_input_df = t(t(input_df)/reference_gene_expr)
  scaled_input_df = input_df
  ######
  long_scaled_input_df = melt(as.matrix(scaled_input_df))
  colnames(long_scaled_input_df) = c("GeneID", "Tissue", "Scaled_expr")
  long_scaled_input_df$Species = rep(species, nrow(long_scaled_input_df))
  all_species_expr_df = rbind(all_species_expr_df, long_scaled_input_df)
}

##############################
########## FGF17 #############
##############################
my_OG_ID = "GF_005320"
tissue = "Neural"
my_species = "Hs2"

tissue_tau_df = read.delim(paste0(ts_gains_losses_dir, my_version, "/Bilateria_conserved_orthogroups-All_genes-", tissue, "_specific_OGs.tab"), header=FALSE, col.names=c("OG_ID", "Species", "GeneID", "Tau", "Tissue", "Expr_cutoff"))

my_OG_ID_genes = as.vector(subset(tissue_tau_df, OG_ID == my_OG_ID)$GeneID)
selected_expr_df = subset(all_species_expr_df, GeneID %in% my_OG_ID_genes)
selected_expr_df$Category = rep("NO_TS", nrow(selected_expr_df))
selected_expr_df$Category[selected_expr_df$Species == my_species] = "TS"
selected_expr_df$Tissue = factor(selected_expr_df$Tissue, levels=ordered_tissues)

#Get query species dotplot
fgf17_dotplot = ggplot(data=subset(selected_expr_df, Category=="TS"), aes(x=Tissue, y=Scaled_expr, color=Tissue, fill=Tissue)) +
  geom_point(shape=21, alpha=0.8, size=4) +
  scale_fill_manual(values=my_color_vector) +
  scale_color_manual(values=my_color_vector) +
  theme_bw() +
  theme(axis.text.x = element_blank(),
        axis.text.y = element_text(color="black"),
        axis.ticks.x = element_blank(),
        axis.title = element_blank()
        #axis.title.y = element_text(color="black")
        ) +
  coord_cartesian(ylim = c(0,4.5)) +
  scale_y_continuous(position="right") +
  guides(fill="none", color="none")
print(fgf17_dotplot)

#Get other species boxplots
fgf17_boxplots = ggplot(data=subset(selected_expr_df, Category!="TS"), aes(x=Tissue, y=Scaled_expr, fill=Tissue)) +
  geom_boxplot(alpha=0.8, color="black", outlier.size = 0.5, stroke=0.8) +
  theme_bw() +
  theme(#axis.title.y = element_text(color="black"),
        axis.title = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y = element_text(color="black"),
        axis.text.x = element_blank(),
        plot.title = element_text(color="black", face="bold")) +
  scale_fill_manual(values=my_color_vector) +
  #ggtitle(paste0(ancestor)) +
  #scale_y_continuous(limits=c(0,1)) +
  coord_cartesian(ylim = c(0,4.5)) +
  scale_y_continuous(position="right") +
  guides(fill=FALSE)
print(fgf17_boxplots)

panel_Q = plot_grid(fgf17_dotplot, fgf17_boxplots, ncol=2, rel_widths = c(1,1), align="h")

pdf("/Users/federica/Documents/CRG_PhD/bilaterian_GE/Images/Images_from_R/new_Fig5_v1/Fig5_panelQ.pdf", width=4, height=3)
print(panel_Q)
dev.off()

#egg::ggarrange(query_species_dotplot, no_query_expr_plot, ncol=2, widths = c(1,1))
#final_clade_expr_plot = clade_expr_plot + facet_nested_wrap(~Query_tissue+Category, labeller = #labeller(Query_tissue=Query_tissue_number), drop=FALSE, nrow=2
```

```{r fgf17_expr_raw_TPMs, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions"), eval=FALSE, fig.height=3, fig.width=7}
#### Build expression table: all genes from all species across all tissues
all_species_expr_df = vector()
for (species in ordered_species) {
  input_file = paste0(home, "projects/bilaterian_GE/data/preprocessing_all/norm_counts_quant/tissues_average_quantification/", species, "-tissue_average_expr-NOSVA-NOlog2-TPMs.tab")
  input_df = read.delim(input_file, header=TRUE, row=1)
  scaled_input_df = input_df
  ######
  long_scaled_input_df = melt(as.matrix(scaled_input_df))
  colnames(long_scaled_input_df) = c("GeneID", "Tissue", "Scaled_expr")
  long_scaled_input_df$Species = rep(species, nrow(long_scaled_input_df))
  all_species_expr_df = rbind(all_species_expr_df, long_scaled_input_df)
}

##############################
########## FGF17 #############
##############################
my_OG_ID = "GF_005320"
tissue = "Neural"
my_species = "Hs2"

tissue_tau_df = read.delim(paste0(ts_gains_losses_dir, my_version, "/Bilateria_conserved_orthogroups-All_genes-", tissue, "_specific_OGs.tab"), header=FALSE, col.names=c("OG_ID", "Species", "GeneID", "Tau", "Tissue", "Expr_cutoff"))

my_OG_ID_genes = as.vector(subset(tissue_tau_df, OG_ID == my_OG_ID)$GeneID)
selected_expr_df = subset(all_species_expr_df, GeneID %in% my_OG_ID_genes)
selected_expr_df$Category = rep("NO_TS", nrow(selected_expr_df))
selected_expr_df$Category[selected_expr_df$Species == my_species] = "TS"
selected_expr_df$Tissue = factor(selected_expr_df$Tissue, levels=ordered_tissues)

#Get query species dotplot
#Get query species dotplot
fgf17_dotplot = ggplot(data=subset(selected_expr_df, Category=="TS"), aes(x=Tissue, y=Scaled_expr, color=Tissue, fill=Tissue)) +
  geom_point(shape=21, alpha=0.8, size=4) +
  scale_fill_manual(values=my_color_vector) +
  scale_color_manual(values=my_color_vector) +
  theme_bw() +
  theme(axis.text.x = element_blank(),
        axis.text.y = element_text(color="black"),
        axis.ticks.x = element_blank(),
        axis.title = element_blank()
        #axis.title.y = element_text(color="black")
        ) +
  coord_cartesian(ylim = c(0,35)) +
  scale_y_continuous(position="right") +
  guides(fill="none", color="none")
print(fgf17_dotplot)

#Get other species boxplots
fgf17_boxplots = ggplot(data=subset(selected_expr_df, Category!="TS" & Species %in% Deuterostoma), aes(x=Tissue, y=Scaled_expr, fill=Tissue)) +
  geom_boxplot(alpha=0.8, color="black", outlier.size = 0.5, stroke=0.8) +
  theme_bw() +
  theme(#axis.title.y = element_text(color="black"),
        axis.title = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.y = element_text(color="black"),
        axis.text.x = element_blank(),
        plot.title = element_text(color="black", face="bold")) +
  scale_fill_manual(values=my_color_vector) +
  #ggtitle(paste0(ancestor)) +
  #scale_y_continuous(limits=c(0,1)) +
  coord_cartesian(ylim = c(0,35)) +
  scale_y_continuous(position="right") +
  guides(fill=FALSE)
print(fgf17_boxplots)

panel_Q = plot_grid(fgf17_dotplot, fgf17_boxplots, ncol=2, rel_widths = c(1,1), align="h")
print(panel_Q)

pdf("/Users/federica/Documents/CRG_PhD/bilaterian_GE/Images/Images_from_R/new_Fig5_v1/Fig5_panelQ-TPM_deut_expr.pdf", width=4, height=3)
print(panel_Q)
dev.off()

#egg::ggarrange(query_species_dotplot, no_query_expr_plot, ncol=2, widths = c(1,1))
#final_clade_expr_plot = clade_expr_plot + facet_nested_wrap(~Query_tissue+Category, labeller = #labeller(Query_tissue=Query_tissue_number), drop=FALSE, nrow=2
```




<!-- Backup -->
```{r GO_enrichments_multiple_all_tissues_Hs2_annot_observed_vs_expected, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions"), fig.width=10, fig.height=5, eval=FALSE}
####### Upload GO transfer df
GO_transfers_df = read.delim(paste0(GO_transfers_dir, my_version, "/GO_annot/Hs2_transfers/orthogroups_from_Hs2-transferred_GO-reduced.txt"),
                             header=FALSE, col.names=c("OG_ID", "GO_term", "GO_name"))
### Generate id to term dictionary
name_id_transfers_df = unique(GO_transfers_df[,c("GO_term", "GO_name")])
GO_ID_term_dict = hashmap(as.vector(name_id_transfers_df$GO_term), as.vector(name_id_transfers_df$GO_name))

###### Upload TS-gains df
#### Upload gains info
tissue_gains_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, my_version, "/inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  input_df$Node_Tissue = paste0(input_df$TS_gains, "_", input_df$Tissue)
  tissue_gains_df = rbind(tissue_gains_df, input_df)
}

##### For each species and each ancestors, get the total number of gains (to use for the proportion computation)
tissue_gains_counts = table(as.vector(tissue_gains_df$Node_Tissue))
all_combinations_df = expand.grid(c(all_ancestors, ordered_species), ordered_tissues)
all_combinations = paste0(all_combinations_df$Var1, "_", all_combinations_df$Var2)
missing_IDs = all_combinations[!(all_combinations %in% names(tissue_gains_counts))]
if (length(missing_IDs) > 0) {
  for (missing_ID in missing_IDs) {
    tissue_gains_counts = c(tissue_gains_counts, 0)
    names(tissue_gains_counts)[length(tissue_gains_counts)] = missing_ID
  }
}
tissue_gains_counts_dict = hashmap(names(tissue_gains_counts), as.vector(tissue_gains_counts))

##### Cycle on all tissues
selected_ids = vector()
annotation_row_df = vector()
for (tissue in ordered_tissues) {
  final_GO_res_df = vector()
  for (node in ordered_species)) {
    for (inference_type in c("gains")) {
      GO_res_file = paste0(ts_gains_losses_dir, my_version, "/inferences/GO_enrichments/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_", inference_type, "-", node, "-orthogroups_from_Hs2-GO_res.tab")
      if (!file.exists(GO_res_file) | file.size(GO_res_file) == 1 | is.na(file.size(GO_res_file))) { #If the file does not exists
        next
      }
      else { #if file exists
        GO_res_df = read.delim(GO_res_file, header=TRUE)
        #Select entries with at least 5 genes in the set.
        GO_res_df = subset(GO_res_df, significant=="TRUE" & intersection_size>=5)
        #If there are no significant entries, continue
        if (nrow(GO_res_df)==0) {
          next
        }
        #Modify term names
        #GO_res_df$term_name = str_trunc(firstup(gsub("_", " ", GO_res_df$term_name)), 75, side="right", ellipsis = "...")
        #Add tissue
        GO_res_df$node_tissue_inference_type = rep(paste0(tissue, "_", node), nrow(GO_res_df))
        #Add to final df
        final_GO_res_df = rbind(final_GO_res_df, GO_res_df)
      }
    }
  }
  heatmap_input_df = final_GO_res_df[,c("term_name", "term_id", "node_tissue_inference_type")]
  heatmap_input_df$significant = rep(1, nrow(heatmap_input_df))
  ### Generate wide dataframe
  heatmap_input_wide_tibble = heatmap_input_df %>% pivot_wider(names_from="node_tissue_inference_type", values_from="significant", values_fill=0)
  heatmap_input_wide_df = as.data.frame(heatmap_input_wide_tibble)
  rownames(heatmap_input_wide_df) = heatmap_input_wide_df$term_name
  heatmap_input_wide_df$term_name = NULL
  heatmap_input_wide_df[is.na(heatmap_input_wide_df)] = 0
  
  #Select categories significant in at least two nodes
  selected_terms = names(table(heatmap_input_df$term_id)[table(heatmap_input_df$term_id)>3])
  #tissue_selected_ids = unique(as.vector(subset(heatmap_input_df, term_name %in% selected_terms)$term_id))
  tissue_selected_ids = selected_terms
  selected_ids = c(selected_ids, tissue_selected_ids)
  annotation_row_df = rbind(annotation_row_df, data.frame(GO_term=tissue_selected_ids, GO_class=rep(tissue, length(tissue_selected_ids))))
}


#### get annotation row_df
annotation_row_df$GO_class = as.vector(annotation_row_df$GO_class)
annotation_row_df$GO_term = as.vector(annotation_row_df$GO_term)
annotation_row_df$GO_class[annotation_row_df$GO_term %in% selected_ids[duplicated(selected_ids)]] = "Multiple"
annotation_row_df = unique(annotation_row_df)

selected_ids = selected_ids[!(duplicated(selected_ids))] #Remove extra copies of the duplicated ids

#### Cycle on the selected categories, and compute the proportion of annotated gains
selected_GO_transfer_df = subset(GO_transfers_df, GO_term %in% selected_ids)
selected_GO_transfer_df$GO_term = as.vector(selected_GO_transfer_df$GO_term)
selected_GO_transfer_df$OG_ID = as.vector(selected_GO_transfer_df$OG_ID)

##### For each selected IDs, compute proportion of genes belonging to each category over the total numbers of genes with a functional association
#This will be used to compute the expected number of gains in each category, if these gains were random.
genes_with_annot = unique(as.vector(GO_transfers_df$OG_ID))
#Subset for the ones that among the bilaterian conserved
all_orthogroups_df = read.delim(paste0(gene_sets_dir, my_version, "/STRICT/Bilateria/conserved/Bilateria_conserved_orthogroups-EXPR_genes.txt"),
                                 header=FALSE, col.names = c("OG_ID", "Species", "GeneID"))
bilaterian_conserved_OGs = unique(as.vector(all_orthogroups_df$OG_ID))
Tot_genes_with_annot = length(genes_with_annot[genes_with_annot %in% bilaterian_conserved_OGs])

category_proportion_df = as.data.frame(table(selected_GO_transfer_df$GO_term))
colnames(category_proportion_df) = c("GO_term", "Freq")
category_proportion_df$proportion = category_proportion_df$Freq/Tot_genes_with_annot
category_proportion_dict = hashmap(as.vector(category_proportion_df$GO_term), as.vector(category_proportion_df$proportion))

#### For each node
all_node_category_proportion_df = vector()
for (node in all_combinations) {
  node_gains_OGs = as.vector(subset(tissue_gains_df, Node_Tissue==node)$OG_ID) #Select all orthogroups for that node_tissue
  node_gains_GO_transfer_count_df = as.data.frame(table(as.vector(subset(selected_GO_transfer_df, GO_term %in% selected_ids & OG_ID %in% node_gains_OGs)$GO_term)))
  if (nrow(node_gains_GO_transfer_count_df) > 0) { #If there is at least one gain for that node
    colnames(node_gains_GO_transfer_count_df) = c("GO_term", "num") 
    #### Add zero counts for missing categories
    missing_categories = selected_ids[!(selected_ids %in% as.vector(node_gains_GO_transfer_count_df$GO_term))]
    if (length(missing_categories) > 0) {node_gains_GO_transfer_count_df = rbind(node_gains_GO_transfer_count_df, 
                                                                                 data.frame("GO_term"=missing_categories,
                                                                                            num=rep(0,length(missing_categories))))}
    } else { #If there are no gains for that node, add 0 for ALL categories
      node_gains_GO_transfer_count_df = data.frame("GO_term"=selected_ids, num=rep(0, length(selected_ids)))
    }
  #Add the total number of gains for each node
  node_gains_GO_transfer_count_df$background = rep(tissue_gains_counts_dict$find(node), nrow(node_gains_GO_transfer_count_df))
  #Compute the expected distributtion for each category (number of gains in the nodes multiplied by the total proportion of genes in that category)
  node_gains_GO_transfer_count_df$expected = node_gains_GO_transfer_count_df$background * category_proportion_dict$find(node_gains_GO_transfer_count_df$GO_term)
  node_gains_GO_transfer_count_df$GO_name = GO_ID_term_dict$find(node_gains_GO_transfer_count_df$GO_term)
  node_gains_GO_transfer_count_df$Node = rep(node, nrow(node_gains_GO_transfer_count_df))
  #Join into final dataframe
  all_node_category_proportion_df = rbind(all_node_category_proportion_df, node_gains_GO_transfer_count_df)
}
###### Get annotation row df for heatmap annotation
rownames(annotation_row_df) = annotation_row_df$GO_term
annotation_row_df$GO_term = NULL

#Compute observed vs expected ratio
all_node_category_proportion_df$proportion = log2((all_node_category_proportion_df$num / all_node_category_proportion_df$expected)+1)
#Order levels for plotting
all_node_category_proportion_df$Node = factor(all_node_category_proportion_df$Node, levels=all_combinations)
all_node_category_proportion_df$GO_term = factor(all_node_category_proportion_df$GO_term, selected_ids)

#If the total number of gains is lower than 10, set to 0.
#all_node_category_proportion_df$proportion[all_node_category_proportion_df$num <= 2 & all_node_category_proportion_df$background <= 10] = 0
#all_node_category_proportion_df$proportion[all_node_category_proportion_df$num < 5 & all_node_category_proportion_df$background <= 10] = 0

wider_all_node_category_proportion_df = all_node_category_proportion_df %>%
  pivot_wider(GO_term, names_from = Node, values_from=proportion) %>%
  column_to_rownames("GO_term")

wider_all_node_category_proportion_df = wider_all_node_category_proportion_df[selected_ids,]


###### Get annotation column df for heatmap annotation
annotation_col_df = data.frame(Tissues=sub(".*_", "", colnames(wider_all_node_category_proportion_df)))
rownames(annotation_col_df) = colnames(wider_all_node_category_proportion_df)
annotation_col_df$Ancestrality = rep("Ancestral", nrow(annotation_col_df))
annotation_col_df$Ancestrality[grep(paste(ordered_species, collapse="|"), rownames(annotation_col_df))] = "Species_specific"


#Filter for at least one enrichment >=3 in the row
wider_all_node_category_proportion_df = wider_all_node_category_proportion_df[apply(wider_all_node_category_proportion_df, 1, function(x) {max(x, na.rm = TRUE) >= 3}),]
rownames(wider_all_node_category_proportion_df) = GO_ID_term_dict$find(rownames(wider_all_node_category_proportion_df))
rownames(annotation_row_df) = GO_ID_term_dict$find(rownames(annotation_row_df))

wider_all_node_category_proportion_df = wider_all_node_category_proportion_df[,colSums(is.na(wider_all_node_category_proportion_df))<nrow(wider_all_node_category_proportion_df)]
GO_res_heatmap = pheatmap(wider_all_node_category_proportion_df,
                       color = c(brewer.pal(9, "PuRd")[1],brewer.pal(9, "PuRd")[3:9]),
                       breaks = c(0,1.2,2,2.5,3,4,5,6),
                       border_color = "black",
                       main = paste0("Recurrent gains in functional categories"),
                       cluster_cols = FALSE,
                       cluster_rows = TRUE,
                       annotation_col = annotation_col_df,
                       annotation_row = subsetted_annotation_row_df,
                       annotation_colors = list(Tissues = my_color_vector, 
                                                #Branch = my_branch_colors, 
                                                Ancestrality = my_ancestral_colors, 
                                                Recurrent_tissue=c("Most_tissues"="black", "Non_germiline"="gray82", "Germline"="mediumorchid2")),
                       annotation_legend = TRUE, 
                       show_rownames = TRUE,
                       show_colnames = FALSE,
                       #silent = TRUE,
                       #legend = FALSE,
                       #angle_col = 45,
                       #fontsize_col = 8,
                       fontsize_row = 8,
                       angle_col = 0
                       )

#### order columns based on decided order
ordered_columns = c("sequence-specific_double-stranded_DNA_binding",
                    #"DNA-binding_transcription_factor_activity,_RNA_polymerase_II-specific",
                    "sensory_organ_development",
                    "metal_ion_transmembrane_transporter_activity",
                    "integral_component_of_plasma_membrane",
                    "cell_surface",
                    #"extracellular_space",
                    "meiotic_nuclear_division", 
                    "chromosome_organization_involved_in_meiotic_cell_cycle")


wider_all_node_category_proportion_df = wider_all_node_category_proportion_df[ordered_columns,]
subsetted_annotation_row_df = data.frame(Recurrent_tissue = c(rep("Most_tissues", 3),
                                                      rep("Non_germiline", 2),
                                                      rep("Germline", 2)))
rownames(subsetted_annotation_row_df) = ordered_columns
                                         
#pdf("/Users/federica/Documents/CRG_PhD/bilaterian_GE/Images/Images_from_R/new_Fig5/Fig5_panelJ.pdf", width=10, height=5)
GO_res_heatmap = pheatmap(wider_all_node_category_proportion_df,
                       color = c(brewer.pal(9, "PuRd")[1],brewer.pal(9, "PuRd")[3:9]),
                       breaks = c(0,1.2,2,2.5,3,4,5,6),
                       border_color = "black",
                       main = paste0("Recurrent gains in functional categories"),
                       cluster_cols = FALSE,
                       cluster_rows = FALSE,
                       annotation_col = annotation_col_df,
                       annotation_row = subsetted_annotation_row_df,
                       annotation_colors = list(Tissues = my_color_vector, 
                                                #Branch = my_branch_colors, 
                                                Ancestrality = my_ancestral_colors, 
                                                Recurrent_tissue=c("Most_tissues"="black", "Non_germiline"="gray82", "Germline"="mediumorchid2")),
                       annotation_legend = TRUE, 
                       show_rownames = TRUE,
                       show_colnames = FALSE,
                       #silent = TRUE,
                       #legend = FALSE,
                       #angle_col = 45,
                       #fontsize_col = 8,
                       fontsize_row = 8,
                       angle_col = 0
                       )
#dev.off()
```

```{r GO_enrichments_multiple_all_tissues_Hs2_annot_observed_vs_expected_v1, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions"), fig.width=10, fig.height=5, eval=FALSE}
####### Upload GO transfer df
GO_transfers_df = read.delim(paste0(GO_transfers_dir, my_version, "/GO_annot/Hs2_transfers/orthogroups_from_Hs2-transferred_GO-reduced.txt"),
                             header=FALSE, col.names=c("OG_ID", "GO_term", "GO_name"))
### Generate id to term dictionary
name_id_transfers_df = unique(GO_transfers_df[,c("GO_term", "GO_name")])
GO_ID_term_dict = hashmap(as.vector(name_id_transfers_df$GO_term), as.vector(name_id_transfers_df$GO_name))

###### Upload TS-gains df
#### Upload gains info
tissue_gains_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, my_version, "/inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  input_df$Node_Tissue = paste0(input_df$TS_gains, "_", input_df$Tissue)
  tissue_gains_df = rbind(tissue_gains_df, input_df)
}

##### For each species and each ancestors, get the total number of gains (to use for the proportion computation)
tissue_gains_counts = table(as.vector(tissue_gains_df$Node_Tissue))
all_combinations_df = expand.grid(c(all_ancestors, ordered_species), ordered_tissues)
all_combinations = paste0(all_combinations_df$Var1, "_", all_combinations_df$Var2)
missing_IDs = all_combinations[!(all_combinations %in% names(tissue_gains_counts))]
if (length(missing_IDs) > 0) {
  for (missing_ID in missing_IDs) {
    tissue_gains_counts = c(tissue_gains_counts, 0)
    names(tissue_gains_counts)[length(tissue_gains_counts)] = missing_ID
  }
}
tissue_gains_counts_dict = hashmap(names(tissue_gains_counts), as.vector(tissue_gains_counts))

##### Cycle on all tissues
selected_ids = vector()
annotation_row_df = vector()
for (tissue in ordered_tissues) {
  final_GO_res_df = vector()
  for (node in ordered_species) {
    for (inference_type in c("gains")) {
      GO_res_file = paste0(ts_gains_losses_dir, my_version, "/inferences/GO_enrichments/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_", inference_type, "-", node, "-orthogroups_from_Hs2-GO_res.tab")
      if (!file.exists(GO_res_file) | file.size(GO_res_file) == 1 | is.na(file.size(GO_res_file))) { #If the file does not exists
        next
      }
      else { #if file exists
        GO_res_df = read.delim(GO_res_file, header=TRUE)
        #Select entries with at least 5 genes in the set.
        GO_res_df = subset(GO_res_df, significant=="TRUE" & intersection_size>=5)
        #If there are no significant entries, continue
        if (nrow(GO_res_df)==0) {
          next
        }
        #Modify term names
        #GO_res_df$term_name = str_trunc(firstup(gsub("_", " ", GO_res_df$term_name)), 75, side="right", ellipsis = "...")
        #Add tissue
        GO_res_df$node_tissue_inference_type = rep(paste0(tissue, "_", node), nrow(GO_res_df))
        #Add to final df
        final_GO_res_df = rbind(final_GO_res_df, GO_res_df)
      }
    }
  }
  heatmap_input_df = final_GO_res_df[,c("term_name", "term_id", "node_tissue_inference_type")]
  heatmap_input_df$significant = rep(1, nrow(heatmap_input_df))
  ### Generate wide dataframe
  heatmap_input_wide_tibble = heatmap_input_df %>% pivot_wider(names_from="node_tissue_inference_type", values_from="significant", values_fill=0)
  heatmap_input_wide_df = as.data.frame(heatmap_input_wide_tibble)
  rownames(heatmap_input_wide_df) = heatmap_input_wide_df$term_name
  heatmap_input_wide_df$term_name = NULL
  heatmap_input_wide_df[is.na(heatmap_input_wide_df)] = 0
  
  #Select categories significant in at least three species
  selected_terms = names(table(heatmap_input_df$term_id)[table(heatmap_input_df$term_id)>=3])
  #tissue_selected_ids = unique(as.vector(subset(heatmap_input_df, term_name %in% selected_terms)$term_id))
  tissue_selected_ids = selected_terms
  selected_ids = c(selected_ids, tissue_selected_ids)
  annotation_row_df = rbind(annotation_row_df, data.frame(GO_term=tissue_selected_ids, GO_class=rep(tissue, length(tissue_selected_ids))))
}


#### get annotation row_df
annotation_row_df$GO_class = as.vector(annotation_row_df$GO_class)
annotation_row_df$GO_term = as.vector(annotation_row_df$GO_term)
annotation_row_df$GO_class[annotation_row_df$GO_term %in% selected_ids[duplicated(selected_ids)]] = "Multiple"
annotation_row_df = unique(annotation_row_df)

selected_ids = unique(selected_ids) #Remove extra copies of the duplicated ids

#### Cycle on the selected categories, and compute the proportion of annotated gains
selected_GO_transfer_df = subset(GO_transfers_df, GO_term %in% selected_ids)
selected_GO_transfer_df$GO_term = as.vector(selected_GO_transfer_df$GO_term)
selected_GO_transfer_df$OG_ID = as.vector(selected_GO_transfer_df$OG_ID)

##### For each selected IDs, compute proportion of genes belonging to each category over the total numbers of genes with a functional association
#This will be used to compute the expected number of gains in each category, if these gains were random.
genes_with_annot = unique(as.vector(GO_transfers_df$OG_ID))
#Subset for the ones that among the bilaterian conserved
all_orthogroups_df = read.delim(paste0(gene_sets_dir, my_version, "/STRICT/Bilateria/conserved/Bilateria_conserved_orthogroups-EXPR_genes.txt"),
                                 header=FALSE, col.names = c("OG_ID", "Species", "GeneID"))
bilaterian_conserved_OGs = unique(as.vector(all_orthogroups_df$OG_ID))
Tot_genes_with_annot = length(genes_with_annot[genes_with_annot %in% bilaterian_conserved_OGs])

category_proportion_df = as.data.frame(table(selected_GO_transfer_df$GO_term))
colnames(category_proportion_df) = c("GO_term", "Freq")
category_proportion_df$proportion = category_proportion_df$Freq/Tot_genes_with_annot
category_proportion_dict = hashmap(as.vector(category_proportion_df$GO_term), as.vector(category_proportion_df$proportion))

#### For each node
all_node_category_proportion_df = vector()
for (node in all_combinations) {
  node_gains_OGs = as.vector(subset(tissue_gains_df, Node_Tissue==node)$OG_ID) #Select all orthogroups for that node_tissue
  node_gains_GO_transfer_count_df = as.data.frame(table(as.vector(subset(selected_GO_transfer_df, GO_term %in% selected_ids & OG_ID %in% node_gains_OGs)$GO_term)))
  if (nrow(node_gains_GO_transfer_count_df) > 0) { #If there is at least one gain for that node
    colnames(node_gains_GO_transfer_count_df) = c("GO_term", "num") 
    #### Add zero counts for missing categories
    missing_categories = selected_ids[!(selected_ids %in% as.vector(node_gains_GO_transfer_count_df$GO_term))]
    if (length(missing_categories) > 0) {node_gains_GO_transfer_count_df = rbind(node_gains_GO_transfer_count_df, 
                                                                                 data.frame("GO_term"=missing_categories,
                                                                                            num=rep(0,length(missing_categories))))}
    } else { #If there are no gains for that node, add 0 for ALL categories
      node_gains_GO_transfer_count_df = data.frame("GO_term"=selected_ids, num=rep(0, length(selected_ids)))
    }
  #Add the total number of gains for each node
  node_gains_GO_transfer_count_df$background = rep(tissue_gains_counts_dict$find(node), nrow(node_gains_GO_transfer_count_df))
  #Compute the expected distributtion for each category (number of gains in the nodes multiplied by the total proportion of genes in that category)
  node_gains_GO_transfer_count_df$expected = node_gains_GO_transfer_count_df$background * category_proportion_dict$find(node_gains_GO_transfer_count_df$GO_term)
  node_gains_GO_transfer_count_df$GO_name = GO_ID_term_dict$find(node_gains_GO_transfer_count_df$GO_term)
  node_gains_GO_transfer_count_df$Node = rep(node, nrow(node_gains_GO_transfer_count_df))
  #Join into final dataframe
  all_node_category_proportion_df = rbind(all_node_category_proportion_df, node_gains_GO_transfer_count_df)
}
###### Get annotation row df for heatmap annotation
rownames(annotation_row_df) = annotation_row_df$GO_term
annotation_row_df$GO_term = NULL

#Compute observed vs expected ratio
all_node_category_proportion_df$proportion = log2((all_node_category_proportion_df$num / all_node_category_proportion_df$expected)+1)
#Order levels for plotting
all_node_category_proportion_df$Node = factor(all_node_category_proportion_df$Node, levels=all_combinations)
all_node_category_proportion_df$GO_term = factor(all_node_category_proportion_df$GO_term, selected_ids)

#If the total number of gains is lower than 10, set to 0.
#all_node_category_proportion_df$proportion[all_node_category_proportion_df$num <= 2 & all_node_category_proportion_df$background <= 10] = 0
#all_node_category_proportion_df$proportion[all_node_category_proportion_df$num < 5 & all_node_category_proportion_df$background <= 10] = 0

wider_all_node_category_proportion_df = all_node_category_proportion_df %>%
  pivot_wider(GO_term, names_from = Node, values_from=proportion) %>%
  column_to_rownames("GO_term")

wider_all_node_category_proportion_df = wider_all_node_category_proportion_df[selected_ids,]


###### Get annotation column df for heatmap annotation
annotation_col_df = data.frame(Tissues=sub(".*_", "", colnames(wider_all_node_category_proportion_df)))
rownames(annotation_col_df) = colnames(wider_all_node_category_proportion_df)
annotation_col_df$Ancestrality = rep("Ancestral", nrow(annotation_col_df))
annotation_col_df$Ancestrality[grep(paste(ordered_species, collapse="|"), rownames(annotation_col_df))] = "Species_specific"


#Filter for at least one enrichment >=3 in the row
wider_all_node_category_proportion_df = wider_all_node_category_proportion_df[apply(wider_all_node_category_proportion_df, 1, function(x) {max(x, na.rm = TRUE) >= 3}),]
rownames(wider_all_node_category_proportion_df) = GO_ID_term_dict$find(rownames(wider_all_node_category_proportion_df))
rownames(annotation_row_df) = GO_ID_term_dict$find(rownames(annotation_row_df))

wider_all_node_category_proportion_df = wider_all_node_category_proportion_df[,colSums(is.na(wider_all_node_category_proportion_df))<nrow(wider_all_node_category_proportion_df)]
GO_res_heatmap = pheatmap(wider_all_node_category_proportion_df,
                       color = c(brewer.pal(9, "PuRd")[1],brewer.pal(9, "PuRd")[3:9]),
                       breaks = c(0,1.2,2,2.5,3,4,5,6),
                       border_color = "black",
                       main = paste0("Recurrent gains in functional categories"),
                       cluster_cols = FALSE,
                       cluster_rows = TRUE,
                       annotation_col = annotation_col_df,
                       annotation_row = subsetted_annotation_row_df,
                       annotation_colors = list(Tissues = my_color_vector, 
                                                #Branch = my_branch_colors, 
                                                Ancestrality = my_ancestral_colors, 
                                                Recurrent_tissue=c("Most_tissues"="black", "Non_germiline"="gray82", "Germline"="mediumorchid2")),
                       annotation_legend = TRUE, 
                       show_rownames = TRUE,
                       show_colnames = FALSE,
                       #silent = TRUE,
                       #legend = FALSE,
                       #angle_col = 45,
                       #fontsize_col = 8,
                       fontsize_row = 8,
                       angle_col = 0
                       )

#### order columns based on decided order
ordered_columns = c("sequence-specific_double-stranded_DNA_binding",
                    #"DNA-binding_transcription_factor_activity,_RNA_polymerase_II-specific",
                    "sensory_organ_development",
                    "metal_ion_transmembrane_transporter_activity",
                    "integral_component_of_plasma_membrane",
                    "cell_surface",
                    #"extracellular_space",
                    "meiotic_nuclear_division", 
                    "chromosome_organization_involved_in_meiotic_cell_cycle")


wider_all_node_category_proportion_df = wider_all_node_category_proportion_df[ordered_columns,]
subsetted_annotation_row_df = data.frame(Recurrent_tissue = c(rep("Most_tissues", 3),
                                                      rep("Non_germiline", 2),
                                                      rep("Germline", 2)))
rownames(subsetted_annotation_row_df) = ordered_columns
                                         
#pdf("/Users/federica/Documents/CRG_PhD/bilaterian_GE/Images/Images_from_R/new_Fig5/Fig5_panelJ.pdf", width=10, height=5)
GO_res_heatmap = pheatmap(wider_all_node_category_proportion_df,
                       color = c(brewer.pal(9, "PuRd")[1],brewer.pal(9, "PuRd")[3:9]),
                       breaks = c(0,1.2,2,2.5,3,4,5,6),
                       border_color = "black",
                       main = paste0("Recurrent gains in functional categories"),
                       cluster_cols = FALSE,
                       cluster_rows = FALSE,
                       annotation_col = annotation_col_df,
                       annotation_row = subsetted_annotation_row_df,
                       annotation_colors = list(Tissues = my_color_vector, 
                                                #Branch = my_branch_colors, 
                                                Ancestrality = my_ancestral_colors, 
                                                Recurrent_tissue=c("Most_tissues"="black", "Non_germiline"="gray82", "Germline"="mediumorchid2")),
                       annotation_legend = TRUE, 
                       show_rownames = TRUE,
                       show_colnames = FALSE,
                       #silent = TRUE,
                       #legend = FALSE,
                       #angle_col = 45,
                       #fontsize_col = 8,
                       fontsize_row = 8,
                       angle_col = 0
                       )
#dev.off()
```

```{r GO_enrichments_multiple_all_tissues_Hs2_annot_observed_vs_expected_v2, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions"), fig.width=10, fig.height=5, eval=FALSE}
####### Upload GO transfer df
GO_transfers_df = read.delim(paste0(GO_transfers_dir, my_version, "/GO_annot/Hs2_transfers/orthogroups_from_Hs2-transferred_GO-reduced.txt"),
                             header=FALSE, col.names=c("OG_ID", "GO_term", "GO_name"))
### Generate id to term dictionary
name_id_transfers_df = unique(GO_transfers_df[,c("GO_term", "GO_name")])
GO_ID_term_dict = hashmap(as.vector(name_id_transfers_df$GO_term), as.vector(name_id_transfers_df$GO_name))

###### Upload TS-gains df
#### Upload gains info
tissue_gains_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, my_version, "/inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  input_df$Node_Tissue = paste0(input_df$TS_gains, "_", input_df$Tissue)
  tissue_gains_df = rbind(tissue_gains_df, input_df)
}

##### For each species and each ancestors, get the total number of gains (to use for the proportion computation)
tissue_gains_counts = table(as.vector(tissue_gains_df$Node_Tissue))
all_combinations_df = expand.grid(c(all_ancestors, ordered_species), ordered_tissues)
all_combinations = paste0(all_combinations_df$Var1, "_", all_combinations_df$Var2)
missing_IDs = all_combinations[!(all_combinations %in% names(tissue_gains_counts))]
if (length(missing_IDs) > 0) {
  for (missing_ID in missing_IDs) {
    tissue_gains_counts = c(tissue_gains_counts, 0)
    names(tissue_gains_counts)[length(tissue_gains_counts)] = missing_ID
  }
}
tissue_gains_counts_dict = hashmap(names(tissue_gains_counts), as.vector(tissue_gains_counts))

##### Cycle on all tissues
selected_ids = vector()
annotation_row_df = vector()
for (tissue in ordered_tissues) {
  final_GO_res_df = vector()
  for (node in c(ordered_species, all_ancestors)) {
    for (inference_type in c("gains")) {
      GO_res_file = paste0(ts_gains_losses_dir, my_version, "/inferences/GO_enrichments/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_", inference_type, "-", node, "-orthogroups_from_Hs2-GO_res.tab")
      if (!file.exists(GO_res_file) | file.size(GO_res_file) == 1 | is.na(file.size(GO_res_file))) { #If the file does not exists
        next
      }
      else { #if file exists
        GO_res_df = read.delim(GO_res_file, header=TRUE)
        #Select entries with at least 5 genes in the set.
        GO_res_df = subset(GO_res_df, significant=="TRUE" & intersection_size>=5)
        #If there are no significant entries, continue
        if (nrow(GO_res_df)==0) {
          next
        }
        #Modify term names
        #GO_res_df$term_name = str_trunc(firstup(gsub("_", " ", GO_res_df$term_name)), 75, side="right", ellipsis = "...")
        #Add tissue
        GO_res_df$node_tissue_inference_type = rep(paste0(tissue, "_", node), nrow(GO_res_df))
        #Add to final df
        final_GO_res_df = rbind(final_GO_res_df, GO_res_df)
      }
    }
  }
  heatmap_input_df = final_GO_res_df[,c("term_name", "term_id", "node_tissue_inference_type")]
  heatmap_input_df$significant = rep(1, nrow(heatmap_input_df))
  ### Generate wide dataframe
  heatmap_input_wide_tibble = heatmap_input_df %>% pivot_wider(names_from="node_tissue_inference_type", values_from="significant", values_fill=0)
  heatmap_input_wide_df = as.data.frame(heatmap_input_wide_tibble)
  rownames(heatmap_input_wide_df) = heatmap_input_wide_df$term_name
  heatmap_input_wide_df$term_name = NULL
  heatmap_input_wide_df[is.na(heatmap_input_wide_df)] = 0
  
  #Select categories significant in at least three species
  selected_terms = names(table(heatmap_input_df$term_id)[table(heatmap_input_df$term_id)>=4])
  #tissue_selected_ids = unique(as.vector(subset(heatmap_input_df, term_name %in% selected_terms)$term_id))
  tissue_selected_ids = selected_terms
  selected_ids = c(selected_ids, tissue_selected_ids)
  annotation_row_df = rbind(annotation_row_df, data.frame(GO_term=tissue_selected_ids, GO_class=rep(tissue, length(tissue_selected_ids))))
}


#### get annotation row_df
annotation_row_df$GO_class = as.vector(annotation_row_df$GO_class)
annotation_row_df$GO_term = as.vector(annotation_row_df$GO_term)
annotation_row_df$GO_class[annotation_row_df$GO_term %in% selected_ids[duplicated(selected_ids)]] = "Multiple"
annotation_row_df = unique(annotation_row_df)

selected_ids = unique(selected_ids) #Remove extra copies of the duplicated ids

#### Cycle on the selected categories, and compute the proportion of annotated gains
selected_GO_transfer_df = subset(GO_transfers_df, GO_term %in% selected_ids)
selected_GO_transfer_df$GO_term = as.vector(selected_GO_transfer_df$GO_term)
selected_GO_transfer_df$OG_ID = as.vector(selected_GO_transfer_df$OG_ID)

##### For each selected IDs, compute proportion of genes belonging to each category over the total numbers of genes with a functional association
#This will be used to compute the expected number of gains in each category, if these gains were random.
genes_with_annot = unique(as.vector(GO_transfers_df$OG_ID))
#Subset for the ones that among the bilaterian conserved
all_orthogroups_df = read.delim(paste0(gene_sets_dir, my_version, "/STRICT/Bilateria/conserved/Bilateria_conserved_orthogroups-EXPR_genes.txt"),
                                 header=FALSE, col.names = c("OG_ID", "Species", "GeneID"))
bilaterian_conserved_OGs = unique(as.vector(all_orthogroups_df$OG_ID))
Tot_genes_with_annot = length(genes_with_annot[genes_with_annot %in% bilaterian_conserved_OGs])

category_proportion_df = as.data.frame(table(selected_GO_transfer_df$GO_term))
colnames(category_proportion_df) = c("GO_term", "Freq")
category_proportion_df$proportion = category_proportion_df$Freq/Tot_genes_with_annot
category_proportion_dict = hashmap(as.vector(category_proportion_df$GO_term), as.vector(category_proportion_df$proportion))

#### For each node
all_node_category_proportion_df = vector()
for (node in all_combinations) {
  node_gains_OGs = as.vector(subset(tissue_gains_df, Node_Tissue==node)$OG_ID) #Select all orthogroups for that node_tissue
  node_gains_GO_transfer_count_df = as.data.frame(table(as.vector(subset(selected_GO_transfer_df, GO_term %in% selected_ids & OG_ID %in% node_gains_OGs)$GO_term)))
  if (nrow(node_gains_GO_transfer_count_df) > 0) { #If there is at least one gain for that node
    colnames(node_gains_GO_transfer_count_df) = c("GO_term", "num") 
    #### Add zero counts for missing categories
    missing_categories = selected_ids[!(selected_ids %in% as.vector(node_gains_GO_transfer_count_df$GO_term))]
    if (length(missing_categories) > 0) {node_gains_GO_transfer_count_df = rbind(node_gains_GO_transfer_count_df, 
                                                                                 data.frame("GO_term"=missing_categories,
                                                                                            num=rep(0,length(missing_categories))))}
    } else { #If there are no gains for that node, add 0 for ALL categories
      node_gains_GO_transfer_count_df = data.frame("GO_term"=selected_ids, num=rep(0, length(selected_ids)))
    }
  #Add the total number of gains for each node
  node_gains_GO_transfer_count_df$background = rep(tissue_gains_counts_dict$find(node), nrow(node_gains_GO_transfer_count_df))
  #Compute the expected distributtion for each category (number of gains in the nodes multiplied by the total proportion of genes in that category)
  node_gains_GO_transfer_count_df$expected = node_gains_GO_transfer_count_df$background * category_proportion_dict$find(node_gains_GO_transfer_count_df$GO_term)
  node_gains_GO_transfer_count_df$GO_name = GO_ID_term_dict$find(node_gains_GO_transfer_count_df$GO_term)
  node_gains_GO_transfer_count_df$Node = rep(node, nrow(node_gains_GO_transfer_count_df))
  #Join into final dataframe
  all_node_category_proportion_df = rbind(all_node_category_proportion_df, node_gains_GO_transfer_count_df)
}
###### Get annotation row df for heatmap annotation
rownames(annotation_row_df) = annotation_row_df$GO_term
annotation_row_df$GO_term = NULL

#Compute observed vs expected ratio
all_node_category_proportion_df$proportion = log2((all_node_category_proportion_df$num / all_node_category_proportion_df$expected)+1)
#Order levels for plotting
all_node_category_proportion_df$Node = factor(all_node_category_proportion_df$Node, levels=all_combinations)
all_node_category_proportion_df$GO_term = factor(all_node_category_proportion_df$GO_term, selected_ids)

#If the total number of gains is lower than 10, set to 0.
#all_node_category_proportion_df$proportion[all_node_category_proportion_df$num <= 2 & all_node_category_proportion_df$background <= 10] = 0
#all_node_category_proportion_df$proportion[all_node_category_proportion_df$num < 5 & all_node_category_proportion_df$background <= 10] = 0

wider_all_node_category_proportion_df = all_node_category_proportion_df %>%
  pivot_wider(GO_term, names_from = Node, values_from=proportion) %>%
  column_to_rownames("GO_term")

wider_all_node_category_proportion_df = wider_all_node_category_proportion_df[selected_ids,]


###### Get annotation column df for heatmap annotation
annotation_col_df = data.frame(Tissues=sub(".*_", "", colnames(wider_all_node_category_proportion_df)))
rownames(annotation_col_df) = colnames(wider_all_node_category_proportion_df)
annotation_col_df$Ancestrality = rep("Ancestral", nrow(annotation_col_df))
annotation_col_df$Ancestrality[grep(paste(ordered_species, collapse="|"), rownames(annotation_col_df))] = "Species_specific"


#Filter for at least one enrichment >=3 in the row
wider_all_node_category_proportion_df = wider_all_node_category_proportion_df[apply(wider_all_node_category_proportion_df, 1, function(x) {max(x, na.rm = TRUE) >= 3}),]
rownames(wider_all_node_category_proportion_df) = GO_ID_term_dict$find(rownames(wider_all_node_category_proportion_df))
rownames(annotation_row_df) = GO_ID_term_dict$find(rownames(annotation_row_df))

wider_all_node_category_proportion_df = wider_all_node_category_proportion_df[,colSums(is.na(wider_all_node_category_proportion_df))<nrow(wider_all_node_category_proportion_df)]
GO_res_heatmap = pheatmap(wider_all_node_category_proportion_df,
                       color = c(brewer.pal(9, "PuRd")[1],brewer.pal(9, "PuRd")[3:9]),
                       breaks = c(0,1.2,2,2.5,3,4,5,6),
                       border_color = "black",
                       main = paste0("Recurrent gains in functional categories"),
                       cluster_cols = FALSE,
                       cluster_rows = TRUE,
                       annotation_col = annotation_col_df,
                       annotation_row = subsetted_annotation_row_df,
                       annotation_colors = list(Tissues = my_color_vector, 
                                                #Branch = my_branch_colors, 
                                                Ancestrality = my_ancestral_colors, 
                                                Recurrent_tissue=c("Most_tissues"="black", "Non_germiline"="gray82", "Germline"="mediumorchid2")),
                       annotation_legend = TRUE, 
                       show_rownames = TRUE,
                       show_colnames = FALSE,
                       #silent = TRUE,
                       #legend = FALSE,
                       #angle_col = 45,
                       #fontsize_col = 8,
                       fontsize_row = 8,
                       angle_col = 0
                       )

#### order columns based on decided order
ordered_columns = c("sequence-specific_double-stranded_DNA_binding",
                    #"DNA-binding_transcription_factor_activity,_RNA_polymerase_II-specific",
                    "sensory_organ_development",
                    "metal_ion_transmembrane_transporter_activity",
                    "integral_component_of_plasma_membrane",
                    "cell_surface",
                    #"extracellular_space",
                    "meiotic_nuclear_division", 
                    "chromosome_organization_involved_in_meiotic_cell_cycle")


wider_all_node_category_proportion_df = wider_all_node_category_proportion_df[ordered_columns,]
subsetted_annotation_row_df = data.frame(Recurrent_tissue = c(rep("Most_tissues", 3),
                                                      rep("Non_germiline", 2),
                                                      rep("Germline", 2)))
rownames(subsetted_annotation_row_df) = ordered_columns
                                         
#pdf("/Users/federica/Documents/CRG_PhD/bilaterian_GE/Images/Images_from_R/new_Fig5/Fig5_panelJ.pdf", width=10, height=5)
GO_res_heatmap = pheatmap(wider_all_node_category_proportion_df,
                       color = c(brewer.pal(9, "PuRd")[1],brewer.pal(9, "PuRd")[3:9]),
                       breaks = c(0,1.2,2,2.5,3,4,5,6),
                       border_color = "black",
                       main = paste0("Recurrent gains in functional categories"),
                       cluster_cols = FALSE,
                       cluster_rows = FALSE,
                       annotation_col = annotation_col_df,
                       annotation_row = subsetted_annotation_row_df,
                       annotation_colors = list(Tissues = my_color_vector, 
                                                #Branch = my_branch_colors, 
                                                Ancestrality = my_ancestral_colors, 
                                                Recurrent_tissue=c("Most_tissues"="black", "Non_germiline"="gray82", "Germline"="mediumorchid2")),
                       annotation_legend = TRUE, 
                       show_rownames = TRUE,
                       show_colnames = FALSE,
                       #silent = TRUE,
                       #legend = FALSE,
                       #angle_col = 45,
                       #fontsize_col = 8,
                       fontsize_row = 8,
                       angle_col = 0
                       )
#dev.off()
```

```{r repeated_GO_enrichments, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions"), fig.width=10, fig.height=5, eval=FALSE}
####### Upload GO transfer df
GO_transfers_df = read.delim(paste0(GO_transfers_dir, my_version, "/GO_annot/Hs2_transfers/orthogroups_from_Hs2-transferred_GO-reduced.txt"),
                             header=FALSE, col.names=c("OG_ID", "GO_term", "GO_name"))
### Generate dictionary from GO term to GO name
name_id_transfers_df = unique(GO_transfers_df[,c("GO_term", "GO_name")])
GO_ID_term_dict = hashmap(as.vector(name_id_transfers_df$GO_term), as.vector(name_id_transfers_df$GO_name))

###### Upload TS-gains df
tissue_gains_df = vector() #initiate dataframe
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, my_version, "/inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category")) #read file.
  input_df$Tissue = rep(tissue, nrow(input_df)) #add tissue info.
  input_df$Node_Tissue = paste0(input_df$TS_gains, "_", input_df$Tissue) #Add node_tissue info.
  tissue_gains_df = rbind(tissue_gains_df, input_df) #join to general dataframe,
}

##### For each species and each ancestors, get the total number of gains in each tissue (to use for the proportion computation).
tissue_gains_counts = table(as.vector(tissue_gains_df$Node_Tissue)) #Total number of gains in each tissue for each node/species.
all_combinations_df = expand.grid(c(all_ancestors, ordered_species), ordered_tissues) #Create a dataframe with col1=node/species and col2=tissue (all combinations).
all_combinations = paste0(all_combinations_df$Var1, "_", all_combinations_df$Var2) #Create unique variable combining node/species and tissue: node_tissue.
missing_IDs = all_combinations[!(all_combinations %in% names(tissue_gains_counts))] #Isolate all the nodes/species that do not have gains in a given tissue.
if (length(missing_IDs) > 0) { #if there are any such nodes, add zero counts for them to the tissue_gains dataframe
  for (missing_ID in missing_IDs) {
    tissue_gains_counts = c(tissue_gains_counts, 0)
    names(tissue_gains_counts)[length(tissue_gains_counts)] = missing_ID
  }
}
tissue_gains_counts_dict = hashmap(names(tissue_gains_counts), as.vector(tissue_gains_counts))

######################################################
##### Select significant categories ##################
######################################################
all_filtered_significant_terms = vector()
all_significant_terms = vector()
annotation_row_df = vector()
for (tissue in ordered_tissues) {
  #For each tissue, upload all GO results that respect some pre-defined cut-offs.
  final_GO_res_df = vector()
  for (node in c(ordered_species, all_ancestors)) {
    #Upload all GO results from the gain files.
    GO_res_file = paste0(ts_gains_losses_dir, my_version, "/inferences/GO_enrichments/Bilateria_conserved_orthogroups-All_genes-", 
                         tissue, "_inferred_gains-", node, "-orthogroups_from_Hs2-GO_res.tab")
    if (!file.exists(GO_res_file) | file.size(GO_res_file) == 1 | is.na(file.size(GO_res_file))) { #If the file does not exists
      next
    }
    else { #if file exists
      GO_res_df = read.delim(GO_res_file, header=TRUE)
      #Select entries with at least 5 genes in the set.
      GO_res_df = subset(GO_res_df, significant=="TRUE" & intersection_size>=5)
      #If there are no significant entries, continue
      if (nrow(GO_res_df)==0) {
        next
      }
      #Add tissue
      GO_res_df$node_tissue = rep(paste0(tissue, "_", node), nrow(GO_res_df))
      #Add to final df
      final_GO_res_df = rbind(final_GO_res_df, GO_res_df)
    }
  }
  #For each tissue, select the categories that are significant in at least X nodes/species
  heatmap_input_df = final_GO_res_df[,c("term_name", "term_id", "node_tissue")] #Subset to only the needed values
  heatmap_input_df$significant = rep(1, nrow(heatmap_input_df)) #Add value of 1 (it indicates that the category is significant)
  ##### Generate wide dataframe (rows=GO_term, cols=node_tissue)
  heatmap_input_wide_tibble = heatmap_input_df %>% pivot_wider(names_from="node_tissue", values_from="significant", values_fill=0)
  #Transform to dataframe and modify it so that we can work on it.
  heatmap_input_wide_df = as.data.frame(heatmap_input_wide_tibble)
  rownames(heatmap_input_wide_df) = heatmap_input_wide_df$term_name
  heatmap_input_wide_df$term_name = NULL
  heatmap_input_wide_df[is.na(heatmap_input_wide_df)] = 0 #Replace NA with 0
  
  #####Select all the GO categories significant in at least X nodes/species in that tissue
  filtered_significant_terms = names(table(heatmap_input_df$term_id)[table(heatmap_input_df$term_id)>=8])
  all_tissue_significant_terms = names(table(heatmap_input_df$term_id)[table(heatmap_input_df$term_id)>=1])
  #Add the filtered significant terms to the relative final vector
  all_filtered_significant_terms = c(all_filtered_significant_terms, filtered_significant_terms)
  #Add all significant terms to the relative final vector
  all_significant_terms = c(all_significant_terms, all_tissue_significant_terms)
  #annotation_row_df = rbind(annotation_row_df, data.frame(GO_term=tissue_selected_ids, GO_class=rep(tissue, length(tissue_selected_ids))))
}

### select only terms that are significant in only one tissue: if there are duplicated values, it is because they come from different tissues
non_repeated_significant_terms = all_significant_terms[!(duplicated(all_significant_terms)|duplicated(all_significant_terms, fromLast=TRUE))]
enriched_non_repeated_significant_terms = non_repeated_significant_terms[non_repeated_significant_terms %in% all_filtered_significant_terms]
selected_ids = enriched_non_repeated_significant_terms

#######################################################################
#### Compute EXPECTED propotions of annotated gains ##################
#######################################################################
selected_GO_transfer_df = subset(GO_transfers_df, GO_term %in% selected_ids)
selected_GO_transfer_df$GO_term = as.vector(selected_GO_transfer_df$GO_term)
selected_GO_transfer_df$OG_ID = as.vector(selected_GO_transfer_df$OG_ID)

##### For each selected IDs, compute proportion of genes belonging to each category 
#over the total numbers of genes that:
#1. have a functional association.
#2. are bilaterian conserved.
#This will be used to compute the expected number of gains in each category, if these gains were random.
genes_with_annot = unique(as.vector(GO_transfers_df$OG_ID))
#Uload info about bilaterian conserved
all_orthogroups_df = read.delim(paste0(gene_sets_dir, my_version, "/STRICT/Bilateria/conserved/Bilateria_conserved_orthogroups-EXPR_genes.txt"),
                                 header=FALSE, col.names = c("OG_ID", "Species", "GeneID"))
bilaterian_conserved_OGs = unique(as.vector(all_orthogroups_df$OG_ID))
#Count how many annotated genes are also bilaterian conserved.
conserved_genes_with_annot = genes_with_annot[genes_with_annot %in% bilaterian_conserved_OGs]
tot_genes_with_annot = length(conserved_genes_with_annot)

#Initiate dataframe to compute the expected proportion (the dataframe is already subsetted to the enriched categories).
category_proportion_df = as.data.frame(table(selected_GO_transfer_df$GO_term)) 
colnames(category_proportion_df) = c("GO_term", "Freq")
category_proportion_df$proportion = category_proportion_df$Freq/Tot_genes_with_annot
#Create dictionary with GO term to proportion correspondence.
category_proportion_dict = hashmap(as.vector(category_proportion_df$GO_term), as.vector(category_proportion_df$proportion))

#######################################################################
#### Compute OBSERVED propotions of annotated gains ###################
#######################################################################
all_node_category_proportion_df = vector()
#for each node/species_tissue
for (comb in all_combinations) {
  node_gains_OGs = as.vector(subset(tissue_gains_df, Node_Tissue==comb)$OG_ID) #Select all orthogroups with gains in that node_tissue.
  #Count how many gains in the selected categories for that node/species_tissue
  node_gains_GO_transfer_count_df = as.data.frame(table(as.vector(subset(selected_GO_transfer_df, GO_term %in% selected_ids & OG_ID %in% node_gains_OGs)$GO_term)))
  #### If there is at least one gain for that node/species_tissue in any of the categories.
  if (nrow(node_gains_GO_transfer_count_df) > 0) { 
    colnames(node_gains_GO_transfer_count_df) = c("GO_term", "num") #set colnames
    # Add zero counts for any missing categories
    missing_categories = selected_ids[!(selected_ids %in% as.vector(node_gains_GO_transfer_count_df$GO_term))] #isolate missing categories
    #if there are any missing categories
    if (length(missing_categories) > 0) {
      #add one line with zero to the node/species_tissue df for each of the missing categories
      node_gains_GO_transfer_count_df = rbind(node_gains_GO_transfer_count_df, 
                                              data.frame("GO_term"=missing_categories, num=rep(0,length(missing_categories))))
      }
    } else { #If there are no gains for that node/species_tissue, add 0 for ALL categories
      node_gains_GO_transfer_count_df = data.frame("GO_term"=selected_ids, num=rep(0, length(selected_ids)))
    }
  ### Add information about the total number of gains for each node/species_tissue
  node_gains_GO_transfer_count_df$background = rep(tissue_gains_counts_dict$find(comb), nrow(node_gains_GO_transfer_count_df))
  ### Compute the EXPECTED distributtion for each category (number of gains in the nodes multiplied by the total proportion of genes in that category)
  node_gains_GO_transfer_count_df$expected = node_gains_GO_transfer_count_df$background * category_proportion_dict$find(node_gains_GO_transfer_count_df$GO_term)
  node_gains_GO_transfer_count_df$GO_name = GO_ID_term_dict$find(node_gains_GO_transfer_count_df$GO_term) #Add GO name
  node_gains_GO_transfer_count_df$Node = rep(comb, nrow(node_gains_GO_transfer_count_df)) #Add node/species_tissue info
  ### Join into final dataframe
  all_node_category_proportion_df = rbind(all_node_category_proportion_df, node_gains_GO_transfer_count_df)
}

#### Compute observed vs expected ratio for all groups (use the log2 values)
#NB: num is the number of gains in each node/species_tissue belonging to each category.
all_node_category_proportion_df$proportion = log2((all_node_category_proportion_df$num / all_node_category_proportion_df$expected)+1)
#Order levels for plotting
all_node_category_proportion_df$Node = factor(all_node_category_proportion_df$Node, levels=all_combinations)
all_node_category_proportion_df$GO_term = factor(all_node_category_proportion_df$GO_term, selected_ids)

#If the total number of gains is lower than 10, set to 0.
all_node_category_proportion_df$proportion[all_node_category_proportion_df$num <= 2 & all_node_category_proportion_df$background <= 10] = 0
all_node_category_proportion_df$proportion[all_node_category_proportion_df$num < 5 & all_node_category_proportion_df$background <= 10] = 0


#######################################################################
####################### Generate final table ##########################
#######################################################################

wider_all_node_category_proportion_df = all_node_category_proportion_df %>%
  pivot_wider(GO_term, names_from = Node, values_from=proportion) %>%
  column_to_rownames("GO_term")

wider_all_node_category_proportion_df = wider_all_node_category_proportion_df[selected_ids,] #there should be no need to do this, in principle...

#Filter for at least one enrichment >=3 in the row
wider_all_node_category_proportion_df = wider_all_node_category_proportion_df[apply(wider_all_node_category_proportion_df, 1, function(x) {max(x, na.rm = TRUE) >= 3}),]
rownames(wider_all_node_category_proportion_df) = GO_ID_term_dict$find(rownames(wider_all_node_category_proportion_df))
#Remove all cases where there are only NAs (not sure why this should happen).
wider_all_node_category_proportion_df = wider_all_node_category_proportion_df[,colSums(is.na(wider_all_node_category_proportion_df)) < nrow(wider_all_node_category_proportion_df)]


#######################################################################
####################### Get annotation col df #########################
#######################################################################
annotation_col_df = data.frame(Tissues=sub(".*_", "", colnames(wider_all_node_category_proportion_df)))
rownames(annotation_col_df) = colnames(wider_all_node_category_proportion_df)
annotation_col_df$Ancestrality = rep("Ancestral", nrow(annotation_col_df))
annotation_col_df$Ancestrality[grep(paste(ordered_species, collapse="|"), rownames(annotation_col_df))] = "Species_specific"

#######################################################################
####################### Plot ##########################################
#######################################################################

GO_res_heatmap = pheatmap(wider_all_node_category_proportion_df,
                       color = c(brewer.pal(9, "PuRd")[1],brewer.pal(9, "PuRd")[3:9]),
                       breaks = c(0,1.2,2,2.5,3,4,5,6),
                       border_color = "black",
                       main = paste0("Recurrent gains in functional categories"),
                       cluster_cols = FALSE,
                       cluster_rows = TRUE,
                       annotation_col = annotation_col_df,
                       annotation_colors = list(Tissues = my_color_vector, 
                                                #Branch = my_branch_colors, 
                                                Ancestrality = my_ancestral_colors, 
                                                Recurrent_tissue=c("Most_tissues"="black", "Non_germiline"="gray82", "Germline"="mediumorchid2")),
                       annotation_legend = TRUE, 
                       show_rownames = TRUE,
                       show_colnames = FALSE,
                       #silent = TRUE,
                       #legend = FALSE,
                       #angle_col = 45,
                       #fontsize_col = 8,
                       fontsize_row = 8,
                       angle_col = 0
                       )
```

```{r GO_enrichments_multiple_all_tissues_Hs2_annot_observed_vs_expected_v3, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions"), fig.width=10, fig.height=5, eval=FALSE}
####### Upload GO transfer df
GO_transfers_df = read.delim(paste0(GO_transfers_dir, my_version, "/GO_annot/Hs2_transfers/orthogroups_from_Hs2-transferred_GO-reduced.txt"),
                             header=FALSE, col.names=c("OG_ID", "GO_term", "GO_name"))
### Generate id to term dictionary
name_id_transfers_df = unique(GO_transfers_df[,c("GO_term", "GO_name")])
GO_ID_term_dict = hashmap(as.vector(name_id_transfers_df$GO_term), as.vector(name_id_transfers_df$GO_name))

###### Upload TS-gains df
#### Upload gains info
tissue_gains_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, my_version, "/inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  input_df$Node_Tissue = paste0(input_df$TS_gains, "_", input_df$Tissue)
  tissue_gains_df = rbind(tissue_gains_df, input_df)
}

##### For each species and each ancestors, get the total number of gains (to use for the proportion computation)
tissue_gains_counts = table(as.vector(tissue_gains_df$Node_Tissue))
all_combinations_df = expand.grid(c(all_ancestors, ordered_species), ordered_tissues)
all_combinations = paste0(all_combinations_df$Var1, "_", all_combinations_df$Var2)
missing_IDs = all_combinations[!(all_combinations %in% names(tissue_gains_counts))]
if (length(missing_IDs) > 0) {
  for (missing_ID in missing_IDs) {
    tissue_gains_counts = c(tissue_gains_counts, 0)
    names(tissue_gains_counts)[length(tissue_gains_counts)] = missing_ID
  }
}
tissue_gains_counts_dict = hashmap(names(tissue_gains_counts), as.vector(tissue_gains_counts))

##### Cycle on all tissues
selected_ids = vector()
annotation_row_df = vector()
final_GO_res_df = vector()
for (tissue in ordered_tissues) {
  for (node in c(ordered_species, all_ancestors)) {
    for (inference_type in c("gains")) {
      GO_res_file = paste0(ts_gains_losses_dir, my_version, "/inferences/GO_enrichments/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_", inference_type, "-", node, "-orthogroups_from_Hs2-GO_res.tab")
      if (!file.exists(GO_res_file) | file.size(GO_res_file) == 1 | is.na(file.size(GO_res_file))) { #If the file does not exists
        next
      }
      else { #if file exists
        GO_res_df = read.delim(GO_res_file, header=TRUE)
        #Select entries with at least 5 genes in the set.
        GO_res_df = subset(GO_res_df, significant=="TRUE" & intersection_size>=5)
        #If there are no significant entries, continue
        if (nrow(GO_res_df)==0) {
          next
        }
        #Add tissue
        GO_res_df$node_tissue_inference_type = rep(paste0(tissue, "_", node), nrow(GO_res_df))
        #Add to final df
        final_GO_res_df = rbind(final_GO_res_df, GO_res_df)
      }
    }
  }
}
heatmap_input_df = final_GO_res_df[,c("term_name", "term_id", "node_tissue_inference_type")]
heatmap_input_df$significant = rep(1, nrow(heatmap_input_df))
### Generate wide dataframe
heatmap_input_wide_tibble = heatmap_input_df %>% pivot_wider(names_from="node_tissue_inference_type", values_from="significant", values_fill=0)
heatmap_input_wide_df = as.data.frame(heatmap_input_wide_tibble)
rownames(heatmap_input_wide_df) = heatmap_input_wide_df$term_name
heatmap_input_wide_df$term_name = NULL
heatmap_input_wide_df[is.na(heatmap_input_wide_df)] = 0

#Select categories significant in at least 15 nodes/species
selected_terms = names(table(heatmap_input_df$term_id)[table(heatmap_input_df$term_id)>=15])
#tissue_selected_ids = unique(as.vector(subset(heatmap_input_df, term_name %in% selected_terms)$term_id))
tissue_selected_ids = selected_terms
selected_ids = c(selected_ids, tissue_selected_ids)
annotation_row_df = rbind(annotation_row_df, data.frame(GO_term=tissue_selected_ids, GO_class=rep(tissue, length(tissue_selected_ids))))


#### get annotation row_df
annotation_row_df$GO_class = as.vector(annotation_row_df$GO_class)
annotation_row_df$GO_term = as.vector(annotation_row_df$GO_term)
annotation_row_df$GO_class[annotation_row_df$GO_term %in% selected_ids[duplicated(selected_ids)]] = "Multiple"
annotation_row_df = unique(annotation_row_df)

selected_ids = unique(selected_ids) #Remove extra copies of the duplicated ids

#### Cycle on the selected categories, and compute the proportion of annotated gains
selected_GO_transfer_df = subset(GO_transfers_df, GO_term %in% selected_ids)
selected_GO_transfer_df$GO_term = as.vector(selected_GO_transfer_df$GO_term)
selected_GO_transfer_df$OG_ID = as.vector(selected_GO_transfer_df$OG_ID)

##### For each selected IDs, compute proportion of genes belonging to each category over the total numbers of genes with a functional association
#This will be used to compute the expected number of gains in each category, if these gains were random.
genes_with_annot = unique(as.vector(GO_transfers_df$OG_ID))
#Subset for the ones that among the bilaterian conserved
all_orthogroups_df = read.delim(paste0(gene_sets_dir, my_version, "/STRICT/Bilateria/conserved/Bilateria_conserved_orthogroups-EXPR_genes.txt"),
                                 header=FALSE, col.names = c("OG_ID", "Species", "GeneID"))
bilaterian_conserved_OGs = unique(as.vector(all_orthogroups_df$OG_ID))
Tot_genes_with_annot = length(genes_with_annot[genes_with_annot %in% bilaterian_conserved_OGs])

category_proportion_df = as.data.frame(table(selected_GO_transfer_df$GO_term))
colnames(category_proportion_df) = c("GO_term", "Freq")
category_proportion_df$proportion = category_proportion_df$Freq/Tot_genes_with_annot
category_proportion_dict = hashmap(as.vector(category_proportion_df$GO_term), as.vector(category_proportion_df$proportion))

#### For each node
all_node_category_proportion_df = vector()
for (node in all_combinations) {
  node_gains_OGs = as.vector(subset(tissue_gains_df, Node_Tissue==node)$OG_ID) #Select all orthogroups for that node_tissue
  node_gains_GO_transfer_count_df = as.data.frame(table(as.vector(subset(selected_GO_transfer_df, GO_term %in% selected_ids & OG_ID %in% node_gains_OGs)$GO_term)))
  if (nrow(node_gains_GO_transfer_count_df) > 0) { #If there is at least one gain for that node
    colnames(node_gains_GO_transfer_count_df) = c("GO_term", "num") 
    #### Add zero counts for missing categories
    missing_categories = selected_ids[!(selected_ids %in% as.vector(node_gains_GO_transfer_count_df$GO_term))]
    if (length(missing_categories) > 0) {node_gains_GO_transfer_count_df = rbind(node_gains_GO_transfer_count_df, 
                                                                                 data.frame("GO_term"=missing_categories,
                                                                                            num=rep(0,length(missing_categories))))}
    } else { #If there are no gains for that node, add 0 for ALL categories
      node_gains_GO_transfer_count_df = data.frame("GO_term"=selected_ids, num=rep(0, length(selected_ids)))
    }
  #Add the total number of gains for each node
  node_gains_GO_transfer_count_df$background = rep(tissue_gains_counts_dict$find(node), nrow(node_gains_GO_transfer_count_df))
  #Compute the expected distributtion for each category (number of gains in the nodes multiplied by the total proportion of genes in that category)
  node_gains_GO_transfer_count_df$expected = node_gains_GO_transfer_count_df$background * category_proportion_dict$find(node_gains_GO_transfer_count_df$GO_term)
  node_gains_GO_transfer_count_df$GO_name = GO_ID_term_dict$find(node_gains_GO_transfer_count_df$GO_term)
  node_gains_GO_transfer_count_df$Node = rep(node, nrow(node_gains_GO_transfer_count_df))
  #Join into final dataframe
  all_node_category_proportion_df = rbind(all_node_category_proportion_df, node_gains_GO_transfer_count_df)
}
###### Get annotation row df for heatmap annotation
rownames(annotation_row_df) = annotation_row_df$GO_term
annotation_row_df$GO_term = NULL

#Compute observed vs expected ratio
all_node_category_proportion_df$proportion = log2((all_node_category_proportion_df$num / all_node_category_proportion_df$expected)+1)
#Order levels for plotting
all_node_category_proportion_df$Node = factor(all_node_category_proportion_df$Node, levels=all_combinations)
all_node_category_proportion_df$GO_term = factor(all_node_category_proportion_df$GO_term, selected_ids)

#If the total number of gains is lower than 10, set to 0.
#all_node_category_proportion_df$proportion[all_node_category_proportion_df$num <= 2 & all_node_category_proportion_df$background <= 10] = 0
#all_node_category_proportion_df$proportion[all_node_category_proportion_df$num < 5 & all_node_category_proportion_df$background <= 10] = 0

wider_all_node_category_proportion_df = all_node_category_proportion_df %>%
  pivot_wider(GO_term, names_from = Node, values_from=proportion) %>%
  column_to_rownames("GO_term")

wider_all_node_category_proportion_df = wider_all_node_category_proportion_df[selected_ids,]


###### Get annotation column df for heatmap annotation
annotation_col_df = data.frame(Tissues=sub(".*_", "", colnames(wider_all_node_category_proportion_df)))
rownames(annotation_col_df) = colnames(wider_all_node_category_proportion_df)
annotation_col_df$Ancestrality = rep("Ancestral", nrow(annotation_col_df))
annotation_col_df$Ancestrality[grep(paste(ordered_species, collapse="|"), rownames(annotation_col_df))] = "Species_specific"


#Filter for at least one enrichment >=3 in the row
wider_all_node_category_proportion_df = wider_all_node_category_proportion_df[apply(wider_all_node_category_proportion_df, 1, function(x) {max(x, na.rm = TRUE) >= 3}),]
rownames(wider_all_node_category_proportion_df) = GO_ID_term_dict$find(rownames(wider_all_node_category_proportion_df))
rownames(annotation_row_df) = GO_ID_term_dict$find(rownames(annotation_row_df))

wider_all_node_category_proportion_df = wider_all_node_category_proportion_df[,colSums(is.na(wider_all_node_category_proportion_df))<nrow(wider_all_node_category_proportion_df)]
GO_res_heatmap = pheatmap(wider_all_node_category_proportion_df,
                       color = c(brewer.pal(9, "PuRd")[1],brewer.pal(9, "PuRd")[3:9]),
                       breaks = c(0,1.2,2,2.5,3,4,5,6),
                       border_color = "black",
                       main = paste0("Recurrent gains in functional categories"),
                       cluster_cols = FALSE,
                       cluster_rows = TRUE,
                       annotation_col = annotation_col_df,
                       annotation_row = subsetted_annotation_row_df,
                       annotation_colors = list(Tissues = my_color_vector, 
                                                #Branch = my_branch_colors, 
                                                Ancestrality = my_ancestral_colors, 
                                                Recurrent_tissue=c("Most_tissues"="black", "Non_germiline"="gray82", "Germline"="mediumorchid2")),
                       annotation_legend = TRUE, 
                       show_rownames = TRUE,
                       show_colnames = FALSE,
                       #silent = TRUE,
                       #legend = FALSE,
                       #angle_col = 45,
                       #fontsize_col = 8,
                       fontsize_row = 8,
                       angle_col = 0
                       )

#### order columns based on decided order
ordered_columns = c("sequence-specific_double-stranded_DNA_binding",
                    #"DNA-binding_transcription_factor_activity,_RNA_polymerase_II-specific",
                    "sensory_organ_development",
                    "metal_ion_transmembrane_transporter_activity",
                    "integral_component_of_plasma_membrane",
                    "cell_surface",
                    #"extracellular_space",
                    "meiotic_nuclear_division", 
                    "chromosome_organization_involved_in_meiotic_cell_cycle")


wider_all_node_category_proportion_df = wider_all_node_category_proportion_df[ordered_columns,]
subsetted_annotation_row_df = data.frame(Recurrent_tissue = c(rep("Most_tissues", 3),
                                                      rep("Non_germiline", 2),
                                                      rep("Germline", 2)))
rownames(subsetted_annotation_row_df) = ordered_columns
                                         
#pdf("/Users/federica/Documents/CRG_PhD/bilaterian_GE/Images/Images_from_R/new_Fig5/Fig5_panelJ.pdf", width=10, height=5)
GO_res_heatmap = pheatmap(wider_all_node_category_proportion_df,
                       color = c(brewer.pal(9, "PuRd")[1],brewer.pal(9, "PuRd")[3:9]),
                       breaks = c(0,1.2,2,2.5,3,4,5,6),
                       border_color = "black",
                       main = paste0("Recurrent gains in functional categories"),
                       cluster_cols = FALSE,
                       cluster_rows = FALSE,
                       annotation_col = annotation_col_df,
                       annotation_row = subsetted_annotation_row_df,
                       annotation_colors = list(Tissues = my_color_vector, 
                                                #Branch = my_branch_colors, 
                                                Ancestrality = my_ancestral_colors, 
                                                Recurrent_tissue=c("Most_tissues"="black", "Non_germiline"="gray82", "Germline"="mediumorchid2")),
                       annotation_legend = TRUE, 
                       show_rownames = TRUE,
                       show_colnames = FALSE,
                       #silent = TRUE,
                       #legend = FALSE,
                       #angle_col = 45,
                       #fontsize_col = 8,
                       fontsize_row = 8,
                       angle_col = 0
                       )
#dev.off()
```

```{r GO_enrichments_multiple_all_tissues_Hs2_annot_observed_vs_expected_v4, echo=FALSE, message=FALSE, warning=FALSE, dependson=c("setup", "functions"), fig.width=10, fig.height=5, eval=FALSE}
####### Upload GO transfer df
GO_transfers_df = read.delim(paste0(GO_transfers_dir, my_version, "/GO_annot/Hs2_transfers/orthogroups_from_Hs2-transferred_GO-reduced.txt"),
                             header=FALSE, col.names=c("OG_ID", "GO_term", "GO_name"))
### Generate id to term dictionary
name_id_transfers_df = unique(GO_transfers_df[,c("GO_term", "GO_name")])
GO_ID_term_dict = hashmap(as.vector(name_id_transfers_df$GO_term), as.vector(name_id_transfers_df$GO_name))

###### Upload TS-gains df
#### Upload gains info
tissue_gains_df = vector()
for (tissue in ordered_tissues) {
  input_df = read.delim(paste0(ts_gains_losses_dir, my_version, "/inferences/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_gains.tab"), header=TRUE, stringsAsFactors = FALSE, col.names = c("OG_ID", "TS_gains", "Category"))
  input_df$Tissue = rep(tissue, nrow(input_df))
  input_df$Node_Tissue = paste0(input_df$TS_gains, "_", input_df$Tissue)
  tissue_gains_df = rbind(tissue_gains_df, input_df)
}

##### For each species and each ancestors, get the total number of gains (to use for the proportion computation)
tissue_gains_counts = table(as.vector(tissue_gains_df$Node_Tissue))
all_combinations_df = expand.grid(c(all_ancestors, ordered_species), ordered_tissues)
all_combinations = paste0(all_combinations_df$Var1, "_", all_combinations_df$Var2)
missing_IDs = all_combinations[!(all_combinations %in% names(tissue_gains_counts))]
if (length(missing_IDs) > 0) {
  for (missing_ID in missing_IDs) {
    tissue_gains_counts = c(tissue_gains_counts, 0)
    names(tissue_gains_counts)[length(tissue_gains_counts)] = missing_ID
  }
}
tissue_gains_counts_dict = hashmap(names(tissue_gains_counts), as.vector(tissue_gains_counts))

##### Cycle on all tissues
selected_ids = vector()
all_significant_terms = vector()
annotation_row_df = vector()
for (tissue in ordered_tissues) {
  final_GO_res_df = vector()
  for (node in c(ordered_species, all_ancestors)) {
    for (inference_type in c("gains")) {
      GO_res_file = paste0(ts_gains_losses_dir, my_version, "/inferences/GO_enrichments/Bilateria_conserved_orthogroups-All_genes-", tissue, "_inferred_", inference_type, "-", node, "-orthogroups_from_Hs2-GO_res.tab")
      if (!file.exists(GO_res_file) | file.size(GO_res_file) == 1 | is.na(file.size(GO_res_file))) { #If the file does not exists
        next
      }
      else { #if file exists
        GO_res_df = read.delim(GO_res_file, header=TRUE)
        #Select entries with at least 5 genes in the set.
        GO_res_df = subset(GO_res_df, significant=="TRUE" & intersection_size>=5)
        #If there are no significant entries, continue
        if (nrow(GO_res_df)==0) {
          next
        }
        #Add tissue
        GO_res_df$node_tissue_inference_type = rep(paste0(tissue, "_", node), nrow(GO_res_df))
        #Add to final df
        final_GO_res_df = rbind(final_GO_res_df, GO_res_df)
      }
    }
  }
  heatmap_input_df = final_GO_res_df[,c("term_name", "term_id", "node_tissue_inference_type")]
  heatmap_input_df$significant = rep(1, nrow(heatmap_input_df))
  ##### Generate wide dataframe
  heatmap_input_wide_tibble = heatmap_input_df %>% pivot_wider(names_from="node_tissue_inference_type", values_from="significant", values_fill=0)
  heatmap_input_wide_df = as.data.frame(heatmap_input_wide_tibble)
  rownames(heatmap_input_wide_df) = heatmap_input_wide_df$term_name
  heatmap_input_wide_df$term_name = NULL
  heatmap_input_wide_df[is.na(heatmap_input_wide_df)] = 0 #Replace NA with 0
  
  #####Select categories significant in at least three nodes/species
  selected_terms = names(table(heatmap_input_df$term_id)[table(heatmap_input_df$term_id)>=10])
  all_tissue_significant_terms = names(table(heatmap_input_df$term_id)[table(heatmap_input_df$term_id)>=1])
  tissue_selected_ids = selected_terms
  selected_ids = c(selected_ids, tissue_selected_ids)
  all_significant_terms = c(all_significant_terms, all_tissue_significant_terms)
  annotation_row_df = rbind(annotation_row_df, data.frame(GO_term=tissue_selected_ids, GO_class=rep(tissue, length(tissue_selected_ids))))
}

### select only terms that are significant in only one tissue: if there are duplicated values, it is because they come from different tissues
non_repeated_significant_terms = all_significant_terms[!(duplicated(all_significant_terms)|duplicated(all_significant_terms, fromLast=TRUE))]
enriched_non_repeated_significant_terms = non_repeated_significant_terms[non_repeated_significant_terms %in% selected_ids]
selected_ids = enriched_non_repeated_significant_terms

#### get annotation row_df
annotation_row_df$GO_class = as.vector(annotation_row_df$GO_class)
annotation_row_df$GO_term = as.vector(annotation_row_df$GO_term)
#annotation_row_df$GO_class[annotation_row_df$GO_term %in% selected_ids[duplicated(selected_ids)]] = "Multiple"
annotation_row_df = subset(annotation_row_df, GO_term %in% selected_ids)

#selected_ids = unique(selected_ids) #Remove extra copies of the duplicated ids

#### Cycle on the selected categories, and compute the proportion of annotated gains
selected_GO_transfer_df = subset(GO_transfers_df, GO_term %in% selected_ids)
selected_GO_transfer_df$GO_term = as.vector(selected_GO_transfer_df$GO_term)
selected_GO_transfer_df$OG_ID = as.vector(selected_GO_transfer_df$OG_ID)

##### For each selected IDs, compute proportion of genes belonging to each category over the total numbers of genes with a functional association
#This will be used to compute the expected number of gains in each category, if these gains were random.
genes_with_annot = unique(as.vector(GO_transfers_df$OG_ID))
#Subset for the ones that among the bilaterian conserved
all_orthogroups_df = read.delim(paste0(gene_sets_dir, my_version, "/STRICT/Bilateria/conserved/Bilateria_conserved_orthogroups-EXPR_genes.txt"),
                                 header=FALSE, col.names = c("OG_ID", "Species", "GeneID"))
bilaterian_conserved_OGs = unique(as.vector(all_orthogroups_df$OG_ID))
Tot_genes_with_annot = length(genes_with_annot[genes_with_annot %in% bilaterian_conserved_OGs])

category_proportion_df = as.data.frame(table(selected_GO_transfer_df$GO_term))
colnames(category_proportion_df) = c("GO_term", "Freq")
category_proportion_df$proportion = category_proportion_df$Freq/Tot_genes_with_annot
category_proportion_dict = hashmap(as.vector(category_proportion_df$GO_term), as.vector(category_proportion_df$proportion))

#### For each node
all_node_category_proportion_df = vector()
for (node in all_combinations) {
  node_gains_OGs = as.vector(subset(tissue_gains_df, Node_Tissue==node)$OG_ID) #Select all orthogroups for that node_tissue
  node_gains_GO_transfer_count_df = as.data.frame(table(as.vector(subset(selected_GO_transfer_df, GO_term %in% selected_ids & OG_ID %in% node_gains_OGs)$GO_term)))
  if (nrow(node_gains_GO_transfer_count_df) > 0) { #If there is at least one gain for that node
    colnames(node_gains_GO_transfer_count_df) = c("GO_term", "num") 
    #### Add zero counts for missing categories
    missing_categories = selected_ids[!(selected_ids %in% as.vector(node_gains_GO_transfer_count_df$GO_term))]
    if (length(missing_categories) > 0) {node_gains_GO_transfer_count_df = rbind(node_gains_GO_transfer_count_df, 
                                                                                 data.frame("GO_term"=missing_categories,
                                                                                            num=rep(0,length(missing_categories))))}
    } else { #If there are no gains for that node, add 0 for ALL categories
      node_gains_GO_transfer_count_df = data.frame("GO_term"=selected_ids, num=rep(0, length(selected_ids)))
    }
  #Add the total number of gains for each node
  node_gains_GO_transfer_count_df$background = rep(tissue_gains_counts_dict$find(node), nrow(node_gains_GO_transfer_count_df))
  #Compute the expected distributtion for each category (number of gains in the nodes multiplied by the total proportion of genes in that category)
  node_gains_GO_transfer_count_df$expected = node_gains_GO_transfer_count_df$background * category_proportion_dict$find(node_gains_GO_transfer_count_df$GO_term)
  node_gains_GO_transfer_count_df$GO_name = GO_ID_term_dict$find(node_gains_GO_transfer_count_df$GO_term)
  node_gains_GO_transfer_count_df$Node = rep(node, nrow(node_gains_GO_transfer_count_df))
  #Join into final dataframe
  all_node_category_proportion_df = rbind(all_node_category_proportion_df, node_gains_GO_transfer_count_df)
}
###### Get annotation row df for heatmap annotation
rownames(annotation_row_df) = annotation_row_df$GO_term
annotation_row_df$GO_term = NULL

#Compute observed vs expected ratio
all_node_category_proportion_df$proportion = log2((all_node_category_proportion_df$num / all_node_category_proportion_df$expected)+1)
#Order levels for plotting
all_node_category_proportion_df$Node = factor(all_node_category_proportion_df$Node, levels=all_combinations)
all_node_category_proportion_df$GO_term = factor(all_node_category_proportion_df$GO_term, selected_ids)

#If the total number of gains is lower than 10, set to 0.
#all_node_category_proportion_df$proportion[all_node_category_proportion_df$num <= 2 & all_node_category_proportion_df$background <= 10] = 0
#all_node_category_proportion_df$proportion[all_node_category_proportion_df$num < 5 & all_node_category_proportion_df$background <= 10] = 0

wider_all_node_category_proportion_df = all_node_category_proportion_df %>%
  pivot_wider(GO_term, names_from = Node, values_from=proportion) %>%
  column_to_rownames("GO_term")

wider_all_node_category_proportion_df = wider_all_node_category_proportion_df[selected_ids,]


###### Get annotation column df for heatmap annotation
annotation_col_df = data.frame(Tissues=sub(".*_", "", colnames(wider_all_node_category_proportion_df)))
rownames(annotation_col_df) = colnames(wider_all_node_category_proportion_df)
annotation_col_df$Ancestrality = rep("Ancestral", nrow(annotation_col_df))
annotation_col_df$Ancestrality[grep(paste(ordered_species, collapse="|"), rownames(annotation_col_df))] = "Species_specific"


#Filter for at least one enrichment >=3 in the row
wider_all_node_category_proportion_df = wider_all_node_category_proportion_df[apply(wider_all_node_category_proportion_df, 1, function(x) {max(x, na.rm = TRUE) >= 3}),]
rownames(wider_all_node_category_proportion_df) = GO_ID_term_dict$find(rownames(wider_all_node_category_proportion_df))
rownames(annotation_row_df) = GO_ID_term_dict$find(rownames(annotation_row_df))

wider_all_node_category_proportion_df = wider_all_node_category_proportion_df[,colSums(is.na(wider_all_node_category_proportion_df))<nrow(wider_all_node_category_proportion_df)]
GO_res_heatmap = pheatmap(wider_all_node_category_proportion_df,
                       color = c(brewer.pal(9, "PuRd")[1],brewer.pal(9, "PuRd")[3:9]),
                       breaks = c(0,1.2,2,2.5,3,4,5,6),
                       border_color = "black",
                       main = paste0("Recurrent gains in functional categories"),
                       cluster_cols = FALSE,
                       cluster_rows = TRUE,
                       annotation_col = annotation_col_df,
                       annotation_row = subsetted_annotation_row_df,
                       annotation_colors = list(Tissues = my_color_vector, 
                                                #Branch = my_branch_colors, 
                                                Ancestrality = my_ancestral_colors, 
                                                Recurrent_tissue=c("Most_tissues"="black", "Non_germiline"="gray82", "Germline"="mediumorchid2")),
                       annotation_legend = TRUE, 
                       show_rownames = TRUE,
                       show_colnames = FALSE,
                       #silent = TRUE,
                       #legend = FALSE,
                       #angle_col = 45,
                       #fontsize_col = 8,
                       fontsize_row = 8,
                       angle_col = 0
                       )
```
